{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Test imputation on untyped SNPs chrom 18"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "┌ Info: Precompiling MendelImpute [e47305d1-6a61-5370-bc5d-77554d143183]\n",
      "└ @ Base loading.jl:1278\n"
     ]
    }
   ],
   "source": [
    "using Revise\n",
    "using VCFTools\n",
    "using MendelImpute\n",
    "using GeneticVariation\n",
    "using Random\n",
    "using StatsBase\n",
    "using CodecZlib\n",
    "using ProgressMeter\n",
    "using JLSO\n",
    "using BenchmarkTools\n",
    "using GroupSlices\n",
    "using LinearAlgebra\n",
    "# using ProfileView\n",
    "\n",
    "BLAS.set_num_threads(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of threads = 8\n",
      "Importing reference haplotype data...\n",
      "Total windows = 2367, averaging ~ 527 unique haplotypes per window.\n",
      "\n",
      "Timings: \n",
      "    Data import                     = 21.3053 seconds\n",
      "        import target data             = 2.86344 seconds\n",
      "        import compressed haplotypes   = 18.4419 seconds\n",
      "    Computing haplotype pair        = 4.19614 seconds\n",
      "        BLAS3 mul! to get M and N      = 0.298691 seconds per thread\n",
      "        haplopair search               = 3.47523 seconds per thread\n",
      "        initializing missing           = 0.0215332 seconds per thread\n",
      "        allocating and viewing         = 0.0514944 seconds per thread\n",
      "        index conversion               = 0.00119099 seconds per thread\n",
      "    Phasing by win-win intersection = 1.34342 seconds\n",
      "        Window-by-window intersection  = 0.0943886 seconds per thread\n",
      "        Breakpoint search              = 0.973515 seconds per thread\n",
      "        Recording result               = 0.0260911 seconds per thread\n",
      "    Imputation                     = 1.62703 seconds\n",
      "        Imputing missing               = 0.638642 seconds\n",
      "        Writing to file                = 0.988389 seconds\n",
      "\n",
      "    Total time                      = 28.4731 seconds\n",
      "\n",
      " 28.473121 seconds (66.92 M allocations: 5.360 GiB, 8.76% gc time)\n",
      "error overall = 0.005257435473996073 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "# 8 threads, d = 1000\n",
    "Random.seed!(2020)\n",
    "chr = 18\n",
    "maf = 0.1\n",
    "d = 1000\n",
    "overlap = 0.0\n",
    "tgtfile = \"target.chr$chr.typedOnly.maf$maf.masked.vcf.gz\"\n",
    "reffile = \"ref.chr18.maxd$d.overlap$overlap.maf$maf.excludeTarget.jlso\"\n",
    "outfile = \"mendel.imputed.vcf.gz\"\n",
    "@time ph = phase(tgtfile, reffile, outfile);\n",
    "\n",
    "X_complete = convert_gt(UInt8, \"target.chr18.full.vcf.gz\")\n",
    "n, p = size(X_complete)\n",
    "X_mendel = convert_gt(UInt8, outfile)\n",
    "println(\"error overall = $(sum(X_mendel .!= X_complete) / n / p) \\n\")\n",
    "# rm(outfile, force=true)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of threads = 8\n",
      "Importing reference haplotype data...\n",
      "Total windows = 2367, averaging ~ 527 unique haplotypes per window.\n",
      "\n",
      "Timings: \n",
      "    Data import                     = 24.2939 seconds\n",
      "        import target data             = 4.03945 seconds\n",
      "        import compressed haplotypes   = 20.2544 seconds\n",
      "    Computing haplotype pair        = 4.14796 seconds\n",
      "        BLAS3 mul! to get M and N      = 0.316513 seconds per thread\n",
      "        haplopair search               = 3.41119 seconds per thread\n",
      "        initializing missing           = 0.0212078 seconds per thread\n",
      "        allocating and viewing         = 0.0508483 seconds per thread\n",
      "        index conversion               = 0.00139291 seconds per thread\n",
      "    Phasing by win-win intersection = 1.34999 seconds\n",
      "        Window-by-window intersection  = 0.0945875 seconds per thread\n",
      "        Breakpoint search              = 0.982726 seconds per thread\n",
      "        Recording result               = 0.0267201 seconds per thread\n",
      "    Imputation                     = 2.35556 seconds\n",
      "        Imputing missing               = 1.09677 seconds\n",
      "        Writing to file                = 1.25879 seconds\n",
      "\n",
      "    Total time                      = 32.1486 seconds\n",
      "\n",
      " 32.150886 seconds (69.58 M allocations: 5.581 GiB, 11.34% gc time)\n",
      "error overall = 0.005257435473996073 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "# 8 threads, d = 1000\n",
    "Random.seed!(2020)\n",
    "chr = 18\n",
    "maf = 0.1\n",
    "d = 1000\n",
    "overlap = 0.0\n",
    "tgtfile = \"target.chr$chr.typedOnly.maf$maf.masked.vcf.gz\"\n",
    "reffile = \"ref.chr18.maxd$d.overlap$overlap.maf$maf.excludeTarget.jlso\"\n",
    "outfile = \"mendel.imputed.vcf.gz\"\n",
    "@time ph = phase(tgtfile, reffile, outfile);\n",
    "\n",
    "X_complete = convert_gt(UInt8, \"target.chr18.full.vcf.gz\")\n",
    "n, p = size(X_complete)\n",
    "X_mendel = convert_gt(UInt8, outfile)\n",
    "println(\"error overall = $(sum(X_mendel .!= X_complete) / n / p) \\n\")\n",
    "# rm(outfile, force=true)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "using GeneticVariation\n",
    "snpscores = Vector{Float64}(undef, nrecords(outfile))\n",
    "\n",
    "# loop over SNPs\n",
    "for (i, record) in enumerate(VCF.Reader(openvcf(outfile, \"r\")))\n",
    "    snpscores[i] = parse(Float64, VCF.info(record)[1].second)\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"utf-8\"?>\n",
       "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" width=\"600\" height=\"400\" viewBox=\"0 0 2400 1600\">\n",
       "<defs>\n",
       "  <clipPath id=\"clip850\">\n",
       "    <rect x=\"0\" y=\"0\" width=\"2400\" height=\"1600\"/>\n",
       "  </clipPath>\n",
       "</defs>\n",
       "<path clip-path=\"url(#clip850)\" d=\"\n",
       "M0 1600 L2400 1600 L2400 0 L0 0  Z\n",
       "  \" fill=\"#ffffff\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<defs>\n",
       "  <clipPath id=\"clip851\">\n",
       "    <rect x=\"480\" y=\"0\" width=\"1681\" height=\"1600\"/>\n",
       "  </clipPath>\n",
       "</defs>\n",
       "<path clip-path=\"url(#clip850)\" d=\"\n",
       "M279.655 1423.18 L2352.76 1423.18 L2352.76 47.2441 L279.655 47.2441  Z\n",
       "  \" fill=\"#ffffff\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<defs>\n",
       "  <clipPath id=\"clip852\">\n",
       "    <rect x=\"279\" y=\"47\" width=\"2074\" height=\"1377\"/>\n",
       "  </clipPath>\n",
       "</defs>\n",
       "<polyline clip-path=\"url(#clip852)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  393.679,1423.18 393.679,47.2441 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip852)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  677.533,1423.18 677.533,47.2441 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip852)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  961.387,1423.18 961.387,47.2441 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip852)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  1245.24,1423.18 1245.24,47.2441 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip852)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  1529.1,1423.18 1529.1,47.2441 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip852)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  1812.95,1423.18 1812.95,47.2441 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip852)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  2096.8,1423.18 2096.8,47.2441 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip852)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  279.655,1384.24 2352.76,1384.24 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip852)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  279.655,1190.11 2352.76,1190.11 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip852)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  279.655,995.973 2352.76,995.973 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip852)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  279.655,801.841 2352.76,801.841 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip852)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  279.655,607.709 2352.76,607.709 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip852)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  279.655,413.576 2352.76,413.576 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip852)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  279.655,219.444 2352.76,219.444 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip850)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  279.655,1423.18 2352.76,1423.18 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip850)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  279.655,1423.18 279.655,47.2441 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip850)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  393.679,1423.18 393.679,1406.67 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip850)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  677.533,1423.18 677.533,1406.67 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip850)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  961.387,1423.18 961.387,1406.67 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip850)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1245.24,1423.18 1245.24,1406.67 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip850)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1529.1,1423.18 1529.1,1406.67 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip850)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1812.95,1423.18 1812.95,1406.67 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip850)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  2096.8,1423.18 2096.8,1406.67 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip850)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  279.655,1384.24 304.532,1384.24 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip850)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  279.655,1190.11 304.532,1190.11 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip850)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  279.655,995.973 304.532,995.973 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip850)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  279.655,801.841 304.532,801.841 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip850)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  279.655,607.709 304.532,607.709 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip850)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  279.655,413.576 304.532,413.576 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip850)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  279.655,219.444 304.532,219.444 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip850)\" d=\"M 0 0 M361.688 1445.17 Q358.077 1445.17 356.249 1448.74 Q354.443 1452.28 354.443 1459.41 Q354.443 1466.51 356.249 1470.08 Q358.077 1473.62 361.688 1473.62 Q365.323 1473.62 367.128 1470.08 Q368.957 1466.51 368.957 1459.41 Q368.957 1452.28 367.128 1448.74 Q365.323 1445.17 361.688 1445.17 M361.688 1441.47 Q367.499 1441.47 370.554 1446.07 Q373.633 1450.66 373.633 1459.41 Q373.633 1468.13 370.554 1472.74 Q367.499 1477.32 361.688 1477.32 Q355.878 1477.32 352.8 1472.74 Q349.744 1468.13 349.744 1459.41 Q349.744 1450.66 352.8 1446.07 Q355.878 1441.47 361.688 1441.47 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M378.702 1470.77 L383.586 1470.77 L383.586 1476.65 L378.702 1476.65 L378.702 1470.77 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M398.656 1445.17 Q395.045 1445.17 393.216 1448.74 Q391.41 1452.28 391.41 1459.41 Q391.41 1466.51 393.216 1470.08 Q395.045 1473.62 398.656 1473.62 Q402.29 1473.62 404.096 1470.08 Q405.924 1466.51 405.924 1459.41 Q405.924 1452.28 404.096 1448.74 Q402.29 1445.17 398.656 1445.17 M398.656 1441.47 Q404.466 1441.47 407.521 1446.07 Q410.6 1450.66 410.6 1459.41 Q410.6 1468.13 407.521 1472.74 Q404.466 1477.32 398.656 1477.32 Q392.846 1477.32 389.767 1472.74 Q386.711 1468.13 386.711 1459.41 Q386.711 1450.66 389.767 1446.07 Q392.846 1441.47 398.656 1441.47 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M425.67 1445.17 Q422.058 1445.17 420.23 1448.74 Q418.424 1452.28 418.424 1459.41 Q418.424 1466.51 420.23 1470.08 Q422.058 1473.62 425.67 1473.62 Q429.304 1473.62 431.109 1470.08 Q432.938 1466.51 432.938 1459.41 Q432.938 1452.28 431.109 1448.74 Q429.304 1445.17 425.67 1445.17 M425.67 1441.47 Q431.48 1441.47 434.535 1446.07 Q437.614 1450.66 437.614 1459.41 Q437.614 1468.13 434.535 1472.74 Q431.48 1477.32 425.67 1477.32 Q419.859 1477.32 416.781 1472.74 Q413.725 1468.13 413.725 1459.41 Q413.725 1450.66 416.781 1446.07 Q419.859 1441.47 425.67 1441.47 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M646.156 1445.17 Q642.545 1445.17 640.716 1448.74 Q638.911 1452.28 638.911 1459.41 Q638.911 1466.51 640.716 1470.08 Q642.545 1473.62 646.156 1473.62 Q649.79 1473.62 651.596 1470.08 Q653.425 1466.51 653.425 1459.41 Q653.425 1452.28 651.596 1448.74 Q649.79 1445.17 646.156 1445.17 M646.156 1441.47 Q651.966 1441.47 655.022 1446.07 Q658.1 1450.66 658.1 1459.41 Q658.1 1468.13 655.022 1472.74 Q651.966 1477.32 646.156 1477.32 Q640.346 1477.32 637.267 1472.74 Q634.212 1468.13 634.212 1459.41 Q634.212 1450.66 637.267 1446.07 Q640.346 1441.47 646.156 1441.47 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M663.17 1470.77 L668.054 1470.77 L668.054 1476.65 L663.17 1476.65 L663.17 1470.77 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M683.123 1445.17 Q679.512 1445.17 677.684 1448.74 Q675.878 1452.28 675.878 1459.41 Q675.878 1466.51 677.684 1470.08 Q679.512 1473.62 683.123 1473.62 Q686.758 1473.62 688.563 1470.08 Q690.392 1466.51 690.392 1459.41 Q690.392 1452.28 688.563 1448.74 Q686.758 1445.17 683.123 1445.17 M683.123 1441.47 Q688.934 1441.47 691.989 1446.07 Q695.068 1450.66 695.068 1459.41 Q695.068 1468.13 691.989 1472.74 Q688.934 1477.32 683.123 1477.32 Q677.313 1477.32 674.235 1472.74 Q671.179 1468.13 671.179 1459.41 Q671.179 1450.66 674.235 1446.07 Q677.313 1441.47 683.123 1441.47 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M700.947 1472.72 L708.586 1472.72 L708.586 1446.35 L700.276 1448.02 L700.276 1443.76 L708.54 1442.09 L713.216 1442.09 L713.216 1472.72 L720.855 1472.72 L720.855 1476.65 L700.947 1476.65 L700.947 1472.72 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M930.196 1445.17 Q926.584 1445.17 924.756 1448.74 Q922.95 1452.28 922.95 1459.41 Q922.95 1466.51 924.756 1470.08 Q926.584 1473.62 930.196 1473.62 Q933.83 1473.62 935.635 1470.08 Q937.464 1466.51 937.464 1459.41 Q937.464 1452.28 935.635 1448.74 Q933.83 1445.17 930.196 1445.17 M930.196 1441.47 Q936.006 1441.47 939.061 1446.07 Q942.14 1450.66 942.14 1459.41 Q942.14 1468.13 939.061 1472.74 Q936.006 1477.32 930.196 1477.32 Q924.385 1477.32 921.307 1472.74 Q918.251 1468.13 918.251 1459.41 Q918.251 1450.66 921.307 1446.07 Q924.385 1441.47 930.196 1441.47 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M947.209 1470.77 L952.094 1470.77 L952.094 1476.65 L947.209 1476.65 L947.209 1470.77 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M967.163 1445.17 Q963.552 1445.17 961.723 1448.74 Q959.918 1452.28 959.918 1459.41 Q959.918 1466.51 961.723 1470.08 Q963.552 1473.62 967.163 1473.62 Q970.797 1473.62 972.603 1470.08 Q974.431 1466.51 974.431 1459.41 Q974.431 1452.28 972.603 1448.74 Q970.797 1445.17 967.163 1445.17 M967.163 1441.47 Q972.973 1441.47 976.029 1446.07 Q979.107 1450.66 979.107 1459.41 Q979.107 1468.13 976.029 1472.74 Q972.973 1477.32 967.163 1477.32 Q961.353 1477.32 958.274 1472.74 Q955.219 1468.13 955.219 1459.41 Q955.219 1450.66 958.274 1446.07 Q961.353 1441.47 967.163 1441.47 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M988.204 1472.72 L1004.52 1472.72 L1004.52 1476.65 L982.58 1476.65 L982.58 1472.72 Q985.242 1469.96 989.825 1465.33 Q994.431 1460.68 995.612 1459.34 Q997.857 1456.81 998.737 1455.08 Q999.64 1453.32 999.64 1451.63 Q999.64 1448.87 997.695 1447.14 Q995.774 1445.4 992.672 1445.4 Q990.473 1445.4 988.019 1446.17 Q985.589 1446.93 982.811 1448.48 L982.811 1443.76 Q985.635 1442.62 988.089 1442.05 Q990.542 1441.47 992.579 1441.47 Q997.95 1441.47 1001.14 1444.15 Q1004.34 1446.84 1004.34 1451.33 Q1004.34 1453.46 1003.53 1455.38 Q1002.74 1457.28 1000.63 1459.87 Q1000.06 1460.54 996.954 1463.76 Q993.853 1466.95 988.204 1472.72 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M1213.58 1445.17 Q1209.96 1445.17 1208.14 1448.74 Q1206.33 1452.28 1206.33 1459.41 Q1206.33 1466.51 1208.14 1470.08 Q1209.96 1473.62 1213.58 1473.62 Q1217.21 1473.62 1219.02 1470.08 Q1220.84 1466.51 1220.84 1459.41 Q1220.84 1452.28 1219.02 1448.74 Q1217.21 1445.17 1213.58 1445.17 M1213.58 1441.47 Q1219.39 1441.47 1222.44 1446.07 Q1225.52 1450.66 1225.52 1459.41 Q1225.52 1468.13 1222.44 1472.74 Q1219.39 1477.32 1213.58 1477.32 Q1207.77 1477.32 1204.69 1472.74 Q1201.63 1468.13 1201.63 1459.41 Q1201.63 1450.66 1204.69 1446.07 Q1207.77 1441.47 1213.58 1441.47 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M1230.59 1470.77 L1235.47 1470.77 L1235.47 1476.65 L1230.59 1476.65 L1230.59 1470.77 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M1250.54 1445.17 Q1246.93 1445.17 1245.1 1448.74 Q1243.3 1452.28 1243.3 1459.41 Q1243.3 1466.51 1245.1 1470.08 Q1246.93 1473.62 1250.54 1473.62 Q1254.18 1473.62 1255.98 1470.08 Q1257.81 1466.51 1257.81 1459.41 Q1257.81 1452.28 1255.98 1448.74 Q1254.18 1445.17 1250.54 1445.17 M1250.54 1441.47 Q1256.35 1441.47 1259.41 1446.07 Q1262.49 1450.66 1262.49 1459.41 Q1262.49 1468.13 1259.41 1472.74 Q1256.35 1477.32 1250.54 1477.32 Q1244.73 1477.32 1241.65 1472.74 Q1238.6 1468.13 1238.6 1459.41 Q1238.6 1450.66 1241.65 1446.07 Q1244.73 1441.47 1250.54 1441.47 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M1281.72 1458.02 Q1285.08 1458.74 1286.95 1461 Q1288.85 1463.27 1288.85 1466.61 Q1288.85 1471.72 1285.33 1474.52 Q1281.82 1477.32 1275.33 1477.32 Q1273.16 1477.32 1270.84 1476.88 Q1268.55 1476.47 1266.1 1475.61 L1266.1 1471.1 Q1268.04 1472.23 1270.36 1472.81 Q1272.67 1473.39 1275.2 1473.39 Q1279.59 1473.39 1281.89 1471.65 Q1284.2 1469.92 1284.2 1466.61 Q1284.2 1463.55 1282.05 1461.84 Q1279.92 1460.1 1276.1 1460.1 L1272.07 1460.1 L1272.07 1456.26 L1276.28 1456.26 Q1279.73 1456.26 1281.56 1454.89 Q1283.39 1453.5 1283.39 1450.91 Q1283.39 1448.25 1281.49 1446.84 Q1279.62 1445.4 1276.1 1445.4 Q1274.18 1445.4 1271.98 1445.82 Q1269.78 1446.24 1267.14 1447.11 L1267.14 1442.95 Q1269.8 1442.21 1272.12 1441.84 Q1274.45 1441.47 1276.51 1441.47 Q1281.84 1441.47 1284.94 1443.9 Q1288.04 1446.3 1288.04 1450.43 Q1288.04 1453.3 1286.4 1455.29 Q1284.76 1457.25 1281.72 1458.02 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M1496.86 1445.17 Q1493.25 1445.17 1491.42 1448.74 Q1489.62 1452.28 1489.62 1459.41 Q1489.62 1466.51 1491.42 1470.08 Q1493.25 1473.62 1496.86 1473.62 Q1500.5 1473.62 1502.3 1470.08 Q1504.13 1466.51 1504.13 1459.41 Q1504.13 1452.28 1502.3 1448.74 Q1500.5 1445.17 1496.86 1445.17 M1496.86 1441.47 Q1502.67 1441.47 1505.73 1446.07 Q1508.81 1450.66 1508.81 1459.41 Q1508.81 1468.13 1505.73 1472.74 Q1502.67 1477.32 1496.86 1477.32 Q1491.05 1477.32 1487.97 1472.74 Q1484.92 1468.13 1484.92 1459.41 Q1484.92 1450.66 1487.97 1446.07 Q1491.05 1441.47 1496.86 1441.47 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M1513.88 1470.77 L1518.76 1470.77 L1518.76 1476.65 L1513.88 1476.65 L1513.88 1470.77 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M1533.83 1445.17 Q1530.22 1445.17 1528.39 1448.74 Q1526.58 1452.28 1526.58 1459.41 Q1526.58 1466.51 1528.39 1470.08 Q1530.22 1473.62 1533.83 1473.62 Q1537.46 1473.62 1539.27 1470.08 Q1541.1 1466.51 1541.1 1459.41 Q1541.1 1452.28 1539.27 1448.74 Q1537.46 1445.17 1533.83 1445.17 M1533.83 1441.47 Q1539.64 1441.47 1542.7 1446.07 Q1545.77 1450.66 1545.77 1459.41 Q1545.77 1468.13 1542.7 1472.74 Q1539.64 1477.32 1533.83 1477.32 Q1528.02 1477.32 1524.94 1472.74 Q1521.89 1468.13 1521.89 1459.41 Q1521.89 1450.66 1524.94 1446.07 Q1528.02 1441.47 1533.83 1441.47 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M1563.69 1446.17 L1551.89 1464.61 L1563.69 1464.61 L1563.69 1446.17 M1562.46 1442.09 L1568.34 1442.09 L1568.34 1464.61 L1573.27 1464.61 L1573.27 1468.5 L1568.34 1468.5 L1568.34 1476.65 L1563.69 1476.65 L1563.69 1468.5 L1548.09 1468.5 L1548.09 1463.99 L1562.46 1442.09 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M1781.46 1445.17 Q1777.85 1445.17 1776.02 1448.74 Q1774.21 1452.28 1774.21 1459.41 Q1774.21 1466.51 1776.02 1470.08 Q1777.85 1473.62 1781.46 1473.62 Q1785.09 1473.62 1786.9 1470.08 Q1788.73 1466.51 1788.73 1459.41 Q1788.73 1452.28 1786.9 1448.74 Q1785.09 1445.17 1781.46 1445.17 M1781.46 1441.47 Q1787.27 1441.47 1790.32 1446.07 Q1793.4 1450.66 1793.4 1459.41 Q1793.4 1468.13 1790.32 1472.74 Q1787.27 1477.32 1781.46 1477.32 Q1775.65 1477.32 1772.57 1472.74 Q1769.51 1468.13 1769.51 1459.41 Q1769.51 1450.66 1772.57 1446.07 Q1775.65 1441.47 1781.46 1441.47 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M1798.47 1470.77 L1803.36 1470.77 L1803.36 1476.65 L1798.47 1476.65 L1798.47 1470.77 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M1818.42 1445.17 Q1814.81 1445.17 1812.98 1448.74 Q1811.18 1452.28 1811.18 1459.41 Q1811.18 1466.51 1812.98 1470.08 Q1814.81 1473.62 1818.42 1473.62 Q1822.06 1473.62 1823.86 1470.08 Q1825.69 1466.51 1825.69 1459.41 Q1825.69 1452.28 1823.86 1448.74 Q1822.06 1445.17 1818.42 1445.17 M1818.42 1441.47 Q1824.23 1441.47 1827.29 1446.07 Q1830.37 1450.66 1830.37 1459.41 Q1830.37 1468.13 1827.29 1472.74 Q1824.23 1477.32 1818.42 1477.32 Q1812.61 1477.32 1809.54 1472.74 Q1806.48 1468.13 1806.48 1459.41 Q1806.48 1450.66 1809.54 1446.07 Q1812.61 1441.47 1818.42 1441.47 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M1835.48 1442.09 L1853.84 1442.09 L1853.84 1446.03 L1839.77 1446.03 L1839.77 1454.5 Q1840.79 1454.15 1841.8 1453.99 Q1842.82 1453.8 1843.84 1453.8 Q1849.63 1453.8 1853.01 1456.98 Q1856.39 1460.15 1856.39 1465.56 Q1856.39 1471.14 1852.92 1474.24 Q1849.44 1477.32 1843.12 1477.32 Q1840.95 1477.32 1838.68 1476.95 Q1836.43 1476.58 1834.03 1475.84 L1834.03 1471.14 Q1836.11 1472.28 1838.33 1472.83 Q1840.55 1473.39 1843.03 1473.39 Q1847.04 1473.39 1849.37 1471.28 Q1851.71 1469.18 1851.71 1465.56 Q1851.71 1461.95 1849.37 1459.85 Q1847.04 1457.74 1843.03 1457.74 Q1841.16 1457.74 1839.28 1458.16 Q1837.43 1458.57 1835.48 1459.45 L1835.48 1442.09 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M2064.73 1445.17 Q2061.12 1445.17 2059.29 1448.74 Q2057.49 1452.28 2057.49 1459.41 Q2057.49 1466.51 2059.29 1470.08 Q2061.12 1473.62 2064.73 1473.62 Q2068.37 1473.62 2070.17 1470.08 Q2072 1466.51 2072 1459.41 Q2072 1452.28 2070.17 1448.74 Q2068.37 1445.17 2064.73 1445.17 M2064.73 1441.47 Q2070.54 1441.47 2073.6 1446.07 Q2076.68 1450.66 2076.68 1459.41 Q2076.68 1468.13 2073.6 1472.74 Q2070.54 1477.32 2064.73 1477.32 Q2058.92 1477.32 2055.84 1472.74 Q2052.79 1468.13 2052.79 1459.41 Q2052.79 1450.66 2055.84 1446.07 Q2058.92 1441.47 2064.73 1441.47 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M2081.75 1470.77 L2086.63 1470.77 L2086.63 1476.65 L2081.75 1476.65 L2081.75 1470.77 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M2101.7 1445.17 Q2098.09 1445.17 2096.26 1448.74 Q2094.45 1452.28 2094.45 1459.41 Q2094.45 1466.51 2096.26 1470.08 Q2098.09 1473.62 2101.7 1473.62 Q2105.33 1473.62 2107.14 1470.08 Q2108.97 1466.51 2108.97 1459.41 Q2108.97 1452.28 2107.14 1448.74 Q2105.33 1445.17 2101.7 1445.17 M2101.7 1441.47 Q2107.51 1441.47 2110.57 1446.07 Q2113.64 1450.66 2113.64 1459.41 Q2113.64 1468.13 2110.57 1472.74 Q2107.51 1477.32 2101.7 1477.32 Q2095.89 1477.32 2092.81 1472.74 Q2089.76 1468.13 2089.76 1459.41 Q2089.76 1450.66 2092.81 1446.07 Q2095.89 1441.47 2101.7 1441.47 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M2129.29 1457.51 Q2126.14 1457.51 2124.29 1459.66 Q2122.46 1461.81 2122.46 1465.56 Q2122.46 1469.29 2124.29 1471.47 Q2126.14 1473.62 2129.29 1473.62 Q2132.44 1473.62 2134.27 1471.47 Q2136.12 1469.29 2136.12 1465.56 Q2136.12 1461.81 2134.27 1459.66 Q2132.44 1457.51 2129.29 1457.51 M2138.58 1442.86 L2138.58 1447.11 Q2136.82 1446.28 2135.01 1445.84 Q2133.23 1445.4 2131.47 1445.4 Q2126.84 1445.4 2124.39 1448.53 Q2121.95 1451.65 2121.61 1457.97 Q2122.97 1455.96 2125.03 1454.89 Q2127.09 1453.8 2129.57 1453.8 Q2134.78 1453.8 2137.79 1456.98 Q2140.82 1460.12 2140.82 1465.56 Q2140.82 1470.89 2137.67 1474.11 Q2134.52 1477.32 2129.29 1477.32 Q2123.3 1477.32 2120.13 1472.74 Q2116.95 1468.13 2116.95 1459.41 Q2116.95 1451.21 2120.84 1446.35 Q2124.73 1441.47 2131.28 1441.47 Q2133.04 1441.47 2134.83 1441.81 Q2136.63 1442.16 2138.58 1442.86 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M243.71 1370.04 Q240.099 1370.04 238.271 1373.6 Q236.465 1377.14 236.465 1384.27 Q236.465 1391.38 238.271 1394.94 Q240.099 1398.49 243.71 1398.49 Q247.345 1398.49 249.15 1394.94 Q250.979 1391.38 250.979 1384.27 Q250.979 1377.14 249.15 1373.6 Q247.345 1370.04 243.71 1370.04 M243.71 1366.33 Q249.52 1366.33 252.576 1370.94 Q255.655 1375.52 255.655 1384.27 Q255.655 1393 252.576 1397.61 Q249.52 1402.19 243.71 1402.19 Q237.9 1402.19 234.821 1397.61 Q231.766 1393 231.766 1384.27 Q231.766 1375.52 234.821 1370.94 Q237.9 1366.33 243.71 1366.33 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M131.538 1209.9 L139.177 1209.9 L139.177 1183.53 L130.867 1185.2 L130.867 1180.94 L139.13 1179.27 L143.806 1179.27 L143.806 1209.9 L151.445 1209.9 L151.445 1213.83 L131.538 1213.83 L131.538 1209.9 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M184.686 1188.37 L174.107 1199 L184.686 1209.57 L181.931 1212.38 L171.306 1201.75 L160.681 1212.38 L157.95 1209.57 L168.505 1199 L157.95 1188.37 L160.681 1185.57 L171.306 1196.19 L181.931 1185.57 L184.686 1188.37 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M190.565 1209.9 L198.204 1209.9 L198.204 1183.53 L189.894 1185.2 L189.894 1180.94 L198.158 1179.27 L202.834 1179.27 L202.834 1209.9 L210.473 1209.9 L210.473 1213.83 L190.565 1213.83 L190.565 1209.9 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M225.542 1182.35 Q221.931 1182.35 220.102 1185.92 Q218.297 1189.46 218.297 1196.59 Q218.297 1203.69 220.102 1207.26 Q221.931 1210.8 225.542 1210.8 Q229.176 1210.8 230.982 1207.26 Q232.811 1203.69 232.811 1196.59 Q232.811 1189.46 230.982 1185.92 Q229.176 1182.35 225.542 1182.35 M225.542 1178.65 Q231.352 1178.65 234.408 1183.25 Q237.486 1187.84 237.486 1196.59 Q237.486 1205.32 234.408 1209.92 Q231.352 1214.5 225.542 1214.5 Q219.732 1214.5 216.653 1209.92 Q213.598 1205.32 213.598 1196.59 Q213.598 1187.84 216.653 1183.25 Q219.732 1178.65 225.542 1178.65 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M238.671 1158.34 L253.586 1158.34 L253.586 1161.54 L242.151 1161.54 L242.151 1168.42 Q242.978 1168.14 243.806 1168.01 Q244.633 1167.86 245.461 1167.86 Q250.163 1167.86 252.909 1170.44 Q255.655 1173.01 255.655 1177.41 Q255.655 1181.95 252.834 1184.47 Q250.012 1186.97 244.878 1186.97 Q243.11 1186.97 241.267 1186.67 Q239.442 1186.37 237.486 1185.76 L237.486 1181.95 Q239.179 1182.87 240.985 1183.32 Q242.79 1183.77 244.803 1183.77 Q248.056 1183.77 249.956 1182.06 Q251.856 1180.35 251.856 1177.41 Q251.856 1174.48 249.956 1172.77 Q248.056 1171.06 244.803 1171.06 Q243.279 1171.06 241.756 1171.4 Q240.251 1171.73 238.671 1172.45 L238.671 1158.34 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M135.126 1015.77 L151.445 1015.77 L151.445 1019.7 L129.501 1019.7 L129.501 1015.77 Q132.163 1013.01 136.746 1008.38 Q141.353 1003.73 142.533 1002.39 Q144.779 999.863 145.658 998.127 Q146.561 996.368 146.561 994.678 Q146.561 991.923 144.617 990.187 Q142.695 988.451 139.593 988.451 Q137.394 988.451 134.941 989.215 Q132.51 989.979 129.732 991.53 L129.732 986.808 Q132.556 985.674 135.01 985.095 Q137.464 984.516 139.501 984.516 Q144.871 984.516 148.066 987.201 Q151.26 989.886 151.26 994.377 Q151.26 996.507 150.45 998.428 Q149.663 1000.33 147.556 1002.92 Q146.978 1003.59 143.876 1006.81 Q140.774 1010 135.126 1015.77 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M184.686 994.238 L174.107 1004.86 L184.686 1015.44 L181.931 1018.24 L171.306 1007.62 L160.681 1018.24 L157.95 1015.44 L168.505 1004.86 L157.95 994.238 L160.681 991.437 L171.306 1002.06 L181.931 991.437 L184.686 994.238 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M190.565 1015.77 L198.204 1015.77 L198.204 989.4 L189.894 991.067 L189.894 986.808 L198.158 985.141 L202.834 985.141 L202.834 1015.77 L210.473 1015.77 L210.473 1019.7 L190.565 1019.7 L190.565 1015.77 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M225.542 988.22 Q221.931 988.22 220.102 991.785 Q218.297 995.326 218.297 1002.46 Q218.297 1009.56 220.102 1013.13 Q221.931 1016.67 225.542 1016.67 Q229.176 1016.67 230.982 1013.13 Q232.811 1009.56 232.811 1002.46 Q232.811 995.326 230.982 991.785 Q229.176 988.22 225.542 988.22 M225.542 984.516 Q231.352 984.516 234.408 989.123 Q237.486 993.706 237.486 1002.46 Q237.486 1011.18 234.408 1015.79 Q231.352 1020.37 225.542 1020.37 Q219.732 1020.37 216.653 1015.79 Q213.598 1011.18 213.598 1002.46 Q213.598 993.706 216.653 989.123 Q219.732 984.516 225.542 984.516 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M238.671 964.211 L253.586 964.211 L253.586 967.408 L242.151 967.408 L242.151 974.292 Q242.978 974.01 243.806 973.878 Q244.633 973.727 245.461 973.727 Q250.163 973.727 252.909 976.304 Q255.655 978.881 255.655 983.282 Q255.655 987.814 252.834 990.335 Q250.012 992.836 244.878 992.836 Q243.11 992.836 241.267 992.535 Q239.442 992.234 237.486 991.632 L237.486 987.814 Q239.179 988.736 240.985 989.187 Q242.79 989.639 244.803 989.639 Q248.056 989.639 249.956 987.927 Q251.856 986.216 251.856 983.282 Q251.856 980.348 249.956 978.636 Q248.056 976.925 244.803 976.925 Q243.279 976.925 241.756 977.263 Q240.251 977.602 238.671 978.317 L238.671 964.211 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M144.316 806.935 Q147.672 807.652 149.547 809.921 Q151.445 812.189 151.445 815.522 Q151.445 820.638 147.927 823.439 Q144.408 826.24 137.927 826.24 Q135.751 826.24 133.436 825.8 Q131.144 825.384 128.691 824.527 L128.691 820.013 Q130.635 821.147 132.95 821.726 Q135.265 822.305 137.788 822.305 Q142.186 822.305 144.478 820.569 Q146.792 818.833 146.792 815.522 Q146.792 812.467 144.64 810.754 Q142.51 809.018 138.691 809.018 L134.663 809.018 L134.663 805.175 L138.876 805.175 Q142.325 805.175 144.154 803.81 Q145.982 802.421 145.982 799.828 Q145.982 797.166 144.084 795.754 Q142.209 794.319 138.691 794.319 Q136.769 794.319 134.57 794.736 Q132.371 795.152 129.732 796.032 L129.732 791.865 Q132.394 791.124 134.709 790.754 Q137.047 790.384 139.107 790.384 Q144.431 790.384 147.533 792.814 Q150.635 795.222 150.635 799.342 Q150.635 802.212 148.992 804.203 Q147.348 806.171 144.316 806.935 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M184.686 800.106 L174.107 810.731 L184.686 821.309 L181.931 824.11 L171.306 813.485 L160.681 824.11 L157.95 821.309 L168.505 810.731 L157.95 800.106 L160.681 797.305 L171.306 807.93 L181.931 797.305 L184.686 800.106 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M190.565 821.634 L198.204 821.634 L198.204 795.268 L189.894 796.935 L189.894 792.675 L198.158 791.009 L202.834 791.009 L202.834 821.634 L210.473 821.634 L210.473 825.569 L190.565 825.569 L190.565 821.634 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M225.542 794.087 Q221.931 794.087 220.102 797.652 Q218.297 801.194 218.297 808.323 Q218.297 815.43 220.102 818.995 Q221.931 822.536 225.542 822.536 Q229.176 822.536 230.982 818.995 Q232.811 815.43 232.811 808.323 Q232.811 801.194 230.982 797.652 Q229.176 794.087 225.542 794.087 M225.542 790.384 Q231.352 790.384 234.408 794.99 Q237.486 799.573 237.486 808.323 Q237.486 817.05 234.408 821.657 Q231.352 826.24 225.542 826.24 Q219.732 826.24 216.653 821.657 Q213.598 817.05 213.598 808.323 Q213.598 799.573 216.653 794.99 Q219.732 790.384 225.542 790.384 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M238.671 770.078 L253.586 770.078 L253.586 773.276 L242.151 773.276 L242.151 780.159 Q242.978 779.877 243.806 779.746 Q244.633 779.595 245.461 779.595 Q250.163 779.595 252.909 782.172 Q255.655 784.748 255.655 789.149 Q255.655 793.682 252.834 796.202 Q250.012 798.704 244.878 798.704 Q243.11 798.704 241.267 798.403 Q239.442 798.102 237.486 797.5 L237.486 793.682 Q239.179 794.604 240.985 795.055 Q242.79 795.506 244.803 795.506 Q248.056 795.506 249.956 793.795 Q251.856 792.083 251.856 789.149 Q251.856 786.215 249.956 784.504 Q248.056 782.792 244.803 782.792 Q243.279 782.792 241.756 783.131 Q240.251 783.469 238.671 784.184 L238.671 770.078 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M141.862 600.95 L130.056 619.399 L141.862 619.399 L141.862 600.95 M140.635 596.876 L146.515 596.876 L146.515 619.399 L151.445 619.399 L151.445 623.288 L146.515 623.288 L146.515 631.436 L141.862 631.436 L141.862 623.288 L126.26 623.288 L126.26 618.774 L140.635 596.876 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M184.686 605.974 L174.107 616.598 L184.686 627.177 L181.931 629.978 L171.306 619.353 L160.681 629.978 L157.95 627.177 L168.505 616.598 L157.95 605.974 L160.681 603.173 L171.306 613.798 L181.931 603.173 L184.686 605.974 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M190.565 627.501 L198.204 627.501 L198.204 601.136 L189.894 602.802 L189.894 598.543 L198.158 596.876 L202.834 596.876 L202.834 627.501 L210.473 627.501 L210.473 631.436 L190.565 631.436 L190.565 627.501 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M225.542 599.955 Q221.931 599.955 220.102 603.52 Q218.297 607.061 218.297 614.191 Q218.297 621.297 220.102 624.862 Q221.931 628.404 225.542 628.404 Q229.176 628.404 230.982 624.862 Q232.811 621.297 232.811 614.191 Q232.811 607.061 230.982 603.52 Q229.176 599.955 225.542 599.955 M225.542 596.251 Q231.352 596.251 234.408 600.858 Q237.486 605.441 237.486 614.191 Q237.486 622.918 234.408 627.524 Q231.352 632.108 225.542 632.108 Q219.732 632.108 216.653 627.524 Q213.598 622.918 213.598 614.191 Q213.598 605.441 216.653 600.858 Q219.732 596.251 225.542 596.251 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M238.671 575.946 L253.586 575.946 L253.586 579.143 L242.151 579.143 L242.151 586.027 Q242.978 585.745 243.806 585.613 Q244.633 585.463 245.461 585.463 Q250.163 585.463 252.909 588.039 Q255.655 590.616 255.655 595.017 Q255.655 599.55 252.834 602.07 Q250.012 604.571 244.878 604.571 Q243.11 604.571 241.267 604.27 Q239.442 603.97 237.486 603.368 L237.486 599.55 Q239.179 600.471 240.985 600.923 Q242.79 601.374 244.803 601.374 Q248.056 601.374 249.956 599.663 Q251.856 597.951 251.856 595.017 Q251.856 592.083 249.956 590.371 Q248.056 588.66 244.803 588.66 Q243.279 588.66 241.756 588.999 Q240.251 589.337 238.671 590.052 L238.671 575.946 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M130.543 402.744 L148.899 402.744 L148.899 406.679 L134.825 406.679 L134.825 415.151 Q135.843 414.804 136.862 414.642 Q137.88 414.457 138.899 414.457 Q144.686 414.457 148.066 417.628 Q151.445 420.799 151.445 426.216 Q151.445 431.795 147.973 434.897 Q144.501 437.975 138.181 437.975 Q136.005 437.975 133.737 437.605 Q131.492 437.235 129.084 436.494 L129.084 431.795 Q131.168 432.929 133.39 433.485 Q135.612 434.04 138.089 434.04 Q142.093 434.04 144.431 431.934 Q146.769 429.827 146.769 426.216 Q146.769 422.605 144.431 420.498 Q142.093 418.392 138.089 418.392 Q136.214 418.392 134.339 418.809 Q132.487 419.225 130.543 420.105 L130.543 402.744 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M184.686 411.841 L174.107 422.466 L184.686 433.045 L181.931 435.846 L171.306 425.221 L160.681 435.846 L157.95 433.045 L168.505 422.466 L157.95 411.841 L160.681 409.04 L171.306 419.665 L181.931 409.04 L184.686 411.841 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M190.565 433.369 L198.204 433.369 L198.204 407.003 L189.894 408.67 L189.894 404.411 L198.158 402.744 L202.834 402.744 L202.834 433.369 L210.473 433.369 L210.473 437.304 L190.565 437.304 L190.565 433.369 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M225.542 405.823 Q221.931 405.823 220.102 409.387 Q218.297 412.929 218.297 420.059 Q218.297 427.165 220.102 430.73 Q221.931 434.272 225.542 434.272 Q229.176 434.272 230.982 430.73 Q232.811 427.165 232.811 420.059 Q232.811 412.929 230.982 409.387 Q229.176 405.823 225.542 405.823 M225.542 402.119 Q231.352 402.119 234.408 406.725 Q237.486 411.309 237.486 420.059 Q237.486 428.785 234.408 433.392 Q231.352 437.975 225.542 437.975 Q219.732 437.975 216.653 433.392 Q213.598 428.785 213.598 420.059 Q213.598 411.309 216.653 406.725 Q219.732 402.119 225.542 402.119 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M238.671 381.814 L253.586 381.814 L253.586 385.011 L242.151 385.011 L242.151 391.895 Q242.978 391.612 243.806 391.481 Q244.633 391.33 245.461 391.33 Q250.163 391.33 252.909 393.907 Q255.655 396.484 255.655 400.885 Q255.655 405.417 252.834 407.938 Q250.012 410.439 244.878 410.439 Q243.11 410.439 241.267 410.138 Q239.442 409.837 237.486 409.235 L237.486 405.417 Q239.179 406.339 240.985 406.79 Q242.79 407.242 244.803 407.242 Q248.056 407.242 249.956 405.53 Q251.856 403.819 251.856 400.885 Q251.856 397.951 249.956 396.239 Q248.056 394.528 244.803 394.528 Q243.279 394.528 241.756 394.866 Q240.251 395.205 238.671 395.919 L238.671 381.814 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M139.917 224.028 Q136.769 224.028 134.918 226.181 Q133.089 228.334 133.089 232.084 Q133.089 235.81 134.918 237.986 Q136.769 240.139 139.917 240.139 Q143.066 240.139 144.894 237.986 Q146.746 235.81 146.746 232.084 Q146.746 228.334 144.894 226.181 Q143.066 224.028 139.917 224.028 M149.2 209.375 L149.2 213.635 Q147.441 212.801 145.635 212.362 Q143.853 211.922 142.093 211.922 Q137.464 211.922 135.01 215.047 Q132.58 218.172 132.232 224.491 Q133.598 222.477 135.658 221.412 Q137.718 220.324 140.195 220.324 Q145.404 220.324 148.413 223.496 Q151.445 226.644 151.445 232.084 Q151.445 237.408 148.297 240.625 Q145.149 243.843 139.917 243.843 Q133.922 243.843 130.751 239.26 Q127.58 234.653 127.58 225.926 Q127.58 217.732 131.468 212.871 Q135.357 207.987 141.908 207.987 Q143.667 207.987 145.45 208.334 Q147.255 208.681 149.2 209.375 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M184.686 217.709 L174.107 228.334 L184.686 238.912 L181.931 241.713 L171.306 231.088 L160.681 241.713 L157.95 238.912 L168.505 228.334 L157.95 217.709 L160.681 214.908 L171.306 225.533 L181.931 214.908 L184.686 217.709 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M190.565 239.236 L198.204 239.236 L198.204 212.871 L189.894 214.537 L189.894 210.278 L198.158 208.612 L202.834 208.612 L202.834 239.236 L210.473 239.236 L210.473 243.172 L190.565 243.172 L190.565 239.236 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M225.542 211.69 Q221.931 211.69 220.102 215.255 Q218.297 218.797 218.297 225.926 Q218.297 233.033 220.102 236.598 Q221.931 240.139 225.542 240.139 Q229.176 240.139 230.982 236.598 Q232.811 233.033 232.811 225.926 Q232.811 218.797 230.982 215.255 Q229.176 211.69 225.542 211.69 M225.542 207.987 Q231.352 207.987 234.408 212.593 Q237.486 217.176 237.486 225.926 Q237.486 234.653 234.408 239.26 Q231.352 243.843 225.542 243.843 Q219.732 243.843 216.653 239.26 Q213.598 234.653 213.598 225.926 Q213.598 217.176 216.653 212.593 Q219.732 207.987 225.542 207.987 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M238.671 187.681 L253.586 187.681 L253.586 190.878 L242.151 190.878 L242.151 197.762 Q242.978 197.48 243.806 197.348 Q244.633 197.198 245.461 197.198 Q250.163 197.198 252.909 199.775 Q255.655 202.351 255.655 206.752 Q255.655 211.285 252.834 213.805 Q250.012 216.307 244.878 216.307 Q243.11 216.307 241.267 216.006 Q239.442 215.705 237.486 215.103 L237.486 211.285 Q239.179 212.206 240.985 212.658 Q242.79 213.109 244.803 213.109 Q248.056 213.109 249.956 211.398 Q251.856 209.686 251.856 206.752 Q251.856 203.818 249.956 202.107 Q248.056 200.395 244.803 200.395 Q243.279 200.395 241.756 200.734 Q240.251 201.072 238.671 201.787 L238.671 187.681 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M1028.95 1508.52 L1035.38 1508.52 L1035.38 1550.63 L1058.52 1550.63 L1058.52 1556.04 L1028.95 1556.04 L1028.95 1508.52 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M1094.01 1536.76 L1094.01 1539.62 L1067.08 1539.62 Q1067.47 1545.67 1070.71 1548.85 Q1073.99 1552 1079.81 1552 Q1083.19 1552 1086.34 1551.17 Q1089.52 1550.35 1092.64 1548.69 L1092.64 1554.23 Q1089.49 1555.57 1086.18 1556.27 Q1082.87 1556.97 1079.46 1556.97 Q1070.93 1556.97 1065.94 1552 Q1060.97 1547.04 1060.97 1538.57 Q1060.97 1529.82 1065.68 1524.69 Q1070.43 1519.54 1078.45 1519.54 Q1085.64 1519.54 1089.81 1524.18 Q1094.01 1528.8 1094.01 1536.76 M1088.15 1535.04 Q1088.09 1530.23 1085.45 1527.37 Q1082.84 1524.5 1078.51 1524.5 Q1073.61 1524.5 1070.65 1527.27 Q1067.72 1530.04 1067.27 1535.07 L1088.15 1535.04 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M1116.35 1538.12 Q1109.26 1538.12 1106.52 1539.75 Q1103.78 1541.37 1103.78 1545.29 Q1103.78 1548.4 1105.82 1550.25 Q1107.89 1552.07 1111.42 1552.07 Q1116.29 1552.07 1119.22 1548.63 Q1122.18 1545.16 1122.18 1539.43 L1122.18 1538.12 L1116.35 1538.12 M1128.04 1535.71 L1128.04 1556.04 L1122.18 1556.04 L1122.18 1550.63 Q1120.17 1553.88 1117.18 1555.44 Q1114.19 1556.97 1109.86 1556.97 Q1104.39 1556.97 1101.14 1553.91 Q1097.93 1550.82 1097.93 1545.67 Q1097.93 1539.65 1101.94 1536.6 Q1105.98 1533.54 1113.97 1533.54 L1122.18 1533.54 L1122.18 1532.97 Q1122.18 1528.93 1119.5 1526.73 Q1116.86 1524.5 1112.06 1524.5 Q1109 1524.5 1106.11 1525.23 Q1103.21 1525.97 1100.54 1527.43 L1100.54 1522.02 Q1103.75 1520.78 1106.77 1520.17 Q1109.8 1519.54 1112.66 1519.54 Q1120.4 1519.54 1124.22 1523.55 Q1128.04 1527.56 1128.04 1535.71 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M1156.9 1521.45 L1156.9 1526.98 Q1154.42 1525.71 1151.75 1525.07 Q1149.07 1524.44 1146.21 1524.44 Q1141.85 1524.44 1139.65 1525.77 Q1137.49 1527.11 1137.49 1529.79 Q1137.49 1531.82 1139.05 1533 Q1140.61 1534.15 1145.32 1535.2 L1147.32 1535.64 Q1153.56 1536.98 1156.17 1539.43 Q1158.81 1541.85 1158.81 1546.21 Q1158.81 1551.17 1154.87 1554.07 Q1150.95 1556.97 1144.08 1556.97 Q1141.21 1556.97 1138.09 1556.39 Q1135.01 1555.85 1131.57 1554.74 L1131.57 1548.69 Q1134.81 1550.38 1137.97 1551.24 Q1141.12 1552.07 1144.2 1552.07 Q1148.34 1552.07 1150.57 1550.66 Q1152.8 1549.23 1152.8 1546.65 Q1152.8 1544.27 1151.17 1542.99 Q1149.58 1541.72 1144.14 1540.54 L1142.1 1540.07 Q1136.66 1538.92 1134.24 1536.56 Q1131.82 1534.18 1131.82 1530.04 Q1131.82 1525.01 1135.39 1522.27 Q1138.95 1519.54 1145.51 1519.54 Q1148.76 1519.54 1151.62 1520.01 Q1154.48 1520.49 1156.9 1521.45 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M1170.75 1510.27 L1170.75 1520.4 L1182.81 1520.4 L1182.81 1524.95 L1170.75 1524.95 L1170.75 1544.3 Q1170.75 1548.66 1171.93 1549.9 Q1173.14 1551.14 1176.8 1551.14 L1182.81 1551.14 L1182.81 1556.04 L1176.8 1556.04 Q1170.02 1556.04 1167.44 1553.53 Q1164.86 1550.98 1164.86 1544.3 L1164.86 1524.95 L1160.56 1524.95 L1160.56 1520.4 L1164.86 1520.4 L1164.86 1510.27 L1170.75 1510.27 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M1232.4 1521.45 L1232.4 1526.98 Q1229.92 1525.71 1227.24 1525.07 Q1224.57 1524.44 1221.71 1524.44 Q1217.35 1524.44 1215.15 1525.77 Q1212.99 1527.11 1212.99 1529.79 Q1212.99 1531.82 1214.54 1533 Q1216.1 1534.15 1220.82 1535.2 L1222.82 1535.64 Q1229.06 1536.98 1231.67 1539.43 Q1234.31 1541.85 1234.31 1546.21 Q1234.31 1551.17 1230.36 1554.07 Q1226.45 1556.97 1219.57 1556.97 Q1216.71 1556.97 1213.59 1556.39 Q1210.5 1555.85 1207.07 1554.74 L1207.07 1548.69 Q1210.31 1550.38 1213.46 1551.24 Q1216.61 1552.07 1219.7 1552.07 Q1223.84 1552.07 1226.07 1550.66 Q1228.29 1549.23 1228.29 1546.65 Q1228.29 1544.27 1226.67 1542.99 Q1225.08 1541.72 1219.64 1540.54 L1217.6 1540.07 Q1212.16 1538.92 1209.74 1536.56 Q1207.32 1534.18 1207.32 1530.04 Q1207.32 1525.01 1210.88 1522.27 Q1214.45 1519.54 1221.01 1519.54 Q1224.25 1519.54 1227.12 1520.01 Q1229.98 1520.49 1232.4 1521.45 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M1243.95 1538.25 Q1243.95 1544.71 1246.6 1548.4 Q1249.27 1552.07 1253.92 1552.07 Q1258.56 1552.07 1261.24 1548.4 Q1263.91 1544.71 1263.91 1538.25 Q1263.91 1531.79 1261.24 1528.13 Q1258.56 1524.44 1253.92 1524.44 Q1249.27 1524.44 1246.6 1528.13 Q1243.95 1531.79 1243.95 1538.25 M1263.91 1550.7 Q1262.06 1553.88 1259.23 1555.44 Q1256.43 1556.97 1252.48 1556.97 Q1246.02 1556.97 1241.95 1551.81 Q1237.91 1546.65 1237.91 1538.25 Q1237.91 1529.85 1241.95 1524.69 Q1246.02 1519.54 1252.48 1519.54 Q1256.43 1519.54 1259.23 1521.1 Q1262.06 1522.62 1263.91 1525.81 L1263.91 1520.4 L1269.77 1520.4 L1269.77 1569.6 L1263.91 1569.6 L1263.91 1550.7 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M1275.31 1541.98 L1275.31 1520.4 L1281.16 1520.4 L1281.16 1541.75 Q1281.16 1546.81 1283.14 1549.36 Q1285.11 1551.87 1289.06 1551.87 Q1293.8 1551.87 1296.54 1548.85 Q1299.3 1545.83 1299.3 1540.61 L1299.3 1520.4 L1305.16 1520.4 L1305.16 1556.04 L1299.3 1556.04 L1299.3 1550.57 Q1297.17 1553.82 1294.34 1555.41 Q1291.54 1556.97 1287.81 1556.97 Q1281.67 1556.97 1278.49 1553.15 Q1275.31 1549.33 1275.31 1541.98 M1290.04 1519.54 L1290.04 1519.54 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M1327.5 1538.12 Q1320.41 1538.12 1317.67 1539.75 Q1314.93 1541.37 1314.93 1545.29 Q1314.93 1548.4 1316.97 1550.25 Q1319.04 1552.07 1322.57 1552.07 Q1327.44 1552.07 1330.37 1548.63 Q1333.33 1545.16 1333.33 1539.43 L1333.33 1538.12 L1327.5 1538.12 M1339.19 1535.71 L1339.19 1556.04 L1333.33 1556.04 L1333.33 1550.63 Q1331.32 1553.88 1328.33 1555.44 Q1325.34 1556.97 1321.01 1556.97 Q1315.54 1556.97 1312.29 1553.91 Q1309.08 1550.82 1309.08 1545.67 Q1309.08 1539.65 1313.09 1536.6 Q1317.13 1533.54 1325.12 1533.54 L1333.33 1533.54 L1333.33 1532.97 Q1333.33 1528.93 1330.66 1526.73 Q1328.01 1524.5 1323.21 1524.5 Q1320.15 1524.5 1317.26 1525.23 Q1314.36 1525.97 1311.69 1527.43 L1311.69 1522.02 Q1314.9 1520.78 1317.92 1520.17 Q1320.95 1519.54 1323.81 1519.54 Q1331.55 1519.54 1335.37 1523.55 Q1339.19 1527.56 1339.19 1535.71 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M1365.99 1525.87 Q1365 1525.3 1363.82 1525.04 Q1362.67 1524.76 1361.27 1524.76 Q1356.31 1524.76 1353.64 1528 Q1350.99 1531.22 1350.99 1537.27 L1350.99 1556.04 L1345.11 1556.04 L1345.11 1520.4 L1350.99 1520.4 L1350.99 1525.93 Q1352.84 1522.69 1355.8 1521.13 Q1358.76 1519.54 1362.99 1519.54 Q1363.6 1519.54 1364.33 1519.63 Q1365.06 1519.7 1365.95 1519.85 L1365.99 1525.87 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M1401.19 1536.76 L1401.19 1539.62 L1374.26 1539.62 Q1374.64 1545.67 1377.89 1548.85 Q1381.17 1552 1386.99 1552 Q1390.37 1552 1393.52 1551.17 Q1396.7 1550.35 1399.82 1548.69 L1399.82 1554.23 Q1396.67 1555.57 1393.36 1556.27 Q1390.05 1556.97 1386.64 1556.97 Q1378.11 1556.97 1373.11 1552 Q1368.15 1547.04 1368.15 1538.57 Q1368.15 1529.82 1372.86 1524.69 Q1377.6 1519.54 1385.62 1519.54 Q1392.82 1519.54 1396.99 1524.18 Q1401.19 1528.8 1401.19 1536.76 M1395.33 1535.04 Q1395.27 1530.23 1392.63 1527.37 Q1390.02 1524.5 1385.69 1524.5 Q1380.79 1524.5 1377.83 1527.27 Q1374.9 1530.04 1374.45 1535.07 L1395.33 1535.04 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M1430.06 1521.45 L1430.06 1526.98 Q1427.57 1525.71 1424.9 1525.07 Q1422.23 1524.44 1419.36 1524.44 Q1415 1524.44 1412.8 1525.77 Q1410.64 1527.11 1410.64 1529.79 Q1410.64 1531.82 1412.2 1533 Q1413.76 1534.15 1418.47 1535.2 L1420.48 1535.64 Q1426.71 1536.98 1429.32 1539.43 Q1431.97 1541.85 1431.97 1546.21 Q1431.97 1551.17 1428.02 1554.07 Q1424.1 1556.97 1417.23 1556.97 Q1414.36 1556.97 1411.25 1556.39 Q1408.16 1555.85 1404.72 1554.74 L1404.72 1548.69 Q1407.97 1550.38 1411.12 1551.24 Q1414.27 1552.07 1417.36 1552.07 Q1421.49 1552.07 1423.72 1550.66 Q1425.95 1549.23 1425.95 1546.65 Q1425.95 1544.27 1424.33 1542.99 Q1422.74 1541.72 1417.29 1540.54 L1415.26 1540.07 Q1409.81 1538.92 1407.39 1536.56 Q1404.98 1534.18 1404.98 1530.04 Q1404.98 1525.01 1408.54 1522.27 Q1412.1 1519.54 1418.66 1519.54 Q1421.91 1519.54 1424.77 1520.01 Q1427.64 1520.49 1430.06 1521.45 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M1489.32 1536.76 L1489.32 1539.62 L1462.39 1539.62 Q1462.78 1545.67 1466.02 1548.85 Q1469.3 1552 1475.13 1552 Q1478.5 1552 1481.65 1551.17 Q1484.83 1550.35 1487.95 1548.69 L1487.95 1554.23 Q1484.8 1555.57 1481.49 1556.27 Q1478.18 1556.97 1474.78 1556.97 Q1466.25 1556.97 1461.25 1552 Q1456.28 1547.04 1456.28 1538.57 Q1456.28 1529.82 1460.99 1524.69 Q1465.74 1519.54 1473.76 1519.54 Q1480.95 1519.54 1485.12 1524.18 Q1489.32 1528.8 1489.32 1536.76 M1483.46 1535.04 Q1483.4 1530.23 1480.76 1527.37 Q1478.15 1524.5 1473.82 1524.5 Q1468.92 1524.5 1465.96 1527.27 Q1463.03 1530.04 1462.58 1535.07 L1483.46 1535.04 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M1516.12 1525.87 Q1515.13 1525.3 1513.96 1525.04 Q1512.81 1524.76 1511.41 1524.76 Q1506.44 1524.76 1503.77 1528 Q1501.13 1531.22 1501.13 1537.27 L1501.13 1556.04 L1495.24 1556.04 L1495.24 1520.4 L1501.13 1520.4 L1501.13 1525.93 Q1502.98 1522.69 1505.94 1521.13 Q1508.9 1519.54 1513.13 1519.54 Q1513.73 1519.54 1514.47 1519.63 Q1515.2 1519.7 1516.09 1519.85 L1516.12 1525.87 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M1541.77 1525.87 Q1540.79 1525.3 1539.61 1525.04 Q1538.46 1524.76 1537.06 1524.76 Q1532.1 1524.76 1529.42 1528 Q1526.78 1531.22 1526.78 1537.27 L1526.78 1556.04 L1520.89 1556.04 L1520.89 1520.4 L1526.78 1520.4 L1526.78 1525.93 Q1528.63 1522.69 1531.59 1521.13 Q1534.55 1519.54 1538.78 1519.54 Q1539.39 1519.54 1540.12 1519.63 Q1540.85 1519.7 1541.74 1519.85 L1541.77 1525.87 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M1560.3 1524.5 Q1555.59 1524.5 1552.85 1528.19 Q1550.11 1531.85 1550.11 1538.25 Q1550.11 1544.65 1552.82 1548.34 Q1555.56 1552 1560.3 1552 Q1564.98 1552 1567.71 1548.31 Q1570.45 1544.62 1570.45 1538.25 Q1570.45 1531.92 1567.71 1528.23 Q1564.98 1524.5 1560.3 1524.5 M1560.3 1519.54 Q1567.94 1519.54 1572.3 1524.5 Q1576.66 1529.47 1576.66 1538.25 Q1576.66 1547 1572.3 1552 Q1567.94 1556.97 1560.3 1556.97 Q1552.63 1556.97 1548.27 1552 Q1543.94 1547 1543.94 1538.25 Q1543.94 1529.47 1548.27 1524.5 Q1552.63 1519.54 1560.3 1519.54 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M1603.46 1525.87 Q1602.47 1525.3 1601.29 1525.04 Q1600.15 1524.76 1598.75 1524.76 Q1593.78 1524.76 1591.11 1528 Q1588.47 1531.22 1588.47 1537.27 L1588.47 1556.04 L1582.58 1556.04 L1582.58 1520.4 L1588.47 1520.4 L1588.47 1525.93 Q1590.31 1522.69 1593.27 1521.13 Q1596.23 1519.54 1600.47 1519.54 Q1601.07 1519.54 1601.8 1519.63 Q1602.53 1519.7 1603.43 1519.85 L1603.46 1525.87 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M42.0438 868.462 L48.314 868.462 Q46.5634 872.122 45.704 875.369 Q44.8447 878.615 44.8447 881.639 Q44.8447 886.891 46.8817 889.755 Q48.9187 892.588 52.6745 892.588 Q55.8255 892.588 57.4488 890.71 Q59.0402 888.801 60.0269 883.517 L60.8226 879.634 Q62.1912 872.441 65.6605 869.035 Q69.098 865.598 74.8908 865.598 Q81.7976 865.598 85.3624 870.245 Q88.9272 874.86 88.9272 883.803 Q88.9272 887.177 88.1633 890.997 Q87.3994 894.784 85.9035 898.858 L79.2831 898.858 Q81.4793 894.943 82.5933 891.188 Q83.7073 887.432 83.7073 883.803 Q83.7073 878.297 81.543 875.305 Q79.3786 872.313 75.3682 872.313 Q71.8671 872.313 69.8937 874.478 Q67.9204 876.61 66.9337 881.512 L66.1698 885.427 Q64.7375 892.62 61.682 895.835 Q58.6264 899.049 53.1837 899.049 Q46.8817 899.049 43.2532 894.625 Q39.6248 890.169 39.6248 882.371 Q39.6248 879.029 40.2295 875.56 Q40.8343 872.091 42.0438 868.462 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M40.4842 859.2 L40.4842 850.543 L80.238 829.472 L40.4842 829.472 L40.4842 823.234 L88.0042 823.234 L88.0042 831.891 L48.2503 852.962 L88.0042 852.962 L88.0042 859.2 L40.4842 859.2 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M45.7677 810.407 L63.6235 810.407 L63.6235 802.322 Q63.6235 797.835 61.3 795.384 Q58.9765 792.933 54.6797 792.933 Q50.4147 792.933 48.0912 795.384 Q45.7677 797.835 45.7677 802.322 L45.7677 810.407 M40.4842 816.836 L40.4842 802.322 Q40.4842 794.333 44.1126 790.259 Q47.7092 786.154 54.6797 786.154 Q61.7138 786.154 65.3104 790.259 Q68.907 794.333 68.907 802.322 L68.907 810.407 L88.0042 810.407 L88.0042 816.836 L40.4842 816.836 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M53.7248 733.636 L59.1993 733.636 Q57.8307 736.119 57.1623 738.634 Q56.4621 741.116 56.4621 743.662 Q56.4621 749.36 60.0905 752.511 Q63.6872 755.662 70.212 755.662 Q76.7369 755.662 80.3653 752.511 Q83.9619 749.36 83.9619 743.662 Q83.9619 741.116 83.2935 738.634 Q82.5933 736.119 81.2247 733.636 L86.6355 733.636 Q87.7814 736.087 88.3543 738.729 Q88.9272 741.339 88.9272 744.299 Q88.9272 752.352 83.8664 757.094 Q78.8057 761.837 70.212 761.837 Q61.491 761.837 56.4939 757.062 Q51.4968 752.256 51.4968 743.917 Q51.4968 741.212 52.0697 738.634 Q52.6108 736.055 53.7248 733.636 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M56.4621 713.68 Q56.4621 718.391 60.1542 721.128 Q63.8145 723.865 70.212 723.865 Q76.6095 723.865 80.3017 721.16 Q83.9619 718.422 83.9619 713.68 Q83.9619 709.001 80.2698 706.264 Q76.5777 703.527 70.212 703.527 Q63.8781 703.527 60.186 706.264 Q56.4621 709.001 56.4621 713.68 M51.4968 713.68 Q51.4968 706.041 56.4621 701.681 Q61.4273 697.32 70.212 697.32 Q78.9649 697.32 83.9619 701.681 Q88.9272 706.041 88.9272 713.68 Q88.9272 721.351 83.9619 725.711 Q78.9649 730.04 70.212 730.04 Q61.4273 730.04 56.4621 725.711 Q51.4968 721.351 51.4968 713.68 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M73.9359 691.782 L52.3562 691.782 L52.3562 685.925 L73.7131 685.925 Q78.7739 685.925 81.3202 683.952 Q83.8346 681.979 83.8346 678.032 Q83.8346 673.29 80.8109 670.552 Q77.7872 667.783 72.5673 667.783 L52.3562 667.783 L52.3562 661.927 L88.0042 661.927 L88.0042 667.783 L82.5296 667.783 Q85.7762 669.916 87.3676 672.748 Q88.9272 675.549 88.9272 679.273 Q88.9272 685.416 85.1078 688.599 Q81.2883 691.782 73.9359 691.782 M51.4968 677.045 L51.4968 677.045 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M66.4881 626.151 L88.0042 626.151 L88.0042 632.008 L66.679 632.008 Q61.6183 632.008 59.1038 633.981 Q56.5894 635.955 56.5894 639.901 Q56.5894 644.644 59.6131 647.381 Q62.6368 650.118 67.8567 650.118 L88.0042 650.118 L88.0042 656.007 L52.3562 656.007 L52.3562 650.118 L57.8944 650.118 Q54.6797 648.018 53.0883 645.185 Q51.4968 642.32 51.4968 638.596 Q51.4968 632.454 55.3163 629.303 Q59.1038 626.151 66.4881 626.151 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M42.2347 614.216 L52.3562 614.216 L52.3562 602.153 L56.9077 602.153 L56.9077 614.216 L76.2594 614.216 Q80.6199 614.216 81.8613 613.038 Q83.1026 611.829 83.1026 608.168 L83.1026 602.153 L88.0042 602.153 L88.0042 608.168 Q88.0042 614.948 85.4897 617.526 Q82.9434 620.104 76.2594 620.104 L56.9077 620.104 L56.9077 624.401 L52.3562 624.401 L52.3562 620.104 L42.2347 620.104 L42.2347 614.216 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip850)\" d=\"M 0 0 M53.4065 573.284 L58.9447 573.284 Q57.6716 575.767 57.035 578.441 Q56.3984 581.114 56.3984 583.979 Q56.3984 588.339 57.7352 590.535 Q59.072 592.7 61.7456 592.7 Q63.7826 592.7 64.9603 591.14 Q66.1061 589.581 67.1565 584.87 L67.6021 582.865 Q68.9389 576.626 71.3897 574.016 Q73.8086 571.375 78.1691 571.375 Q83.1344 571.375 86.0308 575.321 Q88.9272 579.236 88.9272 586.111 Q88.9272 588.976 88.3543 592.095 Q87.8132 595.182 86.6992 598.62 L80.6518 598.62 Q82.3387 595.373 83.198 592.222 Q84.0256 589.071 84.0256 585.984 Q84.0256 581.846 82.6251 579.618 Q81.1929 577.39 78.6147 577.39 Q76.2276 577.39 74.9545 579.013 Q73.6813 580.605 72.5037 586.048 L72.0262 588.085 Q70.8804 593.527 68.5251 595.946 Q66.138 598.365 62.0002 598.365 Q56.9713 598.365 54.2341 594.8 Q51.4968 591.236 51.4968 584.679 Q51.4968 581.432 51.9743 578.568 Q52.4517 575.703 53.4065 573.284 Z\" fill=\"#000000\" fill-rule=\"evenodd\" fill-opacity=\"1\" /><path clip-path=\"url(#clip852)\" d=\"\n",
       "M393.679 86.1857 L393.679 1384.24 L535.606 1384.24 L535.606 86.1857 L393.679 86.1857 L393.679 86.1857  Z\n",
       "  \" fill=\"#009af9\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip852)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  393.679,86.1857 393.679,1384.24 535.606,1384.24 535.606,86.1857 393.679,86.1857 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip852)\" d=\"\n",
       "M535.606 1155.36 L535.606 1384.24 L677.533 1384.24 L677.533 1155.36 L535.606 1155.36 L535.606 1155.36  Z\n",
       "  \" fill=\"#009af9\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip852)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  535.606,1155.36 535.606,1384.24 677.533,1384.24 677.533,1155.36 535.606,1155.36 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip852)\" d=\"\n",
       "M677.533 1282.45 L677.533 1384.24 L819.46 1384.24 L819.46 1282.45 L677.533 1282.45 L677.533 1282.45  Z\n",
       "  \" fill=\"#009af9\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip852)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  677.533,1282.45 677.533,1384.24 819.46,1384.24 819.46,1282.45 677.533,1282.45 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip852)\" d=\"\n",
       "M819.46 1369.51 L819.46 1384.24 L961.387 1384.24 L961.387 1369.51 L819.46 1369.51 L819.46 1369.51  Z\n",
       "  \" fill=\"#009af9\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip852)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  819.46,1369.51 819.46,1384.24 961.387,1384.24 961.387,1369.51 819.46,1369.51 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip852)\" d=\"\n",
       "M961.387 1374.76 L961.387 1384.24 L1103.31 1384.24 L1103.31 1374.76 L961.387 1374.76 L961.387 1374.76  Z\n",
       "  \" fill=\"#009af9\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip852)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  961.387,1374.76 961.387,1384.24 1103.31,1384.24 1103.31,1374.76 961.387,1374.76 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip852)\" d=\"\n",
       "M1103.31 1383.27 L1103.31 1384.24 L1245.24 1384.24 L1245.24 1383.27 L1103.31 1383.27 L1103.31 1383.27  Z\n",
       "  \" fill=\"#009af9\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip852)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1103.31,1383.27 1103.31,1384.24 1245.24,1384.24 1245.24,1383.27 1103.31,1383.27 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip852)\" d=\"\n",
       "M1245.24 1383.19 L1245.24 1384.24 L1387.17 1384.24 L1387.17 1383.19 L1245.24 1383.19 L1245.24 1383.19  Z\n",
       "  \" fill=\"#009af9\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip852)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1245.24,1383.19 1245.24,1384.24 1387.17,1384.24 1387.17,1383.19 1245.24,1383.19 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip852)\" d=\"\n",
       "M1387.17 1384.22 L1387.17 1384.24 L1529.1 1384.24 L1529.1 1384.22 L1387.17 1384.22 L1387.17 1384.22  Z\n",
       "  \" fill=\"#009af9\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip852)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1387.17,1384.22 1387.17,1384.24 1529.1,1384.24 1529.1,1384.22 1387.17,1384.22 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip852)\" d=\"\n",
       "M1529.1 1384.07 L1529.1 1384.24 L1671.02 1384.24 L1671.02 1384.07 L1529.1 1384.07 L1529.1 1384.07  Z\n",
       "  \" fill=\"#009af9\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip852)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1529.1,1384.07 1529.1,1384.24 1671.02,1384.24 1671.02,1384.07 1529.1,1384.07 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip852)\" d=\"\n",
       "M1671.02 1384.24 L1671.02 1384.24 L1812.95 1384.24 L1812.95 1384.24 L1671.02 1384.24 L1671.02 1384.24  Z\n",
       "  \" fill=\"#009af9\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip852)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1671.02,1384.24 1671.02,1384.24 1812.95,1384.24 1671.02,1384.24 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip852)\" d=\"\n",
       "M1812.95 1384.21 L1812.95 1384.24 L1954.88 1384.24 L1954.88 1384.21 L1812.95 1384.21 L1812.95 1384.21  Z\n",
       "  \" fill=\"#009af9\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip852)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1812.95,1384.21 1812.95,1384.24 1954.88,1384.24 1954.88,1384.21 1812.95,1384.21 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip852)\" d=\"\n",
       "M1954.88 1384.24 L1954.88 1384.24 L2096.8 1384.24 L2096.8 1384.24 L1954.88 1384.24 L1954.88 1384.24  Z\n",
       "  \" fill=\"#009af9\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip852)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1954.88,1384.24 1954.88,1384.24 2096.8,1384.24 1954.88,1384.24 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip852)\" d=\"\n",
       "M2096.8 1384.23 L2096.8 1384.24 L2238.73 1384.24 L2238.73 1384.23 L2096.8 1384.23 L2096.8 1384.23  Z\n",
       "  \" fill=\"#009af9\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip852)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  2096.8,1384.23 2096.8,1384.24 2238.73,1384.24 2238.73,1384.23 2096.8,1384.23 \n",
       "  \"/>\n",
       "</svg>\n"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# plot histogram of SNP scores\n",
    "using Plots\n",
    "histogram(snpscores, label=:none, xlabel=\"Least squares error\", ylabel=\"SNP counts\", bins=20)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Generate subset of markers for prephasing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "752.076342 seconds (6.51 G allocations: 615.353 GiB, 7.01% gc time)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32mProgress: 100%|█████████████████████████████████████████| Time: 0:08:17\u001b[39m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "549.100648 seconds (6.86 G allocations: 650.723 GiB, 10.18% gc time)\n",
      "1139.017251 seconds (16.68 G allocations: 1.249 TiB, 10.94% gc time)\n",
      "599.893049 seconds (6.86 G allocations: 650.911 GiB, 10.61% gc time)\n",
      " 36.158916 seconds (267.39 M allocations: 27.452 GiB, 6.96% gc time)\n",
      "608.676819 seconds (6.83 G allocations: 648.780 GiB, 10.30% gc time)\n",
      " 35.610763 seconds (249.89 M allocations: 25.655 GiB, 6.86% gc time)\n",
      "560.687780 seconds (6.61 G allocations: 626.620 GiB, 10.27% gc time)\n",
      " 16.352906 seconds (119.38 M allocations: 12.252 GiB, 6.40% gc time)\n",
      "520.047088 seconds (6.51 G allocations: 615.447 GiB, 10.21% gc time)\n",
      "  8.014085 seconds (53.29 M allocations: 5.482 GiB, 5.36% gc time)\n",
      "516.890704 seconds (6.43 G allocations: 607.361 GiB, 10.60% gc time)\n",
      "  0.831218 seconds (5.35 M allocations: 563.018 MiB, 5.66% gc time)\n",
      "6410.931663 seconds (76.92 G allocations: 6.823 TiB, 10.06% gc time)\n"
     ]
    }
   ],
   "source": [
    "cd(\"/Users/biona001/.julia/dev/MendelImpute/data/1000_genome_phase3_v5/filtered\")\n",
    "function filter_and_mask()\n",
    "    for chr in [18]\n",
    "        # filter chromosome data for unique snps\n",
    "        data = \"../beagle_raw/chr$chr.1kg.phase3.v5a.vcf.gz\"\n",
    "        full_record_index = .!find_duplicate_marker(data)\n",
    "        @time VCFTools.filter(data, full_record_index, 1:nsamples(data), \n",
    "            des = \"chr$chr.uniqueSNPs.vcf.gz\")\n",
    "\n",
    "        # summarize data\n",
    "        total_snps, samples, _, _, _, maf_by_record, _ = gtstats(\"chr$chr.uniqueSNPs.vcf.gz\")\n",
    "\n",
    "        # generate target panel with all snps\n",
    "        n = 100\n",
    "        sample_idx = falses(samples)\n",
    "        sample_idx[1:n] .= true\n",
    "        shuffle!(sample_idx)\n",
    "        @time VCFTools.filter(\"chr$chr.uniqueSNPs.vcf.gz\", 1:total_snps, \n",
    "            sample_idx, des = \"target.chr$chr.full.vcf.gz\", allow_multiallelic=false)\n",
    "\n",
    "        # also generate reference panel without target samples\n",
    "        @time VCFTools.filter(\"chr$chr.uniqueSNPs.vcf.gz\", 1:total_snps, \n",
    "            .!sample_idx, des = \"ref.chr$chr.excludeTarget.vcf.gz\", allow_multiallelic=false)\n",
    "        \n",
    "        for maf in [0.0005, 0.001, 0.01, 0.1, 0.45]\n",
    "            \n",
    "            # generate target file with 100 samples and typed snps with certain maf\n",
    "            my_maf = findall(x -> x > maf, maf_by_record)  \n",
    "            p = length(my_maf)\n",
    "            record_idx = falses(total_snps)\n",
    "            record_idx[my_maf] .= true\n",
    "            @time VCFTools.filter(\"chr$chr.uniqueSNPs.vcf.gz\", record_idx, sample_idx, \n",
    "                des = \"target.chr$chr.typedOnly.maf$maf.vcf.gz\", allow_multiallelic=false)\n",
    "\n",
    "            # unphase and mask 1% entries in target file\n",
    "            masks = falses(p, n)\n",
    "            missingprop = 0.001\n",
    "            for j in 1:n, i in 1:p\n",
    "                rand() < missingprop && (masks[i, j] = true)\n",
    "            end\n",
    "            @time mask_gt(\"target.chr$chr.typedOnly.maf$maf.vcf.gz\", masks, \n",
    "                des=\"target.chr$chr.typedOnly.maf$maf.masked.vcf.gz\", unphase=true)\n",
    "        end\n",
    "    end \n",
    "end\n",
    "Random.seed!(2020)\n",
    "@time filter_and_mask()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "nrecords(\"ref.chr18.excludeTarget.vcf.gz\") = 852602\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "852602"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "@show nrecords(\"ref.chr18.excludeTarget.vcf.gz\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "nrecords(\"target.chr18.typedOnly.maf0.1.masked.vcf.gz\") = 169914\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "169914"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "@show nrecords(\"target.chr18.typedOnly.maf0.1.masked.vcf.gz\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Try jlso/jld2 compression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "295.751250 seconds (50.45 M allocations: 72.668 GiB, 3.59% gc time)\n",
      "236.530047 seconds (21.33 M allocations: 38.930 GiB, 2.25% gc time)\n",
      "202.031370 seconds (14.28 M allocations: 21.577 GiB, 1.75% gc time)\n",
      "188.489177 seconds (14.28 M allocations: 21.577 GiB, 2.12% gc time)\n"
     ]
    }
   ],
   "source": [
    "# jlso format\n",
    "maf = 0.1\n",
    "max_d = [1000]\n",
    "minwidths = [100, 250, 400, 600]\n",
    "# overlaps = [0.25, 0.5, 0.75, 1.0]\n",
    "overlap = 0.0\n",
    "\n",
    "# reffile = \"ref.chr18.excludeTarget.vcf.gz\"\n",
    "# tgtfile = \"target.chr18.typedOnly.maf$maf.masked.vcf.gz\"\n",
    "# H, H_sampleID, H_chr, H_pos, H_ids, H_ref, H_alt = convert_ht(Bool, reffile, trans=true, save_snp_info=true, msg=\"importing reference data...\")\n",
    "# X, X_sampleID, X_chr, X_pos, X_ids, X_ref, X_alt = VCFTools.convert_gt(UInt8, tgtfile, trans=true, save_snp_info=true, msg = \"Importing genotype file...\")\n",
    "for d in max_d, minwidth in minwidths\n",
    "    outfile = \"ref.chr18.maxd$d.minwidth$minwidth.maf$maf.excludeTarget.jlso\"\n",
    "    @time compress_haplotypes(H, X, outfile, X_pos, H_sampleID, H_chr, H_pos, H_ids, \n",
    "        H_ref, H_alt, d, minwidth, overlap)\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-rw-r--r--@ 1 biona001  staff  247072174 May 19 23:43 ref.chr18.excludeTarget.vcf.gz\n"
     ]
    }
   ],
   "source": [
    ";ls -al ref.chr18.excludeTarget.vcf.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-rw-r--r--  1 biona001  staff  130476381 Aug  4 23:43 ref.chr18.maxd1000.maf0.1.excludeTarget.jlso\n"
     ]
    }
   ],
   "source": [
    ";ls -al ref.chr18.maxd1000.maf0.1.excludeTarget.jlso"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-rw-r--r--  1 biona001  staff  130259083 Sep  6 16:53 ref.chr18.maxd1000.minwidth20.maf0.1.excludeTarget.jlso\n"
     ]
    }
   ],
   "source": [
    ";ls -al ref.chr18.maxd1000.minwidth20.maf0.1.excludeTarget.jlso"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# MendelImpute error rate and timing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "8"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Threads.nthreads()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32mComputing optimal haplotype pairs...100%|███████████████| Time: 0:01:06\u001b[39m\n",
      "\u001b[32mWriting to file...100%|█████████████████████████████████| Time: 0:00:06\u001b[39m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total windows = 331, averaging ~ 4096 unique haplotypes per window.\n",
      "\n",
      "Timings: \n",
      "    Data import                     = 10.6795 seconds\n",
      "    Computing haplotype pair        = 66.8235 seconds\n",
      "        BLAS3 mul! to get M and N      = 8.81538 seconds per thread\n",
      "        haplopair search               = 43.7844 seconds per thread\n",
      "        supplying constant terms       = 0.00405961 seconds per thread\n",
      "        finding redundant happairs     = 0.0471254 seconds per thread\n",
      "    Phasing by dynamic programming  = 0.325226 seconds\n",
      "    Imputation                      = 9.08613 seconds\n",
      "\n",
      " 87.094466 seconds (53.87 M allocations: 29.474 GiB, 6.25% gc time)\n",
      "error overall = 0.006451474427693109 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "# 8 threads, keep best happair\n",
    "Random.seed!(2020)\n",
    "chr = 18\n",
    "maf = 0.1\n",
    "width = 512\n",
    "tgtfile = \"target.chr$chr.typedOnly.maf$maf.masked.vcf.gz\"\n",
    "reffile = \"ref.chr18.$width.maf$maf.excludeTarget.jlso\"\n",
    "outfile = \"mendel.imputed.dp$width.maf$maf.vcf.gz\"\n",
    "@time ph, hs = phase(tgtfile, reffile, outfile=outfile, impute=true, width=width)\n",
    "\n",
    "X_complete = convert_gt(UInt8, \"target.chr18.full.vcf.gz\")\n",
    "n, p = size(X_complete)\n",
    "X_mendel = convert_gt(UInt8, outfile)\n",
    "println(\"error overall = $(sum(X_mendel .!= X_complete) / n / p) \\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Imputation of typed SNPs only"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of threads = 8\n",
      "Importing reference haplotype data...\n",
      "Total windows = 2367, averaging ~ 527 unique haplotypes per window.\n",
      "\n",
      "Timings: \n",
      "    Data import                     = 22.0836 seconds\n",
      "        import target data             = 3.89646 seconds\n",
      "        import compressed haplotypes   = 18.1871 seconds\n",
      "    Computing haplotype pair        = 4.13377 seconds\n",
      "        BLAS3 mul! to get M and N      = 0.309622 seconds per thread\n",
      "        haplopair search               = 3.42762 seconds per thread\n",
      "        initializing missing           = 0.021281 seconds per thread\n",
      "        allocating and viewing         = 0.0506452 seconds per thread\n",
      "        index conversion               = 0.00123344 seconds per thread\n",
      "    Phasing by win-win intersection = 1.28608 seconds\n",
      "        Window-by-window intersection  = 0.0921849 seconds per thread\n",
      "        Breakpoint search              = 0.952759 seconds per thread\n",
      "        Recording result               = 0.0200383 seconds per thread\n",
      "    Imputation                     = 0.597237 seconds\n",
      "        Imputing missing               = 0.186438 seconds\n",
      "        Writing to file                = 0.410799 seconds\n",
      "\n",
      "    Total time                      = 28.1017 seconds\n",
      "\n",
      " 28.101751 seconds (62.61 M allocations: 4.682 GiB, 9.90% gc time)\n",
      "error overall = 0.0014766293536730346 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "# 8 threads, d = 1000\n",
    "Random.seed!(2020)\n",
    "tgtfile = \"target.chr18.typedOnly.maf0.1.masked.vcf.gz\"\n",
    "reffile = \"ref.chr18.maxd1000.overlap$overlap.maf0.1.excludeTarget.jlso\"\n",
    "outfile = \"mendel.imputed.vcf.gz\"\n",
    "@time ph = phase(tgtfile, reffile, outfile, impute=false);\n",
    "\n",
    "X_complete = convert_gt(UInt8, \"target.chr18.typedOnly.maf0.1.vcf.gz\")\n",
    "n, p = size(X_complete)\n",
    "X_mendel = convert_gt(UInt8, outfile)\n",
    "println(\"error overall = $(sum(X_mendel .!= X_complete) / n / p) \\n\")\n",
    "rm(outfile, force=true)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# MendelImpute with intersecting win-by-win"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Importing reference haplotype data...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32mWriting to file...100%|█████████████████████████████████| Time: 0:00:05\u001b[39m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total windows = 2367, averaging ~ 527 unique haplotypes per window.\n",
      "\n",
      "Timings: \n",
      "    Data import                     = 13.8701 seconds\n",
      "    Computing haplotype pair        = 4.40941 seconds\n",
      "        BLAS3 mul! to get M and N      = 0.3572 seconds per thread\n",
      "        haplopair search               = 3.40097 seconds per thread\n",
      "        initializing missing           = 0.0227854 seconds per thread\n",
      "        allocating and viewing         = 0.0552744 seconds per thread\n",
      "        index conversion               = 0.00175442 seconds per thread\n",
      "    Phasing by win-win intersection = 0.150916 seconds\n",
      "        Window-by-window intersection  = 0.0993135 seconds per thread\n",
      "        Breakpoint search              = 0.0233533 seconds per thread\n",
      "        Recording result               = 0.0133309 seconds per thread\n",
      "    Imputation                      = 6.13664 seconds\n",
      "\n",
      " 24.567881 seconds (62.44 M allocations: 4.900 GiB, 6.45% gc time)\n",
      "error overall = 0.005046844834987485 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "# 8 threads, d = 1000, keep best happair\n",
    "Random.seed!(2020)\n",
    "chr = 18\n",
    "maf = 0.1\n",
    "d = 1000\n",
    "tgtfile = \"target.chr$chr.typedOnly.maf$maf.masked.vcf.gz\"\n",
    "reffile = \"ref.chr18.maxd$d.maf$maf.excludeTarget.jlso\"\n",
    "outfile = \"mendel.imputed.vcf.gz\"\n",
    "@time ph = phase(tgtfile, reffile, outfile=outfile, impute=true, max_d = d,\n",
    "    dynamic_programming = false)\n",
    "\n",
    "X_complete = convert_gt(UInt8, \"target.chr18.full.vcf.gz\")\n",
    "n, p = size(X_complete)\n",
    "X_mendel = convert_gt(UInt8, outfile)\n",
    "println(\"error overall = $(sum(X_mendel .!= X_complete) / n / p) \\n\")\n",
    "rm(outfile, force=true)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Importing reference haplotype data...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32mPhasing chunk 1/1...100%|███████████████████████████████| Time: 0:01:25\u001b[39m\n",
      "\u001b[32mWriting to file...100%|█████████████████████████████████| Time: 0:00:06\u001b[39m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total windows = 331, averaging ~ 4096 unique haplotypes per window.\n",
      "\n",
      "Timings: \n",
      "    Data import                     = 9.22319 seconds\n",
      "    Computing haplotype pair        = 85.1134 seconds\n",
      "        BLAS3 mul! to get M and N      = 40.7029 seconds per thread\n",
      "        haplopair search               = 38.7822 seconds per thread\n",
      "        finding redundant happairs     = 0.0075832 seconds per thread\n",
      "    Phasing by win-win intersection = 0.037751 seconds\n",
      "    Imputation                      = 8.35549 seconds\n",
      "\n",
      "102.829221 seconds (57.19 M allocations: 29.451 GiB, 4.87% gc time)\n",
      "error overall = 0.006448636057621258 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "# 8 threads with Gaius.jl\n",
    "Random.seed!(2020)\n",
    "chr = 18\n",
    "maf = 0.1\n",
    "width = 512\n",
    "tgtfile = \"target.chr$chr.typedOnly.maf$maf.masked.vcf.gz\"\n",
    "reffile = \"ref.chr18.$width.maf$maf.excludeTarget.jlso\"\n",
    "outfile = \"mendel.imputed.dp$width.maf$maf.vcf.gz\"\n",
    "@time ph = phase(tgtfile, reffile, outfile=outfile, impute=true, width=width,\n",
    "    dynamic_programming = false)\n",
    "\n",
    "X_complete = convert_gt(UInt8, \"target.chr18.full.vcf.gz\")\n",
    "n, p = size(X_complete)\n",
    "X_mendel = convert_gt(UInt8, outfile)\n",
    "println(\"error overall = $(sum(X_mendel .!= X_complete) / n / p) \\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Importing reference haplotype data...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32mPhasing chunk 1/1...100%|███████████████████████████████| Time: 0:00:57\u001b[39m\n",
      "\u001b[32mWriting to file...100%|█████████████████████████████████| Time: 0:00:07\u001b[39m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total windows = 331, averaging ~ 4096 unique haplotypes per window.\n",
      "\n",
      "Timings: \n",
      "    Data import                     = 9.23164 seconds\n",
      "    Computing haplotype pair        = 57.9442 seconds\n",
      "        BLAS3 mul! to get M and N      = 14.6295 seconds per thread\n",
      "        haplopair search               = 34.661 seconds per thread\n",
      "        finding redundant happairs     = 0.00644955 seconds per thread\n",
      "    Phasing by win-win intersection = 0.0497861 seconds\n",
      "    Imputation                      = 7.60605 seconds\n",
      "\n",
      " 75.064026 seconds (52.82 M allocations: 29.050 GiB, 6.18% gc time)\n",
      "error overall = 0.0064488354472544045 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "# 8 threads with PaddedMatrices.jl\n",
    "Random.seed!(2020)\n",
    "chr = 18\n",
    "maf = 0.1\n",
    "width = 512\n",
    "tgtfile = \"target.chr$chr.typedOnly.maf$maf.masked.vcf.gz\"\n",
    "reffile = \"ref.chr18.$width.maf$maf.excludeTarget.jlso\"\n",
    "outfile = \"mendel.imputed.dp$width.maf$maf.vcf.gz\"\n",
    "@time ph = phase(tgtfile, reffile, outfile=outfile, impute=true, width=width,\n",
    "    dynamic_programming = false)\n",
    "\n",
    "X_complete = convert_gt(UInt8, \"target.chr18.full.vcf.gz\")\n",
    "n, p = size(X_complete)\n",
    "X_mendel = convert_gt(UInt8, outfile)\n",
    "println(\"error overall = $(sum(X_mendel .!= X_complete) / n / p) \\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# MendelPhase"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Importing reference haplotype data...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32mPhasing chunk 1/1...100%|███████████████████████████████| Time: 0:00:50\u001b[39m\n",
      "\u001b[32mWriting to file...100%|█████████████████████████████████| Time: 0:00:06\u001b[39m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total windows = 331, averaging ~ 4096 unique haplotypes per window.\n",
      "\n",
      "Timings: \n",
      "    Data import                     = 9.3601 seconds\n",
      "    Computing haplotype pair        = 50.7812 seconds\n",
      "        BLAS3 mul! to get M and N      = 11.7251 seconds per thread\n",
      "        haplopair search               = 33.0267 seconds per thread\n",
      "        finding redundant happairs     = 0.0309747 seconds per thread\n",
      "    Phasing by win-win intersection = 0.0394528 seconds\n",
      "    Imputation                      = 7.16706 seconds\n",
      "\n",
      " 67.442743 seconds (54.67 M allocations: 29.016 GiB, 6.78% gc time)\n",
      "error overall = 0.012744234707401578 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "# 8 threads, keep best happair\n",
    "Random.seed!(2020)\n",
    "chr = 18\n",
    "maf = 0.1\n",
    "width = 512\n",
    "tgtfile = \"target.chr$chr.typedOnly.maf$maf.masked.vcf.gz\"\n",
    "reffile = \"ref.chr18.$width.maf$maf.excludeTarget.jlso\"\n",
    "outfile = \"mendel.imputed.dp$width.maf$maf.vcf.gz\"\n",
    "@time ph = phase(tgtfile, reffile, outfile=outfile, impute=true, width=width,\n",
    "    dynamic_programming = false, phase=true)\n",
    "\n",
    "X_complete = convert_gt(UInt8, \"target.chr18.full.vcf.gz\")\n",
    "n, p = size(X_complete)\n",
    "X_mendel = convert_gt(UInt8, outfile)\n",
    "println(\"error overall = $(sum(X_mendel .!= X_complete) / n / p) \\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Ultra compressed output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of threads = 8\n",
      "Importing reference haplotype data...\n",
      "Total windows = 2367, averaging ~ 527 unique haplotypes per window.\n",
      "\n",
      "Timings: \n",
      "    Data import                     = 15.7743 seconds\n",
      "        import target data             = 3.88648 seconds\n",
      "        import compressed haplotypes   = 11.8878 seconds\n",
      "    Computing haplotype pair        = 4.2701 seconds\n",
      "        BLAS3 mul! to get M and N      = 0.331793 seconds per thread\n",
      "        haplopair search               = 3.49273 seconds per thread\n",
      "        initializing missing           = 0.0225607 seconds per thread\n",
      "        allocating and viewing         = 0.0592994 seconds per thread\n",
      "        index conversion               = 0.00190718 seconds per thread\n",
      "    Phasing by win-win intersection = 1.60634 seconds\n",
      "        Window-by-window intersection  = 0.0996143 seconds per thread\n",
      "        Breakpoint search              = 1.05511 seconds per thread\n",
      "        Recording result               = 0.0110744 seconds per thread\n",
      "    Imputation                     = 2.9571 seconds\n",
      "        Imputing missing               = 0.0794888 seconds\n",
      "        Writing to file                = 2.87761 seconds\n",
      "\n",
      "    Total time                      = 24.6088 seconds\n",
      "\n",
      " 24.608961 seconds (66.66 M allocations: 4.629 GiB, 10.72% gc time)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32mimporting H100%|████████████████████████████████████████| Time: 0:04:07\u001b[39m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "error overall = 0.005259159607882693 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "# 8 threads, d = 1000, keep best happair\n",
    "Random.seed!(2020)\n",
    "chr = 18\n",
    "maf = 0.1\n",
    "d = 1000\n",
    "tgtfile = \"target.chr$chr.typedOnly.maf$maf.masked.vcf.gz\"\n",
    "reffile = \"ref.chr18.maxd$d.maf$maf.excludeTarget.jlso\"\n",
    "outfile = \"mendel.imputed.jlso\"\n",
    "@time ph = phase(tgtfile, reffile, outfile=outfile, impute=true, max_d = d,\n",
    "    phase = true);\n",
    "\n",
    "X_complete = convert_gt(UInt8, \"target.chr18.full.vcf.gz\", trans=true)\n",
    "H = convert_ht(Bool, \"ref.chr18.excludeTarget.vcf.gz\", trans=true, msg=\"importing H\")\n",
    "n, p = size(X_complete)\n",
    "X1, X2 = convert_compressed(UInt8, ph, H)\n",
    "X_mendel = X1 + X2\n",
    "println(\"error overall = $(sum(X_mendel .!= X_complete) / n / p) \\n\")\n",
    "# rm(outfile, force=true)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-rw-r--r--  1 biona001  staff  983248 Sep  7 14:40 mendel.imputed.jlso\n"
     ]
    }
   ],
   "source": [
    ";ls -al mendel.imputed.jlso"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-rw-r--r--  1 biona001  staff  17876617 Sep  7 14:36 mendel.imputed.vcf.gz\n"
     ]
    }
   ],
   "source": [
    ";ls -al mendel.imputed.vcf.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "18.18118826582917"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "17876617 / 983248"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Minimum window widths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of threads = 8\n",
      "Importing reference haplotype data...\n",
      "Total windows = 2367, averaging ~ 527 unique haplotypes per window.\n",
      "\n",
      "Timings: \n",
      "    Data import                     = 16.692 seconds\n",
      "        import target data             = 4.87187 seconds\n",
      "        import compressed haplotypes   = 11.8201 seconds\n",
      "    Computing haplotype pair        = 4.26173 seconds\n",
      "        BLAS3 mul! to get M and N      = 0.321671 seconds per thread\n",
      "        haplopair search               = 3.45778 seconds per thread\n",
      "        initializing missing           = 0.0235874 seconds per thread\n",
      "        allocating and viewing         = 0.0624623 seconds per thread\n",
      "        index conversion               = 0.00164948 seconds per thread\n",
      "    Phasing by win-win intersection = 1.50478 seconds\n",
      "        Window-by-window intersection  = 0.100988 seconds per thread\n",
      "        Breakpoint search              = 1.11491 seconds per thread\n",
      "        Recording result               = 0.0143302 seconds per thread\n",
      "    Imputation                     = 2.47542 seconds\n",
      "        Imputing missing               = 0.999966 seconds\n",
      "        Writing to file                = 1.47545 seconds\n",
      "\n",
      "    Total time                      = 24.9351 seconds\n",
      "\n",
      " 24.931055 seconds (62.19 M allocations: 4.806 GiB, 15.69% gc time)\n",
      "error overall = 0.005259159607882693 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "# 8 threads, global search, d = 1000, minwidth=0\n",
    "Random.seed!(2020)\n",
    "chr = 18\n",
    "maf = 0.1\n",
    "d = 1000\n",
    "tgtfile = \"target.chr$chr.typedOnly.maf$maf.masked.vcf.gz\"\n",
    "reffile = \"ref.chr18.maxd$d.maf$maf.excludeTarget.jlso\"\n",
    "outfile = \"mendel.imputed.vcf.gz\"\n",
    "@time ph = phase(tgtfile, reffile, outfile=outfile, impute=true, max_d=d,\n",
    "    phase = true);\n",
    "\n",
    "# X_complete = convert_gt(UInt8, \"target.chr18.full.vcf.gz\")\n",
    "n, p = size(X_complete)\n",
    "X_mendel = convert_gt(UInt8, outfile)\n",
    "println(\"error overall = $(sum(X_mendel .!= X_complete) / n / p) \\n\")\n",
    "# rm(outfile, force=true)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of threads = 8\n",
      "Importing reference haplotype data...\n",
      "Total windows = 2351, averaging ~ 538 unique haplotypes per window.\n",
      "\n",
      "Timings: \n",
      "    Data import                     = 23.0568 seconds\n",
      "        import target data             = 5.07468 seconds\n",
      "        import compressed haplotypes   = 17.9821 seconds\n",
      "    Computing haplotype pair        = 5.33693 seconds\n",
      "        BLAS3 mul! to get M and N      = 0.333737 seconds per thread\n",
      "        haplopair search               = 3.69525 seconds per thread\n",
      "        initializing missing           = 0.0251059 seconds per thread\n",
      "        allocating and viewing         = 0.184637 seconds per thread\n",
      "        index conversion               = 0.00187568 seconds per thread\n",
      "    Phasing by win-win intersection = 1.53646 seconds\n",
      "        Window-by-window intersection  = 0.1011 seconds per thread\n",
      "        Breakpoint search              = 1.14021 seconds per thread\n",
      "        Recording result               = 0.0135461 seconds per thread\n",
      "    Imputation                     = 1.74412 seconds\n",
      "        Imputing missing               = 0.22515 seconds\n",
      "        Writing to file                = 1.51897 seconds\n",
      "\n",
      "    Total time                      = 31.6756 seconds\n",
      "\n",
      " 31.675214 seconds (62.95 M allocations: 5.066 GiB, 13.52% gc time)\n",
      "error overall = 0.005256720017077136 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "# 8 threads, global search, d = 1000, minwidth=20\n",
    "Random.seed!(2020)\n",
    "chr = 18\n",
    "maf = 0.1\n",
    "d = 1000\n",
    "minwidth = 20\n",
    "tgtfile = \"target.chr$chr.typedOnly.maf$maf.masked.vcf.gz\"\n",
    "reffile = \"ref.chr18.maxd$d.minwidth$minwidth.maf$maf.excludeTarget.jlso\"\n",
    "outfile = \"mendel.imputed.vcf.gz\"\n",
    "@time ph = phase(tgtfile, reffile, outfile=outfile, impute=true, max_d=d,\n",
    "    phase = true);\n",
    "\n",
    "# X_complete = convert_gt(UInt8, \"target.chr18.full.vcf.gz\")\n",
    "n, p = size(X_complete)\n",
    "X_mendel = convert_gt(UInt8, outfile)\n",
    "println(\"error overall = $(sum(X_mendel .!= X_complete) / n / p) \\n\")\n",
    "# rm(outfile, force=true)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of threads = 8\n",
      "Importing reference haplotype data...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32mComputing optimal haplotypes...100%|████████████████████| Time: 0:00:17\u001b[39m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total windows = 1715, averaging ~ 902 unique haplotypes per window.\n",
      "\n",
      "Timings: \n",
      "    Data import                     = 17.6011 seconds\n",
      "        import target data             = 4.61148 seconds\n",
      "        import compressed haplotypes   = 12.9896 seconds\n",
      "    Computing haplotype pair        = 20.3727 seconds\n",
      "        BLAS3 mul! to get M and N      = 1.18181 seconds per thread\n",
      "        haplopair search               = 9.50708 seconds per thread\n",
      "        initializing missing           = 0.025314 seconds per thread\n",
      "        allocating and viewing         = 0.426595 seconds per thread\n",
      "        index conversion               = 0.00288651 seconds per thread\n",
      "    Phasing by win-win intersection = 2.24203 seconds\n",
      "        Window-by-window intersection  = 0.0527586 seconds per thread\n",
      "        Breakpoint search              = 1.88314 seconds per thread\n",
      "        Recording result               = 0.0110326 seconds per thread\n",
      "    Imputation                     = 2.2735 seconds\n",
      "        Imputing missing               = 0.860964 seconds\n",
      "        Writing to file                = 1.41254 seconds\n",
      "\n",
      "    Total time                      = 42.4905 seconds\n",
      "\n",
      " 42.490518 seconds (61.21 M allocations: 5.328 GiB, 14.22% gc time)\n",
      "error overall = 0.005169375628957005 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "# 8 threads, global search, d = 1000, minwidth=50\n",
    "Random.seed!(2020)\n",
    "chr = 18\n",
    "maf = 0.1\n",
    "d = 1000\n",
    "minwidth = 50\n",
    "tgtfile = \"target.chr$chr.typedOnly.maf$maf.masked.vcf.gz\"\n",
    "reffile = \"ref.chr18.maxd$d.minwidth$minwidth.maf$maf.excludeTarget.jlso\"\n",
    "outfile = \"mendel.imputed.vcf.gz\"\n",
    "@time ph = phase(tgtfile, reffile, outfile=outfile, impute=true, max_d=d,\n",
    "    phase = true);\n",
    "\n",
    "# X_complete = convert_gt(UInt8, \"target.chr18.full.vcf.gz\")\n",
    "n, p = size(X_complete)\n",
    "X_mendel = convert_gt(UInt8, outfile)\n",
    "println(\"error overall = $(sum(X_mendel .!= X_complete) / n / p) \\n\")\n",
    "# rm(outfile, force=true)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of threads = 8\n",
      "Importing reference haplotype data...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32mImporting genotype file...100%|█████████████████████████| Time: 0:00:05\u001b[39m\n",
      "\u001b[32mComputing optimal haplotypes...100%|████████████████████| Time: 0:00:37\u001b[39m\n",
      "\u001b[32mPhasing...100%|█████████████████████████████████████████| Time: 0:00:06\u001b[39m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total windows = 1010, averaging ~ 1725 unique haplotypes per window.\n",
      "\n",
      "Timings: \n",
      "    Data import                     = 16.2461 seconds\n",
      "        import target data             = 5.2943 seconds\n",
      "        import compressed haplotypes   = 10.9518 seconds\n",
      "    Computing haplotype pair        = 39.9748 seconds\n",
      "        BLAS3 mul! to get M and N      = 3.35306 seconds per thread\n",
      "        haplopair search               = 19.4401 seconds per thread\n",
      "        initializing missing           = 0.0242503 seconds per thread\n",
      "        allocating and viewing         = 0.517274 seconds per thread\n",
      "        index conversion               = 0.00267529 seconds per thread\n",
      "    Phasing by win-win intersection = 6.63678 seconds\n",
      "        Window-by-window intersection  = 0.0201415 seconds per thread\n",
      "        Breakpoint search              = 6.2152 seconds per thread\n",
      "        Recording result               = 0.0182764 seconds per thread\n",
      "    Imputation                     = 2.45815 seconds\n",
      "        Imputing missing               = 0.854102 seconds\n",
      "        Writing to file                = 1.60404 seconds\n",
      "\n",
      "    Total time                      = 65.3168 seconds\n",
      "\n",
      " 65.316848 seconds (60.12 M allocations: 5.428 GiB, 9.39% gc time)\n",
      "error overall = 0.00559921276281313 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "# 8 threads, d = 1000, minwidth = 100\n",
    "Random.seed!(2020)\n",
    "chr = 18\n",
    "maf = 0.1\n",
    "d = 1000\n",
    "overlap = 0.0\n",
    "minwidth = 100\n",
    "tgtfile = \"target.chr$chr.typedOnly.maf$maf.masked.vcf.gz\"\n",
    "reffile = \"ref.chr18.maxd$d.minwidth$minwidth.maf$maf.excludeTarget.jlso\"\n",
    "outfile = \"mendel.imputed.vcf.gz\"\n",
    "@time ph = phase(tgtfile, reffile, outfile=outfile, phase=true);\n",
    "\n",
    "X_complete = convert_gt(UInt8, \"target.chr18.full.vcf.gz\")\n",
    "n, p = size(X_complete)\n",
    "X_mendel = convert_gt(UInt8, outfile)\n",
    "println(\"error overall = $(sum(X_mendel .!= X_complete) / n / p) \\n\")\n",
    "# rm(outfile, force=true)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of threads = 8\n",
      "Importing reference haplotype data...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32mComputing optimal haplotypes...100%|████████████████████| Time: 0:00:50\u001b[39m\n",
      "\u001b[32mPhasing...100%|█████████████████████████████████████████| Time: 0:00:24\u001b[39m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total windows = 512, averaging ~ 3210 unique haplotypes per window.\n",
      "\n",
      "Timings: \n",
      "    Data import                     = 13.9187 seconds\n",
      "        import target data             = 3.88211 seconds\n",
      "        import compressed haplotypes   = 10.0365 seconds\n",
      "    Computing haplotype pair        = 53.2058 seconds\n",
      "        BLAS3 mul! to get M and N      = 7.18431 seconds per thread\n",
      "        haplopair search               = 30.2405 seconds per thread\n",
      "        initializing missing           = 0.0250812 seconds per thread\n",
      "        allocating and viewing         = 0.682338 seconds per thread\n",
      "        index conversion               = 0.00131225 seconds per thread\n",
      "    Phasing by win-win intersection = 24.3224 seconds\n",
      "        Window-by-window intersection  = 0.00482951 seconds per thread\n",
      "        Breakpoint search              = 23.0015 seconds per thread\n",
      "        Recording result               = 0.0138126 seconds per thread\n",
      "    Imputation                     = 2.11399 seconds\n",
      "        Imputing missing               = 0.778374 seconds\n",
      "        Writing to file                = 1.33561 seconds\n",
      "\n",
      "    Total time                      = 93.5617 seconds\n",
      "\n",
      " 93.561606 seconds (55.49 M allocations: 5.355 GiB, 5.38% gc time)\n",
      "error overall = 0.007557629468380322 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "# 8 threads, d = 1000, minwidth = 250\n",
    "Random.seed!(2020)\n",
    "chr = 18\n",
    "maf = 0.1\n",
    "d = 1000\n",
    "overlap = 0.0\n",
    "minwidth = 250\n",
    "tgtfile = \"target.chr$chr.typedOnly.maf$maf.masked.vcf.gz\"\n",
    "reffile = \"ref.chr18.maxd$d.minwidth$minwidth.maf$maf.excludeTarget.jlso\"\n",
    "outfile = \"mendel.imputed.vcf.gz\"\n",
    "@time ph = phase(tgtfile, reffile, outfile=outfile, phase=true);\n",
    "\n",
    "X_complete = convert_gt(UInt8, \"target.chr18.full.vcf.gz\")\n",
    "n, p = size(X_complete)\n",
    "X_mendel = convert_gt(UInt8, outfile)\n",
    "println(\"error overall = $(sum(X_mendel .!= X_complete) / n / p) \\n\")\n",
    "# rm(outfile, force=true)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of threads = 8\n",
      "Importing reference haplotype data...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32mComputing optimal haplotypes...100%|████████████████████| Time: 0:00:41\u001b[39m\n",
      "\u001b[32mPhasing...100%|█████████████████████████████████████████| Time: 0:01:19\u001b[39m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total windows = 256, averaging ~ 4450 unique haplotypes per window.\n",
      "\n",
      "Timings: \n",
      "    Data import                     = 13.0042 seconds\n",
      "        import target data             = 3.90106 seconds\n",
      "        import compressed haplotypes   = 9.10316 seconds\n",
      "    Computing haplotype pair        = 44.0709 seconds\n",
      "        BLAS3 mul! to get M and N      = 9.22317 seconds per thread\n",
      "        haplopair search               = 27.9716 seconds per thread\n",
      "        initializing missing           = 0.0263227 seconds per thread\n",
      "        allocating and viewing         = 0.831811 seconds per thread\n",
      "        index conversion               = 0.000607927 seconds per thread\n",
      "    Phasing by win-win intersection = 79.4378 seconds\n",
      "        Window-by-window intersection  = 0.0016319 seconds per thread\n",
      "        Breakpoint search              = 75.0606 seconds per thread\n",
      "        Recording result               = 0.0120721 seconds per thread\n",
      "    Imputation                     = 2.84468 seconds\n",
      "        Imputing missing               = 0.889501 seconds\n",
      "        Writing to file                = 1.95518 seconds\n",
      "\n",
      "    Total time                      = 139.358 seconds\n",
      "\n",
      "139.357968 seconds (52.79 M allocations: 5.369 GiB, 3.21% gc time)\n",
      "error overall = 0.012676618164161005 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "# 8 threads, d = 1000, minwidth = 400\n",
    "Random.seed!(2020)\n",
    "chr = 18\n",
    "maf = 0.1\n",
    "d = 1000\n",
    "overlap = 0.0\n",
    "minwidth = 400\n",
    "tgtfile = \"target.chr$chr.typedOnly.maf$maf.masked.vcf.gz\"\n",
    "reffile = \"ref.chr18.maxd$d.minwidth$minwidth.maf$maf.excludeTarget.jlso\"\n",
    "outfile = \"mendel.imputed.vcf.gz\"\n",
    "@time ph = phase(tgtfile, reffile, outfile=outfile, phase=true);\n",
    "\n",
    "X_complete = convert_gt(UInt8, \"target.chr18.full.vcf.gz\")\n",
    "n, p = size(X_complete)\n",
    "X_mendel = convert_gt(UInt8, outfile)\n",
    "println(\"error overall = $(sum(X_mendel .!= X_complete) / n / p) \\n\")\n",
    "# rm(outfile, force=true)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of threads = 8\n",
      "Importing reference haplotype data...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32mComputing optimal haplotypes...100%|████████████████████| Time: 0:00:41\u001b[39m\n",
      "\u001b[32mPhasing...100%|█████████████████████████████████████████| Time: 0:01:11\u001b[39m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total windows = 256, averaging ~ 4450 unique haplotypes per window.\n",
      "\n",
      "Timings: \n",
      "    Data import                     = 13.6892 seconds\n",
      "        import target data             = 4.44948 seconds\n",
      "        import compressed haplotypes   = 9.23969 seconds\n",
      "    Computing haplotype pair        = 43.7362 seconds\n",
      "        BLAS3 mul! to get M and N      = 9.11761 seconds per thread\n",
      "        haplopair search               = 27.6916 seconds per thread\n",
      "        initializing missing           = 0.025862 seconds per thread\n",
      "        allocating and viewing         = 0.842874 seconds per thread\n",
      "        index conversion               = 0.0006032 seconds per thread\n",
      "    Phasing by win-win intersection = 71.5608 seconds\n",
      "        Window-by-window intersection  = 0.00138144 seconds per thread\n",
      "        Breakpoint search              = 67.8754 seconds per thread\n",
      "        Recording result               = 0.0101877 seconds per thread\n",
      "    Imputation                     = 2.13964 seconds\n",
      "        Imputing missing               = 0.727538 seconds\n",
      "        Writing to file                = 1.4121 seconds\n",
      "\n",
      "    Total time                      = 131.126 seconds\n",
      "\n",
      "131.126218 seconds (53.02 M allocations: 5.376 GiB, 4.00% gc time)\n",
      "error overall = 0.012676618164161005 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "# 8 threads, d = 1000, minwidth = 600\n",
    "Random.seed!(2020)\n",
    "chr = 18\n",
    "maf = 0.1\n",
    "d = 1000\n",
    "overlap = 0.0\n",
    "minwidth = 600\n",
    "tgtfile = \"target.chr$chr.typedOnly.maf$maf.masked.vcf.gz\"\n",
    "reffile = \"ref.chr18.maxd$d.minwidth$minwidth.maf$maf.excludeTarget.jlso\"\n",
    "outfile = \"mendel.imputed.vcf.gz\"\n",
    "@time ph = phase(tgtfile, reffile, outfile=outfile, phase=true);\n",
    "\n",
    "X_complete = convert_gt(UInt8, \"target.chr18.full.vcf.gz\")\n",
    "n, p = size(X_complete)\n",
    "X_mendel = convert_gt(UInt8, outfile)\n",
    "println(\"error overall = $(sum(X_mendel .!= X_complete) / n / p) \\n\")\n",
    "# rm(outfile, force=true)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Overlapping windows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of threads = 8\n",
      "Importing reference haplotype data...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32mImporting genotype file...100%|█████████████████████████| Time: 0:00:05\u001b[39m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total windows = 2367, averaging ~ 527 unique haplotypes per window.\n",
      "\n",
      "Timings: \n",
      "    Data import                     = 23.4574 seconds\n",
      "        import target data             = 5.23924 seconds\n",
      "        import compressed haplotypes   = 18.2182 seconds\n",
      "    Computing haplotype pair        = 4.53151 seconds\n",
      "        BLAS3 mul! to get M and N      = 0.380729 seconds per thread\n",
      "        haplopair search               = 3.64957 seconds per thread\n",
      "        initializing missing           = 0.0240004 seconds per thread\n",
      "        allocating and viewing         = 0.0684376 seconds per thread\n",
      "        index conversion               = 0.00326528 seconds per thread\n",
      "    Phasing by win-win intersection = 1.41532 seconds\n",
      "        Window-by-window intersection  = 0.105019 seconds per thread\n",
      "        Breakpoint search              = 1.06294 seconds per thread\n",
      "        Recording result               = 0.0144047 seconds per thread\n",
      "    Imputation                     = 2.3351 seconds\n",
      "        Imputing missing               = 1.08188 seconds\n",
      "        Writing to file                = 1.25321 seconds\n",
      "\n",
      "    Total time                      = 31.7406 seconds\n",
      "\n",
      "overlap 0.0 error = 0.005259159607882693 \n",
      "\n",
      "Number of threads = 8\n",
      "Importing reference haplotype data...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32mComputing optimal haplotypes...100%|████████████████████| Time: 0:00:09\u001b[39m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total windows = 2367, averaging ~ 734 unique haplotypes per window.\n",
      "\n",
      "Timings: \n",
      "    Data import                     = 22.9327 seconds\n",
      "        import target data             = 4.59166 seconds\n",
      "        import compressed haplotypes   = 18.341 seconds\n",
      "    Computing haplotype pair        = 11.1601 seconds\n",
      "        BLAS3 mul! to get M and N      = 0.775174 seconds per thread\n",
      "        haplopair search               = 7.77059 seconds per thread\n",
      "        initializing missing           = 0.0300363 seconds per thread\n",
      "        allocating and viewing         = 0.338323 seconds per thread\n",
      "        index conversion               = 0.00296176 seconds per thread\n",
      "    Phasing by win-win intersection = 2.00624 seconds\n",
      "        Window-by-window intersection  = 0.0857522 seconds per thread\n",
      "        Breakpoint search              = 1.76 seconds per thread\n",
      "        Recording result               = 0.014369 seconds per thread\n",
      "    Imputation                     = 2.88384 seconds\n",
      "        Imputing missing               = 0.21867 seconds\n",
      "        Writing to file                = 2.66517 seconds\n",
      "\n",
      "    Total time                      = 38.9841 seconds\n",
      "\n",
      "overlap 0.1 error = 0.0051601450618225155 \n",
      "\n",
      "Number of threads = 8\n",
      "Importing reference haplotype data...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32mComputing optimal haplotypes...100%|████████████████████| Time: 0:00:21\u001b[39m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total windows = 2367, averaging ~ 1059 unique haplotypes per window.\n",
      "\n",
      "Timings: \n",
      "    Data import                     = 21.1849 seconds\n",
      "        import target data             = 4.79981 seconds\n",
      "        import compressed haplotypes   = 16.3851 seconds\n",
      "    Computing haplotype pair        = 23.43 seconds\n",
      "        BLAS3 mul! to get M and N      = 2.21146 seconds per thread\n",
      "        haplopair search               = 17.5658 seconds per thread\n",
      "        initializing missing           = 0.0387842 seconds per thread\n",
      "        allocating and viewing         = 0.524572 seconds per thread\n",
      "        index conversion               = 0.00475071 seconds per thread\n",
      "    Phasing by win-win intersection = 1.76297 seconds\n",
      "        Window-by-window intersection  = 0.0602452 seconds per thread\n",
      "        Breakpoint search              = 1.52317 seconds per thread\n",
      "        Recording result               = 0.0145165 seconds per thread\n",
      "    Imputation                     = 2.34634 seconds\n",
      "        Imputing missing               = 1.02221 seconds\n",
      "        Writing to file                = 1.32413 seconds\n",
      "\n",
      "    Total time                      = 48.7254 seconds\n",
      "\n",
      "overlap 0.25 error = 0.005138528879829041 \n",
      "\n",
      "Number of threads = 8\n",
      "Importing reference haplotype data...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32mComputing optimal haplotypes...100%|████████████████████| Time: 0:00:47\u001b[39m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total windows = 2367, averaging ~ 1593 unique haplotypes per window.\n",
      "\n",
      "Timings: \n",
      "    Data import                     = 24.5372 seconds\n",
      "        import target data             = 4.77399 seconds\n",
      "        import compressed haplotypes   = 19.7632 seconds\n",
      "    Computing haplotype pair        = 50.4585 seconds\n",
      "        BLAS3 mul! to get M and N      = 6.07232 seconds per thread\n",
      "        haplopair search               = 39.456 seconds per thread\n",
      "        initializing missing           = 0.0527969 seconds per thread\n",
      "        allocating and viewing         = 0.792506 seconds per thread\n",
      "        index conversion               = 0.00632086 seconds per thread\n",
      "    Phasing by win-win intersection = 1.81898 seconds\n",
      "        Window-by-window intersection  = 0.0360357 seconds per thread\n",
      "        Breakpoint search              = 1.63553 seconds per thread\n",
      "        Recording result               = 0.0128104 seconds per thread\n",
      "    Imputation                     = 2.41322 seconds\n",
      "        Imputing missing               = 0.929595 seconds\n",
      "        Writing to file                = 1.48362 seconds\n",
      "\n",
      "    Total time                      = 79.2291 seconds\n",
      "\n",
      "overlap 0.5 error = 0.005149249004811155 \n",
      "\n",
      "Number of threads = 8\n",
      "Importing reference haplotype data...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32mComputing optimal haplotypes...100%|████████████████████| Time: 0:01:29\u001b[39m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total windows = 2367, averaging ~ 2099 unique haplotypes per window.\n",
      "\n",
      "Timings: \n",
      "    Data import                     = 20.5885 seconds\n",
      "        import target data             = 4.75614 seconds\n",
      "        import compressed haplotypes   = 15.8323 seconds\n",
      "    Computing haplotype pair        = 92.3361 seconds\n",
      "        BLAS3 mul! to get M and N      = 12.8246 seconds per thread\n",
      "        haplopair search               = 72.2983 seconds per thread\n",
      "        initializing missing           = 0.0705915 seconds per thread\n",
      "        allocating and viewing         = 1.16329 seconds per thread\n",
      "        index conversion               = 0.00620707 seconds per thread\n",
      "    Phasing by win-win intersection = 2.21282 seconds\n",
      "        Window-by-window intersection  = 0.0263918 seconds per thread\n",
      "        Breakpoint search              = 2.01263 seconds per thread\n",
      "        Recording result               = 0.0134365 seconds per thread\n",
      "    Imputation                     = 2.66718 seconds\n",
      "        Imputing missing               = 0.988753 seconds\n",
      "        Writing to file                = 1.67843 seconds\n",
      "\n",
      "    Total time                      = 117.806 seconds\n",
      "\n",
      "overlap 0.75 error = 0.00530363522487632 \n",
      "\n",
      "Number of threads = 8\n",
      "Importing reference haplotype data...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32mComputing optimal haplotypes...100%|████████████████████| Time: 0:02:15\u001b[39m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total windows = 2367, averaging ~ 2542 unique haplotypes per window.\n",
      "\n",
      "Timings: \n",
      "    Data import                     = 19.8248 seconds\n",
      "        import target data             = 4.70737 seconds\n",
      "        import compressed haplotypes   = 15.1175 seconds\n",
      "    Computing haplotype pair        = 138.896 seconds\n",
      "        BLAS3 mul! to get M and N      = 20.6037 seconds per thread\n",
      "        haplopair search               = 108.035 seconds per thread\n",
      "        initializing missing           = 0.0885541 seconds per thread\n",
      "        allocating and viewing         = 1.66364 seconds per thread\n",
      "        index conversion               = 0.00720214 seconds per thread\n",
      "    Phasing by win-win intersection = 2.29939 seconds\n",
      "        Window-by-window intersection  = 0.0217813 seconds per thread\n",
      "        Breakpoint search              = 2.13475 seconds per thread\n",
      "        Recording result               = 0.014543 seconds per thread\n",
      "    Imputation                     = 2.94009 seconds\n",
      "        Imputing missing               = 1.13223 seconds\n",
      "        Writing to file                = 1.80786 seconds\n",
      "\n",
      "    Total time                      = 163.962 seconds\n",
      "\n",
      "overlap 1.0 error = 0.00551579752334618 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "function overlap_run()\n",
    "    chr = 18\n",
    "    maf = 0.1\n",
    "    max_d = 1000\n",
    "    X_complete = convert_gt(UInt8, \"target.chr$chr.full.vcf.gz\")\n",
    "    n, p = size(X_complete)\n",
    "    for overlap in [0.0, 0.1, 0.25, 0.5, 0.75, 1.0]\n",
    "        tgtfile = \"target.chr$chr.typedOnly.maf$maf.masked.vcf.gz\"\n",
    "        reffile = \"ref.chr18.maxd$max_d.overlap$overlap.maf$maf.excludeTarget.jlso\"\n",
    "        outfile = \"mendel.imputed.chr$chr.vcf.gz\"\n",
    "\n",
    "        ph = phase(tgtfile, reffile, outfile=outfile, impute=true, max_d=max_d, phase=true)\n",
    "        X_mendel = convert_gt(UInt8, outfile)\n",
    "        println(\"overlap $overlap error = $(sum(X_mendel .!= X_complete) / n / p) \\n\")\n",
    "        flush(stdout)\n",
    "    end\n",
    "end\n",
    "Random.seed!(2020)\n",
    "overlap_run()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Haplotype rescreening\n",
    "\n",
    "This is enabled by the \"rescreen\" option. This option saves a number of top haplotype pairs when solving the least squares objective, and re-minimize least squares on just observed data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Importing reference haplotype data...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32mPhasing chunk 1/1...100%|███████████████████████████████| Time: 0:05:41\u001b[39m\n",
      "\u001b[32mWriting to file...100%|█████████████████████████████████| Time: 0:00:05\u001b[39m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total windows = 331, averaging ~ 4096 unique haplotypes per window.\n",
      "\n",
      "Timings: \n",
      "    Data import                     = 9.55672 seconds\n",
      "    Computing haplotype pair        = 341.232 seconds\n",
      "        BLAS3 mul! to get M and N      = 8.06296 seconds per thread\n",
      "        haplopair search               = 33.8217 seconds per thread\n",
      "        min least sq on observed data  = 0.232213 seconds per thread\n",
      "        finding redundant happairs     = 0.00460313 seconds per thread\n",
      "    Phasing by win-win intersection = 0.097636 seconds\n",
      "    Imputation                      = 6.46035 seconds\n",
      "\n",
      "357.448572 seconds (52.97 M allocations: 29.080 GiB, 1.11% gc time)\n",
      "error overall = 0.006447486635030178 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "# 8 threads, keep best happair\n",
    "Random.seed!(2020)\n",
    "chr = 18\n",
    "maf = 0.1\n",
    "width = 512\n",
    "tgtfile = \"target.chr$chr.typedOnly.maf$maf.masked.vcf.gz\"\n",
    "reffile = \"ref.chr18.$width.maf$maf.excludeTarget.jlso\"\n",
    "outfile = \"mendel.imputed.dp$width.maf$maf.vcf.gz\"\n",
    "@time ph = phase(tgtfile, reffile, outfile=outfile, impute=true, width=width,\n",
    "    dynamic_programming = false, rescreen=true)\n",
    "\n",
    "X_complete = convert_gt(UInt8, \"target.chr18.full.vcf.gz\")\n",
    "n, p = size(X_complete)\n",
    "X_mendel = convert_gt(UInt8, outfile)\n",
    "println(\"error overall = $(sum(X_mendel .!= X_complete) / n / p) \\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Haplotype Thinning\n",
    "\n",
    "This is enabled by setting \"thinning_factor\" to some constant. This option solves the least squares objective on only \"thining_factor\" unique haplotypes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "8"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Threads.nthreads()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Importing reference haplotype data...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32mWriting to file...100%|█████████████████████████████████| Time: 0:00:05\u001b[39m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total windows = 331, averaging ~ 4096 unique haplotypes per window.\n",
      "\n",
      "Timings: \n",
      "    Data import                     = 9.54492 seconds\n",
      "    Computing haplotype pair        = 3.22873 seconds\n",
      "        computing dist(X, H)           = 0.359715 seconds per thread\n",
      "        BLAS3 mul! to get M and N      = 1.43287 seconds per thread\n",
      "        haplopair search               = 0.0432943 seconds per thread\n",
      "        finding redundant happairs     = 0.00209051 seconds per thread\n",
      "    Phasing by dynamic programming  = 0.00814605 seconds\n",
      "    Imputation                      = 5.61724 seconds\n",
      "\n",
      " 18.537449 seconds (52.94 M allocations: 8.139 GiB, 10.27% gc time)\n",
      "error overall = 0.011923793282211395 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "# 8 threads, not scaling by freq\n",
    "Random.seed!(2020)\n",
    "chr = 18\n",
    "maf = 0.1\n",
    "width = 512\n",
    "tgtfile = \"target.chr$chr.typedOnly.maf$maf.masked.vcf.gz\"\n",
    "reffile = \"ref.chr18.$width.maf$maf.excludeTarget.jlso\"\n",
    "outfile = \"mendel.imputed.dp$width.maf$maf.vcf.gz\"\n",
    "@time ph, hs = phase(tgtfile, reffile, outfile=outfile, impute=true, width=width, \n",
    "    thinning_factor=100, max_haplotypes = 100)\n",
    "\n",
    "X_complete = convert_gt(UInt8, \"target.chr18.full.vcf.gz\")\n",
    "n, p = size(X_complete)\n",
    "X_mendel = convert_gt(UInt8, outfile)\n",
    "println(\"error overall = $(sum(X_mendel .!= X_complete) / n / p) \\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Importing reference haplotype data...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32mWriting to file...100%|█████████████████████████████████| Time: 0:00:05\u001b[39m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total windows = 331, averaging ~ 4096 unique haplotypes per window.\n",
      "\n",
      "Timings: \n",
      "    Data import                     = 9.23721 seconds\n",
      "    Computing haplotype pair        = 3.19395 seconds\n",
      "        computing dist(X, H)           = 0.344292 seconds per thread\n",
      "        BLAS3 mul! to get M and N      = 1.44139 seconds per thread\n",
      "        haplopair search               = 0.0417762 seconds per thread\n",
      "        finding redundant happairs     = 0.00218276 seconds per thread\n",
      "    Phasing by dynamic programming  = 0.0077219 seconds\n",
      "    Imputation                      = 5.57712 seconds\n",
      "\n",
      " 18.134003 seconds (53.35 M allocations: 8.293 GiB, 10.24% gc time)\n",
      "error overall = 0.012110058385976105 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "# 8 threads, scaling by 1/(2p(1-p))\n",
    "Random.seed!(2020)\n",
    "chr = 18\n",
    "maf = 0.1\n",
    "width = 512\n",
    "tgtfile = \"target.chr$chr.typedOnly.maf$maf.masked.vcf.gz\"\n",
    "reffile = \"ref.chr18.$width.maf$maf.excludeTarget.jlso\"\n",
    "outfile = \"mendel.imputed.dp$width.maf$maf.vcf.gz\"\n",
    "@time ph = phase(tgtfile, reffile, outfile=outfile, impute=true, width=width, \n",
    "    thinning_factor=100, thinning_scale_allelefreq=true, max_haplotypes = 100)\n",
    "\n",
    "# X_complete = convert_gt(UInt8, \"target.chr18.full.vcf.gz\")\n",
    "n, p = size(X_complete)\n",
    "X_mendel = convert_gt(UInt8, outfile)\n",
    "println(\"error overall = $(sum(X_mendel .!= X_complete) / n / p) \\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Importing reference haplotype data...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32mWriting to file...100%|█████████████████████████████████| Time: 0:00:05\u001b[39m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total windows = 331, averaging ~ 4096 unique haplotypes per window.\n",
      "\n",
      "Timings: \n",
      "    Data import                     = 9.97356 seconds\n",
      "    Computing haplotype pair        = 3.45103 seconds\n",
      "        computing dist(X, H)           = 0.386598 seconds per thread\n",
      "        BLAS3 mul! to get M and N      = 1.47565 seconds per thread\n",
      "        haplopair search               = 0.0429367 seconds per thread\n",
      "        finding redundant happairs     = 0.00221398 seconds per thread\n",
      "    Phasing by dynamic programming  = 0.00787711 seconds\n",
      "    Imputation                      = 5.71789 seconds\n",
      "\n",
      " 19.249530 seconds (53.34 M allocations: 8.292 GiB, 10.50% gc time)\n",
      "error overall = 0.012115289431645715 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "# 8 threads, scaling by 1/p\n",
    "Random.seed!(2020)\n",
    "chr = 18\n",
    "maf = 0.1\n",
    "width = 512\n",
    "tgtfile = \"target.chr$chr.typedOnly.maf$maf.masked.vcf.gz\"\n",
    "reffile = \"ref.chr18.$width.maf$maf.excludeTarget.jlso\"\n",
    "outfile = \"mendel.imputed.dp$width.maf$maf.vcf.gz\"\n",
    "@time ph = phase(tgtfile, reffile, outfile=outfile, impute=true, width=width, \n",
    "    thinning_factor=100, thinning_scale_allelefreq=true, max_haplotypes = 100)\n",
    "\n",
    "# X_complete = convert_gt(UInt8, \"target.chr18.full.vcf.gz\")\n",
    "n, p = size(X_complete)\n",
    "X_mendel = convert_gt(UInt8, outfile)\n",
    "println(\"error overall = $(sum(X_mendel .!= X_complete) / n / p) \\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Importing reference haplotype data...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32mPhasing chunk 1/1...100%|███████████████████████████████| Time: 0:00:05\u001b[39m\n",
      "\u001b[32mWriting to file...100%|█████████████████████████████████| Time: 0:00:05\u001b[39m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total windows = 331, averaging ~ 4096 unique haplotypes per window.\n",
      "\n",
      "Timings: \n",
      "    Data import                     = 10.3302 seconds\n",
      "    Computing haplotype pair        = 5.41888 seconds\n",
      "        computing dist(X, H)           = 0.356879 seconds per thread\n",
      "        BLAS3 mul! to get M and N      = 3.14795 seconds per thread\n",
      "        haplopair search               = 0.157715 seconds per thread\n",
      "        finding redundant happairs     = 0.00229817 seconds per thread\n",
      "    Phasing by dynamic programming  = 0.0103331 seconds\n",
      "    Imputation                      = 5.91353 seconds\n",
      "\n",
      " 21.780103 seconds (52.92 M allocations: 8.239 GiB, 11.42% gc time)\n",
      "error overall = 0.010622494434683475 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "# 8 threads, scaling by 1-p\n",
    "Random.seed!(2020)\n",
    "chr = 18\n",
    "maf = 0.1\n",
    "width = 512\n",
    "tgtfile = \"target.chr$chr.typedOnly.maf$maf.masked.vcf.gz\"\n",
    "reffile = \"ref.chr18.$width.maf$maf.excludeTarget.jlso\"\n",
    "outfile = \"mendel.imputed.dp$width.maf$maf.vcf.gz\"\n",
    "@time ph = phase(tgtfile, reffile, outfile=outfile, impute=true, width=width, \n",
    "    thinning_factor=100, thinning_scale_allelefreq=true, max_haplotypes = 100)\n",
    "\n",
    "# X_complete = convert_gt(UInt8, \"target.chr18.full.vcf.gz\")\n",
    "n, p = size(X_complete)\n",
    "X_mendel = convert_gt(UInt8, outfile)\n",
    "println(\"error overall = $(sum(X_mendel .!= X_complete) / n / p) \\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Importing reference haplotype data...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32mWriting to file...100%|█████████████████████████████████| Time: 0:00:06\u001b[39m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total windows = 331, averaging ~ 4096 unique haplotypes per window.\n",
      "\n",
      "Timings: \n",
      "    Data import                     = 9.85995 seconds\n",
      "    Computing haplotype pair        = 4.84689 seconds\n",
      "        computing dist(X, H)           = 0.340418 seconds per thread\n",
      "        BLAS3 mul! to get M and N      = 3.21735 seconds per thread\n",
      "        haplopair search               = 0.160892 seconds per thread\n",
      "        finding redundant happairs     = 0.00239908 seconds per thread\n",
      "    Phasing by dynamic programming  = 0.010365 seconds\n",
      "    Imputation                      = 7.81272 seconds\n",
      "\n",
      " 22.775795 seconds (52.84 M allocations: 8.369 GiB, 8.85% gc time)\n",
      "error overall = 0.010608314313126171 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "# 8 threads, scaling by 1-p\n",
    "Random.seed!(2020)\n",
    "chr = 18\n",
    "maf = 0.1\n",
    "width = 512\n",
    "tgtfile = \"target.chr$chr.typedOnly.maf$maf.masked.vcf.gz\"\n",
    "reffile = \"ref.chr18.$width.maf$maf.excludeTarget.jlso\"\n",
    "outfile = \"mendel.imputed.dp$width.maf$maf.vcf.gz\"\n",
    "@time ph = phase(tgtfile, reffile, outfile=outfile, impute=true, width=width, \n",
    "    thinning_factor=200, thinning_scale_allelefreq=true, max_haplotypes = 100)\n",
    "\n",
    "# X_complete = convert_gt(UInt8, \"target.chr18.full.vcf.gz\")\n",
    "n, p = size(X_complete)\n",
    "X_mendel = convert_gt(UInt8, outfile)\n",
    "println(\"error overall = $(sum(X_mendel .!= X_complete) / n / p) \\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Chunking"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Importing reference haplotype data...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32mComputing haplotype pairs for chunk 1/17...100%|████████| Time: 0:00:05\u001b[39m\n",
      "\u001b[32mComputing haplotype pairs for chunk 10/17...100%|███████| Time: 0:00:05\u001b[39m\n",
      "\u001b[32mComputing haplotype pairs for chunk 13/17...100%|███████| Time: 0:00:05\u001b[39m\n",
      "\u001b[32mComputing haplotype pairs for chunk 16/17...100%|███████| Time: 0:00:06\u001b[39m\n",
      "\u001b[32mWriting to file...100%|█████████████████████████████████| Time: 0:00:06\u001b[39m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total windows = 331, averaging ~ 4096 unique haplotypes per window.\n",
      "\n",
      "Timings: \n",
      "    Data import                     = 9.58859 seconds\n",
      "    Computing haplotype pair        = 80.8308 seconds\n",
      "        BLAS3 mul! to get M and N      = 8.3887 seconds per thread\n",
      "        haplopair search               = 46.2132 seconds per thread\n",
      "        finding redundant happairs     = 2.51446 seconds per thread\n",
      "    Phasing by win-win intersection = 0.0706799 seconds\n",
      "    Imputation                      = 6.71369 seconds\n",
      "\n",
      "120.304220 seconds (67.70 M allocations: 29.608 GiB, 3.67% gc time)\n",
      "error overall = 0.00644915212490705 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "# 8 threads, num_windows_per_chunks = 20\n",
    "Random.seed!(2020)\n",
    "chr = 18\n",
    "maf = 0.1\n",
    "width = 512\n",
    "tgtfile = \"target.chr$chr.typedOnly.maf$maf.masked.vcf.gz\"\n",
    "reffile = \"ref.chr18.$width.maf$maf.excludeTarget.jlso\"\n",
    "outfile = \"mendel.imputed.dp$width.maf$maf.vcf.gz\"\n",
    "@time ph, hs = phase(tgtfile, reffile, outfile=outfile, impute=true, width=width,\n",
    "    dynamic_programming = false)\n",
    "\n",
    "X_complete = convert_gt(UInt8, \"target.chr18.full.vcf.gz\")\n",
    "n, p = size(X_complete)\n",
    "X_mendel = convert_gt(UInt8, outfile)\n",
    "println(\"error overall = $(sum(X_mendel .!= X_complete) / n / p) \\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Importing reference haplotype data...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32mPhasing chunk 3/17...100%|██████████████████████████████| Time: 0:00:05\u001b[39m\n",
      "\u001b[32mWriting to file...100%|█████████████████████████████████| Time: 0:00:06\u001b[39m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total windows = 331, averaging ~ 4096 unique haplotypes per window.\n",
      "\n",
      "Timings: \n",
      "    Data import                     = 10.967 seconds\n",
      "    Computing haplotype pair        = 72.1552 seconds\n",
      "        BLAS3 mul! to get M and N      = 9.06695 seconds per thread\n",
      "        haplopair search               = 46.6832 seconds per thread\n",
      "        finding redundant happairs     = 1.885 seconds per thread\n",
      "    Phasing by win-win intersection = 1.99925 seconds\n",
      "    Imputation                      = 6.80302 seconds\n",
      "\n",
      "123.262407 seconds (68.50 M allocations: 29.626 GiB, 4.41% gc time)\n",
      "error overall = 0.00644915212490705 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "# 8 threads, num_windows_per_chunks = 20\n",
    "Random.seed!(2020)\n",
    "chr = 18\n",
    "maf = 0.1\n",
    "width = 512\n",
    "tgtfile = \"target.chr$chr.typedOnly.maf$maf.masked.vcf.gz\"\n",
    "reffile = \"ref.chr18.$width.maf$maf.excludeTarget.jlso\"\n",
    "outfile = \"mendel.imputed.dp$width.maf$maf.vcf.gz\"\n",
    "@time ph, hs = phase(tgtfile, reffile, outfile=outfile, impute=true, width=width,\n",
    "    dynamic_programming = false)\n",
    "\n",
    "X_complete = convert_gt(UInt8, \"target.chr18.full.vcf.gz\")\n",
    "n, p = size(X_complete)\n",
    "X_mendel = convert_gt(UInt8, outfile)\n",
    "println(\"error overall = $(sum(X_mendel .!= X_complete) / n / p) \\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Beagle 5.0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# convert vcf.gz to bref3\n",
    "java -jar ./bref3.18May20.d20.jar ref.chr18.excludeTarget.vcf.gz > ref.chr18.excludeTarget.bref3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "beagle.18May20.d20.jar (version 5.1)\n",
      "Copyright (C) 2014-2018 Brian L. Browning\n",
      "Enter \"java -jar beagle.18May20.d20.jar\" to list command line argument\n",
      "Start time: 11:58 AM PDT on 30 Jun 2020\n",
      "\n",
      "Command line: java -Xmx3641m -jar beagle.18May20.d20.jar\n",
      "  gt=target.chr18.typedOnly.maf0.1.masked.vcf.gz\n",
      "  ref=ref.chr18.excludeTarget.bref3\n",
      "  out=beagle.chr18.result\n",
      "  nthreads=8\n",
      "\n",
      "No genetic map is specified: using 1 cM = 1 Mb\n",
      "\n",
      "Reference samples:       2,404\n",
      "Study samples:             100\n",
      "\n",
      "Window 1 (18:10644-40010629)\n",
      "Reference markers:     409,667\n",
      "Study markers:          81,157\n",
      "\n",
      "Burnin  iteration 1:           8 seconds\n",
      "Burnin  iteration 2:           7 seconds\n",
      "Burnin  iteration 3:           7 seconds\n",
      "Burnin  iteration 4:           8 seconds\n",
      "Burnin  iteration 5:           7 seconds\n",
      "Burnin  iteration 6:           8 seconds\n",
      "\n",
      "Phasing iteration 1:           8 seconds\n",
      "Phasing iteration 2:           7 seconds\n",
      "Phasing iteration 3:           7 seconds\n",
      "Phasing iteration 4:           7 seconds\n",
      "Phasing iteration 5:           7 seconds\n",
      "Phasing iteration 6:           6 seconds\n",
      "Phasing iteration 7:           6 seconds\n",
      "Phasing iteration 8:           6 seconds\n",
      "Phasing iteration 9:           7 seconds\n",
      "Phasing iteration 10:          6 seconds\n",
      "Phasing iteration 11:          7 seconds\n",
      "Phasing iteration 12:          6 seconds\n",
      "\n",
      "Imputation time:               12 seconds\n",
      "\n",
      "Window 2 (18:36010613-76010634)\n",
      "Reference markers:     455,087\n",
      "Study markers:          90,368\n",
      "\n",
      "Burnin  iteration 1:           9 seconds\n",
      "Burnin  iteration 2:           9 seconds\n",
      "Burnin  iteration 3:           10 seconds\n",
      "Burnin  iteration 4:           9 seconds\n",
      "Burnin  iteration 5:           9 seconds\n",
      "Burnin  iteration 6:           11 seconds\n",
      "\n",
      "Phasing iteration 1:           11 seconds\n",
      "Phasing iteration 2:           9 seconds\n",
      "Phasing iteration 3:           9 seconds\n",
      "Phasing iteration 4:           8 seconds\n",
      "Phasing iteration 5:           8 seconds\n",
      "Phasing iteration 6:           8 seconds\n",
      "Phasing iteration 7:           8 seconds\n",
      "Phasing iteration 8:           9 seconds\n",
      "Phasing iteration 9:           8 seconds\n",
      "Phasing iteration 10:          8 seconds\n",
      "Phasing iteration 11:          8 seconds\n",
      "Phasing iteration 12:          8 seconds\n",
      "\n",
      "Imputation time:               14 seconds\n",
      "\n",
      "Window 3 (18:72010594-78017156)\n",
      "Reference markers:      79,388\n",
      "Study markers:          16,097\n",
      "\n",
      "Burnin  iteration 1:           2 seconds\n",
      "Burnin  iteration 2:           2 seconds\n",
      "Burnin  iteration 3:           1 second\n",
      "Burnin  iteration 4:           2 seconds\n",
      "Burnin  iteration 5:           2 seconds\n",
      "Burnin  iteration 6:           2 seconds\n",
      "\n",
      "Phasing iteration 1:           2 seconds\n",
      "Phasing iteration 2:           2 seconds\n",
      "Phasing iteration 3:           2 seconds\n",
      "Phasing iteration 4:           2 seconds\n",
      "Phasing iteration 5:           2 seconds\n",
      "Phasing iteration 6:           2 seconds\n",
      "Phasing iteration 7:           2 seconds\n",
      "Phasing iteration 8:           2 seconds\n",
      "Phasing iteration 9:           2 seconds\n",
      "Phasing iteration 10:          2 seconds\n",
      "Phasing iteration 11:          2 seconds\n",
      "Phasing iteration 12:          2 seconds\n",
      "\n",
      "Imputation time:               1 second\n",
      "\n",
      "Cumulative Statistics:\n",
      "\n",
      "Reference markers:     852,602\n",
      "Study markers:         169,914\n",
      "\n",
      "Haplotype phasing time:        5 minutes 16 seconds\n",
      "Imputation time:               27 seconds\n",
      "Total time:                    5 minutes 48 seconds\n",
      "\n",
      "End time: 12:04 PM PDT on 30 Jun 2020\n",
      "beagle.18May20.d20.jar finished\n"
     ]
    },
    {
     "ename": "UndefVarError",
     "evalue": "UndefVarError: convert_gt not defined",
     "output_type": "error",
     "traceback": [
      "UndefVarError: convert_gt not defined",
      "",
      "Stacktrace:",
      " [1] top-level scope at In[2]:3"
     ]
    }
   ],
   "source": [
    "# run beagle 5.1 (8 thread)\n",
    "run(`java -jar beagle.18May20.d20.jar gt=target.chr18.typedOnly.maf0.1.masked.vcf.gz ref=ref.chr18.excludeTarget.bref3 out=beagle.chr18.result nthreads=8`)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.002799993431870908"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# beagle 5.1 error rate\n",
    "X_complete = convert_gt(Float32, \"target.chr18.full.vcf.gz\")\n",
    "n, p = size(X_complete)\n",
    "X_beagle = convert_gt(Float32, \"beagle.chr18.result.vcf.gz\")\n",
    "error_rate = sum(X_beagle .!= X_complete) / n / p"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "# Eagle 2 + Minimac4\n",
    "\n",
    "In order to use the reference panel in Eagle 2's prephase option, one must first convert it to `.bcf` format via e.g. `htslib` which is *extremely* difficult to install. Even after we went through all the hard work to obtain the final `.bcf` reference file (see commands below), eagle 2.4 STILL SAYS the file is not acceptable (not bgzipped or some processing error). Therefore, I have no choice but to prephase without the reference panel. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# run eagle 2.4: 3367.79 sec on amd-2382 machine (can only run on linux systems)\n",
    "eagle --vcf=target.chr20.typedOnly.masked.vcf.gz --outPrefix=eagle.phased.chr20 --numThreads=4 --geneticMapFile=../Eagle_v2.4.1/tables/genetic_map_hg19_withX.txt.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# convert ref file to m3vcf format (Total Run completed in 1 hours, 46 mins, 24 seconds)\n",
    "/u/home/b/biona001/haplotype_comparisons/Minimac3/bin/Minimac3 --refHaps ref.chr20.excludeTarget.vcf.gz --processReference --prefix ref.chr20.excludeTarget"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# run minimac4 (2619 seconds)\n",
    "minimac4 --refHaps ref.chr20.excludeTarget.m3vcf.gz --haps eagle.phased.vcf.gz --prefix minimac.imputed.chr20 --format GT --cpus 4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# minimac4 error rate    \n",
    "X_complete = convert_gt(Float32, \"target.chr20.full.vcf.gz\")\n",
    "X_minimac = convert_gt(Float32, \"minimac.imputed.chr20.dose.vcf.gz\")\n",
    "n, p = size(X_complete)\n",
    "println(\"error overall = $(sum(X_minimac .!= X_complete) / n / p) \\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "# LASSO and haplotype thinning "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Importing reference haplotype data...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32mPhasing chunk 1/1...100%|███████████████████████████████| Time: 0:00:21\u001b[39m\n",
      "\u001b[32mWriting to file...100%|█████████████████████████████████| Time: 0:00:07\u001b[39m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total windows = 331, averaging ~ 4096 unique haplotypes per window.\n",
      "\n",
      "Timings: \n",
      "    Data import                     = 9.31658 seconds\n",
      "    Computing haplotype pair        = 21.1465 seconds\n",
      "        BLAS3 mul! to get M and N      = 13.9018 seconds per thread\n",
      "        haplopair search               = 2.455 seconds per thread\n",
      "        finding redundant happairs     = 0.00504225 seconds per thread\n",
      "    Phasing by win-win intersection = 0.044287 seconds\n",
      "    Imputation                      = 8.56951 seconds\n",
      "\n",
      "lasso r = 100: time = 39.170199529, error = 0.007713763279936008 \n",
      "\n",
      "Importing reference haplotype data...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32mPhasing chunk 1/1...100%|███████████████████████████████| Time: 0:00:22\u001b[39m\n",
      "\u001b[32mWriting to file...100%|█████████████████████████████████| Time: 0:00:07\u001b[39m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total windows = 331, averaging ~ 4096 unique haplotypes per window.\n",
      "\n",
      "Timings: \n",
      "    Data import                     = 10.1427 seconds\n",
      "    Computing haplotype pair        = 22.8784 seconds\n",
      "        BLAS3 mul! to get M and N      = 15.2554 seconds per thread\n",
      "        haplopair search               = 2.7129 seconds per thread\n",
      "        finding redundant happairs     = 0.00446146 seconds per thread\n",
      "    Phasing by win-win intersection = 0.0403309 seconds\n",
      "    Imputation                      = 8.36141 seconds\n",
      "\n",
      "lasso r = 100: time = 41.515630959, error = 0.007713763279936008 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "# lasso, 8 threads\n",
    "Random.seed!(2020)\n",
    "chr = 18\n",
    "maf = 0.1\n",
    "width = 512\n",
    "tgtfile = \"target.chr$chr.typedOnly.maf$maf.masked.vcf.gz\"\n",
    "reffile = \"ref.chr18.$width.maf$maf.excludeTarget.jlso\"\n",
    "outfile = \"mendel.imputed.dp$width.maf$maf.vcf.gz\"\n",
    "X_complete = convert_gt(UInt8, \"target.chr18.full.vcf.gz\")\n",
    "n, p = size(X_complete)\n",
    "\n",
    "for ls in 100\n",
    "    # not scale by freq\n",
    "    t = @elapsed ph = phase(tgtfile, reffile, outfile=outfile, impute=true, width=width, \n",
    "        dynamic_programming=false, lasso=ls, max_haplotypes=800)\n",
    "    X_mendel = convert_gt(UInt8, outfile)\n",
    "    println(\"lasso r = $ls: time = $t, error = $(sum(X_mendel .!= X_complete) / n / p) \\n\")\n",
    "    \n",
    "    # scale by freq\n",
    "    t = @elapsed ph = phase(tgtfile, reffile, outfile=outfile, impute=true, width=width, \n",
    "        dynamic_programming=false, lasso=ls, max_haplotypes=800, scale_allelefreq=true)\n",
    "    X_mendel = convert_gt(UInt8, outfile)\n",
    "    println(\"lasso r = $ls: time = $t, error = $(sum(X_mendel .!= X_complete) / n / p) \\n\")\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Importing reference haplotype data...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32mPhasing chunk 1/1...100%|███████████████████████████████| Time: 0:00:22\u001b[39m\n",
      "\u001b[32mWriting to file...100%|█████████████████████████████████| Time: 0:00:06\u001b[39m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total windows = 331, averaging ~ 4096 unique haplotypes per window.\n",
      "\n",
      "Timings: \n",
      "    Data import                     = 13.3922 seconds\n",
      "    Computing haplotype pair        = 22.8901 seconds\n",
      "        BLAS3 mul! to get M and N      = 10.1758 seconds per thread\n",
      "        haplopair search               = 2.58633 seconds per thread\n",
      "        finding redundant happairs     = 0.0903123 seconds per thread\n",
      "    Phasing by win-win intersection = 1.78401 seconds\n",
      "    Imputation                      = 7.47731 seconds\n",
      "\n",
      "lasso r = 100: time = 67.011659731, error = 0.00953075409159256 \n",
      "\n",
      "Importing reference haplotype data...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32mPhasing chunk 1/1...100%|███████████████████████████████| Time: 0:00:18\u001b[39m\n",
      "\u001b[32mWriting to file...100%|█████████████████████████████████| Time: 0:00:06\u001b[39m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total windows = 331, averaging ~ 4096 unique haplotypes per window.\n",
      "\n",
      "Timings: \n",
      "    Data import                     = 10.7892 seconds\n",
      "    Computing haplotype pair        = 18.0653 seconds\n",
      "        BLAS3 mul! to get M and N      = 10.6055 seconds per thread\n",
      "        haplopair search               = 2.70464 seconds per thread\n",
      "        finding redundant happairs     = 0.00483429 seconds per thread\n",
      "    Phasing by win-win intersection = 0.0280049 seconds\n",
      "    Imputation                      = 7.27936 seconds\n",
      "\n",
      "lasso r = 100: time = 36.297164102, error = 0.00953075409159256 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "# lasso, use M = H'H and Nt = H'X\n",
    "Random.seed!(2020)\n",
    "chr = 18\n",
    "maf = 0.1\n",
    "width = 512\n",
    "tgtfile = \"target.chr$chr.typedOnly.maf$maf.masked.vcf.gz\"\n",
    "reffile = \"ref.chr18.$width.maf$maf.excludeTarget.jlso\"\n",
    "outfile = \"mendel.imputed.dp$width.maf$maf.vcf.gz\"\n",
    "X_complete = convert_gt(UInt8, \"target.chr18.full.vcf.gz\")\n",
    "n, p = size(X_complete)\n",
    "\n",
    "for ls in 100\n",
    "    # not scale by freq\n",
    "    t = @elapsed ph = phase(tgtfile, reffile, outfile=outfile, impute=true, width=width, \n",
    "        dynamic_programming=false, lasso=ls, max_haplotypes=800)\n",
    "    X_mendel = convert_gt(UInt8, outfile)\n",
    "    println(\"lasso r = $ls: time = $t, error = $(sum(X_mendel .!= X_complete) / n / p) \\n\")\n",
    "    \n",
    "    # scale by freq\n",
    "    t = @elapsed ph = phase(tgtfile, reffile, outfile=outfile, impute=true, width=width, \n",
    "        dynamic_programming=false, lasso=ls, max_haplotypes=800, scale_allelefreq=true)\n",
    "    X_mendel = convert_gt(UInt8, outfile)\n",
    "    println(\"lasso r = $ls: time = $t, error = $(sum(X_mendel .!= X_complete) / n / p) \\n\")\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Importing reference haplotype data...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32mPhasing chunk 1/1...100%|███████████████████████████████| Time: 0:00:33\u001b[39m\n",
      "\u001b[32mWriting to file...100%|█████████████████████████████████| Time: 0:00:06\u001b[39m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total windows = 331, averaging ~ 4096 unique haplotypes per window.\n",
      "\n",
      "Timings: \n",
      "    Data import                     = 10.4958 seconds\n",
      "    Computing haplotype pair        = 33.2974 seconds\n",
      "        BLAS3 mul! to get M and N      = 25.9967 seconds per thread\n",
      "        haplopair search               = 4.11763 seconds per thread\n",
      "        finding redundant happairs     = 0.060105 seconds per thread\n",
      "    Phasing by dynamic programming  = 0.0296741 seconds\n",
      "    Imputation                      = 6.66472 seconds\n",
      "\n",
      "thinning factor (weighted) = 100: time = 50.612580208, error = 0.007702538816470052 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "# thinning, 8 threads\n",
    "Random.seed!(2020)\n",
    "chr = 18\n",
    "maf = 0.1\n",
    "width = 512\n",
    "tgtfile = \"target.chr$chr.typedOnly.maf$maf.masked.vcf.gz\"\n",
    "reffile = \"ref.chr18.$width.maf$maf.excludeTarget.jlso\"\n",
    "outfile = \"mendel.imputed.dp$width.maf$maf.vcf.gz\"\n",
    "X_complete = convert_gt(UInt8, \"target.chr18.full.vcf.gz\")\n",
    "n, p = size(X_complete)\n",
    "\n",
    "for tf in 100:100:100\n",
    "#     t = @elapsed ph, hs = phase(tgtfile, reffile, outfile=outfile, impute=true, width=width, \n",
    "#         thinning_factor=tf, max_haplotypes=2000)\n",
    "#     X_mendel = convert_gt(UInt8, outfile)\n",
    "#     println(\"thinning factor (unweighted) = $tf: time = $t, error = $(sum(X_mendel .!= X_complete) / n / p) \\n\")\n",
    "    \n",
    "    t = @elapsed ph, hs = phase(tgtfile, reffile, outfile=outfile, impute=true, width=width, \n",
    "        max_haplotypes=100, lasso=tf)\n",
    "    X_mendel = convert_gt(UInt8, outfile)\n",
    "    println(\"thinning factor (weighted) = $tf: time = $t, error = $(sum(X_mendel .!= X_complete) / n / p) \\n\")\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "┌ Info: Precompiling UnicodePlots [b8865327-cd53-5732-bb35-84acbb429228]\n",
      "└ @ Base loading.jl:1273\n",
      "WARNING: using UnicodePlots.width in module Main conflicts with an existing identifier.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "\u001b[90m                    ┌                                        ┐\u001b[39m \n",
       "   \u001b[0m\u001b[90m[\u001b[0m1000.0\u001b[90m, \u001b[0m1500.0\u001b[90m)\u001b[0m\u001b[90m ┤\u001b[39m\u001b[32m▇\u001b[39m\u001b[0m 2                                     \u001b[90m \u001b[39m \n",
       "   \u001b[0m\u001b[90m[\u001b[0m1500.0\u001b[90m, \u001b[0m2000.0\u001b[90m)\u001b[0m\u001b[90m ┤\u001b[39m\u001b[32m▇\u001b[39m\u001b[0m 3                                     \u001b[90m \u001b[39m \n",
       "   \u001b[0m\u001b[90m[\u001b[0m2000.0\u001b[90m, \u001b[0m2500.0\u001b[90m)\u001b[0m\u001b[90m ┤\u001b[39m\u001b[32m▇▇▇▇\u001b[39m\u001b[0m 12                                 \u001b[90m \u001b[39m \n",
       "   \u001b[0m\u001b[90m[\u001b[0m2500.0\u001b[90m, \u001b[0m3000.0\u001b[90m)\u001b[0m\u001b[90m ┤\u001b[39m\u001b[32m▇▇▇▇▇\u001b[39m\u001b[0m 16                                \u001b[90m \u001b[39m \n",
       "   \u001b[0m\u001b[90m[\u001b[0m3000.0\u001b[90m, \u001b[0m3500.0\u001b[90m)\u001b[0m\u001b[90m ┤\u001b[39m\u001b[32m▇▇▇▇▇▇▇\u001b[39m\u001b[0m 24                              \u001b[90m \u001b[39m \n",
       "   \u001b[0m\u001b[90m[\u001b[0m3500.0\u001b[90m, \u001b[0m4000.0\u001b[90m)\u001b[0m\u001b[90m ┤\u001b[39m\u001b[32m▇▇▇▇▇▇▇▇▇▇▇▇▇▇\u001b[39m\u001b[0m 45                       \u001b[90m \u001b[39m \n",
       "   \u001b[0m\u001b[90m[\u001b[0m4000.0\u001b[90m, \u001b[0m4500.0\u001b[90m)\u001b[0m\u001b[90m ┤\u001b[39m\u001b[32m▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇\u001b[39m\u001b[0m 113  \u001b[90m \u001b[39m \n",
       "   \u001b[0m\u001b[90m[\u001b[0m4500.0\u001b[90m, \u001b[0m5000.0\u001b[90m)\u001b[0m\u001b[90m ┤\u001b[39m\u001b[32m▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇\u001b[39m\u001b[0m 116 \u001b[90m \u001b[39m \n",
       "\u001b[90m                    └                                        ┘\u001b[39m \n",
       "\u001b[0m                                    Frequency"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "using UnicodePlots\n",
    "x = count_haplotypes_per_window(\"ref.chr18.512.maf0.1.excludeTarget.jlso\")\n",
    "histogram(x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "@webio": {
   "lastCommId": null,
   "lastKernelId": null
  },
  "kernelspec": {
   "display_name": "Julia 1.5.0",
   "language": "julia",
   "name": "julia-1.5"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.5.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
