{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Compare MendelImpute against Minimac4 and Beagle5 on simulated data\n",
    "\n",
    "Using different % missing and seed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "┌ Info: Precompiling Revise [295af30f-e4ad-537b-8983-00126c2a3abe]\n",
      "└ @ Base loading.jl:1273\n",
      "┌ Warning: no Manifest.toml file found, static paths used\n",
      "└ @ Revise /Users/biona001/.julia/packages/Revise/439di/src/Revise.jl:1108\n",
      "┌ Info: Precompiling VCFTools [a620830f-fdd7-5ebc-8d26-3621ab35fbfe]\n",
      "└ @ Base loading.jl:1273\n",
      "┌ Info: Precompiling MendelImpute [e47305d1-6a61-5370-bc5d-77554d143183]\n",
      "└ @ Base loading.jl:1273\n",
      "┌ Info: Precompiling Plots [91a5bcdd-55d7-5caf-9e0b-520d859cae80]\n",
      "└ @ Base loading.jl:1273\n"
     ]
    }
   ],
   "source": [
    "using Revise\n",
    "using VCFTools\n",
    "using MendelImpute\n",
    "using GeneticVariation\n",
    "using Random\n",
    "using SparseArrays\n",
    "using Plots\n",
    "using StatsBase"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Simulate data\n",
    "\n",
    "### Step 0. Install `msprime`\n",
    "\n",
    "[msprime download Link](https://msprime.readthedocs.io/en/stable/installation.html).\n",
    "\n",
    "Some people might need to activate conda environment via `conda config --set auto_activate_base True`. You can turn it off once simulation is done by executing `conda config --set auto_activate_base False`.\n",
    "\n",
    "\n",
    "### Step 1. Simulate data in terminal\n",
    "\n",
    "```\n",
    "python3 msprime_script.py 20000 10000 5000000 2e-8 2e-8 1111 > ./compare2/full.vcf\n",
    "```\n",
    "\n",
    "Arguments: \n",
    "+ Number of haplotypes = 20000\n",
    "+ Effective population size = 10000 ([source](https://www.the-scientist.com/the-nutshell/ancient-humans-more-diverse-43556))\n",
    "+ Sequence length = 10 million (same as Beagle 5's choice)\n",
    "+ Rrecombination rate = 2e-8 (default)\n",
    "+ mutation rate = 2e-8 (default)\n",
    "+ seed = 1111"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(41897, 10000)"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "records, samples = nrecords(\"./compare2/full.vcf\"), nsamples(\"./compare2/full.vcf\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Step 2: Convert simulated haplotypes to reference haplotypes and target genotype files\n",
    "\n",
    "+ `haplo_ref.vcf.gz`: haplotype reference files\n",
    "+ `target.vcf.gz`: complete genotype information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "129.700641 seconds (1.47 G allocations: 135.227 GiB, 13.06% gc time)\n",
      "249.543273 seconds (3.14 G allocations: 242.252 GiB, 14.38% gc time)\n"
     ]
    }
   ],
   "source": [
    "Random.seed!(2019)\n",
    "# randomly choose 1000 samples as imputation targets. \n",
    "records, samples = nrecords(\"./compare2/full.vcf\"), nsamples(\"./compare2/full.vcf\")\n",
    "tgt_index = falses(samples)\n",
    "tgt_index[sample(1:samples, 1000, replace = false)] .= true\n",
    "ref_index = .!tgt_index\n",
    "record_index = trues(records) # save all records (SNPs) \n",
    "\n",
    "# create target.vcf.gz and haplo_ref.vcf.gz\n",
    "tgtfile = \"./compare2/target.vcf.gz\"\n",
    "reffile = \"./compare2/haplo_ref.vcf.gz\"\n",
    "@time VCFTools.filter(\"./compare2/full.vcf\", record_index, tgt_index, des = tgtfile)\n",
    "@time VCFTools.filter(\"./compare2/full.vcf\", record_index, ref_index, des = reffile)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Error rate as a function of % masked\n",
    "\n",
    "`full.vcf` contains the full data with 20000 haplotypes. We randomly choose 1000 samples as imputation target. The remaining haplotypes become reference panels. In the imputation target file, we randomly mask 0.01%, 0.1%, 1%, 5%, 10%, 25%, 50% data and compare compare the speed, memory, and imputation accuracy of MendelImpute, Minimac4, and Beagle5.\n",
    "\n",
    "### Mask 0.01%, 0.1%, 1%, 5%, 10%, 25%, 50% entries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  5.314990 seconds (84.72 M allocations: 7.291 GiB, 16.36% gc time)\n"
     ]
    }
   ],
   "source": [
    "# import full target matrix. Also transpose so that columns are samples. \n",
    "@time X = convert_gt(Float32, tgtfile; as_minorallele=false)\n",
    "X = copy(X')\n",
    "p, n = size(X)\n",
    "missingprop = [0.0001, 0.001, 0.01, 0.05, 0.1, 0.25, 0.5]\n",
    "for i in missingprop\n",
    "    # mask entries and save result as VCF file. \n",
    "    Random.seed!(2019)\n",
    "    X .= ifelse.(rand(Float32, p, n) .< i, missing, X)\n",
    "    masks = ismissing.(X)\n",
    "    mask_gt(tgtfile, masks, des=\"./compare2/target_masked_$i.vcf.gz\")\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Convert reference haplotypes from vcf to m3vcf on Hoffman (Julia 1.2)\n",
    "\n",
    "This is needed because Minimac4 only runs their own M3VCF files (which requires Minimac3)\n",
    "\n",
    "```Julia\n",
    "minimac3 = \"/u/home/b/biona001/haplotype_comparisons/minimac3/Minimac3/bin/Minimac3\"\n",
    "@time run(`$minimac3 --refHaps haplo_ref.vcf.gz --processReference --prefix haplo_ref`)\n",
    "```\n",
    "Time = 1452.914532 seconds (15.31 k allocations: 861.069 KiB)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Analyze:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "beagle.28Sep18.793.jar (version 5.0)\n",
      "Copyright (C) 2014-2018 Brian L. Browning\n",
      "Enter \"java -jar beagle.28Sep18.793.jar\" to list command line argument\n",
      "Start time: 12:04 PM PST on 16 Dec 2019\n",
      "\n",
      "Command line: java -Xmx3641m -jar beagle.28Sep18.793.jar\n",
      "  gt=./compare2/target_masked_0.0001.vcf.gz\n",
      "  ref=./compare2/haplo_ref.vcf.gz\n",
      "  out=./compare2/beagle.result.0.0001\n",
      "\n",
      "No genetic map is specified: using 1 cM = 1 Mb\n",
      "\n",
      "Reference samples:       9,000\n",
      "Study samples:           1,000\n",
      "\n",
      "Window 1 (1:244-4999928)\n",
      "Reference markers:      41,897\n",
      "Study markers:          41,897\n",
      "\n",
      "Burnin  iteration 1:           54 seconds\n",
      "Burnin  iteration 2:           1 minute 11 seconds\n",
      "Burnin  iteration 3:           51 seconds\n",
      "Burnin  iteration 4:           52 seconds\n",
      "Burnin  iteration 5:           1 minute 1 second\n",
      "Burnin  iteration 6:           1 minute 1 second\n",
      "\n",
      "Phasing iteration 1:           28 seconds\n",
      "Phasing iteration 2:           28 seconds\n",
      "Phasing iteration 3:           28 seconds\n",
      "Phasing iteration 4:           27 seconds\n",
      "Phasing iteration 5:           27 seconds\n",
      "Phasing iteration 6:           27 seconds\n",
      "Phasing iteration 7:           28 seconds\n",
      "Phasing iteration 8:           27 seconds\n",
      "Phasing iteration 9:           25 seconds\n",
      "Phasing iteration 10:          25 seconds\n",
      "Phasing iteration 11:          22 seconds\n",
      "Phasing iteration 12:          22 seconds\n",
      "\n",
      "Cumulative Statistics:\n",
      "\n",
      "Study markers:          41,897\n",
      "\n",
      "Haplotype phasing time:        11 minutes 6 seconds\n",
      "Total time:                    11 minutes 34 seconds\n",
      "\n",
      "End time: 12:15 PM PST on 16 Dec 2019\n",
      "beagle.28Sep18.793.jar finished\n",
      "\n",
      "\n",
      " -------------------------------------------------------------------------------- \n",
      "          Minimac4 - Fast Imputation Based on State Space Reduction HMM\n",
      " --------------------------------------------------------------------------------\n",
      "           (c) 2014 - Sayantan Das, Christian Fuchsberger, David Hinds\n",
      "                             Mary Kate Wing, Goncalo Abecasis \n",
      "\n",
      " Version: 1.0.2;\n",
      " Built: Mon Sep 30 11:52:22 PDT 2019 by biona001\n",
      "\n",
      " Command Line Options: \n",
      "       Reference Haplotypes : --refHaps [./compare2/haplo_ref.m3vcf.gz],\n",
      "                              --passOnly, --rsid, --referenceEstimates [ON],\n",
      "                              --mapFile [docs/geneticMapFile.b38.map.txt.gz]\n",
      "          Target Haplotypes : --haps [./compare2/target_masked_0.0001.vcf.gz]\n",
      "          Output Parameters : --prefix [./compare2/minimac4.result.0.0001],\n",
      "                              --estimate, --nobgzip, --vcfBuffer [200],\n",
      "                              --format [GT,DS], --allTypedSites, --meta,\n",
      "                              --memUsage\n",
      "        Chunking Parameters : --ChunkLengthMb [20.00], --ChunkOverlapMb [3.00]\n",
      "          Subset Parameters : --chr [], --start, --end, --window\n",
      "   Approximation Parameters : --minimac3, --probThreshold [0.01],\n",
      "                              --diffThreshold [0.01], --topThreshold [0.01]\n",
      "           Other Parameters : --log, --help, --cpus [1], --params\n",
      "                  PhoneHome : --noPhoneHome, --phoneHomeThinning [50]\n",
      "\n",
      "\n",
      " URL = http://genome.sph.umich.edu/wiki/Minimac4\n",
      " Starting Main Imputation/Estimation Analysis ... \n",
      "\n",
      " Performing preliminary check on input parameters... \n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                             PRELIMINARY FILE CHECK                            \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Checking GWAS haplotype file : ./compare2/target_masked_0.0001.vcf.gz\n",
      "\n",
      " Gathering variant information ...\n",
      "\n",
      " Successful !!! \n",
      "\n",
      " Checking Reference haplotype file : ./compare2/haplo_ref.m3vcf.gz\n",
      "\n",
      " Gathering variant information ...\n",
      "\n",
      " Successful !!! \n",
      "\n",
      " Reference Panel   : Found 9000 samples (18000 haplotypes) and 41897 variants ...\n",
      "\n",
      " Target/GWAS Panel : Found 1000 samples (2000 haplotypes) and 41897 variants ...\n",
      "                     41897 variants overlap with Reference panel \n",
      "                     0 variants imported that exist only in Target/GWAS panel\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                           CHUNKING INFORMATION                           \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Chunking region into 1 chunk(s) with atleast 41897 variants in each chunk ... \n",
      "\n",
      " Details of chunks is given below ...\n",
      "\n",
      " No   LeftBuffer      LeftEnd   RightPoint  RightBuffer       #Sites(GWAS/Ref/%)\n",
      " -------------------------------------------------------------------------------\n",
      "  1          244          244      4999928      4999928    41897/   41897/ 100.00%\n",
      "\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                           MAIN IMPUTATION ANALYSIS                            \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Starting imputation analysis of 1 chunk(s) ... \n",
      "\n",
      " -------------------------------------------\n",
      " Analyzing Chunk 1/1 [1:244-4999928]\n",
      " -------------------------------------------\n",
      "\n",
      " Reading chunk from reference panel ... \n",
      " Reading chunk from target/GWAS panel ... \n",
      "\n",
      " Compressing reference panel at GWAS sites ... \n",
      " Re-compression successful (658 seconds) !!!\n",
      "\n",
      " Starting Imputation ...\n",
      "\n",
      "  Imputing Samples 1-200 [20%] out of 1000 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 201-400 [40%] out of 1000 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 401-600 [60%] out of 1000 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 601-800 [80%] out of 1000 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 801-1000 [100%] out of 1000 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "\n",
      " Imputation successful (1845 seconds) !!!\n",
      "\n",
      " Appending chunk to final output VCF File :  ./compare2/minimac4.result.0.0001.dose.vcf.gz \n",
      " Appending successful (3 seconds) !!!\n",
      "\n",
      " Time taken for this chunk = 2515 seconds.\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                              SUMMARY OF ANALYSIS                              \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Info file written to                    : ./compare2/minimac4.result.0.0001.info\n",
      " Imputed VCF information written to      : ./compare2/minimac4.result.0.0001.dose.vcf.gz\n",
      "\n",
      " Time Taken for Reading File             = 14 seconds\n",
      " Time Taken for Re-compression           = 658 seconds\n",
      " Time Taken for Imputation               = 1835 seconds\n",
      " Time Taken for Writing File             = 13 seconds\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                                END OF PROGRAM                                 \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Program Successfully Implemented... \n",
      " \n",
      " Total Run completed in 0 hours, 42 mins, 1 seconds.\n",
      "\n",
      " Thank You for using Minimac4 !!! \n",
      "\n",
      "beagle.28Sep18.793.jar (version 5.0)\n",
      "Copyright (C) 2014-2018 Brian L. Browning\n",
      "Enter \"java -jar beagle.28Sep18.793.jar\" to list command line argument\n",
      "Start time: 01:03 PM PST on 16 Dec 2019\n",
      "\n",
      "Command line: java -Xmx3641m -jar beagle.28Sep18.793.jar\n",
      "  gt=./compare2/target_masked_0.001.vcf.gz\n",
      "  ref=./compare2/haplo_ref.vcf.gz\n",
      "  out=./compare2/beagle.result.0.001\n",
      "\n",
      "No genetic map is specified: using 1 cM = 1 Mb\n",
      "\n",
      "Reference samples:       9,000\n",
      "Study samples:           1,000\n",
      "\n",
      "Window 1 (1:244-4999928)\n",
      "Reference markers:      41,897\n",
      "Study markers:          41,897\n",
      "\n",
      "Burnin  iteration 1:           36 seconds\n",
      "Burnin  iteration 2:           37 seconds\n",
      "Burnin  iteration 3:           37 seconds\n",
      "Burnin  iteration 4:           36 seconds\n",
      "Burnin  iteration 5:           44 seconds\n",
      "Burnin  iteration 6:           44 seconds\n",
      "\n",
      "Phasing iteration 1:           21 seconds\n",
      "Phasing iteration 2:           20 seconds\n",
      "Phasing iteration 3:           24 seconds\n",
      "Phasing iteration 4:           21 seconds\n",
      "Phasing iteration 5:           20 seconds\n",
      "Phasing iteration 6:           20 seconds\n",
      "Phasing iteration 7:           20 seconds\n",
      "Phasing iteration 8:           20 seconds\n",
      "Phasing iteration 9:           20 seconds\n",
      "Phasing iteration 10:          19 seconds\n",
      "Phasing iteration 11:          18 seconds\n",
      "Phasing iteration 12:          18 seconds\n",
      "\n",
      "Cumulative Statistics:\n",
      "\n",
      "Study markers:          41,897\n",
      "\n",
      "Haplotype phasing time:        7 minutes 55 seconds\n",
      "Total time:                    8 minutes 24 seconds\n",
      "\n",
      "End time: 01:11 PM PST on 16 Dec 2019\n",
      "beagle.28Sep18.793.jar finished\n",
      "\n",
      "\n",
      " -------------------------------------------------------------------------------- \n",
      "          Minimac4 - Fast Imputation Based on State Space Reduction HMM\n",
      " --------------------------------------------------------------------------------\n",
      "           (c) 2014 - Sayantan Das, Christian Fuchsberger, David Hinds\n",
      "                             Mary Kate Wing, Goncalo Abecasis \n",
      "\n",
      " Version: 1.0.2;\n",
      " Built: Mon Sep 30 11:52:22 PDT 2019 by biona001\n",
      "\n",
      " Command Line Options: \n",
      "       Reference Haplotypes : --refHaps [./compare2/haplo_ref.m3vcf.gz],\n",
      "                              --passOnly, --rsid, --referenceEstimates [ON],\n",
      "                              --mapFile [docs/geneticMapFile.b38.map.txt.gz]\n",
      "          Target Haplotypes : --haps [./compare2/target_masked_0.001.vcf.gz]\n",
      "          Output Parameters : --prefix [./compare2/minimac4.result.0.001],\n",
      "                              --estimate, --nobgzip, --vcfBuffer [200],\n",
      "                              --format [GT,DS], --allTypedSites, --meta,\n",
      "                              --memUsage\n",
      "        Chunking Parameters : --ChunkLengthMb [20.00], --ChunkOverlapMb [3.00]\n",
      "          Subset Parameters : --chr [], --start, --end, --window\n",
      "   Approximation Parameters : --minimac3, --probThreshold [0.01],\n",
      "                              --diffThreshold [0.01], --topThreshold [0.01]\n",
      "           Other Parameters : --log, --help, --cpus [1], --params\n",
      "                  PhoneHome : --noPhoneHome, --phoneHomeThinning [50]\n",
      "\n",
      "\n",
      " URL = http://genome.sph.umich.edu/wiki/Minimac4\n",
      " Starting Main Imputation/Estimation Analysis ... \n",
      "\n",
      " Performing preliminary check on input parameters... \n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                             PRELIMINARY FILE CHECK                            \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Checking GWAS haplotype file : ./compare2/target_masked_0.001.vcf.gz\n",
      "\n",
      " Gathering variant information ...\n",
      "\n",
      " Successful !!! \n",
      "\n",
      " Checking Reference haplotype file : ./compare2/haplo_ref.m3vcf.gz\n",
      "\n",
      " Gathering variant information ...\n",
      "\n",
      " Successful !!! \n",
      "\n",
      " Reference Panel   : Found 9000 samples (18000 haplotypes) and 41897 variants ...\n",
      "\n",
      " Target/GWAS Panel : Found 1000 samples (2000 haplotypes) and 41897 variants ...\n",
      "                     41897 variants overlap with Reference panel \n",
      "                     0 variants imported that exist only in Target/GWAS panel\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                           CHUNKING INFORMATION                           \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Chunking region into 1 chunk(s) with atleast 41897 variants in each chunk ... \n",
      "\n",
      " Details of chunks is given below ...\n",
      "\n",
      " No   LeftBuffer      LeftEnd   RightPoint  RightBuffer       #Sites(GWAS/Ref/%)\n",
      " -------------------------------------------------------------------------------\n",
      "  1          244          244      4999928      4999928    41897/   41897/ 100.00%\n",
      "\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                           MAIN IMPUTATION ANALYSIS                            \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Starting imputation analysis of 1 chunk(s) ... \n",
      "\n",
      " -------------------------------------------\n",
      " Analyzing Chunk 1/1 [1:244-4999928]\n",
      " -------------------------------------------\n",
      "\n",
      " Reading chunk from reference panel ... \n",
      " Reading chunk from target/GWAS panel ... \n",
      "\n",
      " Compressing reference panel at GWAS sites ... \n",
      " Re-compression successful (680 seconds) !!!\n",
      "\n",
      " Starting Imputation ...\n",
      "\n",
      "  Imputing Samples 1-200 [20%] out of 1000 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 201-400 [40%] out of 1000 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 401-600 [60%] out of 1000 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 601-800 [80%] out of 1000 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 801-1000 [100%] out of 1000 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "\n",
      " Imputation successful (1874 seconds) !!!\n",
      "\n",
      " Appending chunk to final output VCF File :  ./compare2/minimac4.result.0.001.dose.vcf.gz \n",
      " Appending successful (3 seconds) !!!\n",
      "\n",
      " Time taken for this chunk = 2566 seconds.\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                              SUMMARY OF ANALYSIS                              \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Info file written to                    : ./compare2/minimac4.result.0.001.info\n",
      " Imputed VCF information written to      : ./compare2/minimac4.result.0.001.dose.vcf.gz\n",
      "\n",
      " Time Taken for Reading File             = 12 seconds\n",
      " Time Taken for Re-compression           = 680 seconds\n",
      " Time Taken for Imputation               = 1864 seconds\n",
      " Time Taken for Writing File             = 13 seconds\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                                END OF PROGRAM                                 \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Program Successfully Implemented... \n",
      " \n",
      " Total Run completed in 0 hours, 42 mins, 51 seconds.\n",
      "\n",
      " Thank You for using Minimac4 !!! \n",
      "\n",
      "beagle.28Sep18.793.jar (version 5.0)\n",
      "Copyright (C) 2014-2018 Brian L. Browning\n",
      "Enter \"java -jar beagle.28Sep18.793.jar\" to list command line argument\n",
      "Start time: 02:00 PM PST on 16 Dec 2019\n",
      "\n",
      "Command line: java -Xmx3641m -jar beagle.28Sep18.793.jar\n",
      "  gt=./compare2/target_masked_0.01.vcf.gz\n",
      "  ref=./compare2/haplo_ref.vcf.gz\n",
      "  out=./compare2/beagle.result.0.01\n",
      "\n",
      "No genetic map is specified: using 1 cM = 1 Mb\n",
      "\n",
      "Reference samples:       9,000\n",
      "Study samples:           1,000\n",
      "\n",
      "Window 1 (1:244-4999928)\n",
      "Reference markers:      41,897\n",
      "Study markers:          41,897\n",
      "\n",
      "Burnin  iteration 1:           22 seconds\n",
      "Burnin  iteration 2:           36 seconds\n",
      "Burnin  iteration 3:           36 seconds\n",
      "Burnin  iteration 4:           36 seconds\n",
      "Burnin  iteration 5:           40 seconds\n",
      "Burnin  iteration 6:           41 seconds\n",
      "\n",
      "Phasing iteration 1:           21 seconds\n",
      "Phasing iteration 2:           21 seconds\n",
      "Phasing iteration 3:           21 seconds\n",
      "Phasing iteration 4:           19 seconds\n",
      "Phasing iteration 5:           20 seconds\n",
      "Phasing iteration 6:           21 seconds\n",
      "Phasing iteration 7:           21 seconds\n",
      "Phasing iteration 8:           22 seconds\n",
      "Phasing iteration 9:           22 seconds\n",
      "Phasing iteration 10:          23 seconds\n",
      "Phasing iteration 11:          22 seconds\n",
      "Phasing iteration 12:          17 seconds\n",
      "\n",
      "Cumulative Statistics:\n",
      "\n",
      "Study markers:          41,897\n",
      "\n",
      "Haplotype phasing time:        7 minutes 41 seconds\n",
      "Total time:                    8 minutes 7 seconds\n",
      "\n",
      "End time: 02:08 PM PST on 16 Dec 2019\n",
      "beagle.28Sep18.793.jar finished\n",
      "\n",
      "\n",
      " -------------------------------------------------------------------------------- \n",
      "          Minimac4 - Fast Imputation Based on State Space Reduction HMM\n",
      " --------------------------------------------------------------------------------\n",
      "           (c) 2014 - Sayantan Das, Christian Fuchsberger, David Hinds\n",
      "                             Mary Kate Wing, Goncalo Abecasis \n",
      "\n",
      " Version: 1.0.2;\n",
      " Built: Mon Sep 30 11:52:22 PDT 2019 by biona001\n",
      "\n",
      " Command Line Options: \n",
      "       Reference Haplotypes : --refHaps [./compare2/haplo_ref.m3vcf.gz],\n",
      "                              --passOnly, --rsid, --referenceEstimates [ON],\n",
      "                              --mapFile [docs/geneticMapFile.b38.map.txt.gz]\n",
      "          Target Haplotypes : --haps [./compare2/target_masked_0.01.vcf.gz]\n",
      "          Output Parameters : --prefix [./compare2/minimac4.result.0.01],\n",
      "                              --estimate, --nobgzip, --vcfBuffer [200],\n",
      "                              --format [GT,DS], --allTypedSites, --meta,\n",
      "                              --memUsage\n",
      "        Chunking Parameters : --ChunkLengthMb [20.00], --ChunkOverlapMb [3.00]\n",
      "          Subset Parameters : --chr [], --start, --end, --window\n",
      "   Approximation Parameters : --minimac3, --probThreshold [0.01],\n",
      "                              --diffThreshold [0.01], --topThreshold [0.01]\n",
      "           Other Parameters : --log, --help, --cpus [1], --params\n",
      "                  PhoneHome : --noPhoneHome, --phoneHomeThinning [50]\n",
      "\n",
      "\n",
      " URL = http://genome.sph.umich.edu/wiki/Minimac4\n",
      " Starting Main Imputation/Estimation Analysis ... \n",
      "\n",
      " Performing preliminary check on input parameters... \n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                             PRELIMINARY FILE CHECK                            \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Checking GWAS haplotype file : ./compare2/target_masked_0.01.vcf.gz\n",
      "\n",
      " Gathering variant information ...\n",
      "\n",
      " Successful !!! \n",
      "\n",
      " Checking Reference haplotype file : ./compare2/haplo_ref.m3vcf.gz\n",
      "\n",
      " Gathering variant information ...\n",
      "\n",
      " Successful !!! \n",
      "\n",
      " Reference Panel   : Found 9000 samples (18000 haplotypes) and 41897 variants ...\n",
      "\n",
      " Target/GWAS Panel : Found 1000 samples (2000 haplotypes) and 41897 variants ...\n",
      "                     41897 variants overlap with Reference panel \n",
      "                     0 variants imported that exist only in Target/GWAS panel\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                           CHUNKING INFORMATION                           \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Chunking region into 1 chunk(s) with atleast 41897 variants in each chunk ... \n",
      "\n",
      " Details of chunks is given below ...\n",
      "\n",
      " No   LeftBuffer      LeftEnd   RightPoint  RightBuffer       #Sites(GWAS/Ref/%)\n",
      " -------------------------------------------------------------------------------\n",
      "  1          244          244      4999928      4999928    41897/   41897/ 100.00%\n",
      "\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                           MAIN IMPUTATION ANALYSIS                            \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Starting imputation analysis of 1 chunk(s) ... \n",
      "\n",
      " -------------------------------------------\n",
      " Analyzing Chunk 1/1 [1:244-4999928]\n",
      " -------------------------------------------\n",
      "\n",
      " Reading chunk from reference panel ... \n",
      " Reading chunk from target/GWAS panel ... \n",
      "\n",
      " Compressing reference panel at GWAS sites ... \n",
      " Re-compression successful (696 seconds) !!!\n",
      "\n",
      " Starting Imputation ...\n",
      "\n",
      "  Imputing Samples 1-200 [20%] out of 1000 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 201-400 [40%] out of 1000 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 401-600 [60%] out of 1000 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 601-800 [80%] out of 1000 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 801-1000 [100%] out of 1000 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "\n",
      " Imputation successful (1913 seconds) !!!\n",
      "\n",
      " Appending chunk to final output VCF File :  ./compare2/minimac4.result.0.01.dose.vcf.gz \n",
      " Appending successful (4 seconds) !!!\n",
      "\n",
      " Time taken for this chunk = 2622 seconds.\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                              SUMMARY OF ANALYSIS                              \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Info file written to                    : ./compare2/minimac4.result.0.01.info\n",
      " Imputed VCF information written to      : ./compare2/minimac4.result.0.01.dose.vcf.gz\n",
      "\n",
      " Time Taken for Reading File             = 13 seconds\n",
      " Time Taken for Re-compression           = 696 seconds\n",
      " Time Taken for Imputation               = 1904 seconds\n",
      " Time Taken for Writing File             = 13 seconds\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                                END OF PROGRAM                                 \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Program Successfully Implemented... \n",
      " \n",
      " Total Run completed in 0 hours, 43 mins, 47 seconds.\n",
      "\n",
      " Thank You for using Minimac4 !!! \n",
      "\n",
      "beagle.28Sep18.793.jar (version 5.0)\n",
      "Copyright (C) 2014-2018 Brian L. Browning\n",
      "Enter \"java -jar beagle.28Sep18.793.jar\" to list command line argument\n",
      "Start time: 02:59 PM PST on 16 Dec 2019\n",
      "\n",
      "Command line: java -Xmx3641m -jar beagle.28Sep18.793.jar\n",
      "  gt=./compare2/target_masked_0.05.vcf.gz\n",
      "  ref=./compare2/haplo_ref.vcf.gz\n",
      "  out=./compare2/beagle.result.0.05\n",
      "\n",
      "No genetic map is specified: using 1 cM = 1 Mb\n",
      "\n",
      "Reference samples:       9,000\n",
      "Study samples:           1,000\n",
      "\n",
      "Window 1 (1:244-4999928)\n",
      "Reference markers:      41,897\n",
      "Study markers:          41,897\n",
      "\n",
      "Burnin  iteration 1:           9 seconds\n",
      "Burnin  iteration 2:           20 seconds\n",
      "Burnin  iteration 3:           36 seconds\n",
      "Burnin  iteration 4:           38 seconds\n",
      "Burnin  iteration 5:           42 seconds\n",
      "Burnin  iteration 6:           43 seconds\n",
      "\n",
      "Phasing iteration 1:           22 seconds\n",
      "Phasing iteration 2:           22 seconds\n",
      "Phasing iteration 3:           22 seconds\n",
      "Phasing iteration 4:           22 seconds\n",
      "Phasing iteration 5:           22 seconds\n",
      "Phasing iteration 6:           21 seconds\n",
      "Phasing iteration 7:           22 seconds\n",
      "Phasing iteration 8:           22 seconds\n",
      "Phasing iteration 9:           21 seconds\n",
      "Phasing iteration 10:          21 seconds\n",
      "Phasing iteration 11:          20 seconds\n",
      "Phasing iteration 12:          19 seconds\n",
      "\n",
      "Cumulative Statistics:\n",
      "\n",
      "Study markers:          41,897\n",
      "\n",
      "Haplotype phasing time:        7 minutes 26 seconds\n",
      "Total time:                    7 minutes 56 seconds\n",
      "\n",
      "End time: 03:07 PM PST on 16 Dec 2019\n",
      "beagle.28Sep18.793.jar finished\n",
      "\n",
      "\n",
      " -------------------------------------------------------------------------------- \n",
      "          Minimac4 - Fast Imputation Based on State Space Reduction HMM\n",
      " --------------------------------------------------------------------------------\n",
      "           (c) 2014 - Sayantan Das, Christian Fuchsberger, David Hinds\n",
      "                             Mary Kate Wing, Goncalo Abecasis \n",
      "\n",
      " Version: 1.0.2;\n",
      " Built: Mon Sep 30 11:52:22 PDT 2019 by biona001\n",
      "\n",
      " Command Line Options: \n",
      "       Reference Haplotypes : --refHaps [./compare2/haplo_ref.m3vcf.gz],\n",
      "                              --passOnly, --rsid, --referenceEstimates [ON],\n",
      "                              --mapFile [docs/geneticMapFile.b38.map.txt.gz]\n",
      "          Target Haplotypes : --haps [./compare2/target_masked_0.05.vcf.gz]\n",
      "          Output Parameters : --prefix [./compare2/minimac4.result.0.05],\n",
      "                              --estimate, --nobgzip, --vcfBuffer [200],\n",
      "                              --format [GT,DS], --allTypedSites, --meta,\n",
      "                              --memUsage\n",
      "        Chunking Parameters : --ChunkLengthMb [20.00], --ChunkOverlapMb [3.00]\n",
      "          Subset Parameters : --chr [], --start, --end, --window\n",
      "   Approximation Parameters : --minimac3, --probThreshold [0.01],\n",
      "                              --diffThreshold [0.01], --topThreshold [0.01]\n",
      "           Other Parameters : --log, --help, --cpus [1], --params\n",
      "                  PhoneHome : --noPhoneHome, --phoneHomeThinning [50]\n",
      "\n",
      "\n",
      " URL = http://genome.sph.umich.edu/wiki/Minimac4\n",
      " Starting Main Imputation/Estimation Analysis ... \n",
      "\n",
      " Performing preliminary check on input parameters... \n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                             PRELIMINARY FILE CHECK                            \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Checking GWAS haplotype file : ./compare2/target_masked_0.05.vcf.gz\n",
      "\n",
      " Gathering variant information ...\n",
      "\n",
      " Successful !!! \n",
      "\n",
      " Checking Reference haplotype file : ./compare2/haplo_ref.m3vcf.gz\n",
      "\n",
      " Gathering variant information ...\n",
      "\n",
      " Successful !!! \n",
      "\n",
      " Reference Panel   : Found 9000 samples (18000 haplotypes) and 41897 variants ...\n",
      "\n",
      " Target/GWAS Panel : Found 1000 samples (2000 haplotypes) and 41897 variants ...\n",
      "                     41897 variants overlap with Reference panel \n",
      "                     0 variants imported that exist only in Target/GWAS panel\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                           CHUNKING INFORMATION                           \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Chunking region into 1 chunk(s) with atleast 41897 variants in each chunk ... \n",
      "\n",
      " Details of chunks is given below ...\n",
      "\n",
      " No   LeftBuffer      LeftEnd   RightPoint  RightBuffer       #Sites(GWAS/Ref/%)\n",
      " -------------------------------------------------------------------------------\n",
      "  1          244          244      4999928      4999928    41897/   41897/ 100.00%\n",
      "\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                           MAIN IMPUTATION ANALYSIS                            \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Starting imputation analysis of 1 chunk(s) ... \n",
      "\n",
      " -------------------------------------------\n",
      " Analyzing Chunk 1/1 [1:244-4999928]\n",
      " -------------------------------------------\n",
      "\n",
      " Reading chunk from reference panel ... \n",
      " Reading chunk from target/GWAS panel ... \n",
      "\n",
      " Compressing reference panel at GWAS sites ... \n",
      " Re-compression successful (690 seconds) !!!\n",
      "\n",
      " Starting Imputation ...\n",
      "\n",
      "  Imputing Samples 1-200 [20%] out of 1000 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 201-400 [40%] out of 1000 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 401-600 [60%] out of 1000 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 601-800 [80%] out of 1000 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 801-1000 [100%] out of 1000 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "\n",
      " Imputation successful (1922 seconds) !!!\n",
      "\n",
      " Appending chunk to final output VCF File :  ./compare2/minimac4.result.0.05.dose.vcf.gz \n",
      " Appending successful (4 seconds) !!!\n",
      "\n",
      " Time taken for this chunk = 2626 seconds.\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                              SUMMARY OF ANALYSIS                              \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Info file written to                    : ./compare2/minimac4.result.0.05.info\n",
      " Imputed VCF information written to      : ./compare2/minimac4.result.0.05.dose.vcf.gz\n",
      "\n",
      " Time Taken for Reading File             = 14 seconds\n",
      " Time Taken for Re-compression           = 690 seconds\n",
      " Time Taken for Imputation               = 1912 seconds\n",
      " Time Taken for Writing File             = 14 seconds\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                                END OF PROGRAM                                 \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Program Successfully Implemented... \n",
      " \n",
      " Total Run completed in 0 hours, 43 mins, 51 seconds.\n",
      "\n",
      " Thank You for using Minimac4 !!! \n",
      "\n",
      "beagle.28Sep18.793.jar (version 5.0)\n",
      "Copyright (C) 2014-2018 Brian L. Browning\n",
      "Enter \"java -jar beagle.28Sep18.793.jar\" to list command line argument\n",
      "Start time: 03:57 PM PST on 16 Dec 2019\n",
      "\n",
      "Command line: java -Xmx3641m -jar beagle.28Sep18.793.jar\n",
      "  gt=./compare2/target_masked_0.1.vcf.gz\n",
      "  ref=./compare2/haplo_ref.vcf.gz\n",
      "  out=./compare2/beagle.result.0.1\n",
      "\n",
      "No genetic map is specified: using 1 cM = 1 Mb\n",
      "\n",
      "Reference samples:       9,000\n",
      "Study samples:           1,000\n",
      "\n",
      "Window 1 (1:244-4999928)\n",
      "Reference markers:      41,897\n",
      "Study markers:          41,897\n",
      "\n",
      "Burnin  iteration 1:           49 seconds\n",
      "Burnin  iteration 2:           40 seconds\n",
      "Burnin  iteration 3:           22 seconds\n",
      "Burnin  iteration 4:           34 seconds\n",
      "Burnin  iteration 5:           41 seconds\n",
      "Burnin  iteration 6:           39 seconds\n",
      "\n",
      "Phasing iteration 1:           20 seconds\n",
      "Phasing iteration 2:           19 seconds\n",
      "Phasing iteration 3:           19 seconds\n",
      "Phasing iteration 4:           19 seconds\n",
      "Phasing iteration 5:           19 seconds\n",
      "Phasing iteration 6:           19 seconds\n",
      "Phasing iteration 7:           19 seconds\n",
      "Phasing iteration 8:           19 seconds\n",
      "Phasing iteration 9:           19 seconds\n",
      "Phasing iteration 10:          20 seconds\n",
      "Phasing iteration 11:          17 seconds\n",
      "Phasing iteration 12:          17 seconds\n",
      "\n",
      "Cumulative Statistics:\n",
      "\n",
      "Study markers:          41,897\n",
      "\n",
      "Haplotype phasing time:        7 minutes 31 seconds\n",
      "Total time:                    7 minutes 59 seconds\n",
      "\n",
      "End time: 04:05 PM PST on 16 Dec 2019\n",
      "beagle.28Sep18.793.jar finished\n",
      "\n",
      "\n",
      " -------------------------------------------------------------------------------- \n",
      "          Minimac4 - Fast Imputation Based on State Space Reduction HMM\n",
      " --------------------------------------------------------------------------------\n",
      "           (c) 2014 - Sayantan Das, Christian Fuchsberger, David Hinds\n",
      "                             Mary Kate Wing, Goncalo Abecasis \n",
      "\n",
      " Version: 1.0.2;\n",
      " Built: Mon Sep 30 11:52:22 PDT 2019 by biona001\n",
      "\n",
      " Command Line Options: \n",
      "       Reference Haplotypes : --refHaps [./compare2/haplo_ref.m3vcf.gz],\n",
      "                              --passOnly, --rsid, --referenceEstimates [ON],\n",
      "                              --mapFile [docs/geneticMapFile.b38.map.txt.gz]\n",
      "          Target Haplotypes : --haps [./compare2/target_masked_0.1.vcf.gz]\n",
      "          Output Parameters : --prefix [./compare2/minimac4.result.0.1],\n",
      "                              --estimate, --nobgzip, --vcfBuffer [200],\n",
      "                              --format [GT,DS], --allTypedSites, --meta,\n",
      "                              --memUsage\n",
      "        Chunking Parameters : --ChunkLengthMb [20.00], --ChunkOverlapMb [3.00]\n",
      "          Subset Parameters : --chr [], --start, --end, --window\n",
      "   Approximation Parameters : --minimac3, --probThreshold [0.01],\n",
      "                              --diffThreshold [0.01], --topThreshold [0.01]\n",
      "           Other Parameters : --log, --help, --cpus [1], --params\n",
      "                  PhoneHome : --noPhoneHome, --phoneHomeThinning [50]\n",
      "\n",
      "\n",
      " URL = http://genome.sph.umich.edu/wiki/Minimac4\n",
      " Starting Main Imputation/Estimation Analysis ... \n",
      "\n",
      " Performing preliminary check on input parameters... \n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                             PRELIMINARY FILE CHECK                            \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Checking GWAS haplotype file : ./compare2/target_masked_0.1.vcf.gz\n",
      "\n",
      " Gathering variant information ...\n",
      "\n",
      " Successful !!! \n",
      "\n",
      " Checking Reference haplotype file : ./compare2/haplo_ref.m3vcf.gz\n",
      "\n",
      " Gathering variant information ...\n",
      "\n",
      " Successful !!! \n",
      "\n",
      " Reference Panel   : Found 9000 samples (18000 haplotypes) and 41897 variants ...\n",
      "\n",
      " Target/GWAS Panel : Found 1000 samples (2000 haplotypes) and 41897 variants ...\n",
      "                     41897 variants overlap with Reference panel \n",
      "                     0 variants imported that exist only in Target/GWAS panel\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                           CHUNKING INFORMATION                           \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Chunking region into 1 chunk(s) with atleast 41897 variants in each chunk ... \n",
      "\n",
      " Details of chunks is given below ...\n",
      "\n",
      " No   LeftBuffer      LeftEnd   RightPoint  RightBuffer       #Sites(GWAS/Ref/%)\n",
      " -------------------------------------------------------------------------------\n",
      "  1          244          244      4999928      4999928    41897/   41897/ 100.00%\n",
      "\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                           MAIN IMPUTATION ANALYSIS                            \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Starting imputation analysis of 1 chunk(s) ... \n",
      "\n",
      " -------------------------------------------\n",
      " Analyzing Chunk 1/1 [1:244-4999928]\n",
      " -------------------------------------------\n",
      "\n",
      " Reading chunk from reference panel ... \n",
      " Reading chunk from target/GWAS panel ... \n",
      "\n",
      " Compressing reference panel at GWAS sites ... \n",
      " Re-compression successful (708 seconds) !!!\n",
      "\n",
      " Starting Imputation ...\n",
      "\n",
      "  Imputing Samples 1-200 [20%] out of 1000 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 201-400 [40%] out of 1000 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 401-600 [60%] out of 1000 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 601-800 [80%] out of 1000 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 801-1000 [100%] out of 1000 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "\n",
      " Imputation successful (1880 seconds) !!!\n",
      "\n",
      " Appending chunk to final output VCF File :  ./compare2/minimac4.result.0.1.dose.vcf.gz \n",
      " Appending successful (4 seconds) !!!\n",
      "\n",
      " Time taken for this chunk = 2601 seconds.\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                              SUMMARY OF ANALYSIS                              \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Info file written to                    : ./compare2/minimac4.result.0.1.info\n",
      " Imputed VCF information written to      : ./compare2/minimac4.result.0.1.dose.vcf.gz\n",
      "\n",
      " Time Taken for Reading File             = 13 seconds\n",
      " Time Taken for Re-compression           = 708 seconds\n",
      " Time Taken for Imputation               = 1869 seconds\n",
      " Time Taken for Writing File             = 15 seconds\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                                END OF PROGRAM                                 \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Program Successfully Implemented... \n",
      " \n",
      " Total Run completed in 0 hours, 43 mins, 27 seconds.\n",
      "\n",
      " Thank You for using Minimac4 !!! \n",
      "\n",
      "beagle.28Sep18.793.jar (version 5.0)\n",
      "Copyright (C) 2014-2018 Brian L. Browning\n",
      "Enter \"java -jar beagle.28Sep18.793.jar\" to list command line argument\n",
      "Start time: 04:55 PM PST on 16 Dec 2019\n",
      "\n",
      "Command line: java -Xmx3641m -jar beagle.28Sep18.793.jar\n",
      "  gt=./compare2/target_masked_0.25.vcf.gz\n",
      "  ref=./compare2/haplo_ref.vcf.gz\n",
      "  out=./compare2/beagle.result.0.25\n",
      "\n",
      "No genetic map is specified: using 1 cM = 1 Mb\n",
      "\n",
      "Reference samples:       9,000\n",
      "Study samples:           1,000\n",
      "\n",
      "Window 1 (1:244-4999928)\n",
      "Reference markers:      41,897\n",
      "Study markers:          41,897\n",
      "\n",
      "Burnin  iteration 1:           2 minutes 13 seconds\n",
      "Burnin  iteration 2:           2 minutes 7 seconds\n",
      "Burnin  iteration 3:           16 seconds\n",
      "Burnin  iteration 4:           28 seconds\n",
      "Burnin  iteration 5:           38 seconds\n",
      "Burnin  iteration 6:           42 seconds\n",
      "\n",
      "Phasing iteration 1:           23 seconds\n",
      "Phasing iteration 2:           22 seconds\n",
      "Phasing iteration 3:           21 seconds\n",
      "Phasing iteration 4:           22 seconds\n",
      "Phasing iteration 5:           23 seconds\n",
      "Phasing iteration 6:           22 seconds\n",
      "Phasing iteration 7:           21 seconds\n",
      "Phasing iteration 8:           22 seconds\n",
      "Phasing iteration 9:           22 seconds\n",
      "Phasing iteration 10:          21 seconds\n",
      "Phasing iteration 11:          20 seconds\n",
      "Phasing iteration 12:          19 seconds\n",
      "\n",
      "Cumulative Statistics:\n",
      "\n",
      "Study markers:          41,897\n",
      "\n",
      "Haplotype phasing time:        10 minutes 43 seconds\n",
      "Total time:                    11 minutes 10 seconds\n",
      "\n",
      "End time: 05:06 PM PST on 16 Dec 2019\n",
      "beagle.28Sep18.793.jar finished\n",
      "\n",
      "\n",
      " -------------------------------------------------------------------------------- \n",
      "          Minimac4 - Fast Imputation Based on State Space Reduction HMM\n",
      " --------------------------------------------------------------------------------\n",
      "           (c) 2014 - Sayantan Das, Christian Fuchsberger, David Hinds\n",
      "                             Mary Kate Wing, Goncalo Abecasis \n",
      "\n",
      " Version: 1.0.2;\n",
      " Built: Mon Sep 30 11:52:22 PDT 2019 by biona001\n",
      "\n",
      " Command Line Options: \n",
      "       Reference Haplotypes : --refHaps [./compare2/haplo_ref.m3vcf.gz],\n",
      "                              --passOnly, --rsid, --referenceEstimates [ON],\n",
      "                              --mapFile [docs/geneticMapFile.b38.map.txt.gz]\n",
      "          Target Haplotypes : --haps [./compare2/target_masked_0.25.vcf.gz]\n",
      "          Output Parameters : --prefix [./compare2/minimac4.result.0.25],\n",
      "                              --estimate, --nobgzip, --vcfBuffer [200],\n",
      "                              --format [GT,DS], --allTypedSites, --meta,\n",
      "                              --memUsage\n",
      "        Chunking Parameters : --ChunkLengthMb [20.00], --ChunkOverlapMb [3.00]\n",
      "          Subset Parameters : --chr [], --start, --end, --window\n",
      "   Approximation Parameters : --minimac3, --probThreshold [0.01],\n",
      "                              --diffThreshold [0.01], --topThreshold [0.01]\n",
      "           Other Parameters : --log, --help, --cpus [1], --params\n",
      "                  PhoneHome : --noPhoneHome, --phoneHomeThinning [50]\n",
      "\n",
      "\n",
      " URL = http://genome.sph.umich.edu/wiki/Minimac4\n",
      " Starting Main Imputation/Estimation Analysis ... \n",
      "\n",
      " Performing preliminary check on input parameters... \n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                             PRELIMINARY FILE CHECK                            \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Checking GWAS haplotype file : ./compare2/target_masked_0.25.vcf.gz\n",
      "\n",
      " Gathering variant information ...\n",
      "\n",
      " Successful !!! \n",
      "\n",
      " Checking Reference haplotype file : ./compare2/haplo_ref.m3vcf.gz\n",
      "\n",
      " Gathering variant information ...\n",
      "\n",
      " Successful !!! \n",
      "\n",
      " Reference Panel   : Found 9000 samples (18000 haplotypes) and 41897 variants ...\n",
      "\n",
      " Target/GWAS Panel : Found 1000 samples (2000 haplotypes) and 41897 variants ...\n",
      "                     41897 variants overlap with Reference panel \n",
      "                     0 variants imported that exist only in Target/GWAS panel\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                           CHUNKING INFORMATION                           \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Chunking region into 1 chunk(s) with atleast 41897 variants in each chunk ... \n",
      "\n",
      " Details of chunks is given below ...\n",
      "\n",
      " No   LeftBuffer      LeftEnd   RightPoint  RightBuffer       #Sites(GWAS/Ref/%)\n",
      " -------------------------------------------------------------------------------\n",
      "  1          244          244      4999928      4999928    41897/   41897/ 100.00%\n",
      "\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                           MAIN IMPUTATION ANALYSIS                            \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Starting imputation analysis of 1 chunk(s) ... \n",
      "\n",
      " -------------------------------------------\n",
      " Analyzing Chunk 1/1 [1:244-4999928]\n",
      " -------------------------------------------\n",
      "\n",
      " Reading chunk from reference panel ... \n",
      " Reading chunk from target/GWAS panel ... \n",
      "\n",
      " Compressing reference panel at GWAS sites ... \n",
      " Re-compression successful (667 seconds) !!!\n",
      "\n",
      " Starting Imputation ...\n",
      "\n",
      "  Imputing Samples 1-200 [20%] out of 1000 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 201-400 [40%] out of 1000 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 401-600 [60%] out of 1000 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 601-800 [80%] out of 1000 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 801-1000 [100%] out of 1000 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "\n",
      " Imputation successful (1786 seconds) !!!\n",
      "\n",
      " Appending chunk to final output VCF File :  ./compare2/minimac4.result.0.25.dose.vcf.gz \n",
      " Appending successful (4 seconds) !!!\n",
      "\n",
      " Time taken for this chunk = 2467 seconds.\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                              SUMMARY OF ANALYSIS                              \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Info file written to                    : ./compare2/minimac4.result.0.25.info\n",
      " Imputed VCF information written to      : ./compare2/minimac4.result.0.25.dose.vcf.gz\n",
      "\n",
      " Time Taken for Reading File             = 13 seconds\n",
      " Time Taken for Re-compression           = 667 seconds\n",
      " Time Taken for Imputation               = 1776 seconds\n",
      " Time Taken for Writing File             = 14 seconds\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                                END OF PROGRAM                                 \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Program Successfully Implemented... \n",
      " \n",
      " Total Run completed in 0 hours, 41 mins, 12 seconds.\n",
      "\n",
      " Thank You for using Minimac4 !!! \n",
      "\n"
     ]
    },
    {
     "ename": "InterruptException",
     "evalue": "InterruptException:",
     "output_type": "error",
     "traceback": [
      "InterruptException:",
      "",
      "Stacktrace:",
      " [1] Array at ./boot.jl:404 [inlined]",
      " [2] Array at ./boot.jl:423 [inlined]",
      " [3] getindex at ./array.jl:367 [inlined]",
      " [4] GeneticVariation.VCF.Record() at /Users/biona001/.julia/packages/GeneticVariation/r8DAL/src/vcf/record.jl:31",
      " [5] read at /Users/biona001/.julia/packages/BioCore/YBJvb/src/IO.jl:61 [inlined]",
      " [6] copy_ht_as_is!(::Array{Float32,2}, ::GeneticVariation.VCF.Reader) at /Users/biona001/.julia/dev/VCFTools/src/convert.jl:303",
      " [7] #convert_ht#9(::Bool, ::typeof(convert_ht), ::Type{Float32}, ::String) at /Users/biona001/.julia/dev/VCFTools/src/convert.jl:218",
      " [8] #convert_ht at ./none:0 [inlined]",
      " [9] #phase#7(::Bool, ::String, ::Int64, ::typeof(phase), ::String, ::String) at /Users/biona001/.julia/dev/MendelImpute/src/haplotyping.jl:25",
      " [10] (::MendelImpute.var\"#kw##phase\")(::NamedTuple{(:impute, :outfile, :width),Tuple{Bool,String,Int64}}, ::typeof(phase), ::String, ::String) at ./none:0",
      " [11] top-level scope at ./util.jl:288"
     ]
    }
   ],
   "source": [
    "n = length(missingprop)\n",
    "beagle5_error  = zeros(n)\n",
    "beagle5_times  = zeros(n) #sec\n",
    "beagle5_mem    = zeros(n) #MB\n",
    "minimac4_error = zeros(n)\n",
    "minimac4_times = zeros(n)\n",
    "minimac4_mem   = zeros(n)\n",
    "mendel_100_error  = zeros(n)\n",
    "mendel_100_times  = zeros(n) \n",
    "mendel_100_mem    = zeros(n) \n",
    "mendel_400_error  = zeros(n)\n",
    "mendel_400_times  = zeros(n)\n",
    "mendel_400_mem    = zeros(n)\n",
    "mendel_1200_error = zeros(n)\n",
    "mendel_1200_times = zeros(n)\n",
    "mendel_1200_mem   = zeros(n)\n",
    "X_complete = convert_gt(Float32, \"./compare2/target.vcf.gz\"; as_minorallele=false)\n",
    "for (i, prop) in enumerate(missingprop)\n",
    "    #\n",
    "    # current run info\n",
    "    #\n",
    "    X_masked = convert_gt(Float32, \"./compare2/target_masked_$prop.vcf.gz\"; as_minorallele=false)\n",
    "    missing_idx = ismissing.(X_masked)\n",
    "    total_missing = sum(missing_idx)\n",
    "    actual_missing_values = X_complete[missing_idx]  #true values of missing entries\n",
    "    #\n",
    "    # MendelImpute (window = 100)\n",
    "    #\n",
    "    width   = 100\n",
    "    tgtfile = \"./compare2/target_masked_$prop.vcf.gz\"\n",
    "    reffile = \"./compare2/haplo_ref.vcf.gz\"\n",
    "    outfile = \"./compare2/mendel.result.$prop.window$width.vcf.gz\"\n",
    "    mendel_result = @timed phase(tgtfile, reffile, impute=true, outfile = outfile, width = width)\n",
    "    X_mendel = convert_gt(Float32, outfile, as_minorallele=false)\n",
    "    imputed_missing_values = X_mendel[missing_idx] #imputed values of missing entries\n",
    "    mendel_err = sum(actual_missing_values .!= imputed_missing_values) / total_missing\n",
    "    mendel_100_error[i] = mendel_err\n",
    "    mendel_100_times[i] = mendel_result[2] \n",
    "    mendel_100_mem[i]   = mendel_result[3] / 1e6\n",
    "    #\n",
    "    # MendelImpute (window = 400)\n",
    "    #\n",
    "    width   = 400\n",
    "    tgtfile = \"./compare2/target_masked_$prop.vcf.gz\"\n",
    "    reffile = \"./compare2/haplo_ref.vcf.gz\"\n",
    "    outfile = \"./compare2/mendel.result.$prop.window$width.vcf.gz\"\n",
    "    mendel_result = @timed phase(tgtfile, reffile, impute=true, outfile = outfile, width = width)\n",
    "    X_mendel = convert_gt(Float32, outfile, as_minorallele=false)\n",
    "    imputed_missing_values = X_mendel[missing_idx] #imputed values of missing entries\n",
    "    mendel_err = sum(actual_missing_values .!= imputed_missing_values) / total_missing\n",
    "    mendel_400_error[i] = mendel_err\n",
    "    mendel_400_times[i] = mendel_result[2] \n",
    "    mendel_400_mem[i]   = mendel_result[3] / 1e6\n",
    "    #\n",
    "    # MendelImpute (window = 1200)\n",
    "    #\n",
    "    width   = 1200\n",
    "    tgtfile = \"./compare2/target_masked_$prop.vcf.gz\"\n",
    "    reffile = \"./compare2/haplo_ref.vcf.gz\"\n",
    "    outfile = \"./compare2/mendel.result.$prop.window$width.vcf.gz\"\n",
    "    mendel_result = @timed phase(tgtfile, reffile, impute=true, outfile = outfile, width = width)\n",
    "    X_mendel = convert_gt(Float32, outfile, as_minorallele=false)\n",
    "    imputed_missing_values = X_mendel[missing_idx] #imputed values of missing entries\n",
    "    mendel_err = sum(actual_missing_values .!= imputed_missing_values) / total_missing\n",
    "    mendel_1200_error[i] = mendel_err\n",
    "    mendel_1200_times[i] = mendel_result[2] \n",
    "    mendel_1200_mem[i]   = mendel_result[3] / 1e6\n",
    "    #\n",
    "    # Beagle 5\n",
    "    #\n",
    "    beagle_result = @timed run(`java -jar beagle.28Sep18.793.jar gt=./compare2/target_masked_$prop.vcf.gz ref=./compare2/haplo_ref.vcf.gz out=./compare2/beagle.result.$prop`)\n",
    "    X_beagle = convert_gt(Float32, \"./compare2/beagle.result.$prop.vcf.gz\", as_minorallele=false)\n",
    "    imputed_missing_values = X_beagle[missing_idx] #imputed values of missing entries\n",
    "    beagle_err = sum(actual_missing_values .!= imputed_missing_values) / total_missing\n",
    "    beagle5_error[i] = beagle_err\n",
    "    beagle5_times[i] = beagle_result[2]\n",
    "    beagle5_mem[i]   = beagle_result[3] / 1e6\n",
    "    #\n",
    "    # minimac 4\n",
    "    #\n",
    "    minimac4 = \"/Users/biona001/Benjamin_Folder/UCLA/research/softwares/Minimac4/build/minimac4\"\n",
    "    minimac_result = @timed run(`$minimac4 --refHaps ./compare2/haplo_ref.m3vcf.gz --haps ./compare2/target_masked_$prop.vcf.gz --prefix ./compare2/minimac4.result.$prop`)\n",
    "    X_minimac = convert_gt(Float32, \"./compare2/minimac4.result.$prop.dose.vcf.gz\", as_minorallele=false)\n",
    "    imputed_missing_values = X_minimac[missing_idx] #imputed values of missing entries\n",
    "    minimac_err = sum(actual_missing_values .!= imputed_missing_values) / total_missing\n",
    "    minimac4_error[i] = minimac_err\n",
    "    minimac4_times[i] = minimac_result[2]\n",
    "    minimac4_mem[i]   = minimac_result[3] / 1e6\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "7×6 Array{Float64,2}:\n",
       " 0.0001  0.00594389  0.000713267  0.000713267  0.000237756  0.0       \n",
       " 0.001   0.00558965  0.000999001  0.000309215  2.37857e-5   2.37857e-5\n",
       " 0.01    0.00575938  0.00088422   0.000477957  4.54059e-5   3.82365e-5\n",
       " 0.05    0.00728215  0.00132082   0.000639891  4.15141e-5   3.67424e-5\n",
       " 0.1     0.00926913  0.00188571   0.000855862  4.31988e-5   3.53228e-5\n",
       " 0.25    0.0177278   0.00479307   0.00206166   0.000203837  3.75212e-5\n",
       " 0.5     0.0         0.0          0.0          0.0          0.0       "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "missingprop = [0.0001, 0.001, 0.01, 0.05, 0.1, 0.25, 0.5]\n",
    "[missingprop mendel_100_error mendel_400_error mendel_1200_error beagle5_error minimac4_error]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "7×6 Array{Float64,2}:\n",
       " 0.0001  112.524  100.581   101.083  695.014  2522.29\n",
       " 0.001   115.213   98.3722  100.386  504.123  2572.75\n",
       " 0.01    113.035   99.6651  106.808  487.818  2628.65\n",
       " 0.05    126.799  116.642   115.961  476.18   2633.13\n",
       " 0.1     125.142  113.851   114.97   479.562  2609.16\n",
       " 0.25    118.378  110.692   114.12   670.077  2473.97\n",
       " 0.5       0.0      0.0       0.0      0.0       0.0 "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[missingprop mendel_100_times mendel_400_times mendel_1200_times beagle5_times minimac4_times]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "@webio": {
   "lastCommId": null,
   "lastKernelId": null
  },
  "kernelspec": {
   "display_name": "Julia 1.3.0-rc4",
   "language": "julia",
   "name": "julia-1.3"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.3.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
