{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Compare MendelImpute against Minimac4 and Beagle5 on simulated data\n",
    "\n",
    "Using different % missing and seed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "┌ Info: Precompiling Revise [295af30f-e4ad-537b-8983-00126c2a3abe]\n",
      "└ @ Base loading.jl:1273\n",
      "┌ Warning: no Manifest.toml file found, static paths used\n",
      "└ @ Revise /Users/biona001/.julia/packages/Revise/439di/src/Revise.jl:1108\n",
      "┌ Info: Precompiling VCFTools [a620830f-fdd7-5ebc-8d26-3621ab35fbfe]\n",
      "└ @ Base loading.jl:1273\n",
      "┌ Info: Precompiling MendelImpute [e47305d1-6a61-5370-bc5d-77554d143183]\n",
      "└ @ Base loading.jl:1273\n",
      "┌ Info: Precompiling Plots [91a5bcdd-55d7-5caf-9e0b-520d859cae80]\n",
      "└ @ Base loading.jl:1273\n"
     ]
    }
   ],
   "source": [
    "using Revise\n",
    "using VCFTools\n",
    "using MendelImpute\n",
    "using GeneticVariation\n",
    "using Random\n",
    "using SparseArrays\n",
    "using Plots\n",
    "using StatsPlots\n",
    "using StatsBase"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Simulate data\n",
    "\n",
    "### Step 0. Install `msprime`\n",
    "\n",
    "[msprime download Link](https://msprime.readthedocs.io/en/stable/installation.html).\n",
    "\n",
    "Some people might need to activate conda environment via `conda config --set auto_activate_base True`. You can turn it off once simulation is done by executing `conda config --set auto_activate_base False`.\n",
    "\n",
    "\n",
    "### Step 1. Simulate data in terminal\n",
    "\n",
    "```\n",
    "python3 msprime_script.py 20000 10000 5000000 2e-8 2e-8 1111 > ./compare2/full.vcf\n",
    "```\n",
    "\n",
    "Arguments: \n",
    "+ Number of haplotypes = 20000\n",
    "+ Effective population size = 10000 ([source](https://www.the-scientist.com/the-nutshell/ancient-humans-more-diverse-43556))\n",
    "+ Sequence length = 10 million (same as Beagle 5's choice)\n",
    "+ Rrecombination rate = 2e-8 (default)\n",
    "+ mutation rate = 2e-8 (default)\n",
    "+ seed = 1111"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(41897, 10000)"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "records, samples = nrecords(\"./compare2/full.vcf\"), nsamples(\"./compare2/full.vcf\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Step 2: Convert simulated haplotypes to reference haplotypes and target genotype files\n",
    "\n",
    "+ `haplo_ref.vcf.gz`: haplotype reference files\n",
    "+ `target.vcf.gz`: complete genotype information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "129.700641 seconds (1.47 G allocations: 135.227 GiB, 13.06% gc time)\n",
      "249.543273 seconds (3.14 G allocations: 242.252 GiB, 14.38% gc time)\n"
     ]
    }
   ],
   "source": [
    "Random.seed!(2019)\n",
    "# randomly choose 1000 samples as imputation targets. \n",
    "records, samples = nrecords(\"./compare2/full.vcf\"), nsamples(\"./compare2/full.vcf\")\n",
    "tgt_index = falses(samples)\n",
    "tgt_index[sample(1:samples, 1000, replace = false)] .= true\n",
    "ref_index = .!tgt_index\n",
    "record_index = trues(records) # save all records (SNPs) \n",
    "\n",
    "# create target.vcf.gz and haplo_ref.vcf.gz\n",
    "tgtfile = \"./compare2/target.vcf.gz\"\n",
    "reffile = \"./compare2/haplo_ref.vcf.gz\"\n",
    "@time VCFTools.filter(\"./compare2/full.vcf\", record_index, tgt_index, des = tgtfile)\n",
    "@time VCFTools.filter(\"./compare2/full.vcf\", record_index, ref_index, des = reffile)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Error rate as a function of % masked\n",
    "\n",
    "`full.vcf` contains the full data with 20000 haplotypes. We randomly choose 1000 samples as imputation target. The remaining haplotypes become reference panels. In the imputation target file, we randomly mask 0.01%, 0.1%, 1%, 5%, 10%, 25%, 50% data and compare compare the speed, memory, and imputation accuracy of MendelImpute, Minimac4, and Beagle5.\n",
    "\n",
    "### Mask 0.01%, 0.1%, 1%, 5%, 10%, 25%, 50% entries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  5.314990 seconds (84.72 M allocations: 7.291 GiB, 16.36% gc time)\n"
     ]
    }
   ],
   "source": [
    "# import full target matrix. Also transpose so that columns are samples. \n",
    "@time X = convert_gt(Float32, tgtfile; as_minorallele=false)\n",
    "X = copy(X')\n",
    "p, n = size(X)\n",
    "missingprop = [0.0001, 0.001, 0.01, 0.05, 0.1, 0.25, 0.5]\n",
    "for i in missingprop\n",
    "    # mask entries and save result as VCF file. \n",
    "    Random.seed!(2019)\n",
    "    X .= ifelse.(rand(Float32, p, n) .< i, missing, X)\n",
    "    masks = ismissing.(X)\n",
    "    mask_gt(tgtfile, masks, des=\"./compare2/target_masked_$i.vcf.gz\")\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Convert reference haplotypes from vcf to m3vcf on Hoffman (Julia 1.2)\n",
    "\n",
    "This is needed because Minimac4 only runs their own M3VCF files (which requires Minimac3)\n",
    "\n",
    "```Julia\n",
    "minimac3 = \"/u/home/b/biona001/haplotype_comparisons/minimac3/Minimac3/bin/Minimac3\"\n",
    "@time run(`$minimac3 --refHaps haplo_ref.vcf.gz --processReference --prefix haplo_ref`)\n",
    "```\n",
    "Time = 1452.914532 seconds (15.31 k allocations: 861.069 KiB)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Analyze:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "beagle.28Sep18.793.jar (version 5.0)\n",
      "Copyright (C) 2014-2018 Brian L. Browning\n",
      "Enter \"java -jar beagle.28Sep18.793.jar\" to list command line argument\n",
      "Start time: 12:04 PM PST on 16 Dec 2019\n",
      "\n",
      "Command line: java -Xmx3641m -jar beagle.28Sep18.793.jar\n",
      "  gt=./compare2/target_masked_0.0001.vcf.gz\n",
      "  ref=./compare2/haplo_ref.vcf.gz\n",
      "  out=./compare2/beagle.result.0.0001\n",
      "\n",
      "No genetic map is specified: using 1 cM = 1 Mb\n",
      "\n",
      "Reference samples:       9,000\n",
      "Study samples:           1,000\n",
      "\n",
      "Window 1 (1:244-4999928)\n",
      "Reference markers:      41,897\n",
      "Study markers:          41,897\n",
      "\n",
      "Burnin  iteration 1:           54 seconds\n",
      "Burnin  iteration 2:           1 minute 11 seconds\n",
      "Burnin  iteration 3:           51 seconds\n",
      "Burnin  iteration 4:           52 seconds\n",
      "Burnin  iteration 5:           1 minute 1 second\n",
      "Burnin  iteration 6:           1 minute 1 second\n",
      "\n",
      "Phasing iteration 1:           28 seconds\n",
      "Phasing iteration 2:           28 seconds\n",
      "Phasing iteration 3:           28 seconds\n",
      "Phasing iteration 4:           27 seconds\n",
      "Phasing iteration 5:           27 seconds\n",
      "Phasing iteration 6:           27 seconds\n",
      "Phasing iteration 7:           28 seconds\n",
      "Phasing iteration 8:           27 seconds\n",
      "Phasing iteration 9:           25 seconds\n",
      "Phasing iteration 10:          25 seconds\n",
      "Phasing iteration 11:          22 seconds\n",
      "Phasing iteration 12:          22 seconds\n",
      "\n",
      "Cumulative Statistics:\n",
      "\n",
      "Study markers:          41,897\n",
      "\n",
      "Haplotype phasing time:        11 minutes 6 seconds\n",
      "Total time:                    11 minutes 34 seconds\n",
      "\n",
      "End time: 12:15 PM PST on 16 Dec 2019\n",
      "beagle.28Sep18.793.jar finished\n",
      "\n",
      "\n",
      " -------------------------------------------------------------------------------- \n",
      "          Minimac4 - Fast Imputation Based on State Space Reduction HMM\n",
      " --------------------------------------------------------------------------------\n",
      "           (c) 2014 - Sayantan Das, Christian Fuchsberger, David Hinds\n",
      "                             Mary Kate Wing, Goncalo Abecasis \n",
      "\n",
      " Version: 1.0.2;\n",
      " Built: Mon Sep 30 11:52:22 PDT 2019 by biona001\n",
      "\n",
      " Command Line Options: \n",
      "       Reference Haplotypes : --refHaps [./compare2/haplo_ref.m3vcf.gz],\n",
      "                              --passOnly, --rsid, --referenceEstimates [ON],\n",
      "                              --mapFile [docs/geneticMapFile.b38.map.txt.gz]\n",
      "          Target Haplotypes : --haps [./compare2/target_masked_0.0001.vcf.gz]\n",
      "          Output Parameters : --prefix [./compare2/minimac4.result.0.0001],\n",
      "                              --estimate, --nobgzip, --vcfBuffer [200],\n",
      "                              --format [GT,DS], --allTypedSites, --meta,\n",
      "                              --memUsage\n",
      "        Chunking Parameters : --ChunkLengthMb [20.00], --ChunkOverlapMb [3.00]\n",
      "          Subset Parameters : --chr [], --start, --end, --window\n",
      "   Approximation Parameters : --minimac3, --probThreshold [0.01],\n",
      "                              --diffThreshold [0.01], --topThreshold [0.01]\n",
      "           Other Parameters : --log, --help, --cpus [1], --params\n",
      "                  PhoneHome : --noPhoneHome, --phoneHomeThinning [50]\n",
      "\n",
      "\n",
      " URL = http://genome.sph.umich.edu/wiki/Minimac4\n",
      " Starting Main Imputation/Estimation Analysis ... \n",
      "\n",
      " Performing preliminary check on input parameters... \n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                             PRELIMINARY FILE CHECK                            \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Checking GWAS haplotype file : ./compare2/target_masked_0.0001.vcf.gz\n",
      "\n",
      " Gathering variant information ...\n",
      "\n",
      " Successful !!! \n",
      "\n",
      " Checking Reference haplotype file : ./compare2/haplo_ref.m3vcf.gz\n",
      "\n",
      " Gathering variant information ...\n",
      "\n",
      " Successful !!! \n",
      "\n",
      " Reference Panel   : Found 9000 samples (18000 haplotypes) and 41897 variants ...\n",
      "\n",
      " Target/GWAS Panel : Found 1000 samples (2000 haplotypes) and 41897 variants ...\n",
      "                     41897 variants overlap with Reference panel \n",
      "                     0 variants imported that exist only in Target/GWAS panel\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                           CHUNKING INFORMATION                           \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Chunking region into 1 chunk(s) with atleast 41897 variants in each chunk ... \n",
      "\n",
      " Details of chunks is given below ...\n",
      "\n",
      " No   LeftBuffer      LeftEnd   RightPoint  RightBuffer       #Sites(GWAS/Ref/%)\n",
      " -------------------------------------------------------------------------------\n",
      "  1          244          244      4999928      4999928    41897/   41897/ 100.00%\n",
      "\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                           MAIN IMPUTATION ANALYSIS                            \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Starting imputation analysis of 1 chunk(s) ... \n",
      "\n",
      " -------------------------------------------\n",
      " Analyzing Chunk 1/1 [1:244-4999928]\n",
      " -------------------------------------------\n",
      "\n",
      " Reading chunk from reference panel ... \n",
      " Reading chunk from target/GWAS panel ... \n",
      "\n",
      " Compressing reference panel at GWAS sites ... \n",
      " Re-compression successful (658 seconds) !!!\n",
      "\n",
      " Starting Imputation ...\n",
      "\n",
      "  Imputing Samples 1-200 [20%] out of 1000 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 201-400 [40%] out of 1000 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 401-600 [60%] out of 1000 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 601-800 [80%] out of 1000 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 801-1000 [100%] out of 1000 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "\n",
      " Imputation successful (1845 seconds) !!!\n",
      "\n",
      " Appending chunk to final output VCF File :  ./compare2/minimac4.result.0.0001.dose.vcf.gz \n",
      " Appending successful (3 seconds) !!!\n",
      "\n",
      " Time taken for this chunk = 2515 seconds.\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                              SUMMARY OF ANALYSIS                              \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Info file written to                    : ./compare2/minimac4.result.0.0001.info\n",
      " Imputed VCF information written to      : ./compare2/minimac4.result.0.0001.dose.vcf.gz\n",
      "\n",
      " Time Taken for Reading File             = 14 seconds\n",
      " Time Taken for Re-compression           = 658 seconds\n",
      " Time Taken for Imputation               = 1835 seconds\n",
      " Time Taken for Writing File             = 13 seconds\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                                END OF PROGRAM                                 \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Program Successfully Implemented... \n",
      " \n",
      " Total Run completed in 0 hours, 42 mins, 1 seconds.\n",
      "\n",
      " Thank You for using Minimac4 !!! \n",
      "\n",
      "beagle.28Sep18.793.jar (version 5.0)\n",
      "Copyright (C) 2014-2018 Brian L. Browning\n",
      "Enter \"java -jar beagle.28Sep18.793.jar\" to list command line argument\n",
      "Start time: 01:03 PM PST on 16 Dec 2019\n",
      "\n",
      "Command line: java -Xmx3641m -jar beagle.28Sep18.793.jar\n",
      "  gt=./compare2/target_masked_0.001.vcf.gz\n",
      "  ref=./compare2/haplo_ref.vcf.gz\n",
      "  out=./compare2/beagle.result.0.001\n",
      "\n",
      "No genetic map is specified: using 1 cM = 1 Mb\n",
      "\n",
      "Reference samples:       9,000\n",
      "Study samples:           1,000\n",
      "\n",
      "Window 1 (1:244-4999928)\n",
      "Reference markers:      41,897\n",
      "Study markers:          41,897\n",
      "\n",
      "Burnin  iteration 1:           36 seconds\n",
      "Burnin  iteration 2:           37 seconds\n",
      "Burnin  iteration 3:           37 seconds\n",
      "Burnin  iteration 4:           36 seconds\n",
      "Burnin  iteration 5:           44 seconds\n",
      "Burnin  iteration 6:           44 seconds\n",
      "\n",
      "Phasing iteration 1:           21 seconds\n",
      "Phasing iteration 2:           20 seconds\n",
      "Phasing iteration 3:           24 seconds\n",
      "Phasing iteration 4:           21 seconds\n",
      "Phasing iteration 5:           20 seconds\n",
      "Phasing iteration 6:           20 seconds\n",
      "Phasing iteration 7:           20 seconds\n",
      "Phasing iteration 8:           20 seconds\n",
      "Phasing iteration 9:           20 seconds\n",
      "Phasing iteration 10:          19 seconds\n",
      "Phasing iteration 11:          18 seconds\n",
      "Phasing iteration 12:          18 seconds\n",
      "\n",
      "Cumulative Statistics:\n",
      "\n",
      "Study markers:          41,897\n",
      "\n",
      "Haplotype phasing time:        7 minutes 55 seconds\n",
      "Total time:                    8 minutes 24 seconds\n",
      "\n",
      "End time: 01:11 PM PST on 16 Dec 2019\n",
      "beagle.28Sep18.793.jar finished\n",
      "\n",
      "\n",
      " -------------------------------------------------------------------------------- \n",
      "          Minimac4 - Fast Imputation Based on State Space Reduction HMM\n",
      " --------------------------------------------------------------------------------\n",
      "           (c) 2014 - Sayantan Das, Christian Fuchsberger, David Hinds\n",
      "                             Mary Kate Wing, Goncalo Abecasis \n",
      "\n",
      " Version: 1.0.2;\n",
      " Built: Mon Sep 30 11:52:22 PDT 2019 by biona001\n",
      "\n",
      " Command Line Options: \n",
      "       Reference Haplotypes : --refHaps [./compare2/haplo_ref.m3vcf.gz],\n",
      "                              --passOnly, --rsid, --referenceEstimates [ON],\n",
      "                              --mapFile [docs/geneticMapFile.b38.map.txt.gz]\n",
      "          Target Haplotypes : --haps [./compare2/target_masked_0.001.vcf.gz]\n",
      "          Output Parameters : --prefix [./compare2/minimac4.result.0.001],\n",
      "                              --estimate, --nobgzip, --vcfBuffer [200],\n",
      "                              --format [GT,DS], --allTypedSites, --meta,\n",
      "                              --memUsage\n",
      "        Chunking Parameters : --ChunkLengthMb [20.00], --ChunkOverlapMb [3.00]\n",
      "          Subset Parameters : --chr [], --start, --end, --window\n",
      "   Approximation Parameters : --minimac3, --probThreshold [0.01],\n",
      "                              --diffThreshold [0.01], --topThreshold [0.01]\n",
      "           Other Parameters : --log, --help, --cpus [1], --params\n",
      "                  PhoneHome : --noPhoneHome, --phoneHomeThinning [50]\n",
      "\n",
      "\n",
      " URL = http://genome.sph.umich.edu/wiki/Minimac4\n",
      " Starting Main Imputation/Estimation Analysis ... \n",
      "\n",
      " Performing preliminary check on input parameters... \n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                             PRELIMINARY FILE CHECK                            \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Checking GWAS haplotype file : ./compare2/target_masked_0.001.vcf.gz\n",
      "\n",
      " Gathering variant information ...\n",
      "\n",
      " Successful !!! \n",
      "\n",
      " Checking Reference haplotype file : ./compare2/haplo_ref.m3vcf.gz\n",
      "\n",
      " Gathering variant information ...\n",
      "\n",
      " Successful !!! \n",
      "\n",
      " Reference Panel   : Found 9000 samples (18000 haplotypes) and 41897 variants ...\n",
      "\n",
      " Target/GWAS Panel : Found 1000 samples (2000 haplotypes) and 41897 variants ...\n",
      "                     41897 variants overlap with Reference panel \n",
      "                     0 variants imported that exist only in Target/GWAS panel\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                           CHUNKING INFORMATION                           \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Chunking region into 1 chunk(s) with atleast 41897 variants in each chunk ... \n",
      "\n",
      " Details of chunks is given below ...\n",
      "\n",
      " No   LeftBuffer      LeftEnd   RightPoint  RightBuffer       #Sites(GWAS/Ref/%)\n",
      " -------------------------------------------------------------------------------\n",
      "  1          244          244      4999928      4999928    41897/   41897/ 100.00%\n",
      "\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                           MAIN IMPUTATION ANALYSIS                            \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Starting imputation analysis of 1 chunk(s) ... \n",
      "\n",
      " -------------------------------------------\n",
      " Analyzing Chunk 1/1 [1:244-4999928]\n",
      " -------------------------------------------\n",
      "\n",
      " Reading chunk from reference panel ... \n",
      " Reading chunk from target/GWAS panel ... \n",
      "\n",
      " Compressing reference panel at GWAS sites ... \n",
      " Re-compression successful (680 seconds) !!!\n",
      "\n",
      " Starting Imputation ...\n",
      "\n",
      "  Imputing Samples 1-200 [20%] out of 1000 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 201-400 [40%] out of 1000 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 401-600 [60%] out of 1000 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 601-800 [80%] out of 1000 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 801-1000 [100%] out of 1000 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "\n",
      " Imputation successful (1874 seconds) !!!\n",
      "\n",
      " Appending chunk to final output VCF File :  ./compare2/minimac4.result.0.001.dose.vcf.gz \n",
      " Appending successful (3 seconds) !!!\n",
      "\n",
      " Time taken for this chunk = 2566 seconds.\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                              SUMMARY OF ANALYSIS                              \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Info file written to                    : ./compare2/minimac4.result.0.001.info\n",
      " Imputed VCF information written to      : ./compare2/minimac4.result.0.001.dose.vcf.gz\n",
      "\n",
      " Time Taken for Reading File             = 12 seconds\n",
      " Time Taken for Re-compression           = 680 seconds\n",
      " Time Taken for Imputation               = 1864 seconds\n",
      " Time Taken for Writing File             = 13 seconds\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                                END OF PROGRAM                                 \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Program Successfully Implemented... \n",
      " \n",
      " Total Run completed in 0 hours, 42 mins, 51 seconds.\n",
      "\n",
      " Thank You for using Minimac4 !!! \n",
      "\n",
      "beagle.28Sep18.793.jar (version 5.0)\n",
      "Copyright (C) 2014-2018 Brian L. Browning\n",
      "Enter \"java -jar beagle.28Sep18.793.jar\" to list command line argument\n",
      "Start time: 02:00 PM PST on 16 Dec 2019\n",
      "\n",
      "Command line: java -Xmx3641m -jar beagle.28Sep18.793.jar\n",
      "  gt=./compare2/target_masked_0.01.vcf.gz\n",
      "  ref=./compare2/haplo_ref.vcf.gz\n",
      "  out=./compare2/beagle.result.0.01\n",
      "\n",
      "No genetic map is specified: using 1 cM = 1 Mb\n",
      "\n",
      "Reference samples:       9,000\n",
      "Study samples:           1,000\n",
      "\n",
      "Window 1 (1:244-4999928)\n",
      "Reference markers:      41,897\n",
      "Study markers:          41,897\n",
      "\n",
      "Burnin  iteration 1:           22 seconds\n",
      "Burnin  iteration 2:           36 seconds\n",
      "Burnin  iteration 3:           36 seconds\n",
      "Burnin  iteration 4:           36 seconds\n",
      "Burnin  iteration 5:           40 seconds\n",
      "Burnin  iteration 6:           41 seconds\n",
      "\n",
      "Phasing iteration 1:           21 seconds\n",
      "Phasing iteration 2:           21 seconds\n",
      "Phasing iteration 3:           21 seconds\n",
      "Phasing iteration 4:           19 seconds\n",
      "Phasing iteration 5:           20 seconds\n",
      "Phasing iteration 6:           21 seconds\n",
      "Phasing iteration 7:           21 seconds\n",
      "Phasing iteration 8:           22 seconds\n",
      "Phasing iteration 9:           22 seconds\n",
      "Phasing iteration 10:          23 seconds\n",
      "Phasing iteration 11:          22 seconds\n",
      "Phasing iteration 12:          17 seconds\n",
      "\n",
      "Cumulative Statistics:\n",
      "\n",
      "Study markers:          41,897\n",
      "\n",
      "Haplotype phasing time:        7 minutes 41 seconds\n",
      "Total time:                    8 minutes 7 seconds\n",
      "\n",
      "End time: 02:08 PM PST on 16 Dec 2019\n",
      "beagle.28Sep18.793.jar finished\n",
      "\n",
      "\n",
      " -------------------------------------------------------------------------------- \n",
      "          Minimac4 - Fast Imputation Based on State Space Reduction HMM\n",
      " --------------------------------------------------------------------------------\n",
      "           (c) 2014 - Sayantan Das, Christian Fuchsberger, David Hinds\n",
      "                             Mary Kate Wing, Goncalo Abecasis \n",
      "\n",
      " Version: 1.0.2;\n",
      " Built: Mon Sep 30 11:52:22 PDT 2019 by biona001\n",
      "\n",
      " Command Line Options: \n",
      "       Reference Haplotypes : --refHaps [./compare2/haplo_ref.m3vcf.gz],\n",
      "                              --passOnly, --rsid, --referenceEstimates [ON],\n",
      "                              --mapFile [docs/geneticMapFile.b38.map.txt.gz]\n",
      "          Target Haplotypes : --haps [./compare2/target_masked_0.01.vcf.gz]\n",
      "          Output Parameters : --prefix [./compare2/minimac4.result.0.01],\n",
      "                              --estimate, --nobgzip, --vcfBuffer [200],\n",
      "                              --format [GT,DS], --allTypedSites, --meta,\n",
      "                              --memUsage\n",
      "        Chunking Parameters : --ChunkLengthMb [20.00], --ChunkOverlapMb [3.00]\n",
      "          Subset Parameters : --chr [], --start, --end, --window\n",
      "   Approximation Parameters : --minimac3, --probThreshold [0.01],\n",
      "                              --diffThreshold [0.01], --topThreshold [0.01]\n",
      "           Other Parameters : --log, --help, --cpus [1], --params\n",
      "                  PhoneHome : --noPhoneHome, --phoneHomeThinning [50]\n",
      "\n",
      "\n",
      " URL = http://genome.sph.umich.edu/wiki/Minimac4\n",
      " Starting Main Imputation/Estimation Analysis ... \n",
      "\n",
      " Performing preliminary check on input parameters... \n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                             PRELIMINARY FILE CHECK                            \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Checking GWAS haplotype file : ./compare2/target_masked_0.01.vcf.gz\n",
      "\n",
      " Gathering variant information ...\n",
      "\n",
      " Successful !!! \n",
      "\n",
      " Checking Reference haplotype file : ./compare2/haplo_ref.m3vcf.gz\n",
      "\n",
      " Gathering variant information ...\n",
      "\n",
      " Successful !!! \n",
      "\n",
      " Reference Panel   : Found 9000 samples (18000 haplotypes) and 41897 variants ...\n",
      "\n",
      " Target/GWAS Panel : Found 1000 samples (2000 haplotypes) and 41897 variants ...\n",
      "                     41897 variants overlap with Reference panel \n",
      "                     0 variants imported that exist only in Target/GWAS panel\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                           CHUNKING INFORMATION                           \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Chunking region into 1 chunk(s) with atleast 41897 variants in each chunk ... \n",
      "\n",
      " Details of chunks is given below ...\n",
      "\n",
      " No   LeftBuffer      LeftEnd   RightPoint  RightBuffer       #Sites(GWAS/Ref/%)\n",
      " -------------------------------------------------------------------------------\n",
      "  1          244          244      4999928      4999928    41897/   41897/ 100.00%\n",
      "\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                           MAIN IMPUTATION ANALYSIS                            \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Starting imputation analysis of 1 chunk(s) ... \n",
      "\n",
      " -------------------------------------------\n",
      " Analyzing Chunk 1/1 [1:244-4999928]\n",
      " -------------------------------------------\n",
      "\n",
      " Reading chunk from reference panel ... \n",
      " Reading chunk from target/GWAS panel ... \n",
      "\n",
      " Compressing reference panel at GWAS sites ... \n",
      " Re-compression successful (696 seconds) !!!\n",
      "\n",
      " Starting Imputation ...\n",
      "\n",
      "  Imputing Samples 1-200 [20%] out of 1000 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 201-400 [40%] out of 1000 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 401-600 [60%] out of 1000 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 601-800 [80%] out of 1000 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 801-1000 [100%] out of 1000 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "\n",
      " Imputation successful (1913 seconds) !!!\n",
      "\n",
      " Appending chunk to final output VCF File :  ./compare2/minimac4.result.0.01.dose.vcf.gz \n",
      " Appending successful (4 seconds) !!!\n",
      "\n",
      " Time taken for this chunk = 2622 seconds.\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                              SUMMARY OF ANALYSIS                              \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Info file written to                    : ./compare2/minimac4.result.0.01.info\n",
      " Imputed VCF information written to      : ./compare2/minimac4.result.0.01.dose.vcf.gz\n",
      "\n",
      " Time Taken for Reading File             = 13 seconds\n",
      " Time Taken for Re-compression           = 696 seconds\n",
      " Time Taken for Imputation               = 1904 seconds\n",
      " Time Taken for Writing File             = 13 seconds\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                                END OF PROGRAM                                 \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Program Successfully Implemented... \n",
      " \n",
      " Total Run completed in 0 hours, 43 mins, 47 seconds.\n",
      "\n",
      " Thank You for using Minimac4 !!! \n",
      "\n",
      "beagle.28Sep18.793.jar (version 5.0)\n",
      "Copyright (C) 2014-2018 Brian L. Browning\n",
      "Enter \"java -jar beagle.28Sep18.793.jar\" to list command line argument\n",
      "Start time: 02:59 PM PST on 16 Dec 2019\n",
      "\n",
      "Command line: java -Xmx3641m -jar beagle.28Sep18.793.jar\n",
      "  gt=./compare2/target_masked_0.05.vcf.gz\n",
      "  ref=./compare2/haplo_ref.vcf.gz\n",
      "  out=./compare2/beagle.result.0.05\n",
      "\n",
      "No genetic map is specified: using 1 cM = 1 Mb\n",
      "\n",
      "Reference samples:       9,000\n",
      "Study samples:           1,000\n",
      "\n",
      "Window 1 (1:244-4999928)\n",
      "Reference markers:      41,897\n",
      "Study markers:          41,897\n",
      "\n",
      "Burnin  iteration 1:           9 seconds\n",
      "Burnin  iteration 2:           20 seconds\n",
      "Burnin  iteration 3:           36 seconds\n",
      "Burnin  iteration 4:           38 seconds\n",
      "Burnin  iteration 5:           42 seconds\n",
      "Burnin  iteration 6:           43 seconds\n",
      "\n",
      "Phasing iteration 1:           22 seconds\n",
      "Phasing iteration 2:           22 seconds\n",
      "Phasing iteration 3:           22 seconds\n",
      "Phasing iteration 4:           22 seconds\n",
      "Phasing iteration 5:           22 seconds\n",
      "Phasing iteration 6:           21 seconds\n",
      "Phasing iteration 7:           22 seconds\n",
      "Phasing iteration 8:           22 seconds\n",
      "Phasing iteration 9:           21 seconds\n",
      "Phasing iteration 10:          21 seconds\n",
      "Phasing iteration 11:          20 seconds\n",
      "Phasing iteration 12:          19 seconds\n",
      "\n",
      "Cumulative Statistics:\n",
      "\n",
      "Study markers:          41,897\n",
      "\n",
      "Haplotype phasing time:        7 minutes 26 seconds\n",
      "Total time:                    7 minutes 56 seconds\n",
      "\n",
      "End time: 03:07 PM PST on 16 Dec 2019\n",
      "beagle.28Sep18.793.jar finished\n",
      "\n",
      "\n",
      " -------------------------------------------------------------------------------- \n",
      "          Minimac4 - Fast Imputation Based on State Space Reduction HMM\n",
      " --------------------------------------------------------------------------------\n",
      "           (c) 2014 - Sayantan Das, Christian Fuchsberger, David Hinds\n",
      "                             Mary Kate Wing, Goncalo Abecasis \n",
      "\n",
      " Version: 1.0.2;\n",
      " Built: Mon Sep 30 11:52:22 PDT 2019 by biona001\n",
      "\n",
      " Command Line Options: \n",
      "       Reference Haplotypes : --refHaps [./compare2/haplo_ref.m3vcf.gz],\n",
      "                              --passOnly, --rsid, --referenceEstimates [ON],\n",
      "                              --mapFile [docs/geneticMapFile.b38.map.txt.gz]\n",
      "          Target Haplotypes : --haps [./compare2/target_masked_0.05.vcf.gz]\n",
      "          Output Parameters : --prefix [./compare2/minimac4.result.0.05],\n",
      "                              --estimate, --nobgzip, --vcfBuffer [200],\n",
      "                              --format [GT,DS], --allTypedSites, --meta,\n",
      "                              --memUsage\n",
      "        Chunking Parameters : --ChunkLengthMb [20.00], --ChunkOverlapMb [3.00]\n",
      "          Subset Parameters : --chr [], --start, --end, --window\n",
      "   Approximation Parameters : --minimac3, --probThreshold [0.01],\n",
      "                              --diffThreshold [0.01], --topThreshold [0.01]\n",
      "           Other Parameters : --log, --help, --cpus [1], --params\n",
      "                  PhoneHome : --noPhoneHome, --phoneHomeThinning [50]\n",
      "\n",
      "\n",
      " URL = http://genome.sph.umich.edu/wiki/Minimac4\n",
      " Starting Main Imputation/Estimation Analysis ... \n",
      "\n",
      " Performing preliminary check on input parameters... \n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                             PRELIMINARY FILE CHECK                            \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Checking GWAS haplotype file : ./compare2/target_masked_0.05.vcf.gz\n",
      "\n",
      " Gathering variant information ...\n",
      "\n",
      " Successful !!! \n",
      "\n",
      " Checking Reference haplotype file : ./compare2/haplo_ref.m3vcf.gz\n",
      "\n",
      " Gathering variant information ...\n",
      "\n",
      " Successful !!! \n",
      "\n",
      " Reference Panel   : Found 9000 samples (18000 haplotypes) and 41897 variants ...\n",
      "\n",
      " Target/GWAS Panel : Found 1000 samples (2000 haplotypes) and 41897 variants ...\n",
      "                     41897 variants overlap with Reference panel \n",
      "                     0 variants imported that exist only in Target/GWAS panel\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                           CHUNKING INFORMATION                           \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Chunking region into 1 chunk(s) with atleast 41897 variants in each chunk ... \n",
      "\n",
      " Details of chunks is given below ...\n",
      "\n",
      " No   LeftBuffer      LeftEnd   RightPoint  RightBuffer       #Sites(GWAS/Ref/%)\n",
      " -------------------------------------------------------------------------------\n",
      "  1          244          244      4999928      4999928    41897/   41897/ 100.00%\n",
      "\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                           MAIN IMPUTATION ANALYSIS                            \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Starting imputation analysis of 1 chunk(s) ... \n",
      "\n",
      " -------------------------------------------\n",
      " Analyzing Chunk 1/1 [1:244-4999928]\n",
      " -------------------------------------------\n",
      "\n",
      " Reading chunk from reference panel ... \n",
      " Reading chunk from target/GWAS panel ... \n",
      "\n",
      " Compressing reference panel at GWAS sites ... \n",
      " Re-compression successful (690 seconds) !!!\n",
      "\n",
      " Starting Imputation ...\n",
      "\n",
      "  Imputing Samples 1-200 [20%] out of 1000 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 201-400 [40%] out of 1000 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 401-600 [60%] out of 1000 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 601-800 [80%] out of 1000 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 801-1000 [100%] out of 1000 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "\n",
      " Imputation successful (1922 seconds) !!!\n",
      "\n",
      " Appending chunk to final output VCF File :  ./compare2/minimac4.result.0.05.dose.vcf.gz \n",
      " Appending successful (4 seconds) !!!\n",
      "\n",
      " Time taken for this chunk = 2626 seconds.\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                              SUMMARY OF ANALYSIS                              \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Info file written to                    : ./compare2/minimac4.result.0.05.info\n",
      " Imputed VCF information written to      : ./compare2/minimac4.result.0.05.dose.vcf.gz\n",
      "\n",
      " Time Taken for Reading File             = 14 seconds\n",
      " Time Taken for Re-compression           = 690 seconds\n",
      " Time Taken for Imputation               = 1912 seconds\n",
      " Time Taken for Writing File             = 14 seconds\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                                END OF PROGRAM                                 \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Program Successfully Implemented... \n",
      " \n",
      " Total Run completed in 0 hours, 43 mins, 51 seconds.\n",
      "\n",
      " Thank You for using Minimac4 !!! \n",
      "\n",
      "beagle.28Sep18.793.jar (version 5.0)\n",
      "Copyright (C) 2014-2018 Brian L. Browning\n",
      "Enter \"java -jar beagle.28Sep18.793.jar\" to list command line argument\n",
      "Start time: 03:57 PM PST on 16 Dec 2019\n",
      "\n",
      "Command line: java -Xmx3641m -jar beagle.28Sep18.793.jar\n",
      "  gt=./compare2/target_masked_0.1.vcf.gz\n",
      "  ref=./compare2/haplo_ref.vcf.gz\n",
      "  out=./compare2/beagle.result.0.1\n",
      "\n",
      "No genetic map is specified: using 1 cM = 1 Mb\n",
      "\n",
      "Reference samples:       9,000\n",
      "Study samples:           1,000\n",
      "\n",
      "Window 1 (1:244-4999928)\n",
      "Reference markers:      41,897\n",
      "Study markers:          41,897\n",
      "\n",
      "Burnin  iteration 1:           49 seconds\n",
      "Burnin  iteration 2:           40 seconds\n",
      "Burnin  iteration 3:           22 seconds\n",
      "Burnin  iteration 4:           34 seconds\n",
      "Burnin  iteration 5:           41 seconds\n",
      "Burnin  iteration 6:           39 seconds\n",
      "\n",
      "Phasing iteration 1:           20 seconds\n",
      "Phasing iteration 2:           19 seconds\n",
      "Phasing iteration 3:           19 seconds\n",
      "Phasing iteration 4:           19 seconds\n",
      "Phasing iteration 5:           19 seconds\n",
      "Phasing iteration 6:           19 seconds\n",
      "Phasing iteration 7:           19 seconds\n",
      "Phasing iteration 8:           19 seconds\n",
      "Phasing iteration 9:           19 seconds\n",
      "Phasing iteration 10:          20 seconds\n",
      "Phasing iteration 11:          17 seconds\n",
      "Phasing iteration 12:          17 seconds\n",
      "\n",
      "Cumulative Statistics:\n",
      "\n",
      "Study markers:          41,897\n",
      "\n",
      "Haplotype phasing time:        7 minutes 31 seconds\n",
      "Total time:                    7 minutes 59 seconds\n",
      "\n",
      "End time: 04:05 PM PST on 16 Dec 2019\n",
      "beagle.28Sep18.793.jar finished\n",
      "\n",
      "\n",
      " -------------------------------------------------------------------------------- \n",
      "          Minimac4 - Fast Imputation Based on State Space Reduction HMM\n",
      " --------------------------------------------------------------------------------\n",
      "           (c) 2014 - Sayantan Das, Christian Fuchsberger, David Hinds\n",
      "                             Mary Kate Wing, Goncalo Abecasis \n",
      "\n",
      " Version: 1.0.2;\n",
      " Built: Mon Sep 30 11:52:22 PDT 2019 by biona001\n",
      "\n",
      " Command Line Options: \n",
      "       Reference Haplotypes : --refHaps [./compare2/haplo_ref.m3vcf.gz],\n",
      "                              --passOnly, --rsid, --referenceEstimates [ON],\n",
      "                              --mapFile [docs/geneticMapFile.b38.map.txt.gz]\n",
      "          Target Haplotypes : --haps [./compare2/target_masked_0.1.vcf.gz]\n",
      "          Output Parameters : --prefix [./compare2/minimac4.result.0.1],\n",
      "                              --estimate, --nobgzip, --vcfBuffer [200],\n",
      "                              --format [GT,DS], --allTypedSites, --meta,\n",
      "                              --memUsage\n",
      "        Chunking Parameters : --ChunkLengthMb [20.00], --ChunkOverlapMb [3.00]\n",
      "          Subset Parameters : --chr [], --start, --end, --window\n",
      "   Approximation Parameters : --minimac3, --probThreshold [0.01],\n",
      "                              --diffThreshold [0.01], --topThreshold [0.01]\n",
      "           Other Parameters : --log, --help, --cpus [1], --params\n",
      "                  PhoneHome : --noPhoneHome, --phoneHomeThinning [50]\n",
      "\n",
      "\n",
      " URL = http://genome.sph.umich.edu/wiki/Minimac4\n",
      " Starting Main Imputation/Estimation Analysis ... \n",
      "\n",
      " Performing preliminary check on input parameters... \n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                             PRELIMINARY FILE CHECK                            \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Checking GWAS haplotype file : ./compare2/target_masked_0.1.vcf.gz\n",
      "\n",
      " Gathering variant information ...\n",
      "\n",
      " Successful !!! \n",
      "\n",
      " Checking Reference haplotype file : ./compare2/haplo_ref.m3vcf.gz\n",
      "\n",
      " Gathering variant information ...\n",
      "\n",
      " Successful !!! \n",
      "\n",
      " Reference Panel   : Found 9000 samples (18000 haplotypes) and 41897 variants ...\n",
      "\n",
      " Target/GWAS Panel : Found 1000 samples (2000 haplotypes) and 41897 variants ...\n",
      "                     41897 variants overlap with Reference panel \n",
      "                     0 variants imported that exist only in Target/GWAS panel\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                           CHUNKING INFORMATION                           \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Chunking region into 1 chunk(s) with atleast 41897 variants in each chunk ... \n",
      "\n",
      " Details of chunks is given below ...\n",
      "\n",
      " No   LeftBuffer      LeftEnd   RightPoint  RightBuffer       #Sites(GWAS/Ref/%)\n",
      " -------------------------------------------------------------------------------\n",
      "  1          244          244      4999928      4999928    41897/   41897/ 100.00%\n",
      "\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                           MAIN IMPUTATION ANALYSIS                            \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Starting imputation analysis of 1 chunk(s) ... \n",
      "\n",
      " -------------------------------------------\n",
      " Analyzing Chunk 1/1 [1:244-4999928]\n",
      " -------------------------------------------\n",
      "\n",
      " Reading chunk from reference panel ... \n",
      " Reading chunk from target/GWAS panel ... \n",
      "\n",
      " Compressing reference panel at GWAS sites ... \n",
      " Re-compression successful (708 seconds) !!!\n",
      "\n",
      " Starting Imputation ...\n",
      "\n",
      "  Imputing Samples 1-200 [20%] out of 1000 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 201-400 [40%] out of 1000 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 401-600 [60%] out of 1000 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 601-800 [80%] out of 1000 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 801-1000 [100%] out of 1000 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "\n",
      " Imputation successful (1880 seconds) !!!\n",
      "\n",
      " Appending chunk to final output VCF File :  ./compare2/minimac4.result.0.1.dose.vcf.gz \n",
      " Appending successful (4 seconds) !!!\n",
      "\n",
      " Time taken for this chunk = 2601 seconds.\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                              SUMMARY OF ANALYSIS                              \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Info file written to                    : ./compare2/minimac4.result.0.1.info\n",
      " Imputed VCF information written to      : ./compare2/minimac4.result.0.1.dose.vcf.gz\n",
      "\n",
      " Time Taken for Reading File             = 13 seconds\n",
      " Time Taken for Re-compression           = 708 seconds\n",
      " Time Taken for Imputation               = 1869 seconds\n",
      " Time Taken for Writing File             = 15 seconds\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                                END OF PROGRAM                                 \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Program Successfully Implemented... \n",
      " \n",
      " Total Run completed in 0 hours, 43 mins, 27 seconds.\n",
      "\n",
      " Thank You for using Minimac4 !!! \n",
      "\n",
      "beagle.28Sep18.793.jar (version 5.0)\n",
      "Copyright (C) 2014-2018 Brian L. Browning\n",
      "Enter \"java -jar beagle.28Sep18.793.jar\" to list command line argument\n",
      "Start time: 04:55 PM PST on 16 Dec 2019\n",
      "\n",
      "Command line: java -Xmx3641m -jar beagle.28Sep18.793.jar\n",
      "  gt=./compare2/target_masked_0.25.vcf.gz\n",
      "  ref=./compare2/haplo_ref.vcf.gz\n",
      "  out=./compare2/beagle.result.0.25\n",
      "\n",
      "No genetic map is specified: using 1 cM = 1 Mb\n",
      "\n",
      "Reference samples:       9,000\n",
      "Study samples:           1,000\n",
      "\n",
      "Window 1 (1:244-4999928)\n",
      "Reference markers:      41,897\n",
      "Study markers:          41,897\n",
      "\n",
      "Burnin  iteration 1:           2 minutes 13 seconds\n",
      "Burnin  iteration 2:           2 minutes 7 seconds\n",
      "Burnin  iteration 3:           16 seconds\n",
      "Burnin  iteration 4:           28 seconds\n",
      "Burnin  iteration 5:           38 seconds\n",
      "Burnin  iteration 6:           42 seconds\n",
      "\n",
      "Phasing iteration 1:           23 seconds\n",
      "Phasing iteration 2:           22 seconds\n",
      "Phasing iteration 3:           21 seconds\n",
      "Phasing iteration 4:           22 seconds\n",
      "Phasing iteration 5:           23 seconds\n",
      "Phasing iteration 6:           22 seconds\n",
      "Phasing iteration 7:           21 seconds\n",
      "Phasing iteration 8:           22 seconds\n",
      "Phasing iteration 9:           22 seconds\n",
      "Phasing iteration 10:          21 seconds\n",
      "Phasing iteration 11:          20 seconds\n",
      "Phasing iteration 12:          19 seconds\n",
      "\n",
      "Cumulative Statistics:\n",
      "\n",
      "Study markers:          41,897\n",
      "\n",
      "Haplotype phasing time:        10 minutes 43 seconds\n",
      "Total time:                    11 minutes 10 seconds\n",
      "\n",
      "End time: 05:06 PM PST on 16 Dec 2019\n",
      "beagle.28Sep18.793.jar finished\n",
      "\n",
      "\n",
      " -------------------------------------------------------------------------------- \n",
      "          Minimac4 - Fast Imputation Based on State Space Reduction HMM\n",
      " --------------------------------------------------------------------------------\n",
      "           (c) 2014 - Sayantan Das, Christian Fuchsberger, David Hinds\n",
      "                             Mary Kate Wing, Goncalo Abecasis \n",
      "\n",
      " Version: 1.0.2;\n",
      " Built: Mon Sep 30 11:52:22 PDT 2019 by biona001\n",
      "\n",
      " Command Line Options: \n",
      "       Reference Haplotypes : --refHaps [./compare2/haplo_ref.m3vcf.gz],\n",
      "                              --passOnly, --rsid, --referenceEstimates [ON],\n",
      "                              --mapFile [docs/geneticMapFile.b38.map.txt.gz]\n",
      "          Target Haplotypes : --haps [./compare2/target_masked_0.25.vcf.gz]\n",
      "          Output Parameters : --prefix [./compare2/minimac4.result.0.25],\n",
      "                              --estimate, --nobgzip, --vcfBuffer [200],\n",
      "                              --format [GT,DS], --allTypedSites, --meta,\n",
      "                              --memUsage\n",
      "        Chunking Parameters : --ChunkLengthMb [20.00], --ChunkOverlapMb [3.00]\n",
      "          Subset Parameters : --chr [], --start, --end, --window\n",
      "   Approximation Parameters : --minimac3, --probThreshold [0.01],\n",
      "                              --diffThreshold [0.01], --topThreshold [0.01]\n",
      "           Other Parameters : --log, --help, --cpus [1], --params\n",
      "                  PhoneHome : --noPhoneHome, --phoneHomeThinning [50]\n",
      "\n",
      "\n",
      " URL = http://genome.sph.umich.edu/wiki/Minimac4\n",
      " Starting Main Imputation/Estimation Analysis ... \n",
      "\n",
      " Performing preliminary check on input parameters... \n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                             PRELIMINARY FILE CHECK                            \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Checking GWAS haplotype file : ./compare2/target_masked_0.25.vcf.gz\n",
      "\n",
      " Gathering variant information ...\n",
      "\n",
      " Successful !!! \n",
      "\n",
      " Checking Reference haplotype file : ./compare2/haplo_ref.m3vcf.gz\n",
      "\n",
      " Gathering variant information ...\n",
      "\n",
      " Successful !!! \n",
      "\n",
      " Reference Panel   : Found 9000 samples (18000 haplotypes) and 41897 variants ...\n",
      "\n",
      " Target/GWAS Panel : Found 1000 samples (2000 haplotypes) and 41897 variants ...\n",
      "                     41897 variants overlap with Reference panel \n",
      "                     0 variants imported that exist only in Target/GWAS panel\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                           CHUNKING INFORMATION                           \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Chunking region into 1 chunk(s) with atleast 41897 variants in each chunk ... \n",
      "\n",
      " Details of chunks is given below ...\n",
      "\n",
      " No   LeftBuffer      LeftEnd   RightPoint  RightBuffer       #Sites(GWAS/Ref/%)\n",
      " -------------------------------------------------------------------------------\n",
      "  1          244          244      4999928      4999928    41897/   41897/ 100.00%\n",
      "\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                           MAIN IMPUTATION ANALYSIS                            \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Starting imputation analysis of 1 chunk(s) ... \n",
      "\n",
      " -------------------------------------------\n",
      " Analyzing Chunk 1/1 [1:244-4999928]\n",
      " -------------------------------------------\n",
      "\n",
      " Reading chunk from reference panel ... \n",
      " Reading chunk from target/GWAS panel ... \n",
      "\n",
      " Compressing reference panel at GWAS sites ... \n",
      " Re-compression successful (667 seconds) !!!\n",
      "\n",
      " Starting Imputation ...\n",
      "\n",
      "  Imputing Samples 1-200 [20%] out of 1000 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 201-400 [40%] out of 1000 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 401-600 [60%] out of 1000 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 601-800 [80%] out of 1000 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 801-1000 [100%] out of 1000 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "\n",
      " Imputation successful (1786 seconds) !!!\n",
      "\n",
      " Appending chunk to final output VCF File :  ./compare2/minimac4.result.0.25.dose.vcf.gz \n",
      " Appending successful (4 seconds) !!!\n",
      "\n",
      " Time taken for this chunk = 2467 seconds.\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                              SUMMARY OF ANALYSIS                              \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Info file written to                    : ./compare2/minimac4.result.0.25.info\n",
      " Imputed VCF information written to      : ./compare2/minimac4.result.0.25.dose.vcf.gz\n",
      "\n",
      " Time Taken for Reading File             = 13 seconds\n",
      " Time Taken for Re-compression           = 667 seconds\n",
      " Time Taken for Imputation               = 1776 seconds\n",
      " Time Taken for Writing File             = 14 seconds\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                                END OF PROGRAM                                 \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Program Successfully Implemented... \n",
      " \n",
      " Total Run completed in 0 hours, 41 mins, 12 seconds.\n",
      "\n",
      " Thank You for using Minimac4 !!! \n",
      "\n"
     ]
    },
    {
     "ename": "InterruptException",
     "evalue": "InterruptException:",
     "output_type": "error",
     "traceback": [
      "InterruptException:",
      "",
      "Stacktrace:",
      " [1] Array at ./boot.jl:404 [inlined]",
      " [2] Array at ./boot.jl:423 [inlined]",
      " [3] getindex at ./array.jl:367 [inlined]",
      " [4] GeneticVariation.VCF.Record() at /Users/biona001/.julia/packages/GeneticVariation/r8DAL/src/vcf/record.jl:31",
      " [5] read at /Users/biona001/.julia/packages/BioCore/YBJvb/src/IO.jl:61 [inlined]",
      " [6] copy_ht_as_is!(::Array{Float32,2}, ::GeneticVariation.VCF.Reader) at /Users/biona001/.julia/dev/VCFTools/src/convert.jl:303",
      " [7] #convert_ht#9(::Bool, ::typeof(convert_ht), ::Type{Float32}, ::String) at /Users/biona001/.julia/dev/VCFTools/src/convert.jl:218",
      " [8] #convert_ht at ./none:0 [inlined]",
      " [9] #phase#7(::Bool, ::String, ::Int64, ::typeof(phase), ::String, ::String) at /Users/biona001/.julia/dev/MendelImpute/src/haplotyping.jl:25",
      " [10] (::MendelImpute.var\"#kw##phase\")(::NamedTuple{(:impute, :outfile, :width),Tuple{Bool,String,Int64}}, ::typeof(phase), ::String, ::String) at ./none:0",
      " [11] top-level scope at ./util.jl:288"
     ]
    }
   ],
   "source": [
    "n = length(missingprop)\n",
    "beagle5_error  = zeros(n)\n",
    "beagle5_times  = zeros(n) #sec\n",
    "beagle5_mem    = zeros(n) #MB\n",
    "minimac4_error = zeros(n)\n",
    "minimac4_times = zeros(n)\n",
    "minimac4_mem   = zeros(n)\n",
    "mendel_100_error  = zeros(n)\n",
    "mendel_100_times  = zeros(n) \n",
    "mendel_100_mem    = zeros(n) \n",
    "mendel_400_error  = zeros(n)\n",
    "mendel_400_times  = zeros(n)\n",
    "mendel_400_mem    = zeros(n)\n",
    "mendel_1200_error = zeros(n)\n",
    "mendel_1200_times = zeros(n)\n",
    "mendel_1200_mem   = zeros(n)\n",
    "X_complete = convert_gt(Float32, \"./compare2/target.vcf.gz\"; as_minorallele=false)\n",
    "for (i, prop) in enumerate(missingprop)\n",
    "    #\n",
    "    # current run info\n",
    "    #\n",
    "    X_masked = convert_gt(Float32, \"./compare2/target_masked_$prop.vcf.gz\"; as_minorallele=false)\n",
    "    missing_idx = ismissing.(X_masked)\n",
    "    total_missing = sum(missing_idx)\n",
    "    actual_missing_values = X_complete[missing_idx]  #true values of missing entries\n",
    "    #\n",
    "    # MendelImpute (window = 100)\n",
    "    #\n",
    "    width   = 100\n",
    "    tgtfile = \"./compare2/target_masked_$prop.vcf.gz\"\n",
    "    reffile = \"./compare2/haplo_ref.vcf.gz\"\n",
    "    outfile = \"./compare2/mendel.result.$prop.window$width.vcf.gz\"\n",
    "    mendel_result = @timed phase(tgtfile, reffile, impute=true, outfile = outfile, width = width)\n",
    "    X_mendel = convert_gt(Float32, outfile, as_minorallele=false)\n",
    "    imputed_missing_values = X_mendel[missing_idx] #imputed values of missing entries\n",
    "    mendel_err = sum(actual_missing_values .!= imputed_missing_values) / total_missing\n",
    "    mendel_100_error[i] = mendel_err\n",
    "    mendel_100_times[i] = mendel_result[2] \n",
    "    mendel_100_mem[i]   = mendel_result[3] / 1e6\n",
    "    #\n",
    "    # MendelImpute (window = 400)\n",
    "    #\n",
    "    width   = 400\n",
    "    tgtfile = \"./compare2/target_masked_$prop.vcf.gz\"\n",
    "    reffile = \"./compare2/haplo_ref.vcf.gz\"\n",
    "    outfile = \"./compare2/mendel.result.$prop.window$width.vcf.gz\"\n",
    "    mendel_result = @timed phase(tgtfile, reffile, impute=true, outfile = outfile, width = width)\n",
    "    X_mendel = convert_gt(Float32, outfile, as_minorallele=false)\n",
    "    imputed_missing_values = X_mendel[missing_idx] #imputed values of missing entries\n",
    "    mendel_err = sum(actual_missing_values .!= imputed_missing_values) / total_missing\n",
    "    mendel_400_error[i] = mendel_err\n",
    "    mendel_400_times[i] = mendel_result[2] \n",
    "    mendel_400_mem[i]   = mendel_result[3] / 1e6\n",
    "    #\n",
    "    # MendelImpute (window = 1200)\n",
    "    #\n",
    "    width   = 1200\n",
    "    tgtfile = \"./compare2/target_masked_$prop.vcf.gz\"\n",
    "    reffile = \"./compare2/haplo_ref.vcf.gz\"\n",
    "    outfile = \"./compare2/mendel.result.$prop.window$width.vcf.gz\"\n",
    "    mendel_result = @timed phase(tgtfile, reffile, impute=true, outfile = outfile, width = width)\n",
    "    X_mendel = convert_gt(Float32, outfile, as_minorallele=false)\n",
    "    imputed_missing_values = X_mendel[missing_idx] #imputed values of missing entries\n",
    "    mendel_err = sum(actual_missing_values .!= imputed_missing_values) / total_missing\n",
    "    mendel_1200_error[i] = mendel_err\n",
    "    mendel_1200_times[i] = mendel_result[2] \n",
    "    mendel_1200_mem[i]   = mendel_result[3] / 1e6\n",
    "    #\n",
    "    # Beagle 5\n",
    "    #\n",
    "    beagle_result = @timed run(`java -jar beagle.28Sep18.793.jar gt=./compare2/target_masked_$prop.vcf.gz ref=./compare2/haplo_ref.vcf.gz out=./compare2/beagle.result.$prop`)\n",
    "    X_beagle = convert_gt(Float32, \"./compare2/beagle.result.$prop.vcf.gz\", as_minorallele=false)\n",
    "    imputed_missing_values = X_beagle[missing_idx] #imputed values of missing entries\n",
    "    beagle_err = sum(actual_missing_values .!= imputed_missing_values) / total_missing\n",
    "    beagle5_error[i] = beagle_err\n",
    "    beagle5_times[i] = beagle_result[2]\n",
    "    beagle5_mem[i]   = beagle_result[3] / 1e6\n",
    "    #\n",
    "    # minimac 4\n",
    "    #\n",
    "    minimac4 = \"/Users/biona001/Benjamin_Folder/UCLA/research/softwares/Minimac4/build/minimac4\"\n",
    "    minimac_result = @timed run(`$minimac4 --refHaps ./compare2/haplo_ref.m3vcf.gz --haps ./compare2/target_masked_$prop.vcf.gz --prefix ./compare2/minimac4.result.$prop`)\n",
    "    X_minimac = convert_gt(Float32, \"./compare2/minimac4.result.$prop.dose.vcf.gz\", as_minorallele=false)\n",
    "    imputed_missing_values = X_minimac[missing_idx] #imputed values of missing entries\n",
    "    minimac_err = sum(actual_missing_values .!= imputed_missing_values) / total_missing\n",
    "    minimac4_error[i] = minimac_err\n",
    "    minimac4_times[i] = minimac_result[2]\n",
    "    minimac4_mem[i]   = minimac_result[3] / 1e6\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "7×6 Array{Float64,2}:\n",
       " 0.0001  0.00594389  0.000713267  0.000713267  0.000237756  0.0       \n",
       " 0.001   0.00558965  0.000999001  0.000309215  2.37857e-5   2.37857e-5\n",
       " 0.01    0.00575938  0.00088422   0.000477957  4.54059e-5   3.82365e-5\n",
       " 0.05    0.00728215  0.00132082   0.000639891  4.15141e-5   3.67424e-5\n",
       " 0.1     0.00926913  0.00188571   0.000855862  4.31988e-5   3.53228e-5\n",
       " 0.25    0.0177278   0.00479307   0.00206166   0.000203837  3.75212e-5\n",
       " 0.5     0.0         0.0          0.0          0.0          0.0       "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "missingprop = [0.0001, 0.001, 0.01, 0.05, 0.1, 0.25, 0.5]\n",
    "[missingprop mendel_100_error mendel_400_error mendel_1200_error beagle5_error minimac4_error]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "7×6 Array{Float64,2}:\n",
       " 0.0001  112.524  100.581   101.083  695.014  2522.29\n",
       " 0.001   115.213   98.3722  100.386  504.123  2572.75\n",
       " 0.01    113.035   99.6651  106.808  487.818  2628.65\n",
       " 0.05    126.799  116.642   115.961  476.18   2633.13\n",
       " 0.1     125.142  113.851   114.97   479.562  2609.16\n",
       " 0.25    118.378  110.692   114.12   670.077  2473.97\n",
       " 0.5       0.0      0.0       0.0      0.0       0.0 "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[missingprop mendel_100_times mendel_400_times mendel_1200_times beagle5_times minimac4_times]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plot results (50% missing was not ran)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"utf-8\"?>\n",
       "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" width=\"600\" height=\"400\" viewBox=\"0 0 2400 1600\">\n",
       "<defs>\n",
       "  <clipPath id=\"clip0200\">\n",
       "    <rect x=\"0\" y=\"0\" width=\"2400\" height=\"1600\"/>\n",
       "  </clipPath>\n",
       "</defs>\n",
       "<path clip-path=\"url(#clip0200)\" d=\"\n",
       "M0 1600 L2400 1600 L2400 0 L0 0  Z\n",
       "  \" fill=\"#ffffff\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<defs>\n",
       "  <clipPath id=\"clip0201\">\n",
       "    <rect x=\"480\" y=\"0\" width=\"1681\" height=\"1600\"/>\n",
       "  </clipPath>\n",
       "</defs>\n",
       "<path clip-path=\"url(#clip0200)\" d=\"\n",
       "M269.279 1425.62 L2352.76 1425.62 L2352.76 121.675 L269.279 121.675  Z\n",
       "  \" fill=\"#ffffff\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<defs>\n",
       "  <clipPath id=\"clip0202\">\n",
       "    <rect x=\"269\" y=\"121\" width=\"2084\" height=\"1305\"/>\n",
       "  </clipPath>\n",
       "</defs>\n",
       "<polyline clip-path=\"url(#clip0202)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  488.568,1425.62 488.568,121.675 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip0202)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  762.718,1425.62 762.718,121.675 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip0202)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  1036.87,1425.62 1036.87,121.675 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip0202)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  1311.02,1425.62 1311.02,121.675 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip0202)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  1585.17,1425.62 1585.17,121.675 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip0202)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  1859.32,1425.62 1859.32,121.675 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip0202)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  2133.47,1425.62 2133.47,121.675 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip0202)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  269.279,1388.71 2352.76,1388.71 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip0202)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  269.279,1041.76 2352.76,1041.76 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip0202)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  269.279,694.812 2352.76,694.812 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip0202)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  269.279,347.86 2352.76,347.86 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip0200)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  269.279,1425.62 2352.76,1425.62 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip0200)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  269.279,1425.62 269.279,121.675 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip0200)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  488.568,1425.62 488.568,1406.06 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip0200)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  762.718,1425.62 762.718,1406.06 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip0200)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1036.87,1425.62 1036.87,1406.06 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip0200)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1311.02,1425.62 1311.02,1406.06 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip0200)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1585.17,1425.62 1585.17,1406.06 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip0200)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1859.32,1425.62 1859.32,1406.06 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip0200)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  2133.47,1425.62 2133.47,1406.06 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip0200)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  269.279,1388.71 300.531,1388.71 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip0200)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  269.279,1041.76 300.531,1041.76 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip0200)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  269.279,694.812 300.531,694.812 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip0200)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  269.279,347.86 300.531,347.86 \n",
       "  \"/>\n",
       "<g clip-path=\"url(#clip0200)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:middle;\" transform=\"rotate(0, 488.568, 1479.62)\" x=\"488.568\" y=\"1479.62\">0.01</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip0200)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:middle;\" transform=\"rotate(0, 762.718, 1479.62)\" x=\"762.718\" y=\"1479.62\">0.1</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip0200)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:middle;\" transform=\"rotate(0, 1036.87, 1479.62)\" x=\"1036.87\" y=\"1479.62\">1</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip0200)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:middle;\" transform=\"rotate(0, 1311.02, 1479.62)\" x=\"1311.02\" y=\"1479.62\">5</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip0200)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:middle;\" transform=\"rotate(0, 1585.17, 1479.62)\" x=\"1585.17\" y=\"1479.62\">10</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip0200)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:middle;\" transform=\"rotate(0, 1859.32, 1479.62)\" x=\"1859.32\" y=\"1479.62\">25</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip0200)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:middle;\" transform=\"rotate(0, 2133.47, 1479.62)\" x=\"2133.47\" y=\"1479.62\">50</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip0200)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:end;\" transform=\"rotate(0, 245.279, 1406.21)\" x=\"245.279\" y=\"1406.21\">0.000</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip0200)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:end;\" transform=\"rotate(0, 245.279, 1059.26)\" x=\"245.279\" y=\"1059.26\">0.005</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip0200)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:end;\" transform=\"rotate(0, 245.279, 712.312)\" x=\"245.279\" y=\"712.312\">0.010</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip0200)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:end;\" transform=\"rotate(0, 245.279, 365.36)\" x=\"245.279\" y=\"365.36\">0.015</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip0200)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:84px; text-anchor:middle;\" transform=\"rotate(0, 1311.02, 73.2)\" x=\"1311.02\" y=\"73.2\">Error rate with varying proportions of data missing</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip0200)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:66px; text-anchor:middle;\" transform=\"rotate(0, 1311.02, 1559.48)\" x=\"1311.02\" y=\"1559.48\">% of data missing</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip0200)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:66px; text-anchor:middle;\" transform=\"rotate(-90, 89.2861, 773.647)\" x=\"89.2861\" y=\"773.647\">Error rate</text>\n",
       "</g>\n",
       "<path clip-path=\"url(#clip0202)\" d=\"\n",
       "M378.908 976.266 L378.908 1388.71 L422.772 1388.71 L422.772 976.266 L378.908 976.266 L378.908 976.266  Z\n",
       "  \" fill=\"#009af9\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip0202)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  378.908,976.266 378.908,1388.71 422.772,1388.71 422.772,976.266 378.908,976.266 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip0202)\" d=\"\n",
       "M653.058 1000.85 L653.058 1388.71 L696.922 1388.71 L696.922 1000.85 L653.058 1000.85 L653.058 1000.85  Z\n",
       "  \" fill=\"#009af9\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip0202)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  653.058,1000.85 653.058,1388.71 696.922,1388.71 696.922,1000.85 653.058,1000.85 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip0202)\" d=\"\n",
       "M927.208 989.07 L927.208 1388.71 L971.072 1388.71 L971.072 989.07 L927.208 989.07 L927.208 989.07  Z\n",
       "  \" fill=\"#009af9\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip0202)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  927.208,989.07 927.208,1388.71 971.072,1388.71 971.072,989.07 927.208,989.07 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip0202)\" d=\"\n",
       "M1201.36 883.404 L1201.36 1388.71 L1245.22 1388.71 L1245.22 883.404 L1201.36 883.404 L1201.36 883.404  Z\n",
       "  \" fill=\"#009af9\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip0202)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1201.36,883.404 1201.36,1388.71 1245.22,1388.71 1245.22,883.404 1201.36,883.404 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip0202)\" d=\"\n",
       "M1475.51 745.527 L1475.51 1388.71 L1519.37 1388.71 L1519.37 745.527 L1475.51 745.527 L1475.51 745.527  Z\n",
       "  \" fill=\"#009af9\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip0202)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1475.51,745.527 1475.51,1388.71 1519.37,1388.71 1519.37,745.527 1475.51,745.527 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip0202)\" d=\"\n",
       "M1749.66 158.579 L1749.66 1388.71 L1793.52 1388.71 L1793.52 158.579 L1749.66 158.579 L1749.66 158.579  Z\n",
       "  \" fill=\"#009af9\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip0202)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1749.66,158.579 1749.66,1388.71 1793.52,1388.71 1793.52,158.579 1749.66,158.579 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip0202)\" d=\"\n",
       "M2023.81 1388.71 L2023.81 1388.71 L2067.67 1388.71 L2067.67 1388.71 L2023.81 1388.71 L2023.81 1388.71  Z\n",
       "  \" fill=\"#009af9\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip0202)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  2023.81,1388.71 2023.81,1388.71 2067.67,1388.71 2023.81,1388.71 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip0202)\" d=\"\n",
       "M422.772 1339.22 L422.772 1388.71 L466.636 1388.71 L466.636 1339.22 L422.772 1339.22 L422.772 1339.22  Z\n",
       "  \" fill=\"#e26f46\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip0202)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  422.772,1339.22 422.772,1388.71 466.636,1388.71 466.636,1339.22 422.772,1339.22 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip0202)\" d=\"\n",
       "M696.922 1319.39 L696.922 1388.71 L740.786 1388.71 L740.786 1319.39 L696.922 1319.39 L696.922 1319.39  Z\n",
       "  \" fill=\"#e26f46\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip0202)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  696.922,1319.39 696.922,1388.71 740.786,1388.71 740.786,1319.39 696.922,1319.39 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip0202)\" d=\"\n",
       "M971.072 1327.36 L971.072 1388.71 L1014.94 1388.71 L1014.94 1327.36 L971.072 1327.36 L971.072 1327.36  Z\n",
       "  \" fill=\"#e26f46\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip0202)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  971.072,1327.36 971.072,1388.71 1014.94,1388.71 1014.94,1327.36 971.072,1327.36 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip0202)\" d=\"\n",
       "M1245.22 1297.06 L1245.22 1388.71 L1289.09 1388.71 L1289.09 1297.06 L1245.22 1297.06 L1245.22 1297.06  Z\n",
       "  \" fill=\"#e26f46\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip0202)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1245.22,1297.06 1245.22,1388.71 1289.09,1388.71 1289.09,1297.06 1245.22,1297.06 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip0202)\" d=\"\n",
       "M1519.37 1257.86 L1519.37 1388.71 L1563.24 1388.71 L1563.24 1257.86 L1519.37 1257.86 L1519.37 1257.86  Z\n",
       "  \" fill=\"#e26f46\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip0202)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1519.37,1257.86 1519.37,1388.71 1563.24,1388.71 1563.24,1257.86 1519.37,1257.86 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip0202)\" d=\"\n",
       "M1793.52 1056.12 L1793.52 1388.71 L1837.38 1388.71 L1837.38 1056.12 L1793.52 1056.12 L1793.52 1056.12  Z\n",
       "  \" fill=\"#e26f46\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip0202)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1793.52,1056.12 1793.52,1388.71 1837.38,1388.71 1837.38,1056.12 1793.52,1056.12 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip0202)\" d=\"\n",
       "M2067.67 1388.71 L2067.67 1388.71 L2111.53 1388.71 L2111.53 1388.71 L2067.67 1388.71 L2067.67 1388.71  Z\n",
       "  \" fill=\"#e26f46\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip0202)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  2067.67,1388.71 2067.67,1388.71 2111.53,1388.71 2067.67,1388.71 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip0202)\" d=\"\n",
       "M466.636 1339.22 L466.636 1388.71 L510.5 1388.71 L510.5 1339.22 L466.636 1339.22 L466.636 1339.22  Z\n",
       "  \" fill=\"#3da44d\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip0202)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  466.636,1339.22 466.636,1388.71 510.5,1388.71 510.5,1339.22 466.636,1339.22 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip0202)\" d=\"\n",
       "M740.786 1367.26 L740.786 1388.71 L784.65 1388.71 L784.65 1367.26 L740.786 1367.26 L740.786 1367.26  Z\n",
       "  \" fill=\"#3da44d\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip0202)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  740.786,1367.26 740.786,1388.71 784.65,1388.71 784.65,1367.26 740.786,1367.26 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip0202)\" d=\"\n",
       "M1014.94 1355.55 L1014.94 1388.71 L1058.8 1388.71 L1058.8 1355.55 L1014.94 1355.55 L1014.94 1355.55  Z\n",
       "  \" fill=\"#3da44d\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip0202)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1014.94,1355.55 1014.94,1388.71 1058.8,1388.71 1058.8,1355.55 1014.94,1355.55 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip0202)\" d=\"\n",
       "M1289.09 1344.31 L1289.09 1388.71 L1332.95 1388.71 L1332.95 1344.31 L1289.09 1344.31 L1289.09 1344.31  Z\n",
       "  \" fill=\"#3da44d\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip0202)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1289.09,1344.31 1289.09,1388.71 1332.95,1388.71 1332.95,1344.31 1289.09,1344.31 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip0202)\" d=\"\n",
       "M1563.24 1329.33 L1563.24 1388.71 L1607.1 1388.71 L1607.1 1329.33 L1563.24 1329.33 L1563.24 1329.33  Z\n",
       "  \" fill=\"#3da44d\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip0202)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1563.24,1329.33 1563.24,1388.71 1607.1,1388.71 1607.1,1329.33 1563.24,1329.33 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip0202)\" d=\"\n",
       "M1837.38 1245.66 L1837.38 1388.71 L1881.25 1388.71 L1881.25 1245.66 L1837.38 1245.66 L1837.38 1245.66  Z\n",
       "  \" fill=\"#3da44d\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip0202)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1837.38,1245.66 1837.38,1388.71 1881.25,1388.71 1881.25,1245.66 1837.38,1245.66 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip0202)\" d=\"\n",
       "M2111.53 1388.71 L2111.53 1388.71 L2155.4 1388.71 L2155.4 1388.71 L2111.53 1388.71 L2111.53 1388.71  Z\n",
       "  \" fill=\"#3da44d\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip0202)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  2111.53,1388.71 2111.53,1388.71 2155.4,1388.71 2111.53,1388.71 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip0202)\" d=\"\n",
       "M510.5 1372.22 L510.5 1388.71 L554.364 1388.71 L554.364 1372.22 L510.5 1372.22 L510.5 1372.22  Z\n",
       "  \" fill=\"#c271d2\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip0202)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  510.5,1372.22 510.5,1388.71 554.364,1388.71 554.364,1372.22 510.5,1372.22 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip0202)\" d=\"\n",
       "M784.65 1387.06 L784.65 1388.71 L828.514 1388.71 L828.514 1387.06 L784.65 1387.06 L784.65 1387.06  Z\n",
       "  \" fill=\"#c271d2\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip0202)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  784.65,1387.06 784.65,1388.71 828.514,1388.71 828.514,1387.06 784.65,1387.06 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip0202)\" d=\"\n",
       "M1058.8 1385.56 L1058.8 1388.71 L1102.66 1388.71 L1102.66 1385.56 L1058.8 1385.56 L1058.8 1385.56  Z\n",
       "  \" fill=\"#c271d2\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip0202)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1058.8,1385.56 1058.8,1388.71 1102.66,1388.71 1102.66,1385.56 1058.8,1385.56 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip0202)\" d=\"\n",
       "M1332.95 1385.83 L1332.95 1388.71 L1376.81 1388.71 L1376.81 1385.83 L1332.95 1385.83 L1332.95 1385.83  Z\n",
       "  \" fill=\"#c271d2\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip0202)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1332.95,1385.83 1332.95,1388.71 1376.81,1388.71 1376.81,1385.83 1332.95,1385.83 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip0202)\" d=\"\n",
       "M1607.1 1385.72 L1607.1 1388.71 L1650.96 1388.71 L1650.96 1385.72 L1607.1 1385.72 L1607.1 1385.72  Z\n",
       "  \" fill=\"#c271d2\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip0202)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1607.1,1385.72 1607.1,1388.71 1650.96,1388.71 1650.96,1385.72 1607.1,1385.72 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip0202)\" d=\"\n",
       "M1881.25 1374.57 L1881.25 1388.71 L1925.11 1388.71 L1925.11 1374.57 L1881.25 1374.57 L1881.25 1374.57  Z\n",
       "  \" fill=\"#c271d2\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip0202)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1881.25,1374.57 1881.25,1388.71 1925.11,1388.71 1925.11,1374.57 1881.25,1374.57 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip0202)\" d=\"\n",
       "M2155.4 1388.71 L2155.4 1388.71 L2199.26 1388.71 L2199.26 1388.71 L2155.4 1388.71 L2155.4 1388.71  Z\n",
       "  \" fill=\"#c271d2\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip0202)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  2155.4,1388.71 2155.4,1388.71 2199.26,1388.71 2155.4,1388.71 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip0202)\" d=\"\n",
       "M554.364 1388.71 L554.364 1388.71 L598.228 1388.71 L598.228 1388.71 L554.364 1388.71 L554.364 1388.71  Z\n",
       "  \" fill=\"#ac8d18\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip0202)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  554.364,1388.71 554.364,1388.71 598.228,1388.71 554.364,1388.71 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip0202)\" d=\"\n",
       "M828.514 1387.06 L828.514 1388.71 L872.378 1388.71 L872.378 1387.06 L828.514 1387.06 L828.514 1387.06  Z\n",
       "  \" fill=\"#ac8d18\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip0202)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  828.514,1387.06 828.514,1388.71 872.378,1388.71 872.378,1387.06 828.514,1387.06 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip0202)\" d=\"\n",
       "M1102.66 1386.06 L1102.66 1388.71 L1146.53 1388.71 L1146.53 1386.06 L1102.66 1386.06 L1102.66 1386.06  Z\n",
       "  \" fill=\"#ac8d18\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip0202)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1102.66,1386.06 1102.66,1388.71 1146.53,1388.71 1146.53,1386.06 1102.66,1386.06 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip0202)\" d=\"\n",
       "M1376.81 1386.16 L1376.81 1388.71 L1420.68 1388.71 L1420.68 1386.16 L1376.81 1386.16 L1376.81 1386.16  Z\n",
       "  \" fill=\"#ac8d18\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip0202)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1376.81,1386.16 1376.81,1388.71 1420.68,1388.71 1420.68,1386.16 1376.81,1386.16 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip0202)\" d=\"\n",
       "M1650.96 1386.26 L1650.96 1388.71 L1694.83 1388.71 L1694.83 1386.26 L1650.96 1386.26 L1650.96 1386.26  Z\n",
       "  \" fill=\"#ac8d18\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip0202)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1650.96,1386.26 1650.96,1388.71 1694.83,1388.71 1694.83,1386.26 1650.96,1386.26 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip0202)\" d=\"\n",
       "M1925.11 1386.11 L1925.11 1388.71 L1968.98 1388.71 L1968.98 1386.11 L1925.11 1386.11 L1925.11 1386.11  Z\n",
       "  \" fill=\"#ac8d18\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip0202)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1925.11,1386.11 1925.11,1388.71 1968.98,1388.71 1968.98,1386.11 1925.11,1386.11 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip0202)\" d=\"\n",
       "M2199.26 1388.71 L2199.26 1388.71 L2243.13 1388.71 L2243.13 1388.71 L2199.26 1388.71 L2199.26 1388.71  Z\n",
       "  \" fill=\"#ac8d18\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip0202)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  2199.26,1388.71 2199.26,1388.71 2243.13,1388.71 2199.26,1388.71 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip0200)\" d=\"\n",
       "M1493.67 568.075 L2280.76 568.075 L2280.76 205.195 L1493.67 205.195  Z\n",
       "  \" fill=\"#ffffff\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip0200)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1493.67,568.075 2280.76,568.075 2280.76,205.195 1493.67,205.195 1493.67,568.075 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip0200)\" d=\"\n",
       "M1517.67 289.867 L1661.67 289.867 L1661.67 241.483 L1517.67 241.483 L1517.67 289.867  Z\n",
       "  \" fill=\"#009af9\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip0200)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1517.67,289.867 1661.67,289.867 1661.67,241.483 1517.67,241.483 1517.67,289.867 \n",
       "  \"/>\n",
       "<g clip-path=\"url(#clip0200)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:start;\" transform=\"rotate(0, 1685.67, 283.175)\" x=\"1685.67\" y=\"283.175\">MendelImpute (w = 100)</text>\n",
       "</g>\n",
       "<path clip-path=\"url(#clip0200)\" d=\"\n",
       "M1517.67 350.347 L1661.67 350.347 L1661.67 301.963 L1517.67 301.963 L1517.67 350.347  Z\n",
       "  \" fill=\"#e26f46\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip0200)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1517.67,350.347 1661.67,350.347 1661.67,301.963 1517.67,301.963 1517.67,350.347 \n",
       "  \"/>\n",
       "<g clip-path=\"url(#clip0200)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:start;\" transform=\"rotate(0, 1685.67, 343.655)\" x=\"1685.67\" y=\"343.655\">MendelImpute (w = 400)</text>\n",
       "</g>\n",
       "<path clip-path=\"url(#clip0200)\" d=\"\n",
       "M1517.67 410.827 L1661.67 410.827 L1661.67 362.443 L1517.67 362.443 L1517.67 410.827  Z\n",
       "  \" fill=\"#3da44d\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip0200)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1517.67,410.827 1661.67,410.827 1661.67,362.443 1517.67,362.443 1517.67,410.827 \n",
       "  \"/>\n",
       "<g clip-path=\"url(#clip0200)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:start;\" transform=\"rotate(0, 1685.67, 404.135)\" x=\"1685.67\" y=\"404.135\">MendelImpute (w = 1200)</text>\n",
       "</g>\n",
       "<path clip-path=\"url(#clip0200)\" d=\"\n",
       "M1517.67 471.307 L1661.67 471.307 L1661.67 422.923 L1517.67 422.923 L1517.67 471.307  Z\n",
       "  \" fill=\"#c271d2\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip0200)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1517.67,471.307 1661.67,471.307 1661.67,422.923 1517.67,422.923 1517.67,471.307 \n",
       "  \"/>\n",
       "<g clip-path=\"url(#clip0200)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:start;\" transform=\"rotate(0, 1685.67, 464.615)\" x=\"1685.67\" y=\"464.615\">Beagle 5.0</text>\n",
       "</g>\n",
       "<path clip-path=\"url(#clip0200)\" d=\"\n",
       "M1517.67 531.787 L1661.67 531.787 L1661.67 483.403 L1517.67 483.403 L1517.67 531.787  Z\n",
       "  \" fill=\"#ac8d18\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip0200)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1517.67,531.787 1661.67,531.787 1661.67,483.403 1517.67,483.403 1517.67,531.787 \n",
       "  \"/>\n",
       "<g clip-path=\"url(#clip0200)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:start;\" transform=\"rotate(0, 1685.67, 525.095)\" x=\"1685.67\" y=\"525.095\">Minimac 4</text>\n",
       "</g>\n",
       "</svg>\n"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "missingprop_name = [\"0.01\", \"0.1\", \"1\", \"5\", \"10\", \"25\", \"50\"]\n",
    "groupedbar(missingprop_name, [mendel_100_error mendel_400_error mendel_1200_error beagle5_error minimac4_error],\n",
    "    label = [\"MendelImpute (w = 100)\", \"MendelImpute (w = 400)\", \"MendelImpute (w = 1200)\", \"Beagle 5.0\", \"Minimac 4\"], \n",
    "    title = \"Error rate with varying proportions of data missing\",\n",
    "    ylabel = \"Error rate\", xlabel = \"% of data missing\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"utf-8\"?>\n",
       "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" width=\"600\" height=\"400\" viewBox=\"0 0 2400 1600\">\n",
       "<defs>\n",
       "  <clipPath id=\"clip9400\">\n",
       "    <rect x=\"0\" y=\"0\" width=\"2400\" height=\"1600\"/>\n",
       "  </clipPath>\n",
       "</defs>\n",
       "<path clip-path=\"url(#clip9400)\" d=\"\n",
       "M0 1600 L2400 1600 L2400 0 L0 0  Z\n",
       "  \" fill=\"#ffffff\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<defs>\n",
       "  <clipPath id=\"clip9401\">\n",
       "    <rect x=\"480\" y=\"0\" width=\"1681\" height=\"1600\"/>\n",
       "  </clipPath>\n",
       "</defs>\n",
       "<path clip-path=\"url(#clip9400)\" d=\"\n",
       "M255.898 1425.62 L2352.76 1425.62 L2352.76 121.675 L255.898 121.675  Z\n",
       "  \" fill=\"#ffffff\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<defs>\n",
       "  <clipPath id=\"clip9402\">\n",
       "    <rect x=\"255\" y=\"121\" width=\"2098\" height=\"1305\"/>\n",
       "  </clipPath>\n",
       "</defs>\n",
       "<polyline clip-path=\"url(#clip9402)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  476.595,1425.62 476.595,121.675 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip9402)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  752.506,1425.62 752.506,121.675 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip9402)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  1028.42,1425.62 1028.42,121.675 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip9402)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  1304.33,1425.62 1304.33,121.675 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip9402)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  1580.24,1425.62 1580.24,121.675 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip9402)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  1856.15,1425.62 1856.15,121.675 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip9402)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  2132.06,1425.62 2132.06,121.675 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip9402)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  255.898,1388.71 2352.76,1388.71 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip9402)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  255.898,1155.13 2352.76,1155.13 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip9402)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  255.898,921.538 2352.76,921.538 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip9402)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  255.898,687.949 2352.76,687.949 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip9402)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  255.898,454.361 2352.76,454.361 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip9402)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  255.898,220.772 2352.76,220.772 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip9400)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  255.898,1425.62 2352.76,1425.62 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip9400)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  255.898,1425.62 255.898,121.675 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip9400)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  476.595,1425.62 476.595,1406.06 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip9400)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  752.506,1425.62 752.506,1406.06 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip9400)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1028.42,1425.62 1028.42,1406.06 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip9400)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1304.33,1425.62 1304.33,1406.06 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip9400)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1580.24,1425.62 1580.24,1406.06 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip9400)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1856.15,1425.62 1856.15,1406.06 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip9400)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  2132.06,1425.62 2132.06,1406.06 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip9400)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  255.898,1388.71 287.35,1388.71 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip9400)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  255.898,1155.13 287.35,1155.13 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip9400)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  255.898,921.538 287.35,921.538 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip9400)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  255.898,687.949 287.35,687.949 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip9400)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  255.898,454.361 287.35,454.361 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip9400)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  255.898,220.772 287.35,220.772 \n",
       "  \"/>\n",
       "<g clip-path=\"url(#clip9400)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:middle;\" transform=\"rotate(0, 476.595, 1479.62)\" x=\"476.595\" y=\"1479.62\">0.01</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip9400)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:middle;\" transform=\"rotate(0, 752.506, 1479.62)\" x=\"752.506\" y=\"1479.62\">0.1</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip9400)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:middle;\" transform=\"rotate(0, 1028.42, 1479.62)\" x=\"1028.42\" y=\"1479.62\">1</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip9400)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:middle;\" transform=\"rotate(0, 1304.33, 1479.62)\" x=\"1304.33\" y=\"1479.62\">5</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip9400)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:middle;\" transform=\"rotate(0, 1580.24, 1479.62)\" x=\"1580.24\" y=\"1479.62\">10</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip9400)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:middle;\" transform=\"rotate(0, 1856.15, 1479.62)\" x=\"1856.15\" y=\"1479.62\">25</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip9400)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:middle;\" transform=\"rotate(0, 2132.06, 1479.62)\" x=\"2132.06\" y=\"1479.62\">50</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip9400)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:end;\" transform=\"rotate(0, 231.898, 1406.21)\" x=\"231.898\" y=\"1406.21\">0</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip9400)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:end;\" transform=\"rotate(0, 231.898, 1172.63)\" x=\"231.898\" y=\"1172.63\">500</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip9400)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:end;\" transform=\"rotate(0, 231.898, 939.038)\" x=\"231.898\" y=\"939.038\">1000</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip9400)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:end;\" transform=\"rotate(0, 231.898, 705.449)\" x=\"231.898\" y=\"705.449\">1500</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip9400)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:end;\" transform=\"rotate(0, 231.898, 471.861)\" x=\"231.898\" y=\"471.861\">2000</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip9400)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:end;\" transform=\"rotate(0, 231.898, 238.272)\" x=\"231.898\" y=\"238.272\">2500</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip9400)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:84px; text-anchor:middle;\" transform=\"rotate(0, 1304.33, 73.2)\" x=\"1304.33\" y=\"73.2\">Computation time with varying proportions of data missing</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip9400)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:66px; text-anchor:middle;\" transform=\"rotate(0, 1304.33, 1559.48)\" x=\"1304.33\" y=\"1559.48\">% of data missing</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip9400)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:66px; text-anchor:middle;\" transform=\"rotate(-90, 89.2861, 773.647)\" x=\"89.2861\" y=\"773.647\">Time (seconds)</text>\n",
       "</g>\n",
       "<path clip-path=\"url(#clip9402)\" d=\"\n",
       "M366.231 1336.15 L366.231 1388.71 L410.377 1388.71 L410.377 1336.15 L366.231 1336.15 L366.231 1336.15  Z\n",
       "  \" fill=\"#009af9\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip9402)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  366.231,1336.15 366.231,1388.71 410.377,1388.71 410.377,1336.15 366.231,1336.15 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip9402)\" d=\"\n",
       "M642.141 1334.89 L642.141 1388.71 L686.287 1388.71 L686.287 1334.89 L642.141 1334.89 L642.141 1334.89  Z\n",
       "  \" fill=\"#009af9\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip9402)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  642.141,1334.89 642.141,1388.71 686.287,1388.71 686.287,1334.89 642.141,1334.89 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip9402)\" d=\"\n",
       "M918.052 1335.91 L918.052 1388.71 L962.198 1388.71 L962.198 1335.91 L918.052 1335.91 L918.052 1335.91  Z\n",
       "  \" fill=\"#009af9\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip9402)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  918.052,1335.91 918.052,1388.71 962.198,1388.71 962.198,1335.91 918.052,1335.91 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip9402)\" d=\"\n",
       "M1193.96 1329.48 L1193.96 1388.71 L1238.11 1388.71 L1238.11 1329.48 L1193.96 1329.48 L1193.96 1329.48  Z\n",
       "  \" fill=\"#009af9\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip9402)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1193.96,1329.48 1193.96,1388.71 1238.11,1388.71 1238.11,1329.48 1193.96,1329.48 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip9402)\" d=\"\n",
       "M1469.87 1330.25 L1469.87 1388.71 L1514.02 1388.71 L1514.02 1330.25 L1469.87 1330.25 L1469.87 1330.25  Z\n",
       "  \" fill=\"#009af9\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip9402)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1469.87,1330.25 1469.87,1388.71 1514.02,1388.71 1514.02,1330.25 1469.87,1330.25 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip9402)\" d=\"\n",
       "M1745.78 1333.41 L1745.78 1388.71 L1789.93 1388.71 L1789.93 1333.41 L1745.78 1333.41 L1745.78 1333.41  Z\n",
       "  \" fill=\"#009af9\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip9402)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1745.78,1333.41 1745.78,1388.71 1789.93,1388.71 1789.93,1333.41 1745.78,1333.41 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip9402)\" d=\"\n",
       "M2021.69 1388.71 L2021.69 1388.71 L2065.84 1388.71 L2065.84 1388.71 L2021.69 1388.71 L2021.69 1388.71  Z\n",
       "  \" fill=\"#009af9\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip9402)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  2021.69,1388.71 2021.69,1388.71 2065.84,1388.71 2021.69,1388.71 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip9402)\" d=\"\n",
       "M410.377 1341.73 L410.377 1388.71 L454.522 1388.71 L454.522 1341.73 L410.377 1341.73 L410.377 1341.73  Z\n",
       "  \" fill=\"#e26f46\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip9402)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  410.377,1341.73 410.377,1388.71 454.522,1388.71 454.522,1341.73 410.377,1341.73 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip9402)\" d=\"\n",
       "M686.287 1342.76 L686.287 1388.71 L730.433 1388.71 L730.433 1342.76 L686.287 1342.76 L686.287 1342.76  Z\n",
       "  \" fill=\"#e26f46\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip9402)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  686.287,1342.76 686.287,1388.71 730.433,1388.71 730.433,1342.76 686.287,1342.76 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip9402)\" d=\"\n",
       "M962.198 1342.15 L962.198 1388.71 L1006.34 1388.71 L1006.34 1342.15 L962.198 1342.15 L962.198 1342.15  Z\n",
       "  \" fill=\"#e26f46\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip9402)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  962.198,1342.15 962.198,1388.71 1006.34,1388.71 1006.34,1342.15 962.198,1342.15 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip9402)\" d=\"\n",
       "M1238.11 1334.22 L1238.11 1388.71 L1282.25 1388.71 L1282.25 1334.22 L1238.11 1334.22 L1238.11 1334.22  Z\n",
       "  \" fill=\"#e26f46\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip9402)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1238.11,1334.22 1238.11,1388.71 1282.25,1388.71 1282.25,1334.22 1238.11,1334.22 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip9402)\" d=\"\n",
       "M1514.02 1335.53 L1514.02 1388.71 L1558.16 1388.71 L1558.16 1335.53 L1514.02 1335.53 L1514.02 1335.53  Z\n",
       "  \" fill=\"#e26f46\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip9402)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1514.02,1335.53 1514.02,1388.71 1558.16,1388.71 1558.16,1335.53 1514.02,1335.53 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip9402)\" d=\"\n",
       "M1789.93 1337 L1789.93 1388.71 L1834.07 1388.71 L1834.07 1337 L1789.93 1337 L1789.93 1337  Z\n",
       "  \" fill=\"#e26f46\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip9402)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1789.93,1337 1789.93,1388.71 1834.07,1388.71 1834.07,1337 1789.93,1337 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip9402)\" d=\"\n",
       "M2065.84 1388.71 L2065.84 1388.71 L2109.99 1388.71 L2109.99 1388.71 L2065.84 1388.71 L2065.84 1388.71  Z\n",
       "  \" fill=\"#e26f46\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip9402)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  2065.84,1388.71 2065.84,1388.71 2109.99,1388.71 2065.84,1388.71 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip9402)\" d=\"\n",
       "M454.522 1341.49 L454.522 1388.71 L498.668 1388.71 L498.668 1341.49 L454.522 1341.49 L454.522 1341.49  Z\n",
       "  \" fill=\"#3da44d\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip9402)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  454.522,1341.49 454.522,1388.71 498.668,1388.71 498.668,1341.49 454.522,1341.49 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip9402)\" d=\"\n",
       "M730.433 1341.82 L730.433 1388.71 L774.579 1388.71 L774.579 1341.82 L730.433 1341.82 L730.433 1341.82  Z\n",
       "  \" fill=\"#3da44d\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip9402)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  730.433,1341.82 730.433,1388.71 774.579,1388.71 774.579,1341.82 730.433,1341.82 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip9402)\" d=\"\n",
       "M1006.34 1338.82 L1006.34 1388.71 L1050.49 1388.71 L1050.49 1338.82 L1006.34 1338.82 L1006.34 1338.82  Z\n",
       "  \" fill=\"#3da44d\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip9402)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1006.34,1338.82 1006.34,1388.71 1050.49,1388.71 1050.49,1338.82 1006.34,1338.82 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip9402)\" d=\"\n",
       "M1282.25 1334.54 L1282.25 1388.71 L1326.4 1388.71 L1326.4 1334.54 L1282.25 1334.54 L1282.25 1334.54  Z\n",
       "  \" fill=\"#3da44d\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip9402)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1282.25,1334.54 1282.25,1388.71 1326.4,1388.71 1326.4,1334.54 1282.25,1334.54 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip9402)\" d=\"\n",
       "M1558.16 1335 L1558.16 1388.71 L1602.31 1388.71 L1602.31 1335 L1558.16 1335 L1558.16 1335  Z\n",
       "  \" fill=\"#3da44d\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip9402)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1558.16,1335 1558.16,1388.71 1602.31,1388.71 1602.31,1335 1558.16,1335 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip9402)\" d=\"\n",
       "M1834.07 1335.4 L1834.07 1388.71 L1878.22 1388.71 L1878.22 1335.4 L1834.07 1335.4 L1834.07 1335.4  Z\n",
       "  \" fill=\"#3da44d\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip9402)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1834.07,1335.4 1834.07,1388.71 1878.22,1388.71 1878.22,1335.4 1834.07,1335.4 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip9402)\" d=\"\n",
       "M2109.99 1388.71 L2109.99 1388.71 L2154.13 1388.71 L2154.13 1388.71 L2109.99 1388.71 L2109.99 1388.71  Z\n",
       "  \" fill=\"#3da44d\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip9402)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  2109.99,1388.71 2109.99,1388.71 2154.13,1388.71 2109.99,1388.71 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip9402)\" d=\"\n",
       "M498.668 1064.02 L498.668 1388.71 L542.814 1388.71 L542.814 1064.02 L498.668 1064.02 L498.668 1064.02  Z\n",
       "  \" fill=\"#c271d2\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip9402)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  498.668,1064.02 498.668,1388.71 542.814,1388.71 542.814,1064.02 498.668,1064.02 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip9402)\" d=\"\n",
       "M774.579 1153.2 L774.579 1388.71 L818.724 1388.71 L818.724 1153.2 L774.579 1153.2 L774.579 1153.2  Z\n",
       "  \" fill=\"#c271d2\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip9402)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  774.579,1153.2 774.579,1388.71 818.724,1388.71 818.724,1153.2 774.579,1153.2 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip9402)\" d=\"\n",
       "M1050.49 1160.82 L1050.49 1388.71 L1094.63 1388.71 L1094.63 1160.82 L1050.49 1160.82 L1050.49 1160.82  Z\n",
       "  \" fill=\"#c271d2\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip9402)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1050.49,1160.82 1050.49,1388.71 1094.63,1388.71 1094.63,1160.82 1050.49,1160.82 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip9402)\" d=\"\n",
       "M1326.4 1166.25 L1326.4 1388.71 L1370.55 1388.71 L1370.55 1166.25 L1326.4 1166.25 L1326.4 1166.25  Z\n",
       "  \" fill=\"#c271d2\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip9402)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1326.4,1166.25 1326.4,1388.71 1370.55,1388.71 1370.55,1166.25 1326.4,1166.25 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip9402)\" d=\"\n",
       "M1602.31 1164.67 L1602.31 1388.71 L1646.46 1388.71 L1646.46 1164.67 L1602.31 1164.67 L1602.31 1164.67  Z\n",
       "  \" fill=\"#c271d2\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip9402)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1602.31,1164.67 1602.31,1388.71 1646.46,1388.71 1646.46,1164.67 1602.31,1164.67 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip9402)\" d=\"\n",
       "M1878.22 1075.67 L1878.22 1388.71 L1922.37 1388.71 L1922.37 1075.67 L1878.22 1075.67 L1878.22 1075.67  Z\n",
       "  \" fill=\"#c271d2\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip9402)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1878.22,1075.67 1878.22,1388.71 1922.37,1388.71 1922.37,1075.67 1878.22,1075.67 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip9402)\" d=\"\n",
       "M2154.13 1388.71 L2154.13 1388.71 L2198.28 1388.71 L2198.28 1388.71 L2154.13 1388.71 L2154.13 1388.71  Z\n",
       "  \" fill=\"#c271d2\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip9402)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  2154.13,1388.71 2154.13,1388.71 2198.28,1388.71 2154.13,1388.71 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip9402)\" d=\"\n",
       "M542.814 210.361 L542.814 1388.71 L586.959 1388.71 L586.959 210.361 L542.814 210.361 L542.814 210.361  Z\n",
       "  \" fill=\"#ac8d18\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip9402)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  542.814,210.361 542.814,1388.71 586.959,1388.71 586.959,210.361 542.814,210.361 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip9402)\" d=\"\n",
       "M818.724 186.786 L818.724 1388.71 L862.87 1388.71 L862.87 186.786 L818.724 186.786 L818.724 186.786  Z\n",
       "  \" fill=\"#ac8d18\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip9402)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  818.724,186.786 818.724,1388.71 862.87,1388.71 862.87,186.786 818.724,186.786 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip9402)\" d=\"\n",
       "M1094.63 160.668 L1094.63 1388.71 L1138.78 1388.71 L1138.78 160.668 L1094.63 160.668 L1094.63 160.668  Z\n",
       "  \" fill=\"#ac8d18\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip9402)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1094.63,160.668 1094.63,1388.71 1138.78,1388.71 1138.78,160.668 1094.63,160.668 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip9402)\" d=\"\n",
       "M1370.55 158.579 L1370.55 1388.71 L1414.69 1388.71 L1414.69 158.579 L1370.55 158.579 L1370.55 158.579  Z\n",
       "  \" fill=\"#ac8d18\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip9402)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1370.55,158.579 1370.55,1388.71 1414.69,1388.71 1414.69,158.579 1370.55,158.579 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip9402)\" d=\"\n",
       "M1646.46 169.774 L1646.46 1388.71 L1690.6 1388.71 L1690.6 169.774 L1646.46 169.774 L1646.46 169.774  Z\n",
       "  \" fill=\"#ac8d18\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip9402)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1646.46,169.774 1646.46,1388.71 1690.6,1388.71 1690.6,169.774 1646.46,169.774 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip9402)\" d=\"\n",
       "M1922.37 232.931 L1922.37 1388.71 L1966.51 1388.71 L1966.51 232.931 L1922.37 232.931 L1922.37 232.931  Z\n",
       "  \" fill=\"#ac8d18\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip9402)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1922.37,232.931 1922.37,1388.71 1966.51,1388.71 1966.51,232.931 1922.37,232.931 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip9402)\" d=\"\n",
       "M2198.28 1388.71 L2198.28 1388.71 L2242.42 1388.71 L2242.42 1388.71 L2198.28 1388.71 L2198.28 1388.71  Z\n",
       "  \" fill=\"#ac8d18\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip9402)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  2198.28,1388.71 2198.28,1388.71 2242.42,1388.71 2198.28,1388.71 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip9400)\" d=\"\n",
       "M1493.67 568.075 L2280.76 568.075 L2280.76 205.195 L1493.67 205.195  Z\n",
       "  \" fill=\"#ffffff\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip9400)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1493.67,568.075 2280.76,568.075 2280.76,205.195 1493.67,205.195 1493.67,568.075 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip9400)\" d=\"\n",
       "M1517.67 289.867 L1661.67 289.867 L1661.67 241.483 L1517.67 241.483 L1517.67 289.867  Z\n",
       "  \" fill=\"#009af9\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip9400)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1517.67,289.867 1661.67,289.867 1661.67,241.483 1517.67,241.483 1517.67,289.867 \n",
       "  \"/>\n",
       "<g clip-path=\"url(#clip9400)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:start;\" transform=\"rotate(0, 1685.67, 283.175)\" x=\"1685.67\" y=\"283.175\">MendelImpute (w = 100)</text>\n",
       "</g>\n",
       "<path clip-path=\"url(#clip9400)\" d=\"\n",
       "M1517.67 350.347 L1661.67 350.347 L1661.67 301.963 L1517.67 301.963 L1517.67 350.347  Z\n",
       "  \" fill=\"#e26f46\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip9400)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1517.67,350.347 1661.67,350.347 1661.67,301.963 1517.67,301.963 1517.67,350.347 \n",
       "  \"/>\n",
       "<g clip-path=\"url(#clip9400)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:start;\" transform=\"rotate(0, 1685.67, 343.655)\" x=\"1685.67\" y=\"343.655\">MendelImpute (w = 400)</text>\n",
       "</g>\n",
       "<path clip-path=\"url(#clip9400)\" d=\"\n",
       "M1517.67 410.827 L1661.67 410.827 L1661.67 362.443 L1517.67 362.443 L1517.67 410.827  Z\n",
       "  \" fill=\"#3da44d\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip9400)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1517.67,410.827 1661.67,410.827 1661.67,362.443 1517.67,362.443 1517.67,410.827 \n",
       "  \"/>\n",
       "<g clip-path=\"url(#clip9400)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:start;\" transform=\"rotate(0, 1685.67, 404.135)\" x=\"1685.67\" y=\"404.135\">MendelImpute (w = 1200)</text>\n",
       "</g>\n",
       "<path clip-path=\"url(#clip9400)\" d=\"\n",
       "M1517.67 471.307 L1661.67 471.307 L1661.67 422.923 L1517.67 422.923 L1517.67 471.307  Z\n",
       "  \" fill=\"#c271d2\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip9400)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1517.67,471.307 1661.67,471.307 1661.67,422.923 1517.67,422.923 1517.67,471.307 \n",
       "  \"/>\n",
       "<g clip-path=\"url(#clip9400)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:start;\" transform=\"rotate(0, 1685.67, 464.615)\" x=\"1685.67\" y=\"464.615\">Beagle 5.0</text>\n",
       "</g>\n",
       "<path clip-path=\"url(#clip9400)\" d=\"\n",
       "M1517.67 531.787 L1661.67 531.787 L1661.67 483.403 L1517.67 483.403 L1517.67 531.787  Z\n",
       "  \" fill=\"#ac8d18\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip9400)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1517.67,531.787 1661.67,531.787 1661.67,483.403 1517.67,483.403 1517.67,531.787 \n",
       "  \"/>\n",
       "<g clip-path=\"url(#clip9400)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:start;\" transform=\"rotate(0, 1685.67, 525.095)\" x=\"1685.67\" y=\"525.095\">Minimac 4</text>\n",
       "</g>\n",
       "</svg>\n"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "groupedbar(missingprop_name, [mendel_100_times mendel_400_times mendel_1200_times beagle5_times minimac4_times],\n",
    "    label = [\"MendelImpute (w = 100)\", \"MendelImpute (w = 400)\", \"MendelImpute (w = 1200)\", \"Beagle 5.0\", \"Minimac 4\"], \n",
    "    title = \"Computation time with varying proportions of data missing\",\n",
    "    ylabel = \"Time (seconds)\", xlabel = \"% of data missing\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "@webio": {
   "lastCommId": null,
   "lastKernelId": null
  },
  "kernelspec": {
   "display_name": "Julia 1.3.0-rc4",
   "language": "julia",
   "name": "julia-1.3"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.3.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
