{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Compare MendelImpute against Minimac4 and Beagle5 on simulated data\n",
    "\n",
    "Here the genotype matrix is simulated using real haplotypes from chromosome 22 of the 1000 Genomes Phase 1 data reference haplotypes. The entire dataset can be [downloaded here](ftp://share.sph.umich.edu/minimac3/G1K_P1_VCF_Files.tar.gz) (4.15GB).\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "┌ Warning: no Manifest.toml file found, static paths used\n",
      "└ @ Revise /Users/biona001/.julia/packages/Revise/439di/src/Revise.jl:1108\n",
      "┌ Info: Precompiling MendelImpute [e47305d1-6a61-5370-bc5d-77554d143183]\n",
      "└ @ Base loading.jl:1273\n"
     ]
    }
   ],
   "source": [
    "using Revise\n",
    "using VCFTools\n",
    "using MendelImpute\n",
    "using GeneticVariation\n",
    "using Random\n",
    "using SparseArrays\n",
    "using Plots"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Simulate data\n",
    "\n",
    "Reference haplotypes are obtained from the 1000 genomes project as above. We use this haplotype matrix to construct a genotype matrix matrix with 100 samples. The genotype matrix are randomly masked to generate the missing values. The result is stored in 3 files:\n",
    "\n",
    "+ `ALL.chr22.phase1_v3.snps_indels_svs.genotypes.all.noSingleton.vcf.gz`: haplotype reference files\n",
    "+ `target.vcf.gz`: complete genotype information\n",
    "+ `target_masked.vcf.gz`: the same as `target.vcf.gz` except some entries are masked"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " 57.531036 seconds (808.09 M allocations: 73.560 GiB, 17.98% gc time)\n",
      "  1.523623 seconds (4.43 k allocations: 405.812 MiB, 1.29% gc time)\n",
      "  0.780748 seconds (23 allocations: 871.764 MiB, 32.07% gc time)\n"
     ]
    }
   ],
   "source": [
    "# import haplotype matrix and simulate genotype matrix\n",
    "@time H = convert_ht(Bool, \"compare4/VCF_Files/ALL.chr22.phase1_v3.snps_indels_svs.genotypes.all.noSingleton.vcf.gz\")\n",
    "@time X = simulate_genotypes(H', 100)\n",
    "\n",
    "# randomly mask entries\n",
    "Random.seed!(2020)\n",
    "missingprop = 0.1\n",
    "snps, people = size(X)\n",
    "@time Xm = ifelse.(rand(snps, people) .< missingprop, missing, X);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  7.101095 seconds (805.33 k allocations: 37.253 MiB)\n",
      " 10.362518 seconds (733.25 k allocations: 33.556 MiB, 0.19% gc time)\n"
     ]
    }
   ],
   "source": [
    "# extract CHROM, POS, ID, REF, ALT info from ref VCF file\n",
    "vcffile = \"compare4/VCF_Files/ALL.chr22.phase1_v3.snps_indels_svs.genotypes.all.noSingleton.vcf.gz\"\n",
    "records = nrecords(vcffile)\n",
    "chrom   = [\"22\" for i in 1:records]\n",
    "pos     = zeros(Int, records)\n",
    "ids     = Vector{String}(undef, records)\n",
    "refs    = Vector{String}(undef, records)\n",
    "alts    = Vector{String}(undef, records)\n",
    "reader  = VCF.Reader(openvcf(vcffile, \"r\"))\n",
    "for (i, record) in enumerate(reader)\n",
    "    pos[i]   = VCF.pos(record)\n",
    "    ids[i]   = VCF.id(record)[1]\n",
    "    refs[i]  = VCF.ref(record)\n",
    "    alts[i]  = VCF.alt(record)[1]\n",
    "end\n",
    "\n",
    "# write simulated genotype matrices to disk\n",
    "@time make_tgtvcf_file(X, vcffilename=\"./compare4/target.vcf.gz\", marker_chrom=chrom, \n",
    "      marker_pos=pos, marker_ID=ids, marker_REF=refs, marker_ALT=alts)\n",
    "@time make_tgtvcf_file(Xm, vcffilename=\"./compare4/target_masked.vcf.gz\", marker_chrom=chrom, \n",
    "      marker_pos=pos, marker_ID=ids, marker_REF=refs, marker_ALT=alts)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# MendelImpute error"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "100×365644 Array{Union{Missing, Float32},2}:\n",
       " 0.0       0.0       0.0       0.0       …  0.0       0.0       0.0     \n",
       "  missing  0.0       0.0       0.0          0.0       0.0       0.0     \n",
       " 0.0        missing  0.0       0.0          0.0        missing  0.0     \n",
       " 0.0       0.0       0.0       0.0          0.0       0.0       0.0     \n",
       " 0.0       0.0       0.0        missing     0.0       0.0        missing\n",
       " 0.0       1.0       0.0       0.0       …  0.0       0.0        missing\n",
       " 0.0       1.0       0.0       0.0          0.0       0.0        missing\n",
       " 0.0       0.0       0.0       0.0          0.0        missing  0.0     \n",
       " 0.0       0.0       0.0       0.0          0.0       0.0       0.0     \n",
       " 0.0       0.0       0.0       0.0          0.0       0.0       0.0     \n",
       " 0.0        missing   missing  0.0       …  0.0       0.0       0.0     \n",
       " 0.0       0.0        missing  0.0          0.0       0.0       0.0     \n",
       " 0.0       0.0       0.0       0.0          0.0        missing  0.0     \n",
       " ⋮                                       ⋱                              \n",
       " 0.0       0.0       0.0       0.0          0.0       0.0       0.0     \n",
       " 0.0       1.0        missing  0.0          0.0        missing  0.0     \n",
       " 0.0       0.0       0.0       0.0       …  0.0       0.0       0.0     \n",
       " 0.0       0.0       0.0       0.0          0.0       0.0       0.0     \n",
       " 0.0       1.0       0.0        missing      missing  0.0       0.0     \n",
       " 0.0       0.0        missing  0.0          0.0       0.0       1.0     \n",
       " 0.0       0.0       0.0       0.0          0.0       0.0       0.0     \n",
       " 0.0       0.0       0.0       0.0       …  0.0       0.0       0.0     \n",
       " 0.0       0.0       0.0       0.0          0.0       0.0       0.0     \n",
       " 0.0       0.0       0.0       0.0          0.0       0.0       0.0     \n",
       " 1.0       1.0       1.0       0.0          0.0       0.0       0.0     \n",
       " 0.0       0.0       0.0       0.0          0.0       0.0       0.0     "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read original target data (all entries observed)\n",
    "X_complete = convert_gt(Float32, \"./compare4/target.vcf.gz\"; as_minorallele=false)\n",
    "Xm = convert_gt(Float32, \"./compare4/target_masked.vcf.gz\"; as_minorallele=false)\n",
    "Xm_original = copy(Xm)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Most accurate code with flanking windows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "362.337703 seconds (1.11 G allocations: 105.757 GiB, 3.03% gc time)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "7.077922788285874e-5"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# impute\n",
    "tgtfile = \"./compare4/target_masked.vcf.gz\"\n",
    "reffile = \"./compare4/VCF_Files/ALL.chr22.phase1_v3.snps_indels_svs.genotypes.all.noSingleton.vcf.gz\"\n",
    "outfile = \"./compare4/imputed_target.vcf.gz\"\n",
    "width   = 400\n",
    "@time hs, ph = phase(tgtfile, reffile, impute=true, outfile = outfile, width = width);\n",
    "\n",
    "# import imputed result and compare with true\n",
    "X_mendel = convert_gt(Float32, outfile, as_minorallele=false)\n",
    "n, p = size(X_mendel)\n",
    "error_rate = sum(X_complete .!= X_mendel) / p / n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Beagle 5 error\n",
    "\n",
    "Beagle requires >14GB of RAM to run. Thus it can only be done on hoffman."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# following took 24 minutes 22 seconds on n7080 of Hoffman2 using 30GB\n",
    "run(`java -Xmx14g -jar beagle.28Sep18.793.jar gt=./compare4/target_masked.vcf.gz ref=./compare4/VCF_Files/ALL.chr22.phase1_v3.snps_indels_svs.genotypes.all.noSingleton.vcf.gz out=./compare4/beagle.result`)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "6.837251534279245e-7"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# beagle 5 error rate\n",
    "X_complete = convert_gt(Float32, \"./compare4/target.vcf.gz\"; as_minorallele=false)\n",
    "X_beagle = convert_gt(Float32, \"./compare4/beagle.result.vcf.gz\", as_minorallele=false)\n",
    "n, p = size(X_complete)\n",
    "error_rate = sum(X_complete .!= X_beagle) / p / n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Minimac4 error\n",
    "\n",
    "Need to first convert reference vcf file to m3vcf using minimac3 (on Hoffman)\n",
    "\n",
    "```Julia\n",
    "minimac3 = \"/u/home/b/biona001/haplotype_comparisons/minimac3/Minimac3/bin/Minimac3\"\n",
    "@time run(`$minimac3 --refHaps ALL.chr22.phase1_v3.snps_indels_svs.genotypes.all.noSingleton.vcf.gz --processReference --prefix haplo_ref`)\n",
    "\n",
    "#on n4034: 2675.055412 seconds (15.31 k allocations: 861.116 KiB)\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      " -------------------------------------------------------------------------------- \n",
      "          Minimac4 - Fast Imputation Based on State Space Reduction HMM\n",
      " --------------------------------------------------------------------------------\n",
      "           (c) 2014 - Sayantan Das, Christian Fuchsberger, David Hinds\n",
      "                             Mary Kate Wing, Goncalo Abecasis \n",
      "\n",
      " Version: 1.0.2;\n",
      " Built: Mon Sep 30 11:52:22 PDT 2019 by biona001\n",
      "\n",
      " Command Line Options: \n",
      "       Reference Haplotypes : --refHaps [./compare4/ALL.chr22.phase1_v3.snps_indels_svs.genotypes.all.noSingleton.m3vcf.gz],\n",
      "                              --passOnly, --rsid, --referenceEstimates [ON],\n",
      "                              --mapFile [docs/geneticMapFile.b38.map.txt.gz]\n",
      "          Target Haplotypes : --haps [./compare4/target_masked.vcf.gz]\n",
      "          Output Parameters : --prefix [./compare4/minimac4.result],\n",
      "                              --estimate, --nobgzip, --vcfBuffer [200],\n",
      "                              --format [GT,DS], --allTypedSites, --meta,\n",
      "                              --memUsage\n",
      "        Chunking Parameters : --ChunkLengthMb [20.00], --ChunkOverlapMb [3.00]\n",
      "          Subset Parameters : --chr [], --start, --end, --window\n",
      "   Approximation Parameters : --minimac3, --probThreshold [0.01],\n",
      "                              --diffThreshold [0.01], --topThreshold [0.01]\n",
      "           Other Parameters : --log, --help, --cpus [1], --params\n",
      "                  PhoneHome : --noPhoneHome, --phoneHomeThinning [50]\n",
      "\n",
      "\n",
      " URL = http://genome.sph.umich.edu/wiki/Minimac4\n",
      " Starting Main Imputation/Estimation Analysis ... \n",
      "\n",
      " Performing preliminary check on input parameters... \n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                             PRELIMINARY FILE CHECK                            \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Checking GWAS haplotype file : ./compare4/target_masked.vcf.gz\n",
      "\n",
      " Gathering variant information ...\n",
      "\n",
      " Successful !!! \n",
      "\n",
      " Checking Reference haplotype file : ./compare4/ALL.chr22.phase1_v3.snps_indels_svs.genotypes.all.noSingleton.m3vcf.gz\n",
      "\n",
      " Gathering variant information ...\n",
      "\n",
      " Successful !!! \n",
      "\n",
      " Reference Panel   : Found 1092 samples (2184 haplotypes) and 365644 variants ...\n",
      "\n",
      " Target/GWAS Panel : Found 100 samples (200 haplotypes) and 365644 variants ...\n",
      "                     365644 variants overlap with Reference panel \n",
      "                     0 variants imported that exist only in Target/GWAS panel\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                           CHUNKING INFORMATION                           \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Chunking region into 1 chunk(s) with atleast 365644 variants in each chunk ... \n",
      "\n",
      " Details of chunks is given below ...\n",
      "\n",
      " No   LeftBuffer      LeftEnd   RightPoint  RightBuffer       #Sites(GWAS/Ref/%)\n",
      " -------------------------------------------------------------------------------\n",
      "  1     16050408     16050408     51243297     51243297   365644/  365644/ 100.00%\n",
      "\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                           MAIN IMPUTATION ANALYSIS                            \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Starting imputation analysis of 1 chunk(s) ... \n",
      "\n",
      " -------------------------------------------\n",
      " Analyzing Chunk 1/1 [22:16050408-51243297]\n",
      " -------------------------------------------\n",
      "\n",
      " Reading chunk from reference panel ... \n",
      " Reading chunk from target/GWAS panel ... \n",
      "\n",
      " Compressing reference panel at GWAS sites ... \n",
      " Re-compression successful (306 seconds) !!!\n",
      "\n",
      " Starting Imputation ...\n",
      "\n",
      "  Imputing Samples 1-100 [100%] out of 100 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "\n",
      " Imputation successful (996 seconds) !!!\n",
      "\n",
      " Appending chunk to final output VCF File :  ./compare4/minimac4.result.dose.vcf.gz \n",
      " Appending successful (4 seconds) !!!\n",
      "\n",
      " Time taken for this chunk = 1317 seconds.\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                              SUMMARY OF ANALYSIS                              \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Info file written to                    : ./compare4/minimac4.result.info\n",
      " Imputed VCF information written to      : ./compare4/minimac4.result.dose.vcf.gz\n",
      "\n",
      " Time Taken for Reading File             = 19 seconds\n",
      " Time Taken for Re-compression           = 306 seconds\n",
      " Time Taken for Imputation               = 986 seconds\n",
      " Time Taken for Writing File             = 14 seconds\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                                END OF PROGRAM                                 \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Program Successfully Implemented... \n",
      " \n",
      " Total Run completed in 0 hours, 22 mins, 12 seconds.\n",
      "\n",
      " Thank You for using Minimac4 !!! \n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Process(`\u001b[4m/Users/biona001/Benjamin_Folder/UCLA/research/softwares/Minimac4/build/minimac4\u001b[24m \u001b[4m--refHaps\u001b[24m \u001b[4m./compare4/ALL.chr22.phase1_v3.snps_indels_svs.genotypes.all.noSingleton.m3vcf.gz\u001b[24m \u001b[4m--haps\u001b[24m \u001b[4m./compare4/target_masked.vcf.gz\u001b[24m \u001b[4m--prefix\u001b[24m \u001b[4m./compare4/minimac4.result\u001b[24m`, ProcessExited(0))"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# run minimac 4\n",
    "minimac4 = \"/Users/biona001/Benjamin_Folder/UCLA/research/softwares/Minimac4/build/minimac4\"\n",
    "run(`$minimac4 --refHaps ./compare4/ALL.chr22.phase1_v3.snps_indels_svs.genotypes.all.noSingleton.m3vcf.gz --haps ./compare4/target_masked.vcf.gz --prefix ./compare4/minimac4.result`)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.011691317237531588"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X_complete = convert_gt(Float32, \"./compare4/target.vcf.gz\"; as_minorallele=false)\n",
    "X_minimac = convert_gt(Float32, \"./compare4/minimac4.result.dose.vcf.gz\", as_minorallele=false)\n",
    "n, p = size(X_complete)\n",
    "error_rate = sum(X_complete .!= X_minimac) / n / p"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Visualize error distribution\n",
    "\n",
    "Minimac4 does not preserve non-missing data! "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sum(X_complete[.!(missing_idx)] .!= X_mendel[.!(missing_idx)]) = 0\n",
      "sum(X_complete[.!(missing_idx)] .!= X_beagle[.!(missing_idx)]) = 0\n",
      "sum(X_complete[.!(missing_idx)] .!= X_minimac[.!(missing_idx)]) = 342631\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "342631"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# did mendel/beagle/minimac preserve non-missing data?\n",
    "missing_idx = ismissing.(Xm_original)\n",
    "X_complete = convert_gt(Float32, \"./compare4/target.vcf.gz\"; as_minorallele=false)\n",
    "@show sum(X_complete[.!missing_idx] .!= X_mendel[.!missing_idx])\n",
    "@show sum(X_complete[.!missing_idx] .!= X_beagle[.!missing_idx])\n",
    "@show sum(X_complete[.!missing_idx] .!= X_minimac[.!missing_idx])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sum(X_complete .!= X_mendel) = 2626\n",
      "sum(X_complete .!= X_beagle) = 25\n",
      "sum(X_complete .!= X_minimac) = 427486\n"
     ]
    }
   ],
   "source": [
    "# total number of wrong values\n",
    "@show sum(X_complete .!= X_mendel)\n",
    "@show sum(X_complete .!= X_beagle)\n",
    "@show sum(X_complete .!= X_minimac);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Mendel \n",
    "Random.seed!(2019)\n",
    "X_complete  = convert_gt(Float32, \"./compare4/target.vcf.gz\"; as_minorallele=false)\n",
    "X_mendel = convert_gt(Float32, \"./compare4/imputed_target.vcf.gz\", as_minorallele=false)\n",
    "mendel_disagreeing_entries = randn(size(X_complete)) .* sparse(X_complete .!= X_mendel)\n",
    "\n",
    "# Beagle 5.0\n",
    "X_beagle = convert_gt(Float32, \"./compare4/beagle.result.vcf.gz\", as_minorallele=false)\n",
    "beagle_disagreeing_entries = randn(size(X_complete)) .* sparse(X_complete .!= X_beagle)\n",
    "\n",
    "# Minimac 4\n",
    "X_minimac = convert_gt(Float32, \"./compare4/minimac4.result.dose.vcf.gz\", as_minorallele=false)\n",
    "minimac_disagreeing_entries = randn(size(X_complete)) .* sparse(X_complete .!= X_minimac);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "A = Plots.spy(mendel_disagreeing_entries, size=(4000,500), title=\"MendelImpute\", title_location=:left)\n",
    "# B = Plots.spy(beagle_disagreeing_entries, size=(4000,500), title=\"Beagle 5.0\", title_location=:left)\n",
    "# C = Plots.spy(minimac_disagreeing_entries, size=(4000,500), title=\"Minimac 4\", title_location=:left)\n",
    "# plot(A, B, C, layout = (3, 1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Error rate as a function of sample size"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "`full.vcf` contains 2000 samples, or 4000 haplotypes. Using this data, we randomly choose the last 100, 200, ..., 1900 samples as imputation targets and use the remaining haplotypes as reference panels. Then we compare the speed, memory, and accuracy of MendelImpute, Minimac4, and Beagle5 on these simulated datasets.\n",
    "\n",
    "### First generate the different datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " 18.239669 seconds (234.13 M allocations: 22.718 GiB, 15.19% gc time)\n",
      " 40.032299 seconds (557.31 M allocations: 43.346 GiB, 13.34% gc time)\n",
      "  0.520760 seconds (7.86 M allocations: 690.876 MiB, 21.26% gc time)\n",
      " 18.806437 seconds (252.12 M allocations: 23.842 GiB, 15.23% gc time)\n",
      " 39.840113 seconds (539.36 M allocations: 42.223 GiB, 13.37% gc time)\n",
      "  0.962408 seconds (15.08 M allocations: 1.312 GiB, 13.27% gc time)\n",
      " 19.772307 seconds (270.07 M allocations: 24.965 GiB, 14.63% gc time)\n",
      " 36.901654 seconds (521.42 M allocations: 41.100 GiB, 13.88% gc time)\n",
      "  1.395913 seconds (22.29 M allocations: 2.014 GiB, 16.38% gc time)\n",
      " 21.279525 seconds (288.01 M allocations: 26.088 GiB, 15.01% gc time)\n",
      " 35.330595 seconds (503.47 M allocations: 39.977 GiB, 13.63% gc time)\n",
      "  2.171800 seconds (29.47 M allocations: 2.585 GiB, 18.65% gc time)\n",
      " 25.166582 seconds (306.00 M allocations: 27.348 GiB, 14.81% gc time)\n",
      " 36.509108 seconds (485.52 M allocations: 38.853 GiB, 13.61% gc time)\n",
      "  2.290493 seconds (36.65 M allocations: 3.147 GiB, 16.88% gc time)\n",
      " 25.618424 seconds (323.95 M allocations: 28.472 GiB, 14.31% gc time)\n",
      " 35.568666 seconds (467.57 M allocations: 37.730 GiB, 14.39% gc time)\n",
      "  3.275990 seconds (43.87 M allocations: 3.987 GiB, 21.52% gc time)\n",
      " 25.755642 seconds (341.90 M allocations: 29.595 GiB, 14.25% gc time)\n",
      " 33.989288 seconds (449.62 M allocations: 36.607 GiB, 13.48% gc time)\n",
      "  3.521830 seconds (51.05 M allocations: 4.554 GiB, 17.25% gc time)\n",
      " 25.998202 seconds (359.84 M allocations: 30.718 GiB, 13.93% gc time)\n",
      " 32.608902 seconds (431.67 M allocations: 35.483 GiB, 13.60% gc time)\n",
      "  4.041639 seconds (58.23 M allocations: 5.118 GiB, 17.05% gc time)\n",
      " 29.123065 seconds (377.79 M allocations: 31.842 GiB, 13.98% gc time)\n",
      " 30.623139 seconds (413.73 M allocations: 34.360 GiB, 14.46% gc time)\n",
      "  3.964555 seconds (65.41 M allocations: 5.683 GiB, 15.90% gc time)\n",
      " 28.759874 seconds (395.78 M allocations: 33.237 GiB, 14.89% gc time)\n",
      " 28.805597 seconds (395.78 M allocations: 33.237 GiB, 14.15% gc time)\n",
      "  5.009076 seconds (72.59 M allocations: 6.247 GiB, 20.40% gc time)\n",
      " 29.997224 seconds (413.73 M allocations: 34.360 GiB, 13.59% gc time)\n",
      " 27.576301 seconds (377.79 M allocations: 31.842 GiB, 13.64% gc time)\n",
      "  5.212575 seconds (79.81 M allocations: 7.364 GiB, 16.07% gc time)\n",
      " 32.132909 seconds (431.67 M allocations: 35.484 GiB, 14.39% gc time)\n",
      " 26.490320 seconds (359.84 M allocations: 30.719 GiB, 13.90% gc time)\n",
      "  5.898000 seconds (86.99 M allocations: 7.928 GiB, 16.60% gc time)\n",
      " 33.655245 seconds (449.62 M allocations: 36.607 GiB, 12.92% gc time)\n",
      " 26.216797 seconds (341.90 M allocations: 29.595 GiB, 13.45% gc time)\n",
      "  6.349447 seconds (94.17 M allocations: 8.493 GiB, 16.42% gc time)\n",
      " 34.971595 seconds (467.57 M allocations: 37.730 GiB, 13.64% gc time)\n",
      " 23.604109 seconds (323.95 M allocations: 28.472 GiB, 14.60% gc time)\n",
      "  6.070191 seconds (101.35 M allocations: 9.057 GiB, 16.87% gc time)\n",
      " 36.766114 seconds (485.52 M allocations: 38.853 GiB, 13.52% gc time)\n",
      " 22.650548 seconds (306.00 M allocations: 27.348 GiB, 14.68% gc time)\n",
      "  7.099656 seconds (108.53 M allocations: 9.624 GiB, 18.77% gc time)\n",
      " 37.162359 seconds (503.47 M allocations: 39.977 GiB, 13.37% gc time)\n",
      " 21.472322 seconds (288.01 M allocations: 26.088 GiB, 14.95% gc time)\n",
      "  7.354280 seconds (115.71 M allocations: 10.189 GiB, 17.13% gc time)\n",
      " 37.161832 seconds (521.42 M allocations: 41.100 GiB, 12.99% gc time)\n",
      " 19.784935 seconds (270.07 M allocations: 24.965 GiB, 14.59% gc time)\n",
      "  7.852530 seconds (122.89 M allocations: 10.753 GiB, 16.53% gc time)\n",
      " 39.879124 seconds (539.36 M allocations: 42.223 GiB, 13.29% gc time)\n",
      " 18.788132 seconds (252.12 M allocations: 23.842 GiB, 14.65% gc time)\n",
      "  8.424203 seconds (130.07 M allocations: 11.318 GiB, 17.05% gc time)\n",
      " 40.704197 seconds (557.31 M allocations: 43.346 GiB, 13.65% gc time)\n",
      " 18.108025 seconds (234.13 M allocations: 22.718 GiB, 15.86% gc time)\n",
      "  8.965865 seconds (137.25 M allocations: 11.884 GiB, 19.05% gc time)\n"
     ]
    }
   ],
   "source": [
    "records, samples = nrecords(\"full.vcf\"), nsamples(\"full.vcf\")\n",
    "for i in 1:19\n",
    "    # compute target and reference index\n",
    "    to_keep = 100i\n",
    "    tgt_index = falses(samples)\n",
    "    tgt_index[samples-to_keep+1:end] .= true\n",
    "    ref_index = .!tgt_index\n",
    "    record_index = trues(records) # save all records (SNPs) \n",
    "\n",
    "    # create target.vcf.gz and haplo_ref.vcf.gz\n",
    "    tgtfile = \"./compare1/$(to_keep)_target.vcf.gz\"\n",
    "    reffile = \"./compare1/$(to_keep)_haplo_ref.vcf.gz\"\n",
    "    @time VCFTools.filter(\"full.vcf\", record_index, tgt_index, des = tgtfile)\n",
    "    @time VCFTools.filter(\"full.vcf\", record_index, ref_index, des = reffile)\n",
    "\n",
    "    # import full target matrix. Also transpose so that columns are samples. \n",
    "    @time X = convert_gt(Float32, tgtfile; as_minorallele=false)\n",
    "    X = copy(X')\n",
    "\n",
    "    # mask 10% entries\n",
    "    p, n = size(X)\n",
    "    Random.seed!(123)\n",
    "    missingprop = 0.1\n",
    "    X .= ifelse.(rand(Float32, p, n) .< missingprop, missing, X)\n",
    "    masks = ismissing.(X)\n",
    "\n",
    "    # save X to new VCF file\n",
    "    mask_gt(tgtfile, masks, des=\"./compare1/$(to_keep)_target_masked.vcf.gz\")\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Convert vcf to m3vcf on Hoffman (Julia 1.2)\n",
    "\n",
    "Each takes ~5 min. This is needed because Minimac4 only runs their own M3VCF files (which requires Minimac3)\n",
    "\n",
    "```Julia\n",
    "for i in 1:19\n",
    "    minimac3 = \"/u/home/b/biona001/haplotype_comparisons/minimac3/Minimac3/bin/Minimac3\"\n",
    "    @time run(`$minimac3 --refHaps $(100i)_haplo_ref.vcf.gz --processReference --prefix $(100i)_haplo_ref`)\n",
    "end\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Analyze:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "mendel_1200_error   = zeros(19)\n",
    "mendel_1200_times   = zeros(19) #sec\n",
    "mendel_1200_mem     = zeros(19) #MB\n",
    "for i in 1:19\n",
    "    #\n",
    "    # current run info\n",
    "    #\n",
    "    X_complete = convert_gt(Float32, \"./compare1/$(100i)_target.vcf.gz\"; as_minorallele=false)\n",
    "    X_masked   = convert_gt(Float32, \"./compare1/$(100i)_target_masked.vcf.gz\"; as_minorallele=false)\n",
    "    missing_idx = ismissing.(X_masked)\n",
    "    total_missing = sum(missing_idx)\n",
    "    actual_missing_values = X_complete[missing_idx]  #true values of missing entries\n",
    "    #\n",
    "    # MendelImpute (window = 1200)\n",
    "    #\n",
    "    tgtfile = \"./compare1/$(100i)_target_masked.vcf.gz\"\n",
    "    reffile = \"./compare1/$(100i)_haplo_ref.vcf.gz\"\n",
    "    outfile = \"./compare1/$(100i)_mendel.result.vcf.gz\"\n",
    "    width   = 1200\n",
    "    mendel_result = @timed phase(tgtfile, reffile, impute=true, outfile = outfile, width = width)\n",
    "    X_mendel = convert_gt(Float32, outfile, as_minorallele=false)\n",
    "    imputed_missing_values = X_mendel[missing_idx] #imputed values of missing entries\n",
    "    mendel_err = sum(actual_missing_values .!= imputed_missing_values) / total_missing\n",
    "    mendel_1200_error[i] = mendel_err\n",
    "    mendel_1200_times[i] = mendel_result[2] \n",
    "    mendel_1200_mem[i]   = mendel_result[3] / 1e6\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "beagle.28Sep18.793.jar (version 5.0)\n",
      "Copyright (C) 2014-2018 Brian L. Browning\n",
      "Enter \"java -jar beagle.28Sep18.793.jar\" to list command line argument\n",
      "Start time: 01:22 AM PST on 11 Dec 2019\n",
      "\n",
      "Command line: java -Xmx3641m -jar beagle.28Sep18.793.jar\n",
      "  gt=./sample_size/100_target_masked.vcf.gz\n",
      "  ref=./sample_size/100_haplo_ref.vcf.gz\n",
      "  out=./sample_size/100_beagle.result\n",
      "\n",
      "No genetic map is specified: using 1 cM = 1 Mb\n",
      "\n",
      "Reference samples:       1,900\n",
      "Study samples:             100\n",
      "\n",
      "Window 1 (1:36-4999683)\n",
      "Reference markers:      35,897\n",
      "Study markers:          35,897\n",
      "\n",
      "Burnin  iteration 1:           9 seconds\n",
      "Burnin  iteration 2:           5 seconds\n",
      "Burnin  iteration 3:           3 seconds\n",
      "Burnin  iteration 4:           4 seconds\n",
      "Burnin  iteration 5:           3 seconds\n",
      "Burnin  iteration 6:           4 seconds\n",
      "\n",
      "Phasing iteration 1:           2 seconds\n",
      "Phasing iteration 2:           2 seconds\n",
      "Phasing iteration 3:           2 seconds\n",
      "Phasing iteration 4:           2 seconds\n",
      "Phasing iteration 5:           2 seconds\n",
      "Phasing iteration 6:           2 seconds\n",
      "Phasing iteration 7:           2 seconds\n",
      "Phasing iteration 8:           2 seconds\n",
      "Phasing iteration 9:           2 seconds\n",
      "Phasing iteration 10:          2 seconds\n",
      "Phasing iteration 11:          2 seconds\n",
      "Phasing iteration 12:          2 seconds\n",
      "\n",
      "Cumulative Statistics:\n",
      "\n",
      "Study markers:          35,897\n",
      "\n",
      "Haplotype phasing time:        52 seconds\n",
      "Total time:                    56 seconds\n",
      "\n",
      "End time: 01:23 AM PST on 11 Dec 2019\n",
      "beagle.28Sep18.793.jar finished\n",
      "\n",
      "\n",
      " -------------------------------------------------------------------------------- \n",
      "          Minimac4 - Fast Imputation Based on State Space Reduction HMM\n",
      " --------------------------------------------------------------------------------\n",
      "           (c) 2014 - Sayantan Das, Christian Fuchsberger, David Hinds\n",
      "                             Mary Kate Wing, Goncalo Abecasis \n",
      "\n",
      " Version: 1.0.2;\n",
      " Built: Mon Sep 30 11:52:22 PDT 2019 by biona001\n",
      "\n",
      " Command Line Options: \n",
      "       Reference Haplotypes : --refHaps [./sample_size/100_haplo_ref.m3vcf.gz],\n",
      "                              --passOnly, --rsid, --referenceEstimates [ON],\n",
      "                              --mapFile [docs/geneticMapFile.b38.map.txt.gz]\n",
      "          Target Haplotypes : --haps [./sample_size/100_target_masked.vcf.gz]\n",
      "          Output Parameters : --prefix [./sample_size/100_minimac4.result],\n",
      "                              --estimate, --nobgzip, --vcfBuffer [200],\n",
      "                              --format [GT,DS], --allTypedSites, --meta,\n",
      "                              --memUsage\n",
      "        Chunking Parameters : --ChunkLengthMb [20.00], --ChunkOverlapMb [3.00]\n",
      "          Subset Parameters : --chr [], --start, --end, --window\n",
      "   Approximation Parameters : --minimac3, --probThreshold [0.01],\n",
      "                              --diffThreshold [0.01], --topThreshold [0.01]\n",
      "           Other Parameters : --log, --help, --cpus [1], --params\n",
      "                  PhoneHome : --noPhoneHome, --phoneHomeThinning [50]\n",
      "\n",
      "\n",
      " URL = http://genome.sph.umich.edu/wiki/Minimac4\n",
      " Starting Main Imputation/Estimation Analysis ... \n",
      "\n",
      " Performing preliminary check on input parameters... \n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                             PRELIMINARY FILE CHECK                            \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Checking GWAS haplotype file : ./sample_size/100_target_masked.vcf.gz\n",
      "\n",
      " Gathering variant information ...\n",
      "\n",
      " Successful !!! \n",
      "\n",
      " Checking Reference haplotype file : ./sample_size/100_haplo_ref.m3vcf.gz\n",
      "\n",
      " Gathering variant information ...\n",
      "\n",
      " Successful !!! \n",
      "\n",
      " Reference Panel   : Found 1900 samples (3800 haplotypes) and 35897 variants ...\n",
      "\n",
      " Target/GWAS Panel : Found 100 samples (200 haplotypes) and 35897 variants ...\n",
      "                     35897 variants overlap with Reference panel \n",
      "                     0 variants imported that exist only in Target/GWAS panel\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                           CHUNKING INFORMATION                           \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Chunking region into 1 chunk(s) with atleast 35897 variants in each chunk ... \n",
      "\n",
      " Details of chunks is given below ...\n",
      "\n",
      " No   LeftBuffer      LeftEnd   RightPoint  RightBuffer       #Sites(GWAS/Ref/%)\n",
      " -------------------------------------------------------------------------------\n",
      "  1           36           36      4999683      4999683    35897/   35897/ 100.00%\n",
      "\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                           MAIN IMPUTATION ANALYSIS                            \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Starting imputation analysis of 1 chunk(s) ... \n",
      "\n",
      " -------------------------------------------\n",
      " Analyzing Chunk 1/1 [1:36-4999683]\n",
      " -------------------------------------------\n",
      "\n",
      " Reading chunk from reference panel ... \n",
      " Reading chunk from target/GWAS panel ... \n",
      "\n",
      " Compressing reference panel at GWAS sites ... \n",
      " Re-compression successful (72 seconds) !!!\n",
      "\n",
      " Starting Imputation ...\n",
      "\n",
      "  Imputing Samples 1-100 [100%] out of 100 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "\n",
      " Imputation successful (77 seconds) !!!\n",
      "\n",
      " Appending chunk to final output VCF File :  ./sample_size/100_minimac4.result.dose.vcf.gz \n",
      " Appending successful (0 seconds) !!!\n",
      "\n",
      " Time taken for this chunk = 150 seconds.\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                              SUMMARY OF ANALYSIS                              \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Info file written to                    : ./sample_size/100_minimac4.result.info\n",
      " Imputed VCF information written to      : ./sample_size/100_minimac4.result.dose.vcf.gz\n",
      "\n",
      " Time Taken for Reading File             = 2 seconds\n",
      " Time Taken for Re-compression           = 72 seconds\n",
      " Time Taken for Imputation               = 76 seconds\n",
      " Time Taken for Writing File             = 1 seconds\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                                END OF PROGRAM                                 \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Program Successfully Implemented... \n",
      " \n",
      " Total Run completed in 0 hours, 2 mins, 31 seconds.\n",
      "\n",
      " Thank You for using Minimac4 !!! \n",
      "\n",
      "beagle.28Sep18.793.jar (version 5.0)\n",
      "Copyright (C) 2014-2018 Brian L. Browning\n",
      "Enter \"java -jar beagle.28Sep18.793.jar\" to list command line argument\n",
      "Start time: 01:26 AM PST on 11 Dec 2019\n",
      "\n",
      "Command line: java -Xmx3641m -jar beagle.28Sep18.793.jar\n",
      "  gt=./sample_size/200_target_masked.vcf.gz\n",
      "  ref=./sample_size/200_haplo_ref.vcf.gz\n",
      "  out=./sample_size/200_beagle.result\n",
      "\n",
      "No genetic map is specified: using 1 cM = 1 Mb\n",
      "\n",
      "Reference samples:       1,800\n",
      "Study samples:             200\n",
      "\n",
      "Window 1 (1:36-4999683)\n",
      "Reference markers:      35,897\n",
      "Study markers:          35,897\n",
      "\n",
      "Burnin  iteration 1:           16 seconds\n",
      "Burnin  iteration 2:           6 seconds\n",
      "Burnin  iteration 3:           4 seconds\n",
      "Burnin  iteration 4:           5 seconds\n",
      "Burnin  iteration 5:           6 seconds\n",
      "Burnin  iteration 6:           6 seconds\n",
      "\n",
      "Phasing iteration 1:           3 seconds\n",
      "Phasing iteration 2:           4 seconds\n",
      "Phasing iteration 3:           3 seconds\n",
      "Phasing iteration 4:           3 seconds\n",
      "Phasing iteration 5:           4 seconds\n",
      "Phasing iteration 6:           4 seconds\n",
      "Phasing iteration 7:           3 seconds\n",
      "Phasing iteration 8:           3 seconds\n",
      "Phasing iteration 9:           3 seconds\n",
      "Phasing iteration 10:          3 seconds\n",
      "Phasing iteration 11:          3 seconds\n",
      "Phasing iteration 12:          3 seconds\n",
      "\n",
      "Cumulative Statistics:\n",
      "\n",
      "Study markers:          35,897\n",
      "\n",
      "Haplotype phasing time:        1 minute 22 seconds\n",
      "Total time:                    1 minute 27 seconds\n",
      "\n",
      "End time: 01:27 AM PST on 11 Dec 2019\n",
      "beagle.28Sep18.793.jar finished\n",
      "\n",
      "\n",
      " -------------------------------------------------------------------------------- \n",
      "          Minimac4 - Fast Imputation Based on State Space Reduction HMM\n",
      " --------------------------------------------------------------------------------\n",
      "           (c) 2014 - Sayantan Das, Christian Fuchsberger, David Hinds\n",
      "                             Mary Kate Wing, Goncalo Abecasis \n",
      "\n",
      " Version: 1.0.2;\n",
      " Built: Mon Sep 30 11:52:22 PDT 2019 by biona001\n",
      "\n",
      " Command Line Options: \n",
      "       Reference Haplotypes : --refHaps [./sample_size/200_haplo_ref.m3vcf.gz],\n",
      "                              --passOnly, --rsid, --referenceEstimates [ON],\n",
      "                              --mapFile [docs/geneticMapFile.b38.map.txt.gz]\n",
      "          Target Haplotypes : --haps [./sample_size/200_target_masked.vcf.gz]\n",
      "          Output Parameters : --prefix [./sample_size/200_minimac4.result],\n",
      "                              --estimate, --nobgzip, --vcfBuffer [200],\n",
      "                              --format [GT,DS], --allTypedSites, --meta,\n",
      "                              --memUsage\n",
      "        Chunking Parameters : --ChunkLengthMb [20.00], --ChunkOverlapMb [3.00]\n",
      "          Subset Parameters : --chr [], --start, --end, --window\n",
      "   Approximation Parameters : --minimac3, --probThreshold [0.01],\n",
      "                              --diffThreshold [0.01], --topThreshold [0.01]\n",
      "           Other Parameters : --log, --help, --cpus [1], --params\n",
      "                  PhoneHome : --noPhoneHome, --phoneHomeThinning [50]\n",
      "\n",
      "\n",
      " URL = http://genome.sph.umich.edu/wiki/Minimac4\n",
      " Starting Main Imputation/Estimation Analysis ... \n",
      "\n",
      " Performing preliminary check on input parameters... \n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                             PRELIMINARY FILE CHECK                            \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Checking GWAS haplotype file : ./sample_size/200_target_masked.vcf.gz\n",
      "\n",
      " Gathering variant information ...\n",
      "\n",
      " Successful !!! \n",
      "\n",
      " Checking Reference haplotype file : ./sample_size/200_haplo_ref.m3vcf.gz\n",
      "\n",
      " Gathering variant information ...\n",
      "\n",
      " Successful !!! \n",
      "\n",
      " Reference Panel   : Found 1800 samples (3600 haplotypes) and 35897 variants ...\n",
      "\n",
      " Target/GWAS Panel : Found 200 samples (400 haplotypes) and 35897 variants ...\n",
      "                     35897 variants overlap with Reference panel \n",
      "                     0 variants imported that exist only in Target/GWAS panel\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                           CHUNKING INFORMATION                           \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Chunking region into 1 chunk(s) with atleast 35897 variants in each chunk ... \n",
      "\n",
      " Details of chunks is given below ...\n",
      "\n",
      " No   LeftBuffer      LeftEnd   RightPoint  RightBuffer       #Sites(GWAS/Ref/%)\n",
      " -------------------------------------------------------------------------------\n",
      "  1           36           36      4999683      4999683    35897/   35897/ 100.00%\n",
      "\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                           MAIN IMPUTATION ANALYSIS                            \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Starting imputation analysis of 1 chunk(s) ... \n",
      "\n",
      " -------------------------------------------\n",
      " Analyzing Chunk 1/1 [1:36-4999683]\n",
      " -------------------------------------------\n",
      "\n",
      " Reading chunk from reference panel ... \n",
      " Reading chunk from target/GWAS panel ... \n",
      "\n",
      " Compressing reference panel at GWAS sites ... \n",
      " Re-compression successful (65 seconds) !!!\n",
      "\n",
      " Starting Imputation ...\n",
      "\n",
      "  Imputing Samples 1-200 [100%] out of 200 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "\n",
      " Imputation successful (141 seconds) !!!\n",
      "\n",
      " Appending chunk to final output VCF File :  ./sample_size/200_minimac4.result.dose.vcf.gz \n",
      " Appending successful (1 seconds) !!!\n",
      "\n",
      " Time taken for this chunk = 209 seconds.\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                              SUMMARY OF ANALYSIS                              \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Info file written to                    : ./sample_size/200_minimac4.result.info\n",
      " Imputed VCF information written to      : ./sample_size/200_minimac4.result.dose.vcf.gz\n",
      "\n",
      " Time Taken for Reading File             = 3 seconds\n",
      " Time Taken for Re-compression           = 65 seconds\n",
      " Time Taken for Imputation               = 139 seconds\n",
      " Time Taken for Writing File             = 3 seconds\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                                END OF PROGRAM                                 \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Program Successfully Implemented... \n",
      " \n",
      " Total Run completed in 0 hours, 3 mins, 30 seconds.\n",
      "\n",
      " Thank You for using Minimac4 !!! \n",
      "\n",
      "beagle.28Sep18.793.jar (version 5.0)\n",
      "Copyright (C) 2014-2018 Brian L. Browning\n",
      "Enter \"java -jar beagle.28Sep18.793.jar\" to list command line argument\n",
      "Start time: 01:31 AM PST on 11 Dec 2019\n",
      "\n",
      "Command line: java -Xmx3641m -jar beagle.28Sep18.793.jar\n",
      "  gt=./sample_size/300_target_masked.vcf.gz\n",
      "  ref=./sample_size/300_haplo_ref.vcf.gz\n",
      "  out=./sample_size/300_beagle.result\n",
      "\n",
      "No genetic map is specified: using 1 cM = 1 Mb\n",
      "\n",
      "Reference samples:       1,700\n",
      "Study samples:             300\n",
      "\n",
      "Window 1 (1:36-4999683)\n",
      "Reference markers:      35,897\n",
      "Study markers:          35,897\n",
      "\n",
      "Burnin  iteration 1:           20 seconds\n",
      "Burnin  iteration 2:           9 seconds\n",
      "Burnin  iteration 3:           6 seconds\n",
      "Burnin  iteration 4:           8 seconds\n",
      "Burnin  iteration 5:           8 seconds\n",
      "Burnin  iteration 6:           8 seconds\n",
      "\n",
      "Phasing iteration 1:           5 seconds\n",
      "Phasing iteration 2:           5 seconds\n",
      "Phasing iteration 3:           5 seconds\n",
      "Phasing iteration 4:           5 seconds\n",
      "Phasing iteration 5:           5 seconds\n",
      "Phasing iteration 6:           5 seconds\n",
      "Phasing iteration 7:           5 seconds\n",
      "Phasing iteration 8:           5 seconds\n",
      "Phasing iteration 9:           5 seconds\n",
      "Phasing iteration 10:          5 seconds\n",
      "Phasing iteration 11:          5 seconds\n",
      "Phasing iteration 12:          4 seconds\n",
      "\n",
      "Cumulative Statistics:\n",
      "\n",
      "Study markers:          35,897\n",
      "\n",
      "Haplotype phasing time:        1 minute 59 seconds\n",
      "Total time:                    2 minutes 6 seconds\n",
      "\n",
      "End time: 01:33 AM PST on 11 Dec 2019\n",
      "beagle.28Sep18.793.jar finished\n",
      "\n",
      "\n",
      " -------------------------------------------------------------------------------- \n",
      "          Minimac4 - Fast Imputation Based on State Space Reduction HMM\n",
      " --------------------------------------------------------------------------------\n",
      "           (c) 2014 - Sayantan Das, Christian Fuchsberger, David Hinds\n",
      "                             Mary Kate Wing, Goncalo Abecasis \n",
      "\n",
      " Version: 1.0.2;\n",
      " Built: Mon Sep 30 11:52:22 PDT 2019 by biona001\n",
      "\n",
      " Command Line Options: \n",
      "       Reference Haplotypes : --refHaps [./sample_size/300_haplo_ref.m3vcf.gz],\n",
      "                              --passOnly, --rsid, --referenceEstimates [ON],\n",
      "                              --mapFile [docs/geneticMapFile.b38.map.txt.gz]\n",
      "          Target Haplotypes : --haps [./sample_size/300_target_masked.vcf.gz]\n",
      "          Output Parameters : --prefix [./sample_size/300_minimac4.result],\n",
      "                              --estimate, --nobgzip, --vcfBuffer [200],\n",
      "                              --format [GT,DS], --allTypedSites, --meta,\n",
      "                              --memUsage\n",
      "        Chunking Parameters : --ChunkLengthMb [20.00], --ChunkOverlapMb [3.00]\n",
      "          Subset Parameters : --chr [], --start, --end, --window\n",
      "   Approximation Parameters : --minimac3, --probThreshold [0.01],\n",
      "                              --diffThreshold [0.01], --topThreshold [0.01]\n",
      "           Other Parameters : --log, --help, --cpus [1], --params\n",
      "                  PhoneHome : --noPhoneHome, --phoneHomeThinning [50]\n",
      "\n",
      "\n",
      " URL = http://genome.sph.umich.edu/wiki/Minimac4\n",
      " Starting Main Imputation/Estimation Analysis ... \n",
      "\n",
      " Performing preliminary check on input parameters... \n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                             PRELIMINARY FILE CHECK                            \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Checking GWAS haplotype file : ./sample_size/300_target_masked.vcf.gz\n",
      "\n",
      " Gathering variant information ...\n",
      "\n",
      " Successful !!! \n",
      "\n",
      " Checking Reference haplotype file : ./sample_size/300_haplo_ref.m3vcf.gz\n",
      "\n",
      " Gathering variant information ...\n",
      "\n",
      " Successful !!! \n",
      "\n",
      " Reference Panel   : Found 1700 samples (3400 haplotypes) and 35897 variants ...\n",
      "\n",
      " Target/GWAS Panel : Found 300 samples (600 haplotypes) and 35897 variants ...\n",
      "                     35897 variants overlap with Reference panel \n",
      "                     0 variants imported that exist only in Target/GWAS panel\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                           CHUNKING INFORMATION                           \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Chunking region into 1 chunk(s) with atleast 35897 variants in each chunk ... \n",
      "\n",
      " Details of chunks is given below ...\n",
      "\n",
      " No   LeftBuffer      LeftEnd   RightPoint  RightBuffer       #Sites(GWAS/Ref/%)\n",
      " -------------------------------------------------------------------------------\n",
      "  1           36           36      4999683      4999683    35897/   35897/ 100.00%\n",
      "\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                           MAIN IMPUTATION ANALYSIS                            \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Starting imputation analysis of 1 chunk(s) ... \n",
      "\n",
      " -------------------------------------------\n",
      " Analyzing Chunk 1/1 [1:36-4999683]\n",
      " -------------------------------------------\n",
      "\n",
      " Reading chunk from reference panel ... \n",
      " Reading chunk from target/GWAS panel ... \n",
      "\n",
      " Compressing reference panel at GWAS sites ... \n",
      " Re-compression successful (60 seconds) !!!\n",
      "\n",
      " Starting Imputation ...\n",
      "\n",
      "  Imputing Samples 1-200 [67%] out of 300 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 201-300 [100%] out of 300 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "\n",
      " Imputation successful (205 seconds) !!!\n",
      "\n",
      " Appending chunk to final output VCF File :  ./sample_size/300_minimac4.result.dose.vcf.gz \n",
      " Appending successful (1 seconds) !!!\n",
      "\n",
      " Time taken for this chunk = 269 seconds.\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                              SUMMARY OF ANALYSIS                              \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Info file written to                    : ./sample_size/300_minimac4.result.info\n",
      " Imputed VCF information written to      : ./sample_size/300_minimac4.result.dose.vcf.gz\n",
      "\n",
      " Time Taken for Reading File             = 4 seconds\n",
      " Time Taken for Re-compression           = 60 seconds\n",
      " Time Taken for Imputation               = 202 seconds\n",
      " Time Taken for Writing File             = 4 seconds\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                                END OF PROGRAM                                 \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Program Successfully Implemented... \n",
      " \n",
      " Total Run completed in 0 hours, 4 mins, 30 seconds.\n",
      "\n",
      " Thank You for using Minimac4 !!! \n",
      "\n",
      "beagle.28Sep18.793.jar (version 5.0)\n",
      "Copyright (C) 2014-2018 Brian L. Browning\n",
      "Enter \"java -jar beagle.28Sep18.793.jar\" to list command line argument\n",
      "Start time: 01:38 AM PST on 11 Dec 2019\n",
      "\n",
      "Command line: java -Xmx3641m -jar beagle.28Sep18.793.jar\n",
      "  gt=./sample_size/400_target_masked.vcf.gz\n",
      "  ref=./sample_size/400_haplo_ref.vcf.gz\n",
      "  out=./sample_size/400_beagle.result\n",
      "\n",
      "No genetic map is specified: using 1 cM = 1 Mb\n",
      "\n",
      "Reference samples:       1,600\n",
      "Study samples:             400\n",
      "\n",
      "Window 1 (1:36-4999683)\n",
      "Reference markers:      35,897\n",
      "Study markers:          35,897\n",
      "\n",
      "Burnin  iteration 1:           23 seconds\n",
      "Burnin  iteration 2:           11 seconds\n",
      "Burnin  iteration 3:           6 seconds\n",
      "Burnin  iteration 4:           9 seconds\n",
      "Burnin  iteration 5:           10 seconds\n",
      "Burnin  iteration 6:           10 seconds\n",
      "\n",
      "Phasing iteration 1:           6 seconds\n",
      "Phasing iteration 2:           6 seconds\n",
      "Phasing iteration 3:           6 seconds\n",
      "Phasing iteration 4:           5 seconds\n",
      "Phasing iteration 5:           6 seconds\n",
      "Phasing iteration 6:           6 seconds\n",
      "Phasing iteration 7:           6 seconds\n",
      "Phasing iteration 8:           5 seconds\n",
      "Phasing iteration 9:           5 seconds\n",
      "Phasing iteration 10:          5 seconds\n",
      "Phasing iteration 11:          5 seconds\n",
      "Phasing iteration 12:          5 seconds\n",
      "\n",
      "Cumulative Statistics:\n",
      "\n",
      "Study markers:          35,897\n",
      "\n",
      "Haplotype phasing time:        2 minutes 14 seconds\n",
      "Total time:                    2 minutes 23 seconds\n",
      "\n",
      "End time: 01:41 AM PST on 11 Dec 2019\n",
      "beagle.28Sep18.793.jar finished\n",
      "\n",
      "\n",
      " -------------------------------------------------------------------------------- \n",
      "          Minimac4 - Fast Imputation Based on State Space Reduction HMM\n",
      " --------------------------------------------------------------------------------\n",
      "           (c) 2014 - Sayantan Das, Christian Fuchsberger, David Hinds\n",
      "                             Mary Kate Wing, Goncalo Abecasis \n",
      "\n",
      " Version: 1.0.2;\n",
      " Built: Mon Sep 30 11:52:22 PDT 2019 by biona001\n",
      "\n",
      " Command Line Options: \n",
      "       Reference Haplotypes : --refHaps [./sample_size/400_haplo_ref.m3vcf.gz],\n",
      "                              --passOnly, --rsid, --referenceEstimates [ON],\n",
      "                              --mapFile [docs/geneticMapFile.b38.map.txt.gz]\n",
      "          Target Haplotypes : --haps [./sample_size/400_target_masked.vcf.gz]\n",
      "          Output Parameters : --prefix [./sample_size/400_minimac4.result],\n",
      "                              --estimate, --nobgzip, --vcfBuffer [200],\n",
      "                              --format [GT,DS], --allTypedSites, --meta,\n",
      "                              --memUsage\n",
      "        Chunking Parameters : --ChunkLengthMb [20.00], --ChunkOverlapMb [3.00]\n",
      "          Subset Parameters : --chr [], --start, --end, --window\n",
      "   Approximation Parameters : --minimac3, --probThreshold [0.01],\n",
      "                              --diffThreshold [0.01], --topThreshold [0.01]\n",
      "           Other Parameters : --log, --help, --cpus [1], --params\n",
      "                  PhoneHome : --noPhoneHome, --phoneHomeThinning [50]\n",
      "\n",
      "\n",
      " URL = http://genome.sph.umich.edu/wiki/Minimac4\n",
      " Starting Main Imputation/Estimation Analysis ... \n",
      "\n",
      " Performing preliminary check on input parameters... \n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                             PRELIMINARY FILE CHECK                            \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Checking GWAS haplotype file : ./sample_size/400_target_masked.vcf.gz\n",
      "\n",
      " Gathering variant information ...\n",
      "\n",
      " Successful !!! \n",
      "\n",
      " Checking Reference haplotype file : ./sample_size/400_haplo_ref.m3vcf.gz\n",
      "\n",
      " Gathering variant information ...\n",
      "\n",
      " Successful !!! \n",
      "\n",
      " Reference Panel   : Found 1600 samples (3200 haplotypes) and 35897 variants ...\n",
      "\n",
      " Target/GWAS Panel : Found 400 samples (800 haplotypes) and 35897 variants ...\n",
      "                     35897 variants overlap with Reference panel \n",
      "                     0 variants imported that exist only in Target/GWAS panel\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                           CHUNKING INFORMATION                           \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Chunking region into 1 chunk(s) with atleast 35897 variants in each chunk ... \n",
      "\n",
      " Details of chunks is given below ...\n",
      "\n",
      " No   LeftBuffer      LeftEnd   RightPoint  RightBuffer       #Sites(GWAS/Ref/%)\n",
      " -------------------------------------------------------------------------------\n",
      "  1           36           36      4999683      4999683    35897/   35897/ 100.00%\n",
      "\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                           MAIN IMPUTATION ANALYSIS                            \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Starting imputation analysis of 1 chunk(s) ... \n",
      "\n",
      " -------------------------------------------\n",
      " Analyzing Chunk 1/1 [1:36-4999683]\n",
      " -------------------------------------------\n",
      "\n",
      " Reading chunk from reference panel ... \n",
      " Reading chunk from target/GWAS panel ... \n",
      "\n",
      " Compressing reference panel at GWAS sites ... \n",
      " Re-compression successful (56 seconds) !!!\n",
      "\n",
      " Starting Imputation ...\n",
      "\n",
      "  Imputing Samples 1-200 [50%] out of 400 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 201-400 [100%] out of 400 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "\n",
      " Imputation successful (263 seconds) !!!\n",
      "\n",
      " Appending chunk to final output VCF File :  ./sample_size/400_minimac4.result.dose.vcf.gz \n",
      " Appending successful (2 seconds) !!!\n",
      "\n",
      " Time taken for this chunk = 324 seconds.\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                              SUMMARY OF ANALYSIS                              \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Info file written to                    : ./sample_size/400_minimac4.result.info\n",
      " Imputed VCF information written to      : ./sample_size/400_minimac4.result.dose.vcf.gz\n",
      "\n",
      " Time Taken for Reading File             = 4 seconds\n",
      " Time Taken for Re-compression           = 56 seconds\n",
      " Time Taken for Imputation               = 260 seconds\n",
      " Time Taken for Writing File             = 5 seconds\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                                END OF PROGRAM                                 \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Program Successfully Implemented... \n",
      " \n",
      " Total Run completed in 0 hours, 5 mins, 26 seconds.\n",
      "\n",
      " Thank You for using Minimac4 !!! \n",
      "\n",
      "beagle.28Sep18.793.jar (version 5.0)\n",
      "Copyright (C) 2014-2018 Brian L. Browning\n",
      "Enter \"java -jar beagle.28Sep18.793.jar\" to list command line argument\n",
      "Start time: 01:47 AM PST on 11 Dec 2019\n",
      "\n",
      "Command line: java -Xmx3641m -jar beagle.28Sep18.793.jar\n",
      "  gt=./sample_size/500_target_masked.vcf.gz\n",
      "  ref=./sample_size/500_haplo_ref.vcf.gz\n",
      "  out=./sample_size/500_beagle.result\n",
      "\n",
      "No genetic map is specified: using 1 cM = 1 Mb\n",
      "\n",
      "Reference samples:       1,500\n",
      "Study samples:             500\n",
      "\n",
      "Window 1 (1:36-4999683)\n",
      "Reference markers:      35,897\n",
      "Study markers:          35,897\n",
      "\n",
      "Burnin  iteration 1:           30 seconds\n",
      "Burnin  iteration 2:           15 seconds\n",
      "Burnin  iteration 3:           7 seconds\n",
      "Burnin  iteration 4:           10 seconds\n",
      "Burnin  iteration 5:           12 seconds\n",
      "Burnin  iteration 6:           12 seconds\n",
      "\n",
      "Phasing iteration 1:           7 seconds\n",
      "Phasing iteration 2:           7 seconds\n",
      "Phasing iteration 3:           7 seconds\n",
      "Phasing iteration 4:           7 seconds\n",
      "Phasing iteration 5:           7 seconds\n",
      "Phasing iteration 6:           7 seconds\n",
      "Phasing iteration 7:           7 seconds\n",
      "Phasing iteration 8:           7 seconds\n",
      "Phasing iteration 9:           7 seconds\n",
      "Phasing iteration 10:          6 seconds\n",
      "Phasing iteration 11:          6 seconds\n",
      "Phasing iteration 12:          6 seconds\n",
      "\n",
      "Cumulative Statistics:\n",
      "\n",
      "Study markers:          35,897\n",
      "\n",
      "Haplotype phasing time:        2 minutes 48 seconds\n",
      "Total time:                    2 minutes 58 seconds\n",
      "\n",
      "End time: 01:49 AM PST on 11 Dec 2019\n",
      "beagle.28Sep18.793.jar finished\n",
      "\n",
      "\n",
      " -------------------------------------------------------------------------------- \n",
      "          Minimac4 - Fast Imputation Based on State Space Reduction HMM\n",
      " --------------------------------------------------------------------------------\n",
      "           (c) 2014 - Sayantan Das, Christian Fuchsberger, David Hinds\n",
      "                             Mary Kate Wing, Goncalo Abecasis \n",
      "\n",
      " Version: 1.0.2;\n",
      " Built: Mon Sep 30 11:52:22 PDT 2019 by biona001\n",
      "\n",
      " Command Line Options: \n",
      "       Reference Haplotypes : --refHaps [./sample_size/500_haplo_ref.m3vcf.gz],\n",
      "                              --passOnly, --rsid, --referenceEstimates [ON],\n",
      "                              --mapFile [docs/geneticMapFile.b38.map.txt.gz]\n",
      "          Target Haplotypes : --haps [./sample_size/500_target_masked.vcf.gz]\n",
      "          Output Parameters : --prefix [./sample_size/500_minimac4.result],\n",
      "                              --estimate, --nobgzip, --vcfBuffer [200],\n",
      "                              --format [GT,DS], --allTypedSites, --meta,\n",
      "                              --memUsage\n",
      "        Chunking Parameters : --ChunkLengthMb [20.00], --ChunkOverlapMb [3.00]\n",
      "          Subset Parameters : --chr [], --start, --end, --window\n",
      "   Approximation Parameters : --minimac3, --probThreshold [0.01],\n",
      "                              --diffThreshold [0.01], --topThreshold [0.01]\n",
      "           Other Parameters : --log, --help, --cpus [1], --params\n",
      "                  PhoneHome : --noPhoneHome, --phoneHomeThinning [50]\n",
      "\n",
      "\n",
      " URL = http://genome.sph.umich.edu/wiki/Minimac4\n",
      " Starting Main Imputation/Estimation Analysis ... \n",
      "\n",
      " Performing preliminary check on input parameters... \n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                             PRELIMINARY FILE CHECK                            \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Checking GWAS haplotype file : ./sample_size/500_target_masked.vcf.gz\n",
      "\n",
      " Gathering variant information ...\n",
      "\n",
      " Successful !!! \n",
      "\n",
      " Checking Reference haplotype file : ./sample_size/500_haplo_ref.m3vcf.gz\n",
      "\n",
      " Gathering variant information ...\n",
      "\n",
      " Successful !!! \n",
      "\n",
      " Reference Panel   : Found 1500 samples (3000 haplotypes) and 35897 variants ...\n",
      "\n",
      " Target/GWAS Panel : Found 500 samples (1000 haplotypes) and 35897 variants ...\n",
      "                     35897 variants overlap with Reference panel \n",
      "                     0 variants imported that exist only in Target/GWAS panel\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                           CHUNKING INFORMATION                           \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Chunking region into 1 chunk(s) with atleast 35897 variants in each chunk ... \n",
      "\n",
      " Details of chunks is given below ...\n",
      "\n",
      " No   LeftBuffer      LeftEnd   RightPoint  RightBuffer       #Sites(GWAS/Ref/%)\n",
      " -------------------------------------------------------------------------------\n",
      "  1           36           36      4999683      4999683    35897/   35897/ 100.00%\n",
      "\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                           MAIN IMPUTATION ANALYSIS                            \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Starting imputation analysis of 1 chunk(s) ... \n",
      "\n",
      " -------------------------------------------\n",
      " Analyzing Chunk 1/1 [1:36-4999683]\n",
      " -------------------------------------------\n",
      "\n",
      " Reading chunk from reference panel ... \n",
      " Reading chunk from target/GWAS panel ... \n",
      "\n",
      " Compressing reference panel at GWAS sites ... \n",
      " Re-compression successful (51 seconds) !!!\n",
      "\n",
      " Starting Imputation ...\n",
      "\n",
      "  Imputing Samples 1-200 [40%] out of 500 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 201-400 [80%] out of 500 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 401-500 [100%] out of 500 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "\n",
      " Imputation successful (318 seconds) !!!\n",
      "\n",
      " Appending chunk to final output VCF File :  ./sample_size/500_minimac4.result.dose.vcf.gz \n",
      " Appending successful (2 seconds) !!!\n",
      "\n",
      " Time taken for this chunk = 375 seconds.\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                              SUMMARY OF ANALYSIS                              \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Info file written to                    : ./sample_size/500_minimac4.result.info\n",
      " Imputed VCF information written to      : ./sample_size/500_minimac4.result.dose.vcf.gz\n",
      "\n",
      " Time Taken for Reading File             = 6 seconds\n",
      " Time Taken for Re-compression           = 51 seconds\n",
      " Time Taken for Imputation               = 314 seconds\n",
      " Time Taken for Writing File             = 6 seconds\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                                END OF PROGRAM                                 \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Program Successfully Implemented... \n",
      " \n",
      " Total Run completed in 0 hours, 6 mins, 18 seconds.\n",
      "\n",
      " Thank You for using Minimac4 !!! \n",
      "\n",
      "beagle.28Sep18.793.jar (version 5.0)\n",
      "Copyright (C) 2014-2018 Brian L. Browning\n",
      "Enter \"java -jar beagle.28Sep18.793.jar\" to list command line argument\n",
      "Start time: 01:57 AM PST on 11 Dec 2019\n",
      "\n",
      "Command line: java -Xmx3641m -jar beagle.28Sep18.793.jar\n",
      "  gt=./sample_size/600_target_masked.vcf.gz\n",
      "  ref=./sample_size/600_haplo_ref.vcf.gz\n",
      "  out=./sample_size/600_beagle.result\n",
      "\n",
      "No genetic map is specified: using 1 cM = 1 Mb\n",
      "\n",
      "Reference samples:       1,400\n",
      "Study samples:             600\n",
      "\n",
      "Window 1 (1:36-4999683)\n",
      "Reference markers:      35,897\n",
      "Study markers:          35,897\n",
      "\n",
      "Burnin  iteration 1:           32 seconds\n",
      "Burnin  iteration 2:           19 seconds\n",
      "Burnin  iteration 3:           8 seconds\n",
      "Burnin  iteration 4:           12 seconds\n",
      "Burnin  iteration 5:           14 seconds\n",
      "Burnin  iteration 6:           15 seconds\n",
      "\n",
      "Phasing iteration 1:           9 seconds\n",
      "Phasing iteration 2:           9 seconds\n",
      "Phasing iteration 3:           8 seconds\n",
      "Phasing iteration 4:           8 seconds\n",
      "Phasing iteration 5:           8 seconds\n",
      "Phasing iteration 6:           8 seconds\n",
      "Phasing iteration 7:           9 seconds\n",
      "Phasing iteration 8:           8 seconds\n",
      "Phasing iteration 9:           8 seconds\n",
      "Phasing iteration 10:          8 seconds\n",
      "Phasing iteration 11:          7 seconds\n",
      "Phasing iteration 12:          7 seconds\n",
      "\n",
      "Cumulative Statistics:\n",
      "\n",
      "Study markers:          35,897\n",
      "\n",
      "Haplotype phasing time:        3 minutes 15 seconds\n",
      "Total time:                    3 minutes 27 seconds\n",
      "\n",
      "End time: 02:00 AM PST on 11 Dec 2019\n",
      "beagle.28Sep18.793.jar finished\n",
      "\n",
      "\n",
      " -------------------------------------------------------------------------------- \n",
      "          Minimac4 - Fast Imputation Based on State Space Reduction HMM\n",
      " --------------------------------------------------------------------------------\n",
      "           (c) 2014 - Sayantan Das, Christian Fuchsberger, David Hinds\n",
      "                             Mary Kate Wing, Goncalo Abecasis \n",
      "\n",
      " Version: 1.0.2;\n",
      " Built: Mon Sep 30 11:52:22 PDT 2019 by biona001\n",
      "\n",
      " Command Line Options: \n",
      "       Reference Haplotypes : --refHaps [./sample_size/600_haplo_ref.m3vcf.gz],\n",
      "                              --passOnly, --rsid, --referenceEstimates [ON],\n",
      "                              --mapFile [docs/geneticMapFile.b38.map.txt.gz]\n",
      "          Target Haplotypes : --haps [./sample_size/600_target_masked.vcf.gz]\n",
      "          Output Parameters : --prefix [./sample_size/600_minimac4.result],\n",
      "                              --estimate, --nobgzip, --vcfBuffer [200],\n",
      "                              --format [GT,DS], --allTypedSites, --meta,\n",
      "                              --memUsage\n",
      "        Chunking Parameters : --ChunkLengthMb [20.00], --ChunkOverlapMb [3.00]\n",
      "          Subset Parameters : --chr [], --start, --end, --window\n",
      "   Approximation Parameters : --minimac3, --probThreshold [0.01],\n",
      "                              --diffThreshold [0.01], --topThreshold [0.01]\n",
      "           Other Parameters : --log, --help, --cpus [1], --params\n",
      "                  PhoneHome : --noPhoneHome, --phoneHomeThinning [50]\n",
      "\n",
      "\n",
      " URL = http://genome.sph.umich.edu/wiki/Minimac4\n",
      " Starting Main Imputation/Estimation Analysis ... \n",
      "\n",
      " Performing preliminary check on input parameters... \n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                             PRELIMINARY FILE CHECK                            \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Checking GWAS haplotype file : ./sample_size/600_target_masked.vcf.gz\n",
      "\n",
      " Gathering variant information ...\n",
      "\n",
      " Successful !!! \n",
      "\n",
      " Checking Reference haplotype file : ./sample_size/600_haplo_ref.m3vcf.gz\n",
      "\n",
      " Gathering variant information ...\n",
      "\n",
      " Successful !!! \n",
      "\n",
      " Reference Panel   : Found 1400 samples (2800 haplotypes) and 35897 variants ...\n",
      "\n",
      " Target/GWAS Panel : Found 600 samples (1200 haplotypes) and 35897 variants ...\n",
      "                     35897 variants overlap with Reference panel \n",
      "                     0 variants imported that exist only in Target/GWAS panel\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                           CHUNKING INFORMATION                           \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Chunking region into 1 chunk(s) with atleast 35897 variants in each chunk ... \n",
      "\n",
      " Details of chunks is given below ...\n",
      "\n",
      " No   LeftBuffer      LeftEnd   RightPoint  RightBuffer       #Sites(GWAS/Ref/%)\n",
      " -------------------------------------------------------------------------------\n",
      "  1           36           36      4999683      4999683    35897/   35897/ 100.00%\n",
      "\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                           MAIN IMPUTATION ANALYSIS                            \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Starting imputation analysis of 1 chunk(s) ... \n",
      "\n",
      " -------------------------------------------\n",
      " Analyzing Chunk 1/1 [1:36-4999683]\n",
      " -------------------------------------------\n",
      "\n",
      " Reading chunk from reference panel ... \n",
      " Reading chunk from target/GWAS panel ... \n",
      "\n",
      " Compressing reference panel at GWAS sites ... \n",
      " Re-compression successful (47 seconds) !!!\n",
      "\n",
      " Starting Imputation ...\n",
      "\n",
      "  Imputing Samples 1-200 [33%] out of 600 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 201-400 [67%] out of 600 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 401-600 [100%] out of 600 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "\n",
      " Imputation successful (370 seconds) !!!\n",
      "\n",
      " Appending chunk to final output VCF File :  ./sample_size/600_minimac4.result.dose.vcf.gz \n",
      " Appending successful (2 seconds) !!!\n",
      "\n",
      " Time taken for this chunk = 424 seconds.\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                              SUMMARY OF ANALYSIS                              \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Info file written to                    : ./sample_size/600_minimac4.result.info\n",
      " Imputed VCF information written to      : ./sample_size/600_minimac4.result.dose.vcf.gz\n",
      "\n",
      " Time Taken for Reading File             = 7 seconds\n",
      " Time Taken for Re-compression           = 47 seconds\n",
      " Time Taken for Imputation               = 365 seconds\n",
      " Time Taken for Writing File             = 7 seconds\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                                END OF PROGRAM                                 \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Program Successfully Implemented... \n",
      " \n",
      " Total Run completed in 0 hours, 7 mins, 7 seconds.\n",
      "\n",
      " Thank You for using Minimac4 !!! \n",
      "\n",
      "beagle.28Sep18.793.jar (version 5.0)\n",
      "Copyright (C) 2014-2018 Brian L. Browning\n",
      "Enter \"java -jar beagle.28Sep18.793.jar\" to list command line argument\n",
      "Start time: 02:08 AM PST on 11 Dec 2019\n",
      "\n",
      "Command line: java -Xmx3641m -jar beagle.28Sep18.793.jar\n",
      "  gt=./sample_size/700_target_masked.vcf.gz\n",
      "  ref=./sample_size/700_haplo_ref.vcf.gz\n",
      "  out=./sample_size/700_beagle.result\n",
      "\n",
      "No genetic map is specified: using 1 cM = 1 Mb\n",
      "\n",
      "Reference samples:       1,300\n",
      "Study samples:             700\n",
      "\n",
      "Window 1 (1:36-4999683)\n",
      "Reference markers:      35,897\n",
      "Study markers:          35,897\n",
      "\n",
      "Burnin  iteration 1:           43 seconds\n",
      "Burnin  iteration 2:           19 seconds\n",
      "Burnin  iteration 3:           10 seconds\n",
      "Burnin  iteration 4:           14 seconds\n",
      "Burnin  iteration 5:           17 seconds\n",
      "Burnin  iteration 6:           17 seconds\n",
      "\n",
      "Phasing iteration 1:           10 seconds\n",
      "Phasing iteration 2:           10 seconds\n",
      "Phasing iteration 3:           10 seconds\n",
      "Phasing iteration 4:           10 seconds\n",
      "Phasing iteration 5:           9 seconds\n",
      "Phasing iteration 6:           9 seconds\n",
      "Phasing iteration 7:           9 seconds\n",
      "Phasing iteration 8:           9 seconds\n",
      "Phasing iteration 9:           9 seconds\n",
      "Phasing iteration 10:          9 seconds\n",
      "Phasing iteration 11:          8 seconds\n",
      "Phasing iteration 12:          8 seconds\n",
      "\n",
      "Cumulative Statistics:\n",
      "\n",
      "Study markers:          35,897\n",
      "\n",
      "Haplotype phasing time:        3 minutes 50 seconds\n",
      "Total time:                    4 minutes 4 seconds\n",
      "\n",
      "End time: 02:12 AM PST on 11 Dec 2019\n",
      "beagle.28Sep18.793.jar finished\n",
      "\n",
      "\n",
      " -------------------------------------------------------------------------------- \n",
      "          Minimac4 - Fast Imputation Based on State Space Reduction HMM\n",
      " --------------------------------------------------------------------------------\n",
      "           (c) 2014 - Sayantan Das, Christian Fuchsberger, David Hinds\n",
      "                             Mary Kate Wing, Goncalo Abecasis \n",
      "\n",
      " Version: 1.0.2;\n",
      " Built: Mon Sep 30 11:52:22 PDT 2019 by biona001\n",
      "\n",
      " Command Line Options: \n",
      "       Reference Haplotypes : --refHaps [./sample_size/700_haplo_ref.m3vcf.gz],\n",
      "                              --passOnly, --rsid, --referenceEstimates [ON],\n",
      "                              --mapFile [docs/geneticMapFile.b38.map.txt.gz]\n",
      "          Target Haplotypes : --haps [./sample_size/700_target_masked.vcf.gz]\n",
      "          Output Parameters : --prefix [./sample_size/700_minimac4.result],\n",
      "                              --estimate, --nobgzip, --vcfBuffer [200],\n",
      "                              --format [GT,DS], --allTypedSites, --meta,\n",
      "                              --memUsage\n",
      "        Chunking Parameters : --ChunkLengthMb [20.00], --ChunkOverlapMb [3.00]\n",
      "          Subset Parameters : --chr [], --start, --end, --window\n",
      "   Approximation Parameters : --minimac3, --probThreshold [0.01],\n",
      "                              --diffThreshold [0.01], --topThreshold [0.01]\n",
      "           Other Parameters : --log, --help, --cpus [1], --params\n",
      "                  PhoneHome : --noPhoneHome, --phoneHomeThinning [50]\n",
      "\n",
      "\n",
      " URL = http://genome.sph.umich.edu/wiki/Minimac4\n",
      " Starting Main Imputation/Estimation Analysis ... \n",
      "\n",
      " Performing preliminary check on input parameters... \n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                             PRELIMINARY FILE CHECK                            \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Checking GWAS haplotype file : ./sample_size/700_target_masked.vcf.gz\n",
      "\n",
      " Gathering variant information ...\n",
      "\n",
      " Successful !!! \n",
      "\n",
      " Checking Reference haplotype file : ./sample_size/700_haplo_ref.m3vcf.gz\n",
      "\n",
      " Gathering variant information ...\n",
      "\n",
      " Successful !!! \n",
      "\n",
      " Reference Panel   : Found 1300 samples (2600 haplotypes) and 35897 variants ...\n",
      "\n",
      " Target/GWAS Panel : Found 700 samples (1400 haplotypes) and 35897 variants ...\n",
      "                     35897 variants overlap with Reference panel \n",
      "                     0 variants imported that exist only in Target/GWAS panel\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                           CHUNKING INFORMATION                           \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Chunking region into 1 chunk(s) with atleast 35897 variants in each chunk ... \n",
      "\n",
      " Details of chunks is given below ...\n",
      "\n",
      " No   LeftBuffer      LeftEnd   RightPoint  RightBuffer       #Sites(GWAS/Ref/%)\n",
      " -------------------------------------------------------------------------------\n",
      "  1           36           36      4999683      4999683    35897/   35897/ 100.00%\n",
      "\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                           MAIN IMPUTATION ANALYSIS                            \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Starting imputation analysis of 1 chunk(s) ... \n",
      "\n",
      " -------------------------------------------\n",
      " Analyzing Chunk 1/1 [1:36-4999683]\n",
      " -------------------------------------------\n",
      "\n",
      " Reading chunk from reference panel ... \n",
      " Reading chunk from target/GWAS panel ... \n",
      "\n",
      " Compressing reference panel at GWAS sites ... \n",
      " Re-compression successful (43 seconds) !!!\n",
      "\n",
      " Starting Imputation ...\n",
      "\n",
      "  Imputing Samples 1-200 [29%] out of 700 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 201-400 [57%] out of 700 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 401-600 [86%] out of 700 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 601-700 [100%] out of 700 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "\n",
      " Imputation successful (413 seconds) !!!\n",
      "\n",
      " Appending chunk to final output VCF File :  ./sample_size/700_minimac4.result.dose.vcf.gz \n",
      " Appending successful (2 seconds) !!!\n",
      "\n",
      " Time taken for this chunk = 463 seconds.\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                              SUMMARY OF ANALYSIS                              \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Info file written to                    : ./sample_size/700_minimac4.result.info\n",
      " Imputed VCF information written to      : ./sample_size/700_minimac4.result.dose.vcf.gz\n",
      "\n",
      " Time Taken for Reading File             = 8 seconds\n",
      " Time Taken for Re-compression           = 43 seconds\n",
      " Time Taken for Imputation               = 408 seconds\n",
      " Time Taken for Writing File             = 7 seconds\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                                END OF PROGRAM                                 \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Program Successfully Implemented... \n",
      " \n",
      " Total Run completed in 0 hours, 7 mins, 46 seconds.\n",
      "\n",
      " Thank You for using Minimac4 !!! \n",
      "\n",
      "beagle.28Sep18.793.jar (version 5.0)\n",
      "Copyright (C) 2014-2018 Brian L. Browning\n",
      "Enter \"java -jar beagle.28Sep18.793.jar\" to list command line argument\n",
      "Start time: 02:20 AM PST on 11 Dec 2019\n",
      "\n",
      "Command line: java -Xmx3641m -jar beagle.28Sep18.793.jar\n",
      "  gt=./sample_size/800_target_masked.vcf.gz\n",
      "  ref=./sample_size/800_haplo_ref.vcf.gz\n",
      "  out=./sample_size/800_beagle.result\n",
      "\n",
      "No genetic map is specified: using 1 cM = 1 Mb\n",
      "\n",
      "Reference samples:       1,200\n",
      "Study samples:             800\n",
      "\n",
      "Window 1 (1:36-4999683)\n",
      "Reference markers:      35,897\n",
      "Study markers:          35,897\n",
      "\n",
      "Burnin  iteration 1:           43 seconds\n",
      "Burnin  iteration 2:           26 seconds\n",
      "Burnin  iteration 3:           10 seconds\n",
      "Burnin  iteration 4:           16 seconds\n",
      "Burnin  iteration 5:           18 seconds\n",
      "Burnin  iteration 6:           19 seconds\n",
      "\n",
      "Phasing iteration 1:           11 seconds\n",
      "Phasing iteration 2:           10 seconds\n",
      "Phasing iteration 3:           11 seconds\n",
      "Phasing iteration 4:           10 seconds\n",
      "Phasing iteration 5:           10 seconds\n",
      "Phasing iteration 6:           10 seconds\n",
      "Phasing iteration 7:           10 seconds\n",
      "Phasing iteration 8:           10 seconds\n",
      "Phasing iteration 9:           10 seconds\n",
      "Phasing iteration 10:          10 seconds\n",
      "Phasing iteration 11:          9 seconds\n",
      "Phasing iteration 12:          9 seconds\n",
      "\n",
      "Cumulative Statistics:\n",
      "\n",
      "Study markers:          35,897\n",
      "\n",
      "Haplotype phasing time:        4 minutes 11 seconds\n",
      "Total time:                    4 minutes 26 seconds\n",
      "\n",
      "End time: 02:25 AM PST on 11 Dec 2019\n",
      "beagle.28Sep18.793.jar finished\n",
      "\n",
      "\n",
      " -------------------------------------------------------------------------------- \n",
      "          Minimac4 - Fast Imputation Based on State Space Reduction HMM\n",
      " --------------------------------------------------------------------------------\n",
      "           (c) 2014 - Sayantan Das, Christian Fuchsberger, David Hinds\n",
      "                             Mary Kate Wing, Goncalo Abecasis \n",
      "\n",
      " Version: 1.0.2;\n",
      " Built: Mon Sep 30 11:52:22 PDT 2019 by biona001\n",
      "\n",
      " Command Line Options: \n",
      "       Reference Haplotypes : --refHaps [./sample_size/800_haplo_ref.m3vcf.gz],\n",
      "                              --passOnly, --rsid, --referenceEstimates [ON],\n",
      "                              --mapFile [docs/geneticMapFile.b38.map.txt.gz]\n",
      "          Target Haplotypes : --haps [./sample_size/800_target_masked.vcf.gz]\n",
      "          Output Parameters : --prefix [./sample_size/800_minimac4.result],\n",
      "                              --estimate, --nobgzip, --vcfBuffer [200],\n",
      "                              --format [GT,DS], --allTypedSites, --meta,\n",
      "                              --memUsage\n",
      "        Chunking Parameters : --ChunkLengthMb [20.00], --ChunkOverlapMb [3.00]\n",
      "          Subset Parameters : --chr [], --start, --end, --window\n",
      "   Approximation Parameters : --minimac3, --probThreshold [0.01],\n",
      "                              --diffThreshold [0.01], --topThreshold [0.01]\n",
      "           Other Parameters : --log, --help, --cpus [1], --params\n",
      "                  PhoneHome : --noPhoneHome, --phoneHomeThinning [50]\n",
      "\n",
      "\n",
      " URL = http://genome.sph.umich.edu/wiki/Minimac4\n",
      " Starting Main Imputation/Estimation Analysis ... \n",
      "\n",
      " Performing preliminary check on input parameters... \n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                             PRELIMINARY FILE CHECK                            \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Checking GWAS haplotype file : ./sample_size/800_target_masked.vcf.gz\n",
      "\n",
      " Gathering variant information ...\n",
      "\n",
      " Successful !!! \n",
      "\n",
      " Checking Reference haplotype file : ./sample_size/800_haplo_ref.m3vcf.gz\n",
      "\n",
      " Gathering variant information ...\n",
      "\n",
      " Successful !!! \n",
      "\n",
      " Reference Panel   : Found 1200 samples (2400 haplotypes) and 35897 variants ...\n",
      "\n",
      " Target/GWAS Panel : Found 800 samples (1600 haplotypes) and 35897 variants ...\n",
      "                     35897 variants overlap with Reference panel \n",
      "                     0 variants imported that exist only in Target/GWAS panel\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                           CHUNKING INFORMATION                           \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Chunking region into 1 chunk(s) with atleast 35897 variants in each chunk ... \n",
      "\n",
      " Details of chunks is given below ...\n",
      "\n",
      " No   LeftBuffer      LeftEnd   RightPoint  RightBuffer       #Sites(GWAS/Ref/%)\n",
      " -------------------------------------------------------------------------------\n",
      "  1           36           36      4999683      4999683    35897/   35897/ 100.00%\n",
      "\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                           MAIN IMPUTATION ANALYSIS                            \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Starting imputation analysis of 1 chunk(s) ... \n",
      "\n",
      " -------------------------------------------\n",
      " Analyzing Chunk 1/1 [1:36-4999683]\n",
      " -------------------------------------------\n",
      "\n",
      " Reading chunk from reference panel ... \n",
      " Reading chunk from target/GWAS panel ... \n",
      "\n",
      " Compressing reference panel at GWAS sites ... \n",
      " Re-compression successful (38 seconds) !!!\n",
      "\n",
      " Starting Imputation ...\n",
      "\n",
      "  Imputing Samples 1-200 [25%] out of 800 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 201-400 [50%] out of 800 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 401-600 [75%] out of 800 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 601-800 [100%] out of 800 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "\n",
      " Imputation successful (451 seconds) !!!\n",
      "\n",
      " Appending chunk to final output VCF File :  ./sample_size/800_minimac4.result.dose.vcf.gz \n",
      " Appending successful (3 seconds) !!!\n",
      "\n",
      " Time taken for this chunk = 498 seconds.\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                              SUMMARY OF ANALYSIS                              \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Info file written to                    : ./sample_size/800_minimac4.result.info\n",
      " Imputed VCF information written to      : ./sample_size/800_minimac4.result.dose.vcf.gz\n",
      "\n",
      " Time Taken for Reading File             = 9 seconds\n",
      " Time Taken for Re-compression           = 38 seconds\n",
      " Time Taken for Imputation               = 444 seconds\n",
      " Time Taken for Writing File             = 10 seconds\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                                END OF PROGRAM                                 \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Program Successfully Implemented... \n",
      " \n",
      " Total Run completed in 0 hours, 8 mins, 21 seconds.\n",
      "\n",
      " Thank You for using Minimac4 !!! \n",
      "\n",
      "beagle.28Sep18.793.jar (version 5.0)\n",
      "Copyright (C) 2014-2018 Brian L. Browning\n",
      "Enter \"java -jar beagle.28Sep18.793.jar\" to list command line argument\n",
      "Start time: 02:34 AM PST on 11 Dec 2019\n",
      "\n",
      "Command line: java -Xmx3641m -jar beagle.28Sep18.793.jar\n",
      "  gt=./sample_size/900_target_masked.vcf.gz\n",
      "  ref=./sample_size/900_haplo_ref.vcf.gz\n",
      "  out=./sample_size/900_beagle.result\n",
      "\n",
      "No genetic map is specified: using 1 cM = 1 Mb\n",
      "\n",
      "Reference samples:       1,100\n",
      "Study samples:             900\n",
      "\n",
      "Window 1 (1:36-4999683)\n",
      "Reference markers:      35,897\n",
      "Study markers:          35,897\n",
      "\n",
      "Burnin  iteration 1:           47 seconds\n",
      "Burnin  iteration 2:           30 seconds\n",
      "Burnin  iteration 3:           11 seconds\n",
      "Burnin  iteration 4:           16 seconds\n",
      "Burnin  iteration 5:           20 seconds\n",
      "Burnin  iteration 6:           21 seconds\n",
      "\n",
      "Phasing iteration 1:           12 seconds\n",
      "Phasing iteration 2:           12 seconds\n",
      "Phasing iteration 3:           12 seconds\n",
      "Phasing iteration 4:           11 seconds\n",
      "Phasing iteration 5:           11 seconds\n",
      "Phasing iteration 6:           11 seconds\n",
      "Phasing iteration 7:           11 seconds\n",
      "Phasing iteration 8:           11 seconds\n",
      "Phasing iteration 9:           11 seconds\n",
      "Phasing iteration 10:          11 seconds\n",
      "Phasing iteration 11:          10 seconds\n",
      "Phasing iteration 12:          10 seconds\n",
      "\n",
      "Cumulative Statistics:\n",
      "\n",
      "Study markers:          35,897\n",
      "\n",
      "Haplotype phasing time:        4 minutes 36 seconds\n",
      "Total time:                    4 minutes 54 seconds\n",
      "\n",
      "End time: 02:39 AM PST on 11 Dec 2019\n",
      "beagle.28Sep18.793.jar finished\n",
      "\n",
      "\n",
      " -------------------------------------------------------------------------------- \n",
      "          Minimac4 - Fast Imputation Based on State Space Reduction HMM\n",
      " --------------------------------------------------------------------------------\n",
      "           (c) 2014 - Sayantan Das, Christian Fuchsberger, David Hinds\n",
      "                             Mary Kate Wing, Goncalo Abecasis \n",
      "\n",
      " Version: 1.0.2;\n",
      " Built: Mon Sep 30 11:52:22 PDT 2019 by biona001\n",
      "\n",
      " Command Line Options: \n",
      "       Reference Haplotypes : --refHaps [./sample_size/900_haplo_ref.m3vcf.gz],\n",
      "                              --passOnly, --rsid, --referenceEstimates [ON],\n",
      "                              --mapFile [docs/geneticMapFile.b38.map.txt.gz]\n",
      "          Target Haplotypes : --haps [./sample_size/900_target_masked.vcf.gz]\n",
      "          Output Parameters : --prefix [./sample_size/900_minimac4.result],\n",
      "                              --estimate, --nobgzip, --vcfBuffer [200],\n",
      "                              --format [GT,DS], --allTypedSites, --meta,\n",
      "                              --memUsage\n",
      "        Chunking Parameters : --ChunkLengthMb [20.00], --ChunkOverlapMb [3.00]\n",
      "          Subset Parameters : --chr [], --start, --end, --window\n",
      "   Approximation Parameters : --minimac3, --probThreshold [0.01],\n",
      "                              --diffThreshold [0.01], --topThreshold [0.01]\n",
      "           Other Parameters : --log, --help, --cpus [1], --params\n",
      "                  PhoneHome : --noPhoneHome, --phoneHomeThinning [50]\n",
      "\n",
      "\n",
      " URL = http://genome.sph.umich.edu/wiki/Minimac4\n",
      " Starting Main Imputation/Estimation Analysis ... \n",
      "\n",
      " Performing preliminary check on input parameters... \n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                             PRELIMINARY FILE CHECK                            \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Checking GWAS haplotype file : ./sample_size/900_target_masked.vcf.gz\n",
      "\n",
      " Gathering variant information ...\n",
      "\n",
      " Successful !!! \n",
      "\n",
      " Checking Reference haplotype file : ./sample_size/900_haplo_ref.m3vcf.gz\n",
      "\n",
      " Gathering variant information ...\n",
      "\n",
      " Successful !!! \n",
      "\n",
      " Reference Panel   : Found 1100 samples (2200 haplotypes) and 35897 variants ...\n",
      "\n",
      " Target/GWAS Panel : Found 900 samples (1800 haplotypes) and 35897 variants ...\n",
      "                     35897 variants overlap with Reference panel \n",
      "                     0 variants imported that exist only in Target/GWAS panel\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                           CHUNKING INFORMATION                           \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Chunking region into 1 chunk(s) with atleast 35897 variants in each chunk ... \n",
      "\n",
      " Details of chunks is given below ...\n",
      "\n",
      " No   LeftBuffer      LeftEnd   RightPoint  RightBuffer       #Sites(GWAS/Ref/%)\n",
      " -------------------------------------------------------------------------------\n",
      "  1           36           36      4999683      4999683    35897/   35897/ 100.00%\n",
      "\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                           MAIN IMPUTATION ANALYSIS                            \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Starting imputation analysis of 1 chunk(s) ... \n",
      "\n",
      " -------------------------------------------\n",
      " Analyzing Chunk 1/1 [1:36-4999683]\n",
      " -------------------------------------------\n",
      "\n",
      " Reading chunk from reference panel ... \n",
      " Reading chunk from target/GWAS panel ... \n",
      "\n",
      " Compressing reference panel at GWAS sites ... \n",
      " Re-compression successful (34 seconds) !!!\n",
      "\n",
      " Starting Imputation ...\n",
      "\n",
      "  Imputing Samples 1-200 [22%] out of 900 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 201-400 [44%] out of 900 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 401-600 [67%] out of 900 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 601-800 [89%] out of 900 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 801-900 [100%] out of 900 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "\n",
      " Imputation successful (484 seconds) !!!\n",
      "\n",
      " Appending chunk to final output VCF File :  ./sample_size/900_minimac4.result.dose.vcf.gz \n",
      " Appending successful (3 seconds) !!!\n",
      "\n",
      " Time taken for this chunk = 528 seconds.\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                              SUMMARY OF ANALYSIS                              \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Info file written to                    : ./sample_size/900_minimac4.result.info\n",
      " Imputed VCF information written to      : ./sample_size/900_minimac4.result.dose.vcf.gz\n",
      "\n",
      " Time Taken for Reading File             = 10 seconds\n",
      " Time Taken for Re-compression           = 34 seconds\n",
      " Time Taken for Imputation               = 475 seconds\n",
      " Time Taken for Writing File             = 12 seconds\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                                END OF PROGRAM                                 \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Program Successfully Implemented... \n",
      " \n",
      " Total Run completed in 0 hours, 8 mins, 51 seconds.\n",
      "\n",
      " Thank You for using Minimac4 !!! \n",
      "\n",
      "beagle.28Sep18.793.jar (version 5.0)\n",
      "Copyright (C) 2014-2018 Brian L. Browning\n",
      "Enter \"java -jar beagle.28Sep18.793.jar\" to list command line argument\n",
      "Start time: 02:49 AM PST on 11 Dec 2019\n",
      "\n",
      "Command line: java -Xmx3641m -jar beagle.28Sep18.793.jar\n",
      "  gt=./sample_size/1000_target_masked.vcf.gz\n",
      "  ref=./sample_size/1000_haplo_ref.vcf.gz\n",
      "  out=./sample_size/1000_beagle.result\n",
      "\n",
      "No genetic map is specified: using 1 cM = 1 Mb\n",
      "\n",
      "Reference samples:       1,000\n",
      "Study samples:           1,000\n",
      "\n",
      "Window 1 (1:36-4999683)\n",
      "Reference markers:      35,897\n",
      "Study markers:          35,897\n",
      "\n",
      "Burnin  iteration 1:           52 seconds\n",
      "Burnin  iteration 2:           33 seconds\n",
      "Burnin  iteration 3:           12 seconds\n",
      "Burnin  iteration 4:           18 seconds\n",
      "Burnin  iteration 5:           22 seconds\n",
      "Burnin  iteration 6:           23 seconds\n",
      "\n",
      "Phasing iteration 1:           13 seconds\n",
      "Phasing iteration 2:           13 seconds\n",
      "Phasing iteration 3:           12 seconds\n",
      "Phasing iteration 4:           12 seconds\n",
      "Phasing iteration 5:           12 seconds\n",
      "Phasing iteration 6:           12 seconds\n",
      "Phasing iteration 7:           12 seconds\n",
      "Phasing iteration 8:           12 seconds\n",
      "Phasing iteration 9:           12 seconds\n",
      "Phasing iteration 10:          12 seconds\n",
      "Phasing iteration 11:          11 seconds\n",
      "Phasing iteration 12:          10 seconds\n",
      "\n",
      "Cumulative Statistics:\n",
      "\n",
      "Study markers:          35,897\n",
      "\n",
      "Haplotype phasing time:        5 minutes 3 seconds\n",
      "Total time:                    5 minutes 22 seconds\n",
      "\n",
      "End time: 02:55 AM PST on 11 Dec 2019\n",
      "beagle.28Sep18.793.jar finished\n",
      "\n",
      "\n",
      " -------------------------------------------------------------------------------- \n",
      "          Minimac4 - Fast Imputation Based on State Space Reduction HMM\n",
      " --------------------------------------------------------------------------------\n",
      "           (c) 2014 - Sayantan Das, Christian Fuchsberger, David Hinds\n",
      "                             Mary Kate Wing, Goncalo Abecasis \n",
      "\n",
      " Version: 1.0.2;\n",
      " Built: Mon Sep 30 11:52:22 PDT 2019 by biona001\n",
      "\n",
      " Command Line Options: \n",
      "       Reference Haplotypes : --refHaps [./sample_size/1000_haplo_ref.m3vcf.gz],\n",
      "                              --passOnly, --rsid, --referenceEstimates [ON],\n",
      "                              --mapFile [docs/geneticMapFile.b38.map.txt.gz]\n",
      "          Target Haplotypes : --haps [./sample_size/1000_target_masked.vcf.gz]\n",
      "          Output Parameters : --prefix [./sample_size/1000_minimac4.result],\n",
      "                              --estimate, --nobgzip, --vcfBuffer [200],\n",
      "                              --format [GT,DS], --allTypedSites, --meta,\n",
      "                              --memUsage\n",
      "        Chunking Parameters : --ChunkLengthMb [20.00], --ChunkOverlapMb [3.00]\n",
      "          Subset Parameters : --chr [], --start, --end, --window\n",
      "   Approximation Parameters : --minimac3, --probThreshold [0.01],\n",
      "                              --diffThreshold [0.01], --topThreshold [0.01]\n",
      "           Other Parameters : --log, --help, --cpus [1], --params\n",
      "                  PhoneHome : --noPhoneHome, --phoneHomeThinning [50]\n",
      "\n",
      "\n",
      " URL = http://genome.sph.umich.edu/wiki/Minimac4\n",
      " Starting Main Imputation/Estimation Analysis ... \n",
      "\n",
      " Performing preliminary check on input parameters... \n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                             PRELIMINARY FILE CHECK                            \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Checking GWAS haplotype file : ./sample_size/1000_target_masked.vcf.gz\n",
      "\n",
      " Gathering variant information ...\n",
      "\n",
      " Successful !!! \n",
      "\n",
      " Checking Reference haplotype file : ./sample_size/1000_haplo_ref.m3vcf.gz\n",
      "\n",
      " Gathering variant information ...\n",
      "\n",
      " Successful !!! \n",
      "\n",
      " Reference Panel   : Found 1000 samples (2000 haplotypes) and 35897 variants ...\n",
      "\n",
      " Target/GWAS Panel : Found 1000 samples (2000 haplotypes) and 35897 variants ...\n",
      "                     35897 variants overlap with Reference panel \n",
      "                     0 variants imported that exist only in Target/GWAS panel\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                           CHUNKING INFORMATION                           \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Chunking region into 1 chunk(s) with atleast 35897 variants in each chunk ... \n",
      "\n",
      " Details of chunks is given below ...\n",
      "\n",
      " No   LeftBuffer      LeftEnd   RightPoint  RightBuffer       #Sites(GWAS/Ref/%)\n",
      " -------------------------------------------------------------------------------\n",
      "  1           36           36      4999683      4999683    35897/   35897/ 100.00%\n",
      "\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                           MAIN IMPUTATION ANALYSIS                            \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Starting imputation analysis of 1 chunk(s) ... \n",
      "\n",
      " -------------------------------------------\n",
      " Analyzing Chunk 1/1 [1:36-4999683]\n",
      " -------------------------------------------\n",
      "\n",
      " Reading chunk from reference panel ... \n",
      " Reading chunk from target/GWAS panel ... \n",
      "\n",
      " Compressing reference panel at GWAS sites ... \n",
      " Re-compression successful (31 seconds) !!!\n",
      "\n",
      " Starting Imputation ...\n",
      "\n",
      "  Imputing Samples 1-200 [20%] out of 1000 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 201-400 [40%] out of 1000 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 401-600 [60%] out of 1000 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 601-800 [80%] out of 1000 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 801-1000 [100%] out of 1000 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "\n",
      " Imputation successful (513 seconds) !!!\n",
      "\n",
      " Appending chunk to final output VCF File :  ./sample_size/1000_minimac4.result.dose.vcf.gz \n",
      " Appending successful (3 seconds) !!!\n",
      "\n",
      " Time taken for this chunk = 554 seconds.\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                              SUMMARY OF ANALYSIS                              \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Info file written to                    : ./sample_size/1000_minimac4.result.info\n",
      " Imputed VCF information written to      : ./sample_size/1000_minimac4.result.dose.vcf.gz\n",
      "\n",
      " Time Taken for Reading File             = 10 seconds\n",
      " Time Taken for Re-compression           = 31 seconds\n",
      " Time Taken for Imputation               = 504 seconds\n",
      " Time Taken for Writing File             = 12 seconds\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                                END OF PROGRAM                                 \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Program Successfully Implemented... \n",
      " \n",
      " Total Run completed in 0 hours, 9 mins, 19 seconds.\n",
      "\n",
      " Thank You for using Minimac4 !!! \n",
      "\n",
      "beagle.28Sep18.793.jar (version 5.0)\n",
      "Copyright (C) 2014-2018 Brian L. Browning\n",
      "Enter \"java -jar beagle.28Sep18.793.jar\" to list command line argument\n",
      "Start time: 03:05 AM PST on 11 Dec 2019\n",
      "\n",
      "Command line: java -Xmx3641m -jar beagle.28Sep18.793.jar\n",
      "  gt=./sample_size/1100_target_masked.vcf.gz\n",
      "  ref=./sample_size/1100_haplo_ref.vcf.gz\n",
      "  out=./sample_size/1100_beagle.result\n",
      "\n",
      "No genetic map is specified: using 1 cM = 1 Mb\n",
      "\n",
      "Reference samples:         900\n",
      "Study samples:           1,100\n",
      "\n",
      "Window 1 (1:36-4999683)\n",
      "Reference markers:      35,897\n",
      "Study markers:          35,897\n",
      "\n",
      "Burnin  iteration 1:           51 seconds\n",
      "Burnin  iteration 2:           43 seconds\n",
      "Burnin  iteration 3:           13 seconds\n",
      "Burnin  iteration 4:           19 seconds\n",
      "Burnin  iteration 5:           24 seconds\n",
      "Burnin  iteration 6:           25 seconds\n",
      "\n",
      "Phasing iteration 1:           14 seconds\n",
      "Phasing iteration 2:           14 seconds\n",
      "Phasing iteration 3:           14 seconds\n",
      "Phasing iteration 4:           13 seconds\n",
      "Phasing iteration 5:           14 seconds\n",
      "Phasing iteration 6:           13 seconds\n",
      "Phasing iteration 7:           13 seconds\n",
      "Phasing iteration 8:           13 seconds\n",
      "Phasing iteration 9:           13 seconds\n",
      "Phasing iteration 10:          13 seconds\n",
      "Phasing iteration 11:          12 seconds\n",
      "Phasing iteration 12:          11 seconds\n",
      "\n",
      "Cumulative Statistics:\n",
      "\n",
      "Study markers:          35,897\n",
      "\n",
      "Haplotype phasing time:        5 minutes 33 seconds\n",
      "Total time:                    5 minutes 54 seconds\n",
      "\n",
      "End time: 03:11 AM PST on 11 Dec 2019\n",
      "beagle.28Sep18.793.jar finished\n",
      "\n",
      "\n",
      " -------------------------------------------------------------------------------- \n",
      "          Minimac4 - Fast Imputation Based on State Space Reduction HMM\n",
      " --------------------------------------------------------------------------------\n",
      "           (c) 2014 - Sayantan Das, Christian Fuchsberger, David Hinds\n",
      "                             Mary Kate Wing, Goncalo Abecasis \n",
      "\n",
      " Version: 1.0.2;\n",
      " Built: Mon Sep 30 11:52:22 PDT 2019 by biona001\n",
      "\n",
      " Command Line Options: \n",
      "       Reference Haplotypes : --refHaps [./sample_size/1100_haplo_ref.m3vcf.gz],\n",
      "                              --passOnly, --rsid, --referenceEstimates [ON],\n",
      "                              --mapFile [docs/geneticMapFile.b38.map.txt.gz]\n",
      "          Target Haplotypes : --haps [./sample_size/1100_target_masked.vcf.gz]\n",
      "          Output Parameters : --prefix [./sample_size/1100_minimac4.result],\n",
      "                              --estimate, --nobgzip, --vcfBuffer [200],\n",
      "                              --format [GT,DS], --allTypedSites, --meta,\n",
      "                              --memUsage\n",
      "        Chunking Parameters : --ChunkLengthMb [20.00], --ChunkOverlapMb [3.00]\n",
      "          Subset Parameters : --chr [], --start, --end, --window\n",
      "   Approximation Parameters : --minimac3, --probThreshold [0.01],\n",
      "                              --diffThreshold [0.01], --topThreshold [0.01]\n",
      "           Other Parameters : --log, --help, --cpus [1], --params\n",
      "                  PhoneHome : --noPhoneHome, --phoneHomeThinning [50]\n",
      "\n",
      "\n",
      " URL = http://genome.sph.umich.edu/wiki/Minimac4\n",
      " Starting Main Imputation/Estimation Analysis ... \n",
      "\n",
      " Performing preliminary check on input parameters... \n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                             PRELIMINARY FILE CHECK                            \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Checking GWAS haplotype file : ./sample_size/1100_target_masked.vcf.gz\n",
      "\n",
      " Gathering variant information ...\n",
      "\n",
      " Successful !!! \n",
      "\n",
      " Checking Reference haplotype file : ./sample_size/1100_haplo_ref.m3vcf.gz\n",
      "\n",
      " Gathering variant information ...\n",
      "\n",
      " Successful !!! \n",
      "\n",
      " Reference Panel   : Found 900 samples (1800 haplotypes) and 35897 variants ...\n",
      "\n",
      " Target/GWAS Panel : Found 1100 samples (2200 haplotypes) and 35897 variants ...\n",
      "                     35897 variants overlap with Reference panel \n",
      "                     0 variants imported that exist only in Target/GWAS panel\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                           CHUNKING INFORMATION                           \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Chunking region into 1 chunk(s) with atleast 35897 variants in each chunk ... \n",
      "\n",
      " Details of chunks is given below ...\n",
      "\n",
      " No   LeftBuffer      LeftEnd   RightPoint  RightBuffer       #Sites(GWAS/Ref/%)\n",
      " -------------------------------------------------------------------------------\n",
      "  1           36           36      4999683      4999683    35897/   35897/ 100.00%\n",
      "\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                           MAIN IMPUTATION ANALYSIS                            \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Starting imputation analysis of 1 chunk(s) ... \n",
      "\n",
      " -------------------------------------------\n",
      " Analyzing Chunk 1/1 [1:36-4999683]\n",
      " -------------------------------------------\n",
      "\n",
      " Reading chunk from reference panel ... \n",
      " Reading chunk from target/GWAS panel ... \n",
      "\n",
      " Compressing reference panel at GWAS sites ... \n",
      " Re-compression successful (27 seconds) !!!\n",
      "\n",
      " Starting Imputation ...\n",
      "\n",
      "  Imputing Samples 1-200 [18%] out of 1100 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 201-400 [36%] out of 1100 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 401-600 [55%] out of 1100 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 601-800 [73%] out of 1100 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 801-1000 [91%] out of 1100 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 1001-1100 [100%] out of 1100 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "\n",
      " Imputation successful (534 seconds) !!!\n",
      "\n",
      " Appending chunk to final output VCF File :  ./sample_size/1100_minimac4.result.dose.vcf.gz \n",
      " Appending successful (4 seconds) !!!\n",
      "\n",
      " Time taken for this chunk = 572 seconds.\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                              SUMMARY OF ANALYSIS                              \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Info file written to                    : ./sample_size/1100_minimac4.result.info\n",
      " Imputed VCF information written to      : ./sample_size/1100_minimac4.result.dose.vcf.gz\n",
      "\n",
      " Time Taken for Reading File             = 10 seconds\n",
      " Time Taken for Re-compression           = 27 seconds\n",
      " Time Taken for Imputation               = 523 seconds\n",
      " Time Taken for Writing File             = 15 seconds\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                                END OF PROGRAM                                 \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Program Successfully Implemented... \n",
      " \n",
      " Total Run completed in 0 hours, 9 mins, 36 seconds.\n",
      "\n",
      " Thank You for using Minimac4 !!! \n",
      "\n",
      "beagle.28Sep18.793.jar (version 5.0)\n",
      "Copyright (C) 2014-2018 Brian L. Browning\n",
      "Enter \"java -jar beagle.28Sep18.793.jar\" to list command line argument\n",
      "Start time: 03:22 AM PST on 11 Dec 2019\n",
      "\n",
      "Command line: java -Xmx3641m -jar beagle.28Sep18.793.jar\n",
      "  gt=./sample_size/1200_target_masked.vcf.gz\n",
      "  ref=./sample_size/1200_haplo_ref.vcf.gz\n",
      "  out=./sample_size/1200_beagle.result\n",
      "\n",
      "No genetic map is specified: using 1 cM = 1 Mb\n",
      "\n",
      "Reference samples:         800\n",
      "Study samples:           1,200\n",
      "\n",
      "Window 1 (1:36-4999683)\n",
      "Reference markers:      35,897\n",
      "Study markers:          35,897\n",
      "\n",
      "Burnin  iteration 1:           53 seconds\n",
      "Burnin  iteration 2:           43 seconds\n",
      "Burnin  iteration 3:           15 seconds\n",
      "Burnin  iteration 4:           22 seconds\n",
      "Burnin  iteration 5:           27 seconds\n",
      "Burnin  iteration 6:           28 seconds\n",
      "\n",
      "Phasing iteration 1:           16 seconds\n",
      "Phasing iteration 2:           16 seconds\n",
      "Phasing iteration 3:           16 seconds\n",
      "Phasing iteration 4:           15 seconds\n",
      "Phasing iteration 5:           15 seconds\n",
      "Phasing iteration 6:           15 seconds\n",
      "Phasing iteration 7:           15 seconds\n",
      "Phasing iteration 8:           15 seconds\n",
      "Phasing iteration 9:           15 seconds\n",
      "Phasing iteration 10:          16 seconds\n",
      "Phasing iteration 11:          14 seconds\n",
      "Phasing iteration 12:          13 seconds\n",
      "\n",
      "Cumulative Statistics:\n",
      "\n",
      "Study markers:          35,897\n",
      "\n",
      "Haplotype phasing time:        6 minutes 8 seconds\n",
      "Total time:                    6 minutes 31 seconds\n",
      "\n",
      "End time: 03:28 AM PST on 11 Dec 2019\n",
      "beagle.28Sep18.793.jar finished\n",
      "\n",
      "\n",
      " -------------------------------------------------------------------------------- \n",
      "          Minimac4 - Fast Imputation Based on State Space Reduction HMM\n",
      " --------------------------------------------------------------------------------\n",
      "           (c) 2014 - Sayantan Das, Christian Fuchsberger, David Hinds\n",
      "                             Mary Kate Wing, Goncalo Abecasis \n",
      "\n",
      " Version: 1.0.2;\n",
      " Built: Mon Sep 30 11:52:22 PDT 2019 by biona001\n",
      "\n",
      " Command Line Options: \n",
      "       Reference Haplotypes : --refHaps [./sample_size/1200_haplo_ref.m3vcf.gz],\n",
      "                              --passOnly, --rsid, --referenceEstimates [ON],\n",
      "                              --mapFile [docs/geneticMapFile.b38.map.txt.gz]\n",
      "          Target Haplotypes : --haps [./sample_size/1200_target_masked.vcf.gz]\n",
      "          Output Parameters : --prefix [./sample_size/1200_minimac4.result],\n",
      "                              --estimate, --nobgzip, --vcfBuffer [200],\n",
      "                              --format [GT,DS], --allTypedSites, --meta,\n",
      "                              --memUsage\n",
      "        Chunking Parameters : --ChunkLengthMb [20.00], --ChunkOverlapMb [3.00]\n",
      "          Subset Parameters : --chr [], --start, --end, --window\n",
      "   Approximation Parameters : --minimac3, --probThreshold [0.01],\n",
      "                              --diffThreshold [0.01], --topThreshold [0.01]\n",
      "           Other Parameters : --log, --help, --cpus [1], --params\n",
      "                  PhoneHome : --noPhoneHome, --phoneHomeThinning [50]\n",
      "\n",
      "\n",
      " URL = http://genome.sph.umich.edu/wiki/Minimac4\n",
      " Starting Main Imputation/Estimation Analysis ... \n",
      "\n",
      " Performing preliminary check on input parameters... \n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                             PRELIMINARY FILE CHECK                            \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Checking GWAS haplotype file : ./sample_size/1200_target_masked.vcf.gz\n",
      "\n",
      " Gathering variant information ...\n",
      "\n",
      " Successful !!! \n",
      "\n",
      " Checking Reference haplotype file : ./sample_size/1200_haplo_ref.m3vcf.gz\n",
      "\n",
      " Gathering variant information ...\n",
      "\n",
      " Successful !!! \n",
      "\n",
      " Reference Panel   : Found 800 samples (1600 haplotypes) and 35897 variants ...\n",
      "\n",
      " Target/GWAS Panel : Found 1200 samples (2400 haplotypes) and 35897 variants ...\n",
      "                     35897 variants overlap with Reference panel \n",
      "                     0 variants imported that exist only in Target/GWAS panel\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                           CHUNKING INFORMATION                           \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Chunking region into 1 chunk(s) with atleast 35897 variants in each chunk ... \n",
      "\n",
      " Details of chunks is given below ...\n",
      "\n",
      " No   LeftBuffer      LeftEnd   RightPoint  RightBuffer       #Sites(GWAS/Ref/%)\n",
      " -------------------------------------------------------------------------------\n",
      "  1           36           36      4999683      4999683    35897/   35897/ 100.00%\n",
      "\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                           MAIN IMPUTATION ANALYSIS                            \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Starting imputation analysis of 1 chunk(s) ... \n",
      "\n",
      " -------------------------------------------\n",
      " Analyzing Chunk 1/1 [1:36-4999683]\n",
      " -------------------------------------------\n",
      "\n",
      " Reading chunk from reference panel ... \n",
      " Reading chunk from target/GWAS panel ... \n",
      "\n",
      " Compressing reference panel at GWAS sites ... \n",
      " Re-compression successful (23 seconds) !!!\n",
      "\n",
      " Starting Imputation ...\n",
      "\n",
      "  Imputing Samples 1-200 [17%] out of 1200 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 201-400 [33%] out of 1200 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 401-600 [50%] out of 1200 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 601-800 [67%] out of 1200 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 801-1000 [83%] out of 1200 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 1001-1200 [100%] out of 1200 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "\n",
      " Imputation successful (545 seconds) !!!\n",
      "\n",
      " Appending chunk to final output VCF File :  ./sample_size/1200_minimac4.result.dose.vcf.gz \n",
      " Appending successful (4 seconds) !!!\n",
      "\n",
      " Time taken for this chunk = 580 seconds.\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                              SUMMARY OF ANALYSIS                              \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Info file written to                    : ./sample_size/1200_minimac4.result.info\n",
      " Imputed VCF information written to      : ./sample_size/1200_minimac4.result.dose.vcf.gz\n",
      "\n",
      " Time Taken for Reading File             = 12 seconds\n",
      " Time Taken for Re-compression           = 23 seconds\n",
      " Time Taken for Imputation               = 534 seconds\n",
      " Time Taken for Writing File             = 15 seconds\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                                END OF PROGRAM                                 \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Program Successfully Implemented... \n",
      " \n",
      " Total Run completed in 0 hours, 9 mins, 45 seconds.\n",
      "\n",
      " Thank You for using Minimac4 !!! \n",
      "\n",
      "beagle.28Sep18.793.jar (version 5.0)\n",
      "Copyright (C) 2014-2018 Brian L. Browning\n",
      "Enter \"java -jar beagle.28Sep18.793.jar\" to list command line argument\n",
      "Start time: 03:39 AM PST on 11 Dec 2019\n",
      "\n",
      "Command line: java -Xmx3641m -jar beagle.28Sep18.793.jar\n",
      "  gt=./sample_size/1300_target_masked.vcf.gz\n",
      "  ref=./sample_size/1300_haplo_ref.vcf.gz\n",
      "  out=./sample_size/1300_beagle.result\n",
      "\n",
      "No genetic map is specified: using 1 cM = 1 Mb\n",
      "\n",
      "Reference samples:         700\n",
      "Study samples:           1,300\n",
      "\n",
      "Window 1 (1:36-4999683)\n",
      "Reference markers:      35,897\n",
      "Study markers:          35,897\n",
      "\n",
      "Burnin  iteration 1:           1 minute 1 second\n",
      "Burnin  iteration 2:           49 seconds\n",
      "Burnin  iteration 3:           14 seconds\n",
      "Burnin  iteration 4:           22 seconds\n",
      "Burnin  iteration 5:           29 seconds\n",
      "Burnin  iteration 6:           30 seconds\n",
      "\n",
      "Phasing iteration 1:           17 seconds\n",
      "Phasing iteration 2:           16 seconds\n",
      "Phasing iteration 3:           16 seconds\n",
      "Phasing iteration 4:           16 seconds\n",
      "Phasing iteration 5:           16 seconds\n",
      "Phasing iteration 6:           16 seconds\n",
      "Phasing iteration 7:           15 seconds\n",
      "Phasing iteration 8:           15 seconds\n",
      "Phasing iteration 9:           16 seconds\n",
      "Phasing iteration 10:          15 seconds\n",
      "Phasing iteration 11:          14 seconds\n",
      "Phasing iteration 12:          14 seconds\n",
      "\n",
      "Cumulative Statistics:\n",
      "\n",
      "Study markers:          35,897\n",
      "\n",
      "Haplotype phasing time:        6 minutes 31 seconds\n",
      "Total time:                    6 minutes 54 seconds\n",
      "\n",
      "End time: 03:46 AM PST on 11 Dec 2019\n",
      "beagle.28Sep18.793.jar finished\n",
      "\n",
      "\n",
      " -------------------------------------------------------------------------------- \n",
      "          Minimac4 - Fast Imputation Based on State Space Reduction HMM\n",
      " --------------------------------------------------------------------------------\n",
      "           (c) 2014 - Sayantan Das, Christian Fuchsberger, David Hinds\n",
      "                             Mary Kate Wing, Goncalo Abecasis \n",
      "\n",
      " Version: 1.0.2;\n",
      " Built: Mon Sep 30 11:52:22 PDT 2019 by biona001\n",
      "\n",
      " Command Line Options: \n",
      "       Reference Haplotypes : --refHaps [./sample_size/1300_haplo_ref.m3vcf.gz],\n",
      "                              --passOnly, --rsid, --referenceEstimates [ON],\n",
      "                              --mapFile [docs/geneticMapFile.b38.map.txt.gz]\n",
      "          Target Haplotypes : --haps [./sample_size/1300_target_masked.vcf.gz]\n",
      "          Output Parameters : --prefix [./sample_size/1300_minimac4.result],\n",
      "                              --estimate, --nobgzip, --vcfBuffer [200],\n",
      "                              --format [GT,DS], --allTypedSites, --meta,\n",
      "                              --memUsage\n",
      "        Chunking Parameters : --ChunkLengthMb [20.00], --ChunkOverlapMb [3.00]\n",
      "          Subset Parameters : --chr [], --start, --end, --window\n",
      "   Approximation Parameters : --minimac3, --probThreshold [0.01],\n",
      "                              --diffThreshold [0.01], --topThreshold [0.01]\n",
      "           Other Parameters : --log, --help, --cpus [1], --params\n",
      "                  PhoneHome : --noPhoneHome, --phoneHomeThinning [50]\n",
      "\n",
      "\n",
      " URL = http://genome.sph.umich.edu/wiki/Minimac4\n",
      " Starting Main Imputation/Estimation Analysis ... \n",
      "\n",
      " Performing preliminary check on input parameters... \n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                             PRELIMINARY FILE CHECK                            \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Checking GWAS haplotype file : ./sample_size/1300_target_masked.vcf.gz\n",
      "\n",
      " Gathering variant information ...\n",
      "\n",
      " Successful !!! \n",
      "\n",
      " Checking Reference haplotype file : ./sample_size/1300_haplo_ref.m3vcf.gz\n",
      "\n",
      " Gathering variant information ...\n",
      "\n",
      " Successful !!! \n",
      "\n",
      " Reference Panel   : Found 700 samples (1400 haplotypes) and 35897 variants ...\n",
      "\n",
      " Target/GWAS Panel : Found 1300 samples (2600 haplotypes) and 35897 variants ...\n",
      "                     35897 variants overlap with Reference panel \n",
      "                     0 variants imported that exist only in Target/GWAS panel\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                           CHUNKING INFORMATION                           \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Chunking region into 1 chunk(s) with atleast 35897 variants in each chunk ... \n",
      "\n",
      " Details of chunks is given below ...\n",
      "\n",
      " No   LeftBuffer      LeftEnd   RightPoint  RightBuffer       #Sites(GWAS/Ref/%)\n",
      " -------------------------------------------------------------------------------\n",
      "  1           36           36      4999683      4999683    35897/   35897/ 100.00%\n",
      "\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                           MAIN IMPUTATION ANALYSIS                            \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Starting imputation analysis of 1 chunk(s) ... \n",
      "\n",
      " -------------------------------------------\n",
      " Analyzing Chunk 1/1 [1:36-4999683]\n",
      " -------------------------------------------\n",
      "\n",
      " Reading chunk from reference panel ... \n",
      " Reading chunk from target/GWAS panel ... \n",
      "\n",
      " Compressing reference panel at GWAS sites ... \n",
      " Re-compression successful (20 seconds) !!!\n",
      "\n",
      " Starting Imputation ...\n",
      "\n",
      "  Imputing Samples 1-200 [15%] out of 1300 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 201-400 [31%] out of 1300 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 401-600 [46%] out of 1300 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 601-800 [62%] out of 1300 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 801-1000 [77%] out of 1300 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 1001-1200 [92%] out of 1300 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 1201-1300 [100%] out of 1300 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "\n",
      " Imputation successful (550 seconds) !!!\n",
      "\n",
      " Appending chunk to final output VCF File :  ./sample_size/1300_minimac4.result.dose.vcf.gz \n",
      " Appending successful (5 seconds) !!!\n",
      "\n",
      " Time taken for this chunk = 584 seconds.\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                              SUMMARY OF ANALYSIS                              \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Info file written to                    : ./sample_size/1300_minimac4.result.info\n",
      " Imputed VCF information written to      : ./sample_size/1300_minimac4.result.dose.vcf.gz\n",
      "\n",
      " Time Taken for Reading File             = 13 seconds\n",
      " Time Taken for Re-compression           = 20 seconds\n",
      " Time Taken for Imputation               = 540 seconds\n",
      " Time Taken for Writing File             = 15 seconds\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                                END OF PROGRAM                                 \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Program Successfully Implemented... \n",
      " \n",
      " Total Run completed in 0 hours, 9 mins, 48 seconds.\n",
      "\n",
      " Thank You for using Minimac4 !!! \n",
      "\n",
      "beagle.28Sep18.793.jar (version 5.0)\n",
      "Copyright (C) 2014-2018 Brian L. Browning\n",
      "Enter \"java -jar beagle.28Sep18.793.jar\" to list command line argument\n",
      "Start time: 03:57 AM PST on 11 Dec 2019\n",
      "\n",
      "Command line: java -Xmx3641m -jar beagle.28Sep18.793.jar\n",
      "  gt=./sample_size/1400_target_masked.vcf.gz\n",
      "  ref=./sample_size/1400_haplo_ref.vcf.gz\n",
      "  out=./sample_size/1400_beagle.result\n",
      "\n",
      "No genetic map is specified: using 1 cM = 1 Mb\n",
      "\n",
      "Reference samples:         600\n",
      "Study samples:           1,400\n",
      "\n",
      "Window 1 (1:36-4999683)\n",
      "Reference markers:      35,897\n",
      "Study markers:          35,897\n",
      "\n",
      "Burnin  iteration 1:           1 minute 6 seconds\n",
      "Burnin  iteration 2:           58 seconds\n",
      "Burnin  iteration 3:           16 seconds\n",
      "Burnin  iteration 4:           23 seconds\n",
      "Burnin  iteration 5:           30 seconds\n",
      "Burnin  iteration 6:           30 seconds\n",
      "\n",
      "Phasing iteration 1:           17 seconds\n",
      "Phasing iteration 2:           17 seconds\n",
      "Phasing iteration 3:           17 seconds\n",
      "Phasing iteration 4:           16 seconds\n",
      "Phasing iteration 5:           17 seconds\n",
      "Phasing iteration 6:           17 seconds\n",
      "Phasing iteration 7:           16 seconds\n",
      "Phasing iteration 8:           16 seconds\n",
      "Phasing iteration 9:           16 seconds\n",
      "Phasing iteration 10:          15 seconds\n",
      "Phasing iteration 11:          14 seconds\n",
      "Phasing iteration 12:          14 seconds\n",
      "\n",
      "Cumulative Statistics:\n",
      "\n",
      "Study markers:          35,897\n",
      "\n",
      "Haplotype phasing time:        6 minutes 56 seconds\n",
      "Total time:                    7 minutes 22 seconds\n",
      "\n",
      "End time: 04:04 AM PST on 11 Dec 2019\n",
      "beagle.28Sep18.793.jar finished\n",
      "\n",
      "\n",
      " -------------------------------------------------------------------------------- \n",
      "          Minimac4 - Fast Imputation Based on State Space Reduction HMM\n",
      " --------------------------------------------------------------------------------\n",
      "           (c) 2014 - Sayantan Das, Christian Fuchsberger, David Hinds\n",
      "                             Mary Kate Wing, Goncalo Abecasis \n",
      "\n",
      " Version: 1.0.2;\n",
      " Built: Mon Sep 30 11:52:22 PDT 2019 by biona001\n",
      "\n",
      " Command Line Options: \n",
      "       Reference Haplotypes : --refHaps [./sample_size/1400_haplo_ref.m3vcf.gz],\n",
      "                              --passOnly, --rsid, --referenceEstimates [ON],\n",
      "                              --mapFile [docs/geneticMapFile.b38.map.txt.gz]\n",
      "          Target Haplotypes : --haps [./sample_size/1400_target_masked.vcf.gz]\n",
      "          Output Parameters : --prefix [./sample_size/1400_minimac4.result],\n",
      "                              --estimate, --nobgzip, --vcfBuffer [200],\n",
      "                              --format [GT,DS], --allTypedSites, --meta,\n",
      "                              --memUsage\n",
      "        Chunking Parameters : --ChunkLengthMb [20.00], --ChunkOverlapMb [3.00]\n",
      "          Subset Parameters : --chr [], --start, --end, --window\n",
      "   Approximation Parameters : --minimac3, --probThreshold [0.01],\n",
      "                              --diffThreshold [0.01], --topThreshold [0.01]\n",
      "           Other Parameters : --log, --help, --cpus [1], --params\n",
      "                  PhoneHome : --noPhoneHome, --phoneHomeThinning [50]\n",
      "\n",
      "\n",
      " URL = http://genome.sph.umich.edu/wiki/Minimac4\n",
      " Starting Main Imputation/Estimation Analysis ... \n",
      "\n",
      " Performing preliminary check on input parameters... \n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                             PRELIMINARY FILE CHECK                            \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Checking GWAS haplotype file : ./sample_size/1400_target_masked.vcf.gz\n",
      "\n",
      " Gathering variant information ...\n",
      "\n",
      " Successful !!! \n",
      "\n",
      " Checking Reference haplotype file : ./sample_size/1400_haplo_ref.m3vcf.gz\n",
      "\n",
      " Gathering variant information ...\n",
      "\n",
      " Successful !!! \n",
      "\n",
      " Reference Panel   : Found 600 samples (1200 haplotypes) and 35897 variants ...\n",
      "\n",
      " Target/GWAS Panel : Found 1400 samples (2800 haplotypes) and 35897 variants ...\n",
      "                     35897 variants overlap with Reference panel \n",
      "                     0 variants imported that exist only in Target/GWAS panel\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                           CHUNKING INFORMATION                           \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Chunking region into 1 chunk(s) with atleast 35897 variants in each chunk ... \n",
      "\n",
      " Details of chunks is given below ...\n",
      "\n",
      " No   LeftBuffer      LeftEnd   RightPoint  RightBuffer       #Sites(GWAS/Ref/%)\n",
      " -------------------------------------------------------------------------------\n",
      "  1           36           36      4999683      4999683    35897/   35897/ 100.00%\n",
      "\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                           MAIN IMPUTATION ANALYSIS                            \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Starting imputation analysis of 1 chunk(s) ... \n",
      "\n",
      " -------------------------------------------\n",
      " Analyzing Chunk 1/1 [1:36-4999683]\n",
      " -------------------------------------------\n",
      "\n",
      " Reading chunk from reference panel ... \n",
      " Reading chunk from target/GWAS panel ... \n",
      "\n",
      " Compressing reference panel at GWAS sites ... \n",
      " Re-compression successful (16 seconds) !!!\n",
      "\n",
      " Starting Imputation ...\n",
      "\n",
      "  Imputing Samples 1-200 [14%] out of 1400 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 201-400 [29%] out of 1400 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 401-600 [43%] out of 1400 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 601-800 [57%] out of 1400 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 801-1000 [71%] out of 1400 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 1001-1200 [86%] out of 1400 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 1201-1400 [100%] out of 1400 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "\n",
      " Imputation successful (548 seconds) !!!\n",
      "\n",
      " Appending chunk to final output VCF File :  ./sample_size/1400_minimac4.result.dose.vcf.gz \n",
      " Appending successful (5 seconds) !!!\n",
      "\n",
      " Time taken for this chunk = 579 seconds.\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                              SUMMARY OF ANALYSIS                              \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Info file written to                    : ./sample_size/1400_minimac4.result.info\n",
      " Imputed VCF information written to      : ./sample_size/1400_minimac4.result.dose.vcf.gz\n",
      "\n",
      " Time Taken for Reading File             = 14 seconds\n",
      " Time Taken for Re-compression           = 16 seconds\n",
      " Time Taken for Imputation               = 535 seconds\n",
      " Time Taken for Writing File             = 18 seconds\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                                END OF PROGRAM                                 \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Program Successfully Implemented... \n",
      " \n",
      " Total Run completed in 0 hours, 9 mins, 44 seconds.\n",
      "\n",
      " Thank You for using Minimac4 !!! \n",
      "\n",
      "beagle.28Sep18.793.jar (version 5.0)\n",
      "Copyright (C) 2014-2018 Brian L. Browning\n",
      "Enter \"java -jar beagle.28Sep18.793.jar\" to list command line argument\n",
      "Start time: 04:15 AM PST on 11 Dec 2019\n",
      "\n",
      "Command line: java -Xmx3641m -jar beagle.28Sep18.793.jar\n",
      "  gt=./sample_size/1500_target_masked.vcf.gz\n",
      "  ref=./sample_size/1500_haplo_ref.vcf.gz\n",
      "  out=./sample_size/1500_beagle.result\n",
      "\n",
      "No genetic map is specified: using 1 cM = 1 Mb\n",
      "\n",
      "Reference samples:         500\n",
      "Study samples:           1,500\n",
      "\n",
      "Window 1 (1:36-4999683)\n",
      "Reference markers:      35,897\n",
      "Study markers:          35,897\n",
      "\n",
      "Burnin  iteration 1:           1 minute 8 seconds\n",
      "Burnin  iteration 2:           1 minute 12 seconds\n",
      "Burnin  iteration 3:           15 seconds\n",
      "Burnin  iteration 4:           24 seconds\n",
      "Burnin  iteration 5:           32 seconds\n",
      "Burnin  iteration 6:           33 seconds\n",
      "\n",
      "Phasing iteration 1:           18 seconds\n",
      "Phasing iteration 2:           18 seconds\n",
      "Phasing iteration 3:           18 seconds\n",
      "Phasing iteration 4:           17 seconds\n",
      "Phasing iteration 5:           18 seconds\n",
      "Phasing iteration 6:           18 seconds\n",
      "Phasing iteration 7:           17 seconds\n",
      "Phasing iteration 8:           18 seconds\n",
      "Phasing iteration 9:           17 seconds\n",
      "Phasing iteration 10:          17 seconds\n",
      "Phasing iteration 11:          16 seconds\n",
      "Phasing iteration 12:          15 seconds\n",
      "\n",
      "Cumulative Statistics:\n",
      "\n",
      "Study markers:          35,897\n",
      "\n",
      "Haplotype phasing time:        7 minutes 30 seconds\n",
      "Total time:                    7 minutes 57 seconds\n",
      "\n",
      "End time: 04:23 AM PST on 11 Dec 2019\n",
      "beagle.28Sep18.793.jar finished\n",
      "\n",
      "\n",
      " -------------------------------------------------------------------------------- \n",
      "          Minimac4 - Fast Imputation Based on State Space Reduction HMM\n",
      " --------------------------------------------------------------------------------\n",
      "           (c) 2014 - Sayantan Das, Christian Fuchsberger, David Hinds\n",
      "                             Mary Kate Wing, Goncalo Abecasis \n",
      "\n",
      " Version: 1.0.2;\n",
      " Built: Mon Sep 30 11:52:22 PDT 2019 by biona001\n",
      "\n",
      " Command Line Options: \n",
      "       Reference Haplotypes : --refHaps [./sample_size/1500_haplo_ref.m3vcf.gz],\n",
      "                              --passOnly, --rsid, --referenceEstimates [ON],\n",
      "                              --mapFile [docs/geneticMapFile.b38.map.txt.gz]\n",
      "          Target Haplotypes : --haps [./sample_size/1500_target_masked.vcf.gz]\n",
      "          Output Parameters : --prefix [./sample_size/1500_minimac4.result],\n",
      "                              --estimate, --nobgzip, --vcfBuffer [200],\n",
      "                              --format [GT,DS], --allTypedSites, --meta,\n",
      "                              --memUsage\n",
      "        Chunking Parameters : --ChunkLengthMb [20.00], --ChunkOverlapMb [3.00]\n",
      "          Subset Parameters : --chr [], --start, --end, --window\n",
      "   Approximation Parameters : --minimac3, --probThreshold [0.01],\n",
      "                              --diffThreshold [0.01], --topThreshold [0.01]\n",
      "           Other Parameters : --log, --help, --cpus [1], --params\n",
      "                  PhoneHome : --noPhoneHome, --phoneHomeThinning [50]\n",
      "\n",
      "\n",
      " URL = http://genome.sph.umich.edu/wiki/Minimac4\n",
      " Starting Main Imputation/Estimation Analysis ... \n",
      "\n",
      " Performing preliminary check on input parameters... \n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                             PRELIMINARY FILE CHECK                            \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Checking GWAS haplotype file : ./sample_size/1500_target_masked.vcf.gz\n",
      "\n",
      " Gathering variant information ...\n",
      "\n",
      " Successful !!! \n",
      "\n",
      " Checking Reference haplotype file : ./sample_size/1500_haplo_ref.m3vcf.gz\n",
      "\n",
      " Gathering variant information ...\n",
      "\n",
      " Successful !!! \n",
      "\n",
      " Reference Panel   : Found 500 samples (1000 haplotypes) and 35897 variants ...\n",
      "\n",
      " Target/GWAS Panel : Found 1500 samples (3000 haplotypes) and 35897 variants ...\n",
      "                     35897 variants overlap with Reference panel \n",
      "                     0 variants imported that exist only in Target/GWAS panel\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                           CHUNKING INFORMATION                           \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Chunking region into 1 chunk(s) with atleast 35897 variants in each chunk ... \n",
      "\n",
      " Details of chunks is given below ...\n",
      "\n",
      " No   LeftBuffer      LeftEnd   RightPoint  RightBuffer       #Sites(GWAS/Ref/%)\n",
      " -------------------------------------------------------------------------------\n",
      "  1           36           36      4999683      4999683    35897/   35897/ 100.00%\n",
      "\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                           MAIN IMPUTATION ANALYSIS                            \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Starting imputation analysis of 1 chunk(s) ... \n",
      "\n",
      " -------------------------------------------\n",
      " Analyzing Chunk 1/1 [1:36-4999683]\n",
      " -------------------------------------------\n",
      "\n",
      " Reading chunk from reference panel ... \n",
      " Reading chunk from target/GWAS panel ... \n",
      "\n",
      " Compressing reference panel at GWAS sites ... \n",
      " Re-compression successful (13 seconds) !!!\n",
      "\n",
      " Starting Imputation ...\n",
      "\n",
      "  Imputing Samples 1-200 [13%] out of 1500 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 201-400 [27%] out of 1500 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 401-600 [40%] out of 1500 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 601-800 [53%] out of 1500 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 801-1000 [67%] out of 1500 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 1001-1200 [80%] out of 1500 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 1201-1400 [93%] out of 1500 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 1401-1500 [100%] out of 1500 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "\n",
      " Imputation successful (537 seconds) !!!\n",
      "\n",
      " Appending chunk to final output VCF File :  ./sample_size/1500_minimac4.result.dose.vcf.gz \n",
      " Appending successful (6 seconds) !!!\n",
      "\n",
      " Time taken for this chunk = 567 seconds.\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                              SUMMARY OF ANALYSIS                              \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Info file written to                    : ./sample_size/1500_minimac4.result.info\n",
      " Imputed VCF information written to      : ./sample_size/1500_minimac4.result.dose.vcf.gz\n",
      "\n",
      " Time Taken for Reading File             = 16 seconds\n",
      " Time Taken for Re-compression           = 13 seconds\n",
      " Time Taken for Imputation               = 524 seconds\n",
      " Time Taken for Writing File             = 19 seconds\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                                END OF PROGRAM                                 \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Program Successfully Implemented... \n",
      " \n",
      " Total Run completed in 0 hours, 9 mins, 32 seconds.\n",
      "\n",
      " Thank You for using Minimac4 !!! \n",
      "\n",
      "beagle.28Sep18.793.jar (version 5.0)\n",
      "Copyright (C) 2014-2018 Brian L. Browning\n",
      "Enter \"java -jar beagle.28Sep18.793.jar\" to list command line argument\n",
      "Start time: 04:34 AM PST on 11 Dec 2019\n",
      "\n",
      "Command line: java -Xmx3641m -jar beagle.28Sep18.793.jar\n",
      "  gt=./sample_size/1600_target_masked.vcf.gz\n",
      "  ref=./sample_size/1600_haplo_ref.vcf.gz\n",
      "  out=./sample_size/1600_beagle.result\n",
      "\n",
      "No genetic map is specified: using 1 cM = 1 Mb\n",
      "\n",
      "Reference samples:         400\n",
      "Study samples:           1,600\n",
      "\n",
      "Window 1 (1:36-4999683)\n",
      "Reference markers:      35,897\n",
      "Study markers:          35,897\n",
      "\n",
      "Burnin  iteration 1:           1 minute 5 seconds\n",
      "Burnin  iteration 2:           1 minute 16 seconds\n",
      "Burnin  iteration 3:           16 seconds\n",
      "Burnin  iteration 4:           26 seconds\n",
      "Burnin  iteration 5:           34 seconds\n",
      "Burnin  iteration 6:           36 seconds\n",
      "\n",
      "Phasing iteration 1:           20 seconds\n",
      "Phasing iteration 2:           20 seconds\n",
      "Phasing iteration 3:           20 seconds\n",
      "Phasing iteration 4:           19 seconds\n",
      "Phasing iteration 5:           19 seconds\n",
      "Phasing iteration 6:           19 seconds\n",
      "Phasing iteration 7:           19 seconds\n",
      "Phasing iteration 8:           19 seconds\n",
      "Phasing iteration 9:           19 seconds\n",
      "Phasing iteration 10:          18 seconds\n",
      "Phasing iteration 11:          17 seconds\n",
      "Phasing iteration 12:          16 seconds\n",
      "\n",
      "Cumulative Statistics:\n",
      "\n",
      "Study markers:          35,897\n",
      "\n",
      "Haplotype phasing time:        7 minutes 57 seconds\n",
      "Total time:                    8 minutes 28 seconds\n",
      "\n",
      "End time: 04:42 AM PST on 11 Dec 2019\n",
      "beagle.28Sep18.793.jar finished\n",
      "\n",
      "\n",
      " -------------------------------------------------------------------------------- \n",
      "          Minimac4 - Fast Imputation Based on State Space Reduction HMM\n",
      " --------------------------------------------------------------------------------\n",
      "           (c) 2014 - Sayantan Das, Christian Fuchsberger, David Hinds\n",
      "                             Mary Kate Wing, Goncalo Abecasis \n",
      "\n",
      " Version: 1.0.2;\n",
      " Built: Mon Sep 30 11:52:22 PDT 2019 by biona001\n",
      "\n",
      " Command Line Options: \n",
      "       Reference Haplotypes : --refHaps [./sample_size/1600_haplo_ref.m3vcf.gz],\n",
      "                              --passOnly, --rsid, --referenceEstimates [ON],\n",
      "                              --mapFile [docs/geneticMapFile.b38.map.txt.gz]\n",
      "          Target Haplotypes : --haps [./sample_size/1600_target_masked.vcf.gz]\n",
      "          Output Parameters : --prefix [./sample_size/1600_minimac4.result],\n",
      "                              --estimate, --nobgzip, --vcfBuffer [200],\n",
      "                              --format [GT,DS], --allTypedSites, --meta,\n",
      "                              --memUsage\n",
      "        Chunking Parameters : --ChunkLengthMb [20.00], --ChunkOverlapMb [3.00]\n",
      "          Subset Parameters : --chr [], --start, --end, --window\n",
      "   Approximation Parameters : --minimac3, --probThreshold [0.01],\n",
      "                              --diffThreshold [0.01], --topThreshold [0.01]\n",
      "           Other Parameters : --log, --help, --cpus [1], --params\n",
      "                  PhoneHome : --noPhoneHome, --phoneHomeThinning [50]\n",
      "\n",
      "\n",
      " URL = http://genome.sph.umich.edu/wiki/Minimac4\n",
      " Starting Main Imputation/Estimation Analysis ... \n",
      "\n",
      " Performing preliminary check on input parameters... \n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                             PRELIMINARY FILE CHECK                            \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Checking GWAS haplotype file : ./sample_size/1600_target_masked.vcf.gz\n",
      "\n",
      " Gathering variant information ...\n",
      "\n",
      " Successful !!! \n",
      "\n",
      " Checking Reference haplotype file : ./sample_size/1600_haplo_ref.m3vcf.gz\n",
      "\n",
      " Gathering variant information ...\n",
      "\n",
      " Successful !!! \n",
      "\n",
      " Reference Panel   : Found 400 samples (800 haplotypes) and 35897 variants ...\n",
      "\n",
      " Target/GWAS Panel : Found 1600 samples (3200 haplotypes) and 35897 variants ...\n",
      "                     35897 variants overlap with Reference panel \n",
      "                     0 variants imported that exist only in Target/GWAS panel\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                           CHUNKING INFORMATION                           \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Chunking region into 1 chunk(s) with atleast 35897 variants in each chunk ... \n",
      "\n",
      " Details of chunks is given below ...\n",
      "\n",
      " No   LeftBuffer      LeftEnd   RightPoint  RightBuffer       #Sites(GWAS/Ref/%)\n",
      " -------------------------------------------------------------------------------\n",
      "  1           36           36      4999683      4999683    35897/   35897/ 100.00%\n",
      "\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                           MAIN IMPUTATION ANALYSIS                            \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Starting imputation analysis of 1 chunk(s) ... \n",
      "\n",
      " -------------------------------------------\n",
      " Analyzing Chunk 1/1 [1:36-4999683]\n",
      " -------------------------------------------\n",
      "\n",
      " Reading chunk from reference panel ... \n",
      " Reading chunk from target/GWAS panel ... \n",
      "\n",
      " Compressing reference panel at GWAS sites ... \n",
      " Re-compression successful (10 seconds) !!!\n",
      "\n",
      " Starting Imputation ...\n",
      "\n",
      "  Imputing Samples 1-200 [12%] out of 1600 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 201-400 [25%] out of 1600 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 401-600 [38%] out of 1600 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 601-800 [50%] out of 1600 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 801-1000 [62%] out of 1600 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 1001-1200 [75%] out of 1600 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 1201-1400 [88%] out of 1600 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 1401-1600 [100%] out of 1600 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "\n",
      " Imputation successful (512 seconds) !!!\n",
      "\n",
      " Appending chunk to final output VCF File :  ./sample_size/1600_minimac4.result.dose.vcf.gz \n",
      " Appending successful (6 seconds) !!!\n",
      "\n",
      " Time taken for this chunk = 539 seconds.\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                              SUMMARY OF ANALYSIS                              \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Info file written to                    : ./sample_size/1600_minimac4.result.info\n",
      " Imputed VCF information written to      : ./sample_size/1600_minimac4.result.dose.vcf.gz\n",
      "\n",
      " Time Taken for Reading File             = 16 seconds\n",
      " Time Taken for Re-compression           = 10 seconds\n",
      " Time Taken for Imputation               = 499 seconds\n",
      " Time Taken for Writing File             = 19 seconds\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                                END OF PROGRAM                                 \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Program Successfully Implemented... \n",
      " \n",
      " Total Run completed in 0 hours, 9 mins, 4 seconds.\n",
      "\n",
      " Thank You for using Minimac4 !!! \n",
      "\n",
      "beagle.28Sep18.793.jar (version 5.0)\n",
      "Copyright (C) 2014-2018 Brian L. Browning\n",
      "Enter \"java -jar beagle.28Sep18.793.jar\" to list command line argument\n",
      "Start time: 04:53 AM PST on 11 Dec 2019\n",
      "\n",
      "Command line: java -Xmx3641m -jar beagle.28Sep18.793.jar\n",
      "  gt=./sample_size/1700_target_masked.vcf.gz\n",
      "  ref=./sample_size/1700_haplo_ref.vcf.gz\n",
      "  out=./sample_size/1700_beagle.result\n",
      "\n",
      "No genetic map is specified: using 1 cM = 1 Mb\n",
      "\n",
      "Reference samples:         300\n",
      "Study samples:           1,700\n",
      "\n",
      "Window 1 (1:36-4999683)\n",
      "Reference markers:      35,897\n",
      "Study markers:          35,897\n",
      "\n",
      "Burnin  iteration 1:           1 minute 10 seconds\n",
      "Burnin  iteration 2:           1 minute 28 seconds\n",
      "Burnin  iteration 3:           16 seconds\n",
      "Burnin  iteration 4:           28 seconds\n",
      "Burnin  iteration 5:           35 seconds\n",
      "Burnin  iteration 6:           37 seconds\n",
      "\n",
      "Phasing iteration 1:           21 seconds\n",
      "Phasing iteration 2:           21 seconds\n",
      "Phasing iteration 3:           20 seconds\n",
      "Phasing iteration 4:           20 seconds\n",
      "Phasing iteration 5:           21 seconds\n",
      "Phasing iteration 6:           20 seconds\n",
      "Phasing iteration 7:           20 seconds\n",
      "Phasing iteration 8:           19 seconds\n",
      "Phasing iteration 9:           20 seconds\n",
      "Phasing iteration 10:          19 seconds\n",
      "Phasing iteration 11:          18 seconds\n",
      "Phasing iteration 12:          17 seconds\n",
      "\n",
      "Cumulative Statistics:\n",
      "\n",
      "Study markers:          35,897\n",
      "\n",
      "Haplotype phasing time:        8 minutes 28 seconds\n",
      "Total time:                    9 minutes 1 second\n",
      "\n",
      "End time: 05:02 AM PST on 11 Dec 2019\n",
      "beagle.28Sep18.793.jar finished\n",
      "\n",
      "\n",
      " -------------------------------------------------------------------------------- \n",
      "          Minimac4 - Fast Imputation Based on State Space Reduction HMM\n",
      " --------------------------------------------------------------------------------\n",
      "           (c) 2014 - Sayantan Das, Christian Fuchsberger, David Hinds\n",
      "                             Mary Kate Wing, Goncalo Abecasis \n",
      "\n",
      " Version: 1.0.2;\n",
      " Built: Mon Sep 30 11:52:22 PDT 2019 by biona001\n",
      "\n",
      " Command Line Options: \n",
      "       Reference Haplotypes : --refHaps [./sample_size/1700_haplo_ref.m3vcf.gz],\n",
      "                              --passOnly, --rsid, --referenceEstimates [ON],\n",
      "                              --mapFile [docs/geneticMapFile.b38.map.txt.gz]\n",
      "          Target Haplotypes : --haps [./sample_size/1700_target_masked.vcf.gz]\n",
      "          Output Parameters : --prefix [./sample_size/1700_minimac4.result],\n",
      "                              --estimate, --nobgzip, --vcfBuffer [200],\n",
      "                              --format [GT,DS], --allTypedSites, --meta,\n",
      "                              --memUsage\n",
      "        Chunking Parameters : --ChunkLengthMb [20.00], --ChunkOverlapMb [3.00]\n",
      "          Subset Parameters : --chr [], --start, --end, --window\n",
      "   Approximation Parameters : --minimac3, --probThreshold [0.01],\n",
      "                              --diffThreshold [0.01], --topThreshold [0.01]\n",
      "           Other Parameters : --log, --help, --cpus [1], --params\n",
      "                  PhoneHome : --noPhoneHome, --phoneHomeThinning [50]\n",
      "\n",
      "\n",
      " URL = http://genome.sph.umich.edu/wiki/Minimac4\n",
      " Starting Main Imputation/Estimation Analysis ... \n",
      "\n",
      " Performing preliminary check on input parameters... \n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                             PRELIMINARY FILE CHECK                            \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Checking GWAS haplotype file : ./sample_size/1700_target_masked.vcf.gz\n",
      "\n",
      " Gathering variant information ...\n",
      "\n",
      " Successful !!! \n",
      "\n",
      " Checking Reference haplotype file : ./sample_size/1700_haplo_ref.m3vcf.gz\n",
      "\n",
      " Gathering variant information ...\n",
      "\n",
      " Successful !!! \n",
      "\n",
      " Reference Panel   : Found 300 samples (600 haplotypes) and 35897 variants ...\n",
      "\n",
      " Target/GWAS Panel : Found 1700 samples (3400 haplotypes) and 35897 variants ...\n",
      "                     35897 variants overlap with Reference panel \n",
      "                     0 variants imported that exist only in Target/GWAS panel\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                           CHUNKING INFORMATION                           \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Chunking region into 1 chunk(s) with atleast 35897 variants in each chunk ... \n",
      "\n",
      " Details of chunks is given below ...\n",
      "\n",
      " No   LeftBuffer      LeftEnd   RightPoint  RightBuffer       #Sites(GWAS/Ref/%)\n",
      " -------------------------------------------------------------------------------\n",
      "  1           36           36      4999683      4999683    35897/   35897/ 100.00%\n",
      "\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                           MAIN IMPUTATION ANALYSIS                            \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Starting imputation analysis of 1 chunk(s) ... \n",
      "\n",
      " -------------------------------------------\n",
      " Analyzing Chunk 1/1 [1:36-4999683]\n",
      " -------------------------------------------\n",
      "\n",
      " Reading chunk from reference panel ... \n",
      " Reading chunk from target/GWAS panel ... \n",
      "\n",
      " Compressing reference panel at GWAS sites ... \n",
      " Re-compression successful (7 seconds) !!!\n",
      "\n",
      " Starting Imputation ...\n",
      "\n",
      "  Imputing Samples 1-200 [12%] out of 1700 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 201-400 [24%] out of 1700 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 401-600 [35%] out of 1700 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 601-800 [47%] out of 1700 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 801-1000 [59%] out of 1700 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 1001-1200 [71%] out of 1700 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 1201-1400 [82%] out of 1700 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 1401-1600 [94%] out of 1700 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 1601-1700 [100%] out of 1700 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "\n",
      " Imputation successful (474 seconds) !!!\n",
      "\n",
      " Appending chunk to final output VCF File :  ./sample_size/1700_minimac4.result.dose.vcf.gz \n",
      " Appending successful (6 seconds) !!!\n",
      "\n",
      " Time taken for this chunk = 499 seconds.\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                              SUMMARY OF ANALYSIS                              \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Info file written to                    : ./sample_size/1700_minimac4.result.info\n",
      " Imputed VCF information written to      : ./sample_size/1700_minimac4.result.dose.vcf.gz\n",
      "\n",
      " Time Taken for Reading File             = 17 seconds\n",
      " Time Taken for Re-compression           = 7 seconds\n",
      " Time Taken for Imputation               = 459 seconds\n",
      " Time Taken for Writing File             = 21 seconds\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                                END OF PROGRAM                                 \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Program Successfully Implemented... \n",
      " \n",
      " Total Run completed in 0 hours, 8 mins, 24 seconds.\n",
      "\n",
      " Thank You for using Minimac4 !!! \n",
      "\n",
      "beagle.28Sep18.793.jar (version 5.0)\n",
      "Copyright (C) 2014-2018 Brian L. Browning\n",
      "Enter \"java -jar beagle.28Sep18.793.jar\" to list command line argument\n",
      "Start time: 05:12 AM PST on 11 Dec 2019\n",
      "\n",
      "Command line: java -Xmx3641m -jar beagle.28Sep18.793.jar\n",
      "  gt=./sample_size/1800_target_masked.vcf.gz\n",
      "  ref=./sample_size/1800_haplo_ref.vcf.gz\n",
      "  out=./sample_size/1800_beagle.result\n",
      "\n",
      "No genetic map is specified: using 1 cM = 1 Mb\n",
      "\n",
      "Reference samples:         200\n",
      "Study samples:           1,800\n",
      "\n",
      "Window 1 (1:36-4999683)\n",
      "Reference markers:      35,897\n",
      "Study markers:          35,897\n",
      "\n",
      "Burnin  iteration 1:           1 minute 11 seconds\n",
      "Burnin  iteration 2:           1 minute 41 seconds\n",
      "Burnin  iteration 3:           17 seconds\n",
      "Burnin  iteration 4:           29 seconds\n",
      "Burnin  iteration 5:           38 seconds\n",
      "Burnin  iteration 6:           40 seconds\n",
      "\n",
      "Phasing iteration 1:           22 seconds\n",
      "Phasing iteration 2:           22 seconds\n",
      "Phasing iteration 3:           22 seconds\n",
      "Phasing iteration 4:           21 seconds\n",
      "Phasing iteration 5:           22 seconds\n",
      "Phasing iteration 6:           21 seconds\n",
      "Phasing iteration 7:           21 seconds\n",
      "Phasing iteration 8:           21 seconds\n",
      "Phasing iteration 9:           21 seconds\n",
      "Phasing iteration 10:          20 seconds\n",
      "Phasing iteration 11:          19 seconds\n",
      "Phasing iteration 12:          18 seconds\n",
      "\n",
      "Cumulative Statistics:\n",
      "\n",
      "Study markers:          35,897\n",
      "\n",
      "Haplotype phasing time:        9 minutes 5 seconds\n",
      "Total time:                    9 minutes 41 seconds\n",
      "\n",
      "End time: 05:21 AM PST on 11 Dec 2019\n",
      "beagle.28Sep18.793.jar finished\n",
      "\n",
      "\n",
      " -------------------------------------------------------------------------------- \n",
      "          Minimac4 - Fast Imputation Based on State Space Reduction HMM\n",
      " --------------------------------------------------------------------------------\n",
      "           (c) 2014 - Sayantan Das, Christian Fuchsberger, David Hinds\n",
      "                             Mary Kate Wing, Goncalo Abecasis \n",
      "\n",
      " Version: 1.0.2;\n",
      " Built: Mon Sep 30 11:52:22 PDT 2019 by biona001\n",
      "\n",
      " Command Line Options: \n",
      "       Reference Haplotypes : --refHaps [./sample_size/1800_haplo_ref.m3vcf.gz],\n",
      "                              --passOnly, --rsid, --referenceEstimates [ON],\n",
      "                              --mapFile [docs/geneticMapFile.b38.map.txt.gz]\n",
      "          Target Haplotypes : --haps [./sample_size/1800_target_masked.vcf.gz]\n",
      "          Output Parameters : --prefix [./sample_size/1800_minimac4.result],\n",
      "                              --estimate, --nobgzip, --vcfBuffer [200],\n",
      "                              --format [GT,DS], --allTypedSites, --meta,\n",
      "                              --memUsage\n",
      "        Chunking Parameters : --ChunkLengthMb [20.00], --ChunkOverlapMb [3.00]\n",
      "          Subset Parameters : --chr [], --start, --end, --window\n",
      "   Approximation Parameters : --minimac3, --probThreshold [0.01],\n",
      "                              --diffThreshold [0.01], --topThreshold [0.01]\n",
      "           Other Parameters : --log, --help, --cpus [1], --params\n",
      "                  PhoneHome : --noPhoneHome, --phoneHomeThinning [50]\n",
      "\n",
      "\n",
      " URL = http://genome.sph.umich.edu/wiki/Minimac4\n",
      " Starting Main Imputation/Estimation Analysis ... \n",
      "\n",
      " Performing preliminary check on input parameters... \n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                             PRELIMINARY FILE CHECK                            \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Checking GWAS haplotype file : ./sample_size/1800_target_masked.vcf.gz\n",
      "\n",
      " Gathering variant information ...\n",
      "\n",
      " Successful !!! \n",
      "\n",
      " Checking Reference haplotype file : ./sample_size/1800_haplo_ref.m3vcf.gz\n",
      "\n",
      " Gathering variant information ...\n",
      "\n",
      " Successful !!! \n",
      "\n",
      " Reference Panel   : Found 200 samples (400 haplotypes) and 35897 variants ...\n",
      "\n",
      " Target/GWAS Panel : Found 1800 samples (3600 haplotypes) and 35897 variants ...\n",
      "                     35897 variants overlap with Reference panel \n",
      "                     0 variants imported that exist only in Target/GWAS panel\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                           CHUNKING INFORMATION                           \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Chunking region into 1 chunk(s) with atleast 35897 variants in each chunk ... \n",
      "\n",
      " Details of chunks is given below ...\n",
      "\n",
      " No   LeftBuffer      LeftEnd   RightPoint  RightBuffer       #Sites(GWAS/Ref/%)\n",
      " -------------------------------------------------------------------------------\n",
      "  1           36           36      4999683      4999683    35897/   35897/ 100.00%\n",
      "\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                           MAIN IMPUTATION ANALYSIS                            \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Starting imputation analysis of 1 chunk(s) ... \n",
      "\n",
      " -------------------------------------------\n",
      " Analyzing Chunk 1/1 [1:36-4999683]\n",
      " -------------------------------------------\n",
      "\n",
      " Reading chunk from reference panel ... \n",
      " Reading chunk from target/GWAS panel ... \n",
      "\n",
      " Compressing reference panel at GWAS sites ... \n",
      " Re-compression successful (4 seconds) !!!\n",
      "\n",
      " Starting Imputation ...\n",
      "\n",
      "  Imputing Samples 1-200 [11%] out of 1800 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 201-400 [22%] out of 1800 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 401-600 [33%] out of 1800 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 601-800 [44%] out of 1800 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 801-1000 [56%] out of 1800 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 1001-1200 [67%] out of 1800 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 1201-1400 [78%] out of 1800 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 1401-1600 [89%] out of 1800 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 1601-1800 [100%] out of 1800 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "\n",
      " Imputation successful (422 seconds) !!!\n",
      "\n",
      " Appending chunk to final output VCF File :  ./sample_size/1800_minimac4.result.dose.vcf.gz \n",
      " Appending successful (6 seconds) !!!\n",
      "\n",
      " Time taken for this chunk = 445 seconds.\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                              SUMMARY OF ANALYSIS                              \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Info file written to                    : ./sample_size/1800_minimac4.result.info\n",
      " Imputed VCF information written to      : ./sample_size/1800_minimac4.result.dose.vcf.gz\n",
      "\n",
      " Time Taken for Reading File             = 18 seconds\n",
      " Time Taken for Re-compression           = 4 seconds\n",
      " Time Taken for Imputation               = 407 seconds\n",
      " Time Taken for Writing File             = 21 seconds\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                                END OF PROGRAM                                 \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Program Successfully Implemented... \n",
      " \n",
      " Total Run completed in 0 hours, 7 mins, 30 seconds.\n",
      "\n",
      " Thank You for using Minimac4 !!! \n",
      "\n",
      "beagle.28Sep18.793.jar (version 5.0)\n",
      "Copyright (C) 2014-2018 Brian L. Browning\n",
      "Enter \"java -jar beagle.28Sep18.793.jar\" to list command line argument\n",
      "Start time: 05:30 AM PST on 11 Dec 2019\n",
      "\n",
      "Command line: java -Xmx3641m -jar beagle.28Sep18.793.jar\n",
      "  gt=./sample_size/1900_target_masked.vcf.gz\n",
      "  ref=./sample_size/1900_haplo_ref.vcf.gz\n",
      "  out=./sample_size/1900_beagle.result\n",
      "\n",
      "No genetic map is specified: using 1 cM = 1 Mb\n",
      "\n",
      "Reference samples:         100\n",
      "Study samples:           1,900\n",
      "\n",
      "Window 1 (1:36-4999683)\n",
      "Reference markers:      35,897\n",
      "Study markers:          35,897\n",
      "\n",
      "Burnin  iteration 1:           1 minute 16 seconds\n",
      "Burnin  iteration 2:           1 minute 47 seconds\n",
      "Burnin  iteration 3:           20 seconds\n",
      "Burnin  iteration 4:           33 seconds\n",
      "Burnin  iteration 5:           42 seconds\n",
      "Burnin  iteration 6:           46 seconds\n",
      "\n",
      "Phasing iteration 1:           26 seconds\n",
      "Phasing iteration 2:           26 seconds\n",
      "Phasing iteration 3:           25 seconds\n",
      "Phasing iteration 4:           24 seconds\n",
      "Phasing iteration 5:           25 seconds\n",
      "Phasing iteration 6:           24 seconds\n",
      "Phasing iteration 7:           24 seconds\n",
      "Phasing iteration 8:           24 seconds\n",
      "Phasing iteration 9:           24 seconds\n",
      "Phasing iteration 10:          23 seconds\n",
      "Phasing iteration 11:          22 seconds\n",
      "Phasing iteration 12:          21 seconds\n",
      "\n",
      "Cumulative Statistics:\n",
      "\n",
      "Study markers:          35,897\n",
      "\n",
      "Haplotype phasing time:        10 minutes 12 seconds\n",
      "Total time:                    10 minutes 50 seconds\n",
      "\n",
      "End time: 05:41 AM PST on 11 Dec 2019\n",
      "beagle.28Sep18.793.jar finished\n",
      "\n",
      "\n",
      " -------------------------------------------------------------------------------- \n",
      "          Minimac4 - Fast Imputation Based on State Space Reduction HMM\n",
      " --------------------------------------------------------------------------------\n",
      "           (c) 2014 - Sayantan Das, Christian Fuchsberger, David Hinds\n",
      "                             Mary Kate Wing, Goncalo Abecasis \n",
      "\n",
      " Version: 1.0.2;\n",
      " Built: Mon Sep 30 11:52:22 PDT 2019 by biona001\n",
      "\n",
      " Command Line Options: \n",
      "       Reference Haplotypes : --refHaps [./sample_size/1900_haplo_ref.m3vcf.gz],\n",
      "                              --passOnly, --rsid, --referenceEstimates [ON],\n",
      "                              --mapFile [docs/geneticMapFile.b38.map.txt.gz]\n",
      "          Target Haplotypes : --haps [./sample_size/1900_target_masked.vcf.gz]\n",
      "          Output Parameters : --prefix [./sample_size/1900_minimac4.result],\n",
      "                              --estimate, --nobgzip, --vcfBuffer [200],\n",
      "                              --format [GT,DS], --allTypedSites, --meta,\n",
      "                              --memUsage\n",
      "        Chunking Parameters : --ChunkLengthMb [20.00], --ChunkOverlapMb [3.00]\n",
      "          Subset Parameters : --chr [], --start, --end, --window\n",
      "   Approximation Parameters : --minimac3, --probThreshold [0.01],\n",
      "                              --diffThreshold [0.01], --topThreshold [0.01]\n",
      "           Other Parameters : --log, --help, --cpus [1], --params\n",
      "                  PhoneHome : --noPhoneHome, --phoneHomeThinning [50]\n",
      "\n",
      "\n",
      " URL = http://genome.sph.umich.edu/wiki/Minimac4\n",
      " Starting Main Imputation/Estimation Analysis ... \n",
      "\n",
      " Performing preliminary check on input parameters... \n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                             PRELIMINARY FILE CHECK                            \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Checking GWAS haplotype file : ./sample_size/1900_target_masked.vcf.gz\n",
      "\n",
      " Gathering variant information ...\n",
      "\n",
      " Successful !!! \n",
      "\n",
      " Checking Reference haplotype file : ./sample_size/1900_haplo_ref.m3vcf.gz\n",
      "\n",
      " Gathering variant information ...\n",
      "\n",
      " Successful !!! \n",
      "\n",
      " Reference Panel   : Found 100 samples (200 haplotypes) and 35897 variants ...\n",
      "\n",
      " Target/GWAS Panel : Found 1900 samples (3800 haplotypes) and 35897 variants ...\n",
      "                     35897 variants overlap with Reference panel \n",
      "                     0 variants imported that exist only in Target/GWAS panel\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                           CHUNKING INFORMATION                           \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Chunking region into 1 chunk(s) with atleast 35897 variants in each chunk ... \n",
      "\n",
      " Details of chunks is given below ...\n",
      "\n",
      " No   LeftBuffer      LeftEnd   RightPoint  RightBuffer       #Sites(GWAS/Ref/%)\n",
      " -------------------------------------------------------------------------------\n",
      "  1           36           36      4999683      4999683    35897/   35897/ 100.00%\n",
      "\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                           MAIN IMPUTATION ANALYSIS                            \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Starting imputation analysis of 1 chunk(s) ... \n",
      "\n",
      " -------------------------------------------\n",
      " Analyzing Chunk 1/1 [1:36-4999683]\n",
      " -------------------------------------------\n",
      "\n",
      " Reading chunk from reference panel ... \n",
      " Reading chunk from target/GWAS panel ... \n",
      "\n",
      " Compressing reference panel at GWAS sites ... \n",
      " Re-compression successful (2 seconds) !!!\n",
      "\n",
      " Starting Imputation ...\n",
      "\n",
      "  Imputing Samples 1-200 [11%] out of 1900 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 201-400 [21%] out of 1900 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 401-600 [32%] out of 1900 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 601-800 [42%] out of 1900 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 801-1000 [53%] out of 1900 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 1001-1200 [63%] out of 1900 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 1201-1400 [74%] out of 1900 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 1401-1600 [84%] out of 1900 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 1601-1800 [95%] out of 1900 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "  Imputing Samples 1801-1900 [100%] out of 1900 samples ...\n",
      "       Saving Samples in temporary VCF file ... \n",
      "\n",
      " Imputation successful (336 seconds) !!!\n",
      "\n",
      " Appending chunk to final output VCF File :  ./sample_size/1900_minimac4.result.dose.vcf.gz \n",
      " Appending successful (7 seconds) !!!\n",
      "\n",
      " Time taken for this chunk = 358 seconds.\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                              SUMMARY OF ANALYSIS                              \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Info file written to                    : ./sample_size/1900_minimac4.result.info\n",
      " Imputed VCF information written to      : ./sample_size/1900_minimac4.result.dose.vcf.gz\n",
      "\n",
      " Time Taken for Reading File             = 19 seconds\n",
      " Time Taken for Re-compression           = 2 seconds\n",
      " Time Taken for Imputation               = 318 seconds\n",
      " Time Taken for Writing File             = 25 seconds\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "                                END OF PROGRAM                                 \n",
      " ------------------------------------------------------------------------------\n",
      "\n",
      " Program Successfully Implemented... \n",
      " \n",
      " Total Run completed in 0 hours, 6 mins, 4 seconds.\n",
      "\n",
      " Thank You for using Minimac4 !!! \n",
      "\n"
     ]
    }
   ],
   "source": [
    "mendel_error   = zeros(19)\n",
    "mendel_times   = zeros(19) #sec\n",
    "mendel_mem     = zeros(19) #MB\n",
    "beagle5_error  = zeros(19)\n",
    "beagle5_times  = zeros(19)\n",
    "beagle5_mem    = zeros(19)\n",
    "minimac4_error = zeros(19)\n",
    "minimac4_times = zeros(19)\n",
    "minimac4_mem   = zeros(19)\n",
    "for i in 1:19\n",
    "    #\n",
    "    # current run info\n",
    "    #\n",
    "    X_complete = convert_gt(Float32, \"./compare1/$(100i)_target.vcf.gz\"; as_minorallele=false)\n",
    "    X_masked   = convert_gt(Float32, \"./compare1/$(100i)_target_masked.vcf.gz\"; as_minorallele=false)\n",
    "    missing_idx = ismissing.(X_masked)\n",
    "    total_missing = sum(missing_idx)\n",
    "    actual_missing_values = X_complete[missing_idx]  #true values of missing entries\n",
    "    #\n",
    "    # MendelImpute (window = 400)\n",
    "    #\n",
    "    tgtfile = \"./compare1/$(100i)_target_masked.vcf.gz\"\n",
    "    reffile = \"./compare1/$(100i)_haplo_ref.vcf.gz\"\n",
    "    outfile = \"./compare1/$(100i)_mendel.result.vcf.gz\"\n",
    "    width   = 400\n",
    "    mendel_result = @timed phase(tgtfile, reffile, impute=true, outfile = outfile, width = width)\n",
    "    X_mendel = convert_gt(Float32, outfile, as_minorallele=false)\n",
    "    imputed_missing_values = X_mendel[missing_idx] #imputed values of missing entries\n",
    "    mendel_err = sum(actual_missing_values .!= imputed_missing_values) / total_missing\n",
    "    mendel_error[i] = mendel_err\n",
    "    mendel_times[i] = mendel_result[2] \n",
    "    mendel_mem[i]   = mendel_result[3] / 1e6\n",
    "    #\n",
    "    # Beagle 5\n",
    "    #\n",
    "    beagle_result = @timed run(`java -jar beagle.28Sep18.793.jar gt=./compare1/$(100i)_target_masked.vcf.gz ref=./compare1/$(100i)_haplo_ref.vcf.gz out=./compare1/$(100i)_beagle.result`)\n",
    "    X_beagle = convert_gt(Float32, \"./compare1/$(100i)_beagle.result.vcf.gz\", as_minorallele=false)\n",
    "    imputed_missing_values = X_beagle[missing_idx] #imputed values of missing entries\n",
    "    beagle_err = sum(actual_missing_values .!= imputed_missing_values) / total_missing\n",
    "    beagle5_error[i] = beagle_err\n",
    "    beagle5_times[i] = beagle_result[2]\n",
    "    beagle5_mem[i]   = beagle_result[3] / 1e6\n",
    "    #\n",
    "    # minimac 4\n",
    "    #\n",
    "    minimac4 = \"/Users/biona001/Benjamin_Folder/UCLA/research/softwares/Minimac4/build/minimac4\"\n",
    "    minimac_result = @timed run(`$minimac4 --refHaps ./compare1/$(100i)_haplo_ref.m3vcf.gz --haps ./compare1/$(100i)_target_masked.vcf.gz --prefix ./compare1/$(100i)_minimac4.result`)\n",
    "    X_minimac = convert_gt(Float32, \"./compare1/$(100i)_minimac4.result.dose.vcf.gz\", as_minorallele=false)\n",
    "    imputed_missing_values = X_minimac[missing_idx] #imputed values of missing entries\n",
    "    minimac_err = sum(actual_missing_values .!= imputed_missing_values) / total_missing\n",
    "    minimac4_error[i] = minimac_err\n",
    "    minimac4_times[i] = minimac_result[2]\n",
    "    minimac4_mem[i]   = minimac_result[3] / 1e6\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "19×5 Array{Float64,2}:\n",
       "  100.0  0.00221891  0.0023776   0.000203238  0.000139204\n",
       "  200.0  0.00199824  0.00219166  0.000193423  0.000151677\n",
       "  300.0  0.00207736  0.00236603  0.000194926  0.000160582\n",
       "  400.0  0.00209315  0.00264185  0.000192882  0.000165725\n",
       "  500.0  0.00216476  0.00258356  0.000203277  0.000182671\n",
       "  600.0  0.00219264  0.00273848  0.000207763  0.000199879\n",
       "  700.0  0.00231082  0.00301655  0.000325232  0.000222255\n",
       "  800.0  0.00230644  0.0030015   0.000306829  0.000253256\n",
       "  900.0  0.00237092  0.00357184  0.000355452  0.000292034\n",
       " 1000.0  0.00237504  0.003796    0.000222999  0.000304014\n",
       " 1100.0  0.00258552  0.00416094  0.000282526  0.000345563\n",
       " 1200.0  0.0027198   0.00451528  0.000276924  0.000392058\n",
       " 1300.0  0.00286455  0.00493401  0.000262245  0.000429576\n",
       " 1400.0  0.00313687  0.00590106  0.000382832  0.000525101\n",
       " 1500.0  0.00342885  0.00667756  0.00028193   0.00059302 \n",
       " 1600.0  0.0039391   0.00811048  0.000361835  0.000764765\n",
       " 1700.0  0.00495996  0.0108217   0.000301035  0.00108501 \n",
       " 1800.0  0.00692264  0.0152656   0.000352911  0.00158253 \n",
       " 1900.0  0.0123286   0.0269323   0.000391532  0.00321922 "
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "haplotype_reference_size = reverse(collect(200:200:3800))\n",
    "sample_size = (4000 .- haplotype_reference_size) ./ 2\n",
    "[sample_size mendel_error mendel_1200_error beagle5_error minimac4_error]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "19×5 Array{Float64,2}:\n",
       "  100.0  14.0551  14.9088   56.7424  151.583\n",
       "  200.0  16.2895  18.4679   87.6731  211.091\n",
       "  300.0  16.7011  19.5794  126.033   271.342\n",
       "  400.0  18.8638  20.8112  143.121   327.088\n",
       "  500.0  20.2512  22.8566  178.576   383.784\n",
       "  600.0  22.5655  23.8698  207.525   428.037\n",
       "  700.0  24.7581  28.3699  244.148   472.338\n",
       "  800.0  26.3045  28.0556  266.5     502.989\n",
       "  900.0  27.4628  32.357   294.07    562.578\n",
       " 1000.0  28.8264  32.5215  322.526   560.161\n",
       " 1100.0  30.4397  35.3225  354.228   578.269\n",
       " 1200.0  31.9834  35.5846  391.811   586.907\n",
       " 1300.0  32.9212  36.2763  414.914   589.967\n",
       " 1400.0  34.4399  39.7434  442.836   586.003\n",
       " 1500.0  35.168   39.3551  477.496   574.747\n",
       " 1600.0  36.6773  40.5079  508.32    546.31 \n",
       " 1700.0  37.6205  40.7102  541.092   506.69 \n",
       " 1800.0  38.2893  41.0959  581.237   453.177\n",
       " 1900.0  39.4661  43.7754  650.147   366.959"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[sample_size mendel_times mendel_1200_times beagle5_times minimac4_times]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "19×5 Array{Float64,2}:\n",
       " 3800.0  15532.2  15551.6  0.270384  0.291456\n",
       " 3600.0  17149.8  17155.0  0.299632  0.34152 \n",
       " 3400.0  18322.8  18322.4  0.339152  0.422752\n",
       " 3200.0  19291.7  19286.8  0.343488  0.460464\n",
       " 3000.0  20393.4  20376.3  0.375648  0.552896\n",
       " 2800.0  22445.3  22450.0  0.398048  0.563216\n",
       " 2600.0  24612.7  24580.7  0.447408  0.62712 \n",
       " 2400.0  25746.6  25713.0  0.450192  0.652592\n",
       " 2200.0  26698.5  26659.0  0.474176  0.728304\n",
       " 2000.0  27054.0  27009.5  0.498496  0.718784\n",
       " 1800.0  28604.7  28552.0  0.524752  0.742224\n",
       " 1600.0  29547.8  29493.1  0.564912  0.753552\n",
       " 1400.0  30495.1  30437.3  0.575632  0.782336\n",
       " 1200.0  31441.0  31378.2  0.6024    0.768928\n",
       " 1000.0  32094.8  32029.8  0.630352  0.781536\n",
       "  800.0  33044.6  32975.0  0.655648  0.766336\n",
       "  600.0  33981.5  33911.3  0.684736  0.735168\n",
       "  400.0  34778.0  34704.6  0.720272  0.701008\n",
       "  200.0  35650.6  35571.7  0.777568  0.659888"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[haplotype_reference_size mendel_mem mendel_1200_mem beagle5_mem minimac4_mem] # the real memory usage is not correctly benchmarked in all 3 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"utf-8\"?>\n",
       "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" width=\"600\" height=\"400\" viewBox=\"0 0 2400 1600\">\n",
       "<defs>\n",
       "  <clipPath id=\"clip5700\">\n",
       "    <rect x=\"0\" y=\"0\" width=\"2400\" height=\"1600\"/>\n",
       "  </clipPath>\n",
       "</defs>\n",
       "<path clip-path=\"url(#clip5700)\" d=\"\n",
       "M0 1600 L2400 1600 L2400 0 L0 0  Z\n",
       "  \" fill=\"#ffffff\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<defs>\n",
       "  <clipPath id=\"clip5701\">\n",
       "    <rect x=\"480\" y=\"0\" width=\"1681\" height=\"1600\"/>\n",
       "  </clipPath>\n",
       "</defs>\n",
       "<path clip-path=\"url(#clip5700)\" d=\"\n",
       "M269.279 1425.62 L2352.76 1425.62 L2352.76 121.675 L269.279 121.675  Z\n",
       "  \" fill=\"#ffffff\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<defs>\n",
       "  <clipPath id=\"clip5702\">\n",
       "    <rect x=\"269\" y=\"121\" width=\"2084\" height=\"1305\"/>\n",
       "  </clipPath>\n",
       "</defs>\n",
       "<polyline clip-path=\"url(#clip5702)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  492.04,1425.62 492.04,121.675 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip5702)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  765.033,1425.62 765.033,121.675 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip5702)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  1038.03,1425.62 1038.03,121.675 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip5702)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  1311.02,1425.62 1311.02,121.675 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip5702)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  1584.01,1425.62 1584.01,121.675 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip5702)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  1857,1425.62 1857,121.675 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip5702)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  2129.99,1425.62 2129.99,121.675 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip5702)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  269.279,1395.11 2352.76,1395.11 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip5702)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  269.279,1165.54 2352.76,1165.54 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip5702)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  269.279,935.982 2352.76,935.982 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip5702)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  269.279,706.42 2352.76,706.42 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip5702)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  269.279,476.858 2352.76,476.858 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip5702)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  269.279,247.296 2352.76,247.296 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip5700)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  269.279,1425.62 2352.76,1425.62 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip5700)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  269.279,1425.62 269.279,121.675 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip5700)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  492.04,1425.62 492.04,1406.06 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip5700)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  765.033,1425.62 765.033,1406.06 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip5700)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1038.03,1425.62 1038.03,1406.06 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip5700)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1311.02,1425.62 1311.02,1406.06 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip5700)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1584.01,1425.62 1584.01,1406.06 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip5700)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1857,1425.62 1857,1406.06 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip5700)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  2129.99,1425.62 2129.99,1406.06 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip5700)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  269.279,1395.11 300.531,1395.11 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip5700)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  269.279,1165.54 300.531,1165.54 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip5700)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  269.279,935.982 300.531,935.982 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip5700)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  269.279,706.42 300.531,706.42 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip5700)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  269.279,476.858 300.531,476.858 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip5700)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  269.279,247.296 300.531,247.296 \n",
       "  \"/>\n",
       "<g clip-path=\"url(#clip5700)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:middle;\" transform=\"rotate(0, 492.04, 1479.62)\" x=\"492.04\" y=\"1479.62\">500</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip5700)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:middle;\" transform=\"rotate(0, 765.033, 1479.62)\" x=\"765.033\" y=\"1479.62\">1000</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip5700)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:middle;\" transform=\"rotate(0, 1038.03, 1479.62)\" x=\"1038.03\" y=\"1479.62\">1500</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip5700)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:middle;\" transform=\"rotate(0, 1311.02, 1479.62)\" x=\"1311.02\" y=\"1479.62\">2000</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip5700)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:middle;\" transform=\"rotate(0, 1584.01, 1479.62)\" x=\"1584.01\" y=\"1479.62\">2500</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip5700)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:middle;\" transform=\"rotate(0, 1857, 1479.62)\" x=\"1857\" y=\"1479.62\">3000</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip5700)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:middle;\" transform=\"rotate(0, 2129.99, 1479.62)\" x=\"2129.99\" y=\"1479.62\">3500</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip5700)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:end;\" transform=\"rotate(0, 245.279, 1412.61)\" x=\"245.279\" y=\"1412.61\">0.000</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip5700)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:end;\" transform=\"rotate(0, 245.279, 1183.04)\" x=\"245.279\" y=\"1183.04\">0.005</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip5700)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:end;\" transform=\"rotate(0, 245.279, 953.482)\" x=\"245.279\" y=\"953.482\">0.010</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip5700)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:end;\" transform=\"rotate(0, 245.279, 723.92)\" x=\"245.279\" y=\"723.92\">0.015</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip5700)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:end;\" transform=\"rotate(0, 245.279, 494.358)\" x=\"245.279\" y=\"494.358\">0.020</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip5700)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:end;\" transform=\"rotate(0, 245.279, 264.796)\" x=\"245.279\" y=\"264.796\">0.025</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip5700)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:84px; text-anchor:middle;\" transform=\"rotate(0, 1311.02, 73.2)\" x=\"1311.02\" y=\"73.2\">Error rate with varying reference panel size</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip5700)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:66px; text-anchor:middle;\" transform=\"rotate(0, 1311.02, 1559.48)\" x=\"1311.02\" y=\"1559.48\">Total number of haplotypes in reference panel</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip5700)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:66px; text-anchor:middle;\" transform=\"rotate(-90, 89.2861, 773.647)\" x=\"89.2861\" y=\"773.647\">Error rate</text>\n",
       "</g>\n",
       "<polyline clip-path=\"url(#clip5702)\" style=\"stroke:#009af9; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  2293.79,1293.23 2184.59,1303.36 2075.4,1299.73 1966.2,1299 1857,1295.72 1747.8,1294.44 1638.61,1289.01 1529.41,1289.21 1420.21,1286.25 1311.02,1286.06 \n",
       "  1201.82,1276.4 1092.62,1270.23 983.427,1263.59 874.23,1251.08 765.033,1237.68 655.836,1214.25 546.639,1167.38 437.442,1077.27 328.245,829.071 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip5702)\" style=\"stroke:#e26f46; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  2293.79,1285.94 2184.59,1294.48 2075.4,1286.48 1966.2,1273.81 1857,1276.49 1747.8,1269.38 1638.61,1256.61 1529.41,1257.3 1420.21,1231.11 1311.02,1220.82 \n",
       "  1201.82,1204.07 1092.62,1187.8 983.427,1168.57 874.23,1124.17 765.033,1088.52 655.836,1022.73 546.639,898.255 437.442,694.226 328.245,158.579 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip5702)\" style=\"stroke:#3da44d; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  2293.79,1385.77 2184.59,1386.23 2075.4,1386.16 1966.2,1386.25 1857,1385.77 1747.8,1385.57 1638.61,1380.17 1529.41,1381.02 1420.21,1378.79 1311.02,1384.87 \n",
       "  1201.82,1382.13 1092.62,1382.39 983.427,1383.07 874.23,1377.53 765.033,1382.16 655.836,1378.49 546.639,1381.28 437.442,1378.9 328.245,1377.13 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip5702)\" style=\"stroke:#c271d2; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  2293.79,1388.71 2184.59,1388.14 2075.4,1387.73 1966.2,1387.5 1857,1386.72 1747.8,1385.93 1638.61,1384.9 1529.41,1383.48 1420.21,1381.7 1311.02,1381.15 \n",
       "  1201.82,1379.24 1092.62,1377.11 983.427,1375.38 874.23,1371 765.033,1367.88 655.836,1359.99 546.639,1345.29 437.442,1322.45 328.245,1247.3 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip5700)\" d=\"\n",
       "M1493.67 507.595 L2280.76 507.595 L2280.76 205.195 L1493.67 205.195  Z\n",
       "  \" fill=\"#ffffff\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip5700)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1493.67,507.595 2280.76,507.595 2280.76,205.195 1493.67,205.195 1493.67,507.595 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip5700)\" style=\"stroke:#009af9; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1517.67,265.675 1661.67,265.675 \n",
       "  \"/>\n",
       "<g clip-path=\"url(#clip5700)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:start;\" transform=\"rotate(0, 1685.67, 283.175)\" x=\"1685.67\" y=\"283.175\">MendelImpute (w = 400)</text>\n",
       "</g>\n",
       "<polyline clip-path=\"url(#clip5700)\" style=\"stroke:#e26f46; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1517.67,326.155 1661.67,326.155 \n",
       "  \"/>\n",
       "<g clip-path=\"url(#clip5700)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:start;\" transform=\"rotate(0, 1685.67, 343.655)\" x=\"1685.67\" y=\"343.655\">MendelImpute (w = 1200)</text>\n",
       "</g>\n",
       "<polyline clip-path=\"url(#clip5700)\" style=\"stroke:#3da44d; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1517.67,386.635 1661.67,386.635 \n",
       "  \"/>\n",
       "<g clip-path=\"url(#clip5700)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:start;\" transform=\"rotate(0, 1685.67, 404.135)\" x=\"1685.67\" y=\"404.135\">Beagle 5.0</text>\n",
       "</g>\n",
       "<polyline clip-path=\"url(#clip5700)\" style=\"stroke:#c271d2; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1517.67,447.115 1661.67,447.115 \n",
       "  \"/>\n",
       "<g clip-path=\"url(#clip5700)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:start;\" transform=\"rotate(0, 1685.67, 464.615)\" x=\"1685.67\" y=\"464.615\">Minimac 4</text>\n",
       "</g>\n",
       "</svg>\n"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "plot(haplotype_reference_size, [mendel_error mendel_1200_error beagle5_error minimac4_error],\n",
    "    label = [\"MendelImpute (w = 400)\", \"MendelImpute (w = 1200)\", \"Beagle 5.0\", \"Minimac 4\"], \n",
    "    title = \"Error rate with varying reference panel size\",\n",
    "    ylabel = \"Error rate\", xlabel = \"Total number of haplotypes in reference panel\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"utf-8\"?>\n",
       "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" width=\"600\" height=\"400\" viewBox=\"0 0 2400 1600\">\n",
       "<defs>\n",
       "  <clipPath id=\"clip4900\">\n",
       "    <rect x=\"0\" y=\"0\" width=\"2400\" height=\"1600\"/>\n",
       "  </clipPath>\n",
       "</defs>\n",
       "<path clip-path=\"url(#clip4900)\" d=\"\n",
       "M0 1600 L2400 1600 L2400 0 L0 0  Z\n",
       "  \" fill=\"#ffffff\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<defs>\n",
       "  <clipPath id=\"clip4901\">\n",
       "    <rect x=\"480\" y=\"0\" width=\"1681\" height=\"1600\"/>\n",
       "  </clipPath>\n",
       "</defs>\n",
       "<path clip-path=\"url(#clip4900)\" d=\"\n",
       "M229.135 1425.62 L2352.76 1425.62 L2352.76 121.675 L229.135 121.675  Z\n",
       "  \" fill=\"#ffffff\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<defs>\n",
       "  <clipPath id=\"clip4902\">\n",
       "    <rect x=\"229\" y=\"121\" width=\"2125\" height=\"1305\"/>\n",
       "  </clipPath>\n",
       "</defs>\n",
       "<polyline clip-path=\"url(#clip4902)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  456.189,1425.62 456.189,121.675 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip4902)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  734.441,1425.62 734.441,121.675 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip4902)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  1012.69,1425.62 1012.69,121.675 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip4902)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  1290.95,1425.62 1290.95,121.675 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip4902)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  1569.2,1425.62 1569.2,121.675 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip4902)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  1847.45,1425.62 1847.45,121.675 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip4902)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  2125.7,1425.62 2125.7,121.675 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip4902)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  229.135,1415.9 2352.76,1415.9 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip4902)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  229.135,1222.51 2352.76,1222.51 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip4902)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  229.135,1029.12 2352.76,1029.12 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip4902)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  229.135,835.727 2352.76,835.727 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip4902)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  229.135,642.338 2352.76,642.338 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip4902)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  229.135,448.948 2352.76,448.948 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip4902)\" style=\"stroke:#000000; stroke-width:2; stroke-opacity:0.1; fill:none\" points=\"\n",
       "  229.135,255.559 2352.76,255.559 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip4900)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  229.135,1425.62 2352.76,1425.62 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip4900)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  229.135,1425.62 229.135,121.675 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip4900)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  456.189,1425.62 456.189,1406.06 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip4900)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  734.441,1425.62 734.441,1406.06 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip4900)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1012.69,1425.62 1012.69,1406.06 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip4900)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1290.95,1425.62 1290.95,1406.06 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip4900)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1569.2,1425.62 1569.2,1406.06 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip4900)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1847.45,1425.62 1847.45,1406.06 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip4900)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  2125.7,1425.62 2125.7,1406.06 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip4900)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  229.135,1415.9 260.99,1415.9 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip4900)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  229.135,1222.51 260.99,1222.51 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip4900)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  229.135,1029.12 260.99,1029.12 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip4900)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  229.135,835.727 260.99,835.727 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip4900)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  229.135,642.338 260.99,642.338 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip4900)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  229.135,448.948 260.99,448.948 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip4900)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  229.135,255.559 260.99,255.559 \n",
       "  \"/>\n",
       "<g clip-path=\"url(#clip4900)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:middle;\" transform=\"rotate(0, 456.189, 1479.62)\" x=\"456.189\" y=\"1479.62\">250</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip4900)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:middle;\" transform=\"rotate(0, 734.441, 1479.62)\" x=\"734.441\" y=\"1479.62\">500</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip4900)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:middle;\" transform=\"rotate(0, 1012.69, 1479.62)\" x=\"1012.69\" y=\"1479.62\">750</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip4900)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:middle;\" transform=\"rotate(0, 1290.95, 1479.62)\" x=\"1290.95\" y=\"1479.62\">1000</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip4900)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:middle;\" transform=\"rotate(0, 1569.2, 1479.62)\" x=\"1569.2\" y=\"1479.62\">1250</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip4900)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:middle;\" transform=\"rotate(0, 1847.45, 1479.62)\" x=\"1847.45\" y=\"1479.62\">1500</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip4900)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:middle;\" transform=\"rotate(0, 2125.7, 1479.62)\" x=\"2125.7\" y=\"1479.62\">1750</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip4900)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:end;\" transform=\"rotate(0, 205.135, 1433.4)\" x=\"205.135\" y=\"1433.4\">0</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip4900)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:end;\" transform=\"rotate(0, 205.135, 1240.01)\" x=\"205.135\" y=\"1240.01\">100</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip4900)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:end;\" transform=\"rotate(0, 205.135, 1046.62)\" x=\"205.135\" y=\"1046.62\">200</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip4900)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:end;\" transform=\"rotate(0, 205.135, 853.227)\" x=\"205.135\" y=\"853.227\">300</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip4900)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:end;\" transform=\"rotate(0, 205.135, 659.838)\" x=\"205.135\" y=\"659.838\">400</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip4900)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:end;\" transform=\"rotate(0, 205.135, 466.448)\" x=\"205.135\" y=\"466.448\">500</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip4900)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:end;\" transform=\"rotate(0, 205.135, 273.059)\" x=\"205.135\" y=\"273.059\">600</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip4900)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:84px; text-anchor:middle;\" transform=\"rotate(0, 1290.95, 73.2)\" x=\"1290.95\" y=\"73.2\">Computational times with varying imputation sample size</text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip4900)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:66px; text-anchor:middle;\" transform=\"rotate(0, 1290.95, 1559.48)\" x=\"1290.95\" y=\"1559.48\">Sample size </text>\n",
       "</g>\n",
       "<g clip-path=\"url(#clip4900)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:66px; text-anchor:middle;\" transform=\"rotate(-90, 89.2861, 773.647)\" x=\"89.2861\" y=\"773.647\">Time (seconds)</text>\n",
       "</g>\n",
       "<polyline clip-path=\"url(#clip4902)\" style=\"stroke:#009af9; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  289.238,1388.71 400.539,1384.39 511.839,1383.6 623.14,1379.41 734.441,1376.73 845.742,1372.26 957.043,1368.02 1068.34,1365.03 1179.64,1362.79 1290.95,1360.15 \n",
       "  1402.25,1357.03 1513.55,1354.04 1624.85,1352.23 1736.15,1349.29 1847.45,1347.88 1958.75,1344.97 2070.05,1343.14 2181.35,1341.85 2292.65,1339.57 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip4902)\" style=\"stroke:#e26f46; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  289.238,1387.06 400.539,1380.18 511.839,1378.03 623.14,1375.65 734.441,1371.69 845.742,1369.73 957.043,1361.03 1068.34,1361.64 1179.64,1353.32 1290.95,1353 \n",
       "  1402.25,1347.59 1513.55,1347.08 1624.85,1345.74 1736.15,1339.04 1847.45,1339.79 1958.75,1337.56 2070.05,1337.17 2181.35,1336.42 2292.65,1331.24 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip4902)\" style=\"stroke:#3da44d; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  289.238,1306.16 400.539,1246.34 511.839,1172.16 623.14,1139.11 734.441,1070.55 845.742,1014.56 957.043,943.74 1068.34,900.512 1179.64,847.195 1290.95,792.165 \n",
       "  1402.25,730.856 1513.55,658.175 1624.85,613.496 1736.15,559.497 1847.45,492.469 1958.75,432.858 2070.05,369.481 2181.35,291.844 2292.65,158.579 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip4902)\" style=\"stroke:#c271d2; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  289.238,1122.75 400.539,1007.67 511.839,891.148 623.14,783.343 734.441,673.697 845.742,588.117 957.043,502.444 1068.34,443.168 1179.64,327.93 1290.95,332.603 \n",
       "  1402.25,297.584 1513.55,280.878 1624.85,274.962 1736.15,282.628 1847.45,304.396 1958.75,359.389 2070.05,436.01 2181.35,539.498 2292.65,706.235 \n",
       "  \"/>\n",
       "<path clip-path=\"url(#clip4900)\" d=\"\n",
       "M1493.67 507.595 L2280.76 507.595 L2280.76 205.195 L1493.67 205.195  Z\n",
       "  \" fill=\"#ffffff\" fill-rule=\"evenodd\" fill-opacity=\"1\"/>\n",
       "<polyline clip-path=\"url(#clip4900)\" style=\"stroke:#000000; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1493.67,507.595 2280.76,507.595 2280.76,205.195 1493.67,205.195 1493.67,507.595 \n",
       "  \"/>\n",
       "<polyline clip-path=\"url(#clip4900)\" style=\"stroke:#009af9; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1517.67,265.675 1661.67,265.675 \n",
       "  \"/>\n",
       "<g clip-path=\"url(#clip4900)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:start;\" transform=\"rotate(0, 1685.67, 283.175)\" x=\"1685.67\" y=\"283.175\">MendelImpute (w = 400)</text>\n",
       "</g>\n",
       "<polyline clip-path=\"url(#clip4900)\" style=\"stroke:#e26f46; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1517.67,326.155 1661.67,326.155 \n",
       "  \"/>\n",
       "<g clip-path=\"url(#clip4900)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:start;\" transform=\"rotate(0, 1685.67, 343.655)\" x=\"1685.67\" y=\"343.655\">MendelImpute (w = 1200)</text>\n",
       "</g>\n",
       "<polyline clip-path=\"url(#clip4900)\" style=\"stroke:#3da44d; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1517.67,386.635 1661.67,386.635 \n",
       "  \"/>\n",
       "<g clip-path=\"url(#clip4900)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:start;\" transform=\"rotate(0, 1685.67, 404.135)\" x=\"1685.67\" y=\"404.135\">Beagle 5.0</text>\n",
       "</g>\n",
       "<polyline clip-path=\"url(#clip4900)\" style=\"stroke:#c271d2; stroke-width:4; stroke-opacity:1; fill:none\" points=\"\n",
       "  1517.67,447.115 1661.67,447.115 \n",
       "  \"/>\n",
       "<g clip-path=\"url(#clip4900)\">\n",
       "<text style=\"fill:#000000; fill-opacity:1; font-family:Arial,Helvetica Neue,Helvetica,sans-serif; font-size:48px; text-anchor:start;\" transform=\"rotate(0, 1685.67, 464.615)\" x=\"1685.67\" y=\"464.615\">Minimac 4</text>\n",
       "</g>\n",
       "</svg>\n"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "plot(sample_size, [mendel_times mendel_1200_times beagle5_times minimac4_times],\n",
    "    label = [\"MendelImpute (w = 400)\", \"MendelImpute (w = 1200)\", \"Beagle 5.0\", \"Minimac 4\"], \n",
    "    title = \"Computational times with varying imputation sample size\",\n",
    "    ylabel = \"Time (seconds)\", xlabel = \"Sample size \")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "@webio": {
   "lastCommId": null,
   "lastKernelId": null
  },
  "kernelspec": {
   "display_name": "Julia 1.3.0-rc4",
   "language": "julia",
   "name": "julia-1.3"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.3.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
