<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Phasing and Imputation · MendelImpute</title><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.11.1/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit">MendelImpute</span></div><form class="docs-search" action="../../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li class="is-active"><a class="tocitem" href>Phasing and Imputation</a><ul class="internal"><li class="toplevel"><a class="tocitem" href="#Preparing-Reference-Haplotype-Panel"><span>Preparing Reference Haplotype Panel</span></a></li><li class="toplevel"><a class="tocitem" href="#Detailed-Example"><span>Detailed Example</span></a></li><li><a class="tocitem" href="#Step-1:-generating-realistic-reference-and-target-data"><span>Step 1: generating realistic reference and target data</span></a></li><li><a class="tocitem" href="#Step-2:-generating-.jlso-compressed-reference-panel"><span>Step 2: generating <code>.jlso</code> compressed reference panel</span></a></li><li><a class="tocitem" href="#Step-3:-Run-imputation-and-phasing"><span>Step 3: Run imputation and phasing</span></a></li><li><a class="tocitem" href="#Step-3.5:-(only-for-simulated-data)-check-imputation-accuracy"><span>Step 3.5: (only for simulated data) check imputation accuracy</span></a></li><li><a class="tocitem" href="#Post-imputation:-per-SNP-Imputation-Quality-Score"><span>Post-imputation: per-SNP Imputation Quality Score</span></a></li><li><a class="tocitem" href="#Post-imputation:-per-sample-Imputation-Quality-score"><span>Post-imputation: per-sample Imputation Quality score</span></a></li></ul></li><li><a class="tocitem" href="../performance/">Performance Gotchas</a></li><li><a class="tocitem" href="../painting/">Estimating ancestry</a></li><li><a class="tocitem" href="../ultra+compress/">Ultra compression</a></li><li><a class="tocitem" href="../script/">Run as script</a></li><li><a class="tocitem" href="../api/">API</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Phasing and Imputation</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Phasing and Imputation</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/OpenMendel/MendelImpute.jl/blob/master/docs/src/man/Phasing_and_Imputation.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Preparing-Target-Data"><a class="docs-heading-anchor" href="#Preparing-Target-Data">Preparing Target Data</a><a id="Preparing-Target-Data-1"></a><a class="docs-heading-anchor-permalink" href="#Preparing-Target-Data" title="Permalink"></a></h1><p>MendelImpute accepts <a href="https://samtools.github.io/hts-specs/VCFv4.3.pdf">VCF</a> and <a href="https://www.cog-genomics.org/plink2/formats#bed">PLINK</a> files. Please make sure the following are true:</p><ul><li>VCF file ends in <code>.vcf</code> or <code>.vcf.gz</code> (phased or unphased and may contain missing data)</li><li>For PLINK files, all trios (<code>.bim</code>, <code>.bed</code>, <code>.fam</code>) are present in the same directory</li><li>Each file contains only 1 (non-sex) chromosome</li><li>Every record (SNP) is present in the reference panel. If this is untrue, you must <a href="https://openmendel.github.io/VCFTools.jl/dev/man/conformgt/">match markers in 2 VCF files</a>. </li><li>Given a SNP, it&#39;s CHROM, POS, REF, and  ALT fields are the same in target data and reference panel. MendelImpute use SNP position internally to align markers. Note this is not explicitly checked. </li><li>The position of every SNP is unique: so multiallelic markers should be excluded instead of split (this requirement will eventually be lifted). </li></ul><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>We recommend <strong>not</strong> doing pre-imputation quality control (i.e. do not remove SNPs out of Hardy-Weinburg equilibrium or SNPs with low minor allele frequencies...etc). Instead, reserve quality control to post-imputation. This is because every new set of typed SNPs requires a compressed reference haplotype panel that matches the set of typed SNPs. </p></div></div><h1 id="Preparing-Reference-Haplotype-Panel"><a class="docs-heading-anchor" href="#Preparing-Reference-Haplotype-Panel">Preparing Reference Haplotype Panel</a><a id="Preparing-Reference-Haplotype-Panel-1"></a><a class="docs-heading-anchor-permalink" href="#Preparing-Reference-Haplotype-Panel" title="Permalink"></a></h1><p>Reference samples must all be phased and contain no missing genotypes. Reference VCF panels must be compressed into <code>.jlso</code> format first using the <a href="https://OpenMendel.github.io/MendelImpute.jl/dev/man/api/#MendelImpute.compress_haplotypes">compress_haplotypes</a> function. One must specify <code>d</code>: the maximum number of unique haplotypes per window. Larger <code>d</code> slows down computation, but increases accuracy. For most purposes, we recommend <span>$d \approx 1000$</span>. </p><h1 id="Detailed-Example"><a class="docs-heading-anchor" href="#Detailed-Example">Detailed Example</a><a id="Detailed-Example-1"></a><a class="docs-heading-anchor-permalink" href="#Detailed-Example" title="Permalink"></a></h1><p>We use the <a href="http://bochet.gcc.biostat.washington.edu/beagle/1000_Genomes_phase3_v5a/b37.vcf/">1000 genomes chromosome 22</a> as an example. As show below, this data contains 424147 SNPs and 2504 samples.</p><pre><code class="language-julia"># load necessary packages in Julia
using VCFTools
using MendelImpute
using VCFTools
using Random

# compute simple summary statistics
data = &quot;chr22.1kg.phase3.v5a.vcf.gz&quot;
@show nrecords(data)
@show nsamples(data);</code></pre><pre><code class="language-none">┌ Info: Precompiling MendelImpute [e47305d1-6a61-5370-bc5d-77554d143183]
└ @ Base loading.jl:1278


nrecords(data) = 424147
nsamples(data) = 2504</code></pre><p>More summary statistics can be computed using the <a href="https://openmendel.github.io/VCFTools.jl/dev/man/api/#VCFTools.gtstats">gtstats</a> function in <code>VCFTools.jl</code>, with example usage <a href="https://openmendel.github.io/VCFTools.jl/dev/man/summaryinfo/#Summary-statistics">here</a>.</p><h2 id="Step-1:-generating-realistic-reference-and-target-data"><a class="docs-heading-anchor" href="#Step-1:-generating-realistic-reference-and-target-data">Step 1: generating realistic reference and target data</a><a id="Step-1:-generating-realistic-reference-and-target-data-1"></a><a class="docs-heading-anchor-permalink" href="#Step-1:-generating-realistic-reference-and-target-data" title="Permalink"></a></h2><p>First we generate a reference panel and imputation target based on the 1000 genomes data. More specifically, we take the 1000 genomes chromosome 22 and divide it so that </p><ul><li>100 samples are randomly selected as imputation targets, where<ul><li>100k SNPs with minor allele frequency <span>$\ge 0.05$</span> are randomly selected to be the typed positions. </li><li>0.1% of typed SNPs are masked (mimicking GWAS errors)</li><li>Genotypes are unphased</li></ul></li><li>The remaining 2404 samples are used as reference haplotypes. </li><li>SNPs with duplicate positions are filtered out.</li><li>All multiallelic markers are filtered out.</li></ul><p><strong>Instruction: execute the code below in a Julia session or a Jupyter notebook:</strong></p><pre><code class="language-julia"># set random seed for reproducibility
Random.seed!(2020)

# download example data 
data = &quot;chr22.1kg.phase3.v5a.vcf.gz&quot;
if !isfile(data) 
    download(&quot;http://bochet.gcc.biostat.washington.edu/beagle/1000_Genomes_phase3_v5a/b37.vcf/chr22.1kg.phase3.v5a.vcf.gz&quot;)
end

# remove SNPs with the same positions, keep all samples, save result into new file
SNPs_to_keep = .!find_duplicate_marker(data) 
@time VCFTools.filter(data, SNPs_to_keep, 1:nsamples(data), des = &quot;chr22.uniqueSNPs.vcf.gz&quot;)

# summarize data
total_snps, samples, _, _, _, maf_by_record, _ = gtstats(&quot;chr22.uniqueSNPs.vcf.gz&quot;)

# generate target file with 100 samples and 100k snps with maf&gt;0.05
n = 100
p = 100000
record_idx = falses(total_snps)
large_maf = findall(x -&gt; x &gt; 0.05, maf_by_record)  
Random.shuffle!(large_maf)
record_idx[large_maf[1:p]] .= true
sample_idx = falses(samples)
sample_idx[1:n] .= true
Random.shuffle!(sample_idx)
@time VCFTools.filter(&quot;chr22.uniqueSNPs.vcf.gz&quot;, record_idx, sample_idx, 
    des = &quot;target.chr22.typedOnly.vcf.gz&quot;, allow_multiallelic=false)

# unphase and mask 0.1% entries in target file
masks = falses(p, n)
missingprop = 0.001
for j in 1:n, i in 1:p
    rand() &lt; missingprop &amp;&amp; (masks[i, j] = true)
end
@time mask_gt(&quot;target.chr22.typedOnly.vcf.gz&quot;, masks, 
    des=&quot;target.chr22.typedOnly.masked.vcf.gz&quot;, unphase=true)

# generate target panel with all snps (this file contains true phase and genotypes)
@time VCFTools.filter(&quot;chr22.uniqueSNPs.vcf.gz&quot;, 1:total_snps, 
    sample_idx, des = &quot;target.chr22.full.vcf.gz&quot;, allow_multiallelic=false)

# generate reference panel with 2404 samples
@time VCFTools.filter(&quot;chr22.uniqueSNPs.vcf.gz&quot;, 1:total_snps, .!sample_idx, 
    des = &quot;ref.chr22.excludeTarget.vcf.gz&quot;, allow_multiallelic=false)</code></pre><pre><code class="language-none">┌ Info: Precompiling MendelImpute [e47305d1-6a61-5370-bc5d-77554d143183]
└ @ Base loading.jl:1278
[32mfinding duplicate markers...100%|███████████████████████| Time: 0:03:56[39m
[32mfiltering vcf file...100%|██████████████████████████████| Time: 0:04:46[39m


292.131527 seconds (3.20 G allocations: 301.789 GiB, 7.89% gc time)


[32mProgress: 100%|█████████████████████████████████████████| Time: 0:04:02[39m
[32mfiltering vcf file...100%|██████████████████████████████| Time: 0:03:59[39m


244.425505 seconds (3.18 G allocations: 301.694 GiB, 9.69% gc time)
  1.935526 seconds (20.00 M allocations: 1.491 GiB, 6.33% gc time)


[32mfiltering vcf file...100%|██████████████████████████████| Time: 0:04:10[39m


255.505399 seconds (3.27 G allocations: 317.749 GiB, 9.95% gc time)


[32mfiltering vcf file...100%|██████████████████████████████| Time: 0:07:27[39m


453.383147 seconds (6.16 G allocations: 566.535 GiB, 10.16% gc time)</code></pre><h3 id="Output-explanation:"><a class="docs-heading-anchor" href="#Output-explanation:">Output explanation:</a><a id="Output-explanation:-1"></a><a class="docs-heading-anchor-permalink" href="#Output-explanation:" title="Permalink"></a></h3><p>You just generated reference and target VCF files:</p><ul><li><code>ref.chr22.excludeTarget.vcf.gz</code>: Reference haplotype panel with 2404 samples</li><li><code>target.chr22.typedOnly.masked.vcf.gz</code>: Imputation target file containing 100 samples at 100k SNPs. All genotypes are unphased and contains 0.1% missing data. </li></ul><p>You also generated/downloaded:</p><ul><li><code>chr22.1kg.phase3.v5a.vcf.gz</code>: The original chromosome 22 data downloaded from Beagle&#39;s website.</li><li><code>chr22.uniqueSNPs.vcf.gz</code>: This is the original chromosome 22 data excluding duplicate records (SNPs) by checking marker positions. The first SNP is included but all subsequent SNPs are removed. </li><li><code>target.chr22.full.vcf.gz</code>: The complete data for imputation target, used for checking imputation accuracy. All genotypes are phased and non-missing. </li><li><code>target.chr22.typedOnly.vcf.gz</code>: Complete target data on just the typed SNPs. All genotypes are phased and non-missing. Just by-producted for generating other files; not used for anything downstream.</li></ul><h2 id="Step-2:-generating-.jlso-compressed-reference-panel"><a class="docs-heading-anchor" href="#Step-2:-generating-.jlso-compressed-reference-panel">Step 2: generating <code>.jlso</code> compressed reference panel</a><a id="Step-2:-generating-.jlso-compressed-reference-panel-1"></a><a class="docs-heading-anchor-permalink" href="#Step-2:-generating-.jlso-compressed-reference-panel" title="Permalink"></a></h2><p>MendelImpute requires one to pre-process the reference panel for faster reading. This is achieved via the <a href="https://openmendel.github.io/MendelImpute.jl/dev/man/api/#MendelImpute.compress_haplotypes">compress_haplotypes</a> function.</p><pre><code class="language-julia"># load necessary packages in Julia
using MendelImpute

max_d = 1000 # maximum number of unique haplotypes per window
reffile = &quot;ref.chr22.excludeTarget.vcf.gz&quot;
tgtfile = &quot;target.chr22.typedOnly.masked.vcf.gz&quot;
outfile = &quot;ref.chr22.maxd1000.excludeTarget.jlso&quot;
@time compress_haplotypes(reffile, tgtfile, outfile, max_d)</code></pre><pre><code class="language-none">[32mimporting reference data...100%|████████████████████████| Time: 0:02:00[39m


295.546081 seconds (2.09 G allocations: 209.215 GiB, 10.53% gc time)</code></pre><h2 id="Step-3:-Run-imputation-and-phasing"><a class="docs-heading-anchor" href="#Step-3:-Run-imputation-and-phasing">Step 3: Run imputation and phasing</a><a id="Step-3:-Run-imputation-and-phasing-1"></a><a class="docs-heading-anchor-permalink" href="#Step-3:-Run-imputation-and-phasing" title="Permalink"></a></h2><p>Below runs the main <a href="https://openmendel.github.io/MendelImpute.jl/dev/man/api/#MendelImpute.phase">phase</a> function in a single thread. By default all output genotypes will be phased and non-missing. A list of optional inputs can be found in the <a href="https://openmendel.github.io/MendelImpute.jl/dev/man/api/#MendelImpute.phase">API</a>.</p><pre><code class="language-julia"># note: run twice for more accurate timing
reffile = &quot;ref.chr22.maxd1000.excludeTarget.jlso&quot; # jlso reference file
tgtfile = &quot;target.chr22.typedOnly.masked.vcf.gz&quot;  # target genotype file
outfile = &quot;mendel.imputed.chr22.vcf.gz&quot;           # output file name
phase(tgtfile, reffile, outfile);</code></pre><pre><code class="language-none">Number of threads = 8
Importing reference haplotype data...
Total windows = 1634, averaging ~ 508 unique haplotypes per window.

Timings: 
    Data import                     = 9.9879 seconds
        import target data             = 1.59052 seconds
        import compressed haplotypes   = 8.39737 seconds
    Computing haplotype pair        = 4.487 seconds
        BLAS3 mul! to get M and N      = 0.203134 seconds per thread
        haplopair search               = 3.47789 seconds per thread
        initializing missing           = 0.0193766 seconds per thread
        allocating and viewing         = 0.0948972 seconds per thread
        index conversion               = 0.00162558 seconds per thread
    Phasing by win-win intersection = 0.661786 seconds
        Window-by-window intersection  = 0.0792241 seconds per thread
        Breakpoint search              = 0.48882 seconds per thread
        Recording result               = 0.0196465 seconds per thread
    Imputation                     = 1.10519 seconds
        Imputing missing               = 0.509987 seconds
        Writing to file                = 0.595203 seconds

    Total time                      = 16.243 seconds</code></pre><p>The <code>;</code> hides the output, or else the screen will be too jammed. </p><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>To run MendelImpute in parallel, type <code>export JULIA_NUM_THREADS=4</code> <strong>before</strong> starting Julia or Jupyter notebooks. See Performance Gotchas #1 on the left for details.</p></div></div><h2 id="Step-3.5:-(only-for-simulated-data)-check-imputation-accuracy"><a class="docs-heading-anchor" href="#Step-3.5:-(only-for-simulated-data)-check-imputation-accuracy">Step 3.5: (only for simulated data) check imputation accuracy</a><a id="Step-3.5:-(only-for-simulated-data)-check-imputation-accuracy-1"></a><a class="docs-heading-anchor-permalink" href="#Step-3.5:-(only-for-simulated-data)-check-imputation-accuracy" title="Permalink"></a></h2><p>Since we simulated data, we can check imputation accuracy.</p><pre><code class="language-julia">X_truth  = convert_gt(Float64, &quot;target.chr22.full.vcf.gz&quot;)    # import true genotypes
X_mendel = convert_gt(Float64, &quot;mendel.imputed.chr22.vcf.gz&quot;) # import imputed genotypes
n, p = size(X_mendel)
println(&quot;error overall = &quot;, sum(X_mendel .!= X_truth) / n / p)</code></pre><pre><code class="language-none">error overall = 0.005273778941849357</code></pre><h2 id="Post-imputation:-per-SNP-Imputation-Quality-Score"><a class="docs-heading-anchor" href="#Post-imputation:-per-SNP-Imputation-Quality-Score">Post-imputation: per-SNP Imputation Quality Score</a><a id="Post-imputation:-per-SNP-Imputation-Quality-Score-1"></a><a class="docs-heading-anchor-permalink" href="#Post-imputation:-per-SNP-Imputation-Quality-Score" title="Permalink"></a></h2><p>By default, MendelImpute outputs a per-SNP imputation quality score. A score of 0 is best, and larger is worst. At the <span>$k$</span>th SNP, this score is the average least squares error between the observed genotype and the 2 selected haplotypes <span>$\mathbf{h}_1, \mathbf{h}_2$</span> over all <span>$n$</span> samples:</p><div>\[e_k = \frac{1}{n}\sum_{i=1}^n ||x_k - h_{1k} - h_{2k}||_2^2\]</div><p>For untyped (i.e. imputed) SNPs, the quality score is the average of the closest 2 typed SNP&#39;s score. Note samples with missing genotypes are skipped, and <span>$n$</span> adjusted accordingly, by default. </p><p>To extract this score from a VCF file, one can do:</p><pre><code class="language-julia">using GeneticVariation, Plots
reader = VCF.Reader(openvcf(outfile, &quot;r&quot;))
snpscores = Vector{Float64}(undef, nrecords(outfile))

# loop over SNPs
for (i, record) in enumerate(reader)
    snpscores[i] = parse(Float64, VCF.info(record)[1].second)
end
close(reader)

# plot histogram of SNP scores
histogram(snpscores, label=:none, xlabel=&quot;error&quot;, ylabel=&quot;SNP counts&quot;, bins=30)</code></pre><p><img src="../output_16_0.svg" alt="svg"/></p><p><strong>Conclusion:</strong> Most SNPs are well imputed (least squares error close to <span>$0.0$</span>), but a few have quality score above <span>$0.01$</span>. We recommend such SNPs to be <a href="https://openmendel.github.io/VCFTools.jl/dev/man/filter/#Subsetting-VCF-files-using-array-masks">filtered out</a>. </p><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>The popular correlation metric <span>$r^2$</span> such as <a href="https://genome.sph.umich.edu/wiki/Minimac3_Info_File#Rsq">this one</a> is NOT a good metric for measuring imputation accuracy under MendelImpute&#39;s model, because all imputed SNPs will have <span>$r^2 = 1$</span>. For details, please see our paper. </p></div></div><h2 id="Post-imputation:-per-sample-Imputation-Quality-score"><a class="docs-heading-anchor" href="#Post-imputation:-per-sample-Imputation-Quality-score">Post-imputation: per-sample Imputation Quality score</a><a id="Post-imputation:-per-sample-Imputation-Quality-score-1"></a><a class="docs-heading-anchor-permalink" href="#Post-imputation:-per-sample-Imputation-Quality-score" title="Permalink"></a></h2><p>MendelImpute also computes a rough quality score (file ending in <code>sample.error</code>) for measuring how well each sample is imputed. This value is the sum of the least squares error for all typed SNPs scaled by the total number of observed SNPs. A value of 0 is best, and high values mean worse. </p><pre><code class="language-julia">using CSV
quality = CSV.read(&quot;mendel.imputed.chr22.sample.error&quot;) # import quality score 

# visualize error distribution
histogram(quality[:error], label=:none, xlabel=&quot;error&quot;, ylabel=&quot;Number of samples&quot;) </code></pre><p><img src="../output_20_0.svg" alt="svg"/></p><p><strong>Conclusion:</strong> Most samples have small error, but some samples are indeed poorly imputed. From the histogram, we can safely <a href="https://openmendel.github.io/VCFTools.jl/dev/man/filter/#Subsetting-VCF-files-using-array-masks">filtered out</a> samples with error <span>$&gt; 0.004$</span> as that would remove poorly imputed individuals without reducing sample size too much.</p></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../../">« Home</a><a class="docs-footer-nextpage" href="../performance/">Performance Gotchas »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> on <span class="colophon-date" title="Wednesday 21 October 2020 23:02">Wednesday 21 October 2020</span>. Using Julia version 1.5.2.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
